{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Parsing the output from CAZyme prediction tools\n",
    "\n",
    "The desired output is a Pandas dataframe\n",
    "Columns:\n",
    "- Protein ID (unique identifier)\n",
    "- Predicated CAZy family\n",
    "- Predicated CAZy subfamily\n",
    "These columns will be standard across all the dataframes and have the same names.\n",
    "\n",
    "Each prediction tool will also have additional columns repsective of its output."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Requirement already satisfied: pandas in /home/em/anaconda3/lib/python3.8/site-packages (1.0.5)\n",
      "Requirement already satisfied: pytz>=2017.2 in /home/em/anaconda3/lib/python3.8/site-packages (from pandas) (2020.1)\n",
      "Requirement already satisfied: numpy>=1.13.3 in /home/em/anaconda3/lib/python3.8/site-packages (from pandas) (1.18.5)\n",
      "Requirement already satisfied: python-dateutil>=2.6.1 in /home/em/anaconda3/lib/python3.8/site-packages (from pandas) (2.8.1)\n",
      "Requirement already satisfied: six>=1.5 in /home/em/anaconda3/lib/python3.8/site-packages (from python-dateutil>=2.6.1->pandas) (1.15.0)\n",
      "Requirement already satisfied: numpy in /home/em/anaconda3/lib/python3.8/site-packages (1.18.5)\n"
     ]
    }
   ],
   "source": [
    "!pip3 install pandas\n",
    "!pip3 install numpy"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Parsing CUPP\n",
    "\n",
    "### Understanding the output\n",
    "\n",
    "The output FASTA file and log file produce the same output, but present the data in different formats.\n",
    "If output data is missing, becuase it was not calculated, nothing is provided in its place. This can make it difficult to determine specifically what data has been produced by CUPP.\n",
    "\n",
    "#### FASTA File\n",
    "The format of the data is as follows for each protein\n",
    "```\n",
    "> accession | raw_function | best_function | additional_info | subfam \\n seq\n",
    "```\n",
    "\n",
    "For example:\n",
    "```\n",
    ">QIX00022.1 AMS68_005539|&|AA1:1.10.3.2|AA1:1.10.3.2|AA1:6.1(20.3,124..357)|AA1_3\n",
    "mvrsvnfplgvyfqykfikigtdttltweadpnreytvpsdcntspvlsgpwqyppvstsaavaiaatsttaqasavvtsqptycpntpttrncwadgfdintdfdnhwpttgrtvsytfeitnttlApDgvSRPvfaingqyPGPtIYADWgDMISVtvinklqdngttIHFHGVrQWHSNtqdgvpGLTECpIAPGTTRTyTfqatQfGTSwYHSHFSCQYGdGILGPiviqGpAtANYDIDVGpLpitdwyyptvnyhasitehtngsppvadnalINgTMTSSsggayaltklttgkrhRLrLINTSVDNHFMVSLdghqievitaDFvpIVPYNtTWimlgigqRYDVIITADQAPgnywfraevqdgcganannkniksmfaypghenetpistgsgytgrctdetqlvphwnsyvppkagfaeptffgtalnqstaadgtltlywqvngsslniqwdkptlqyvqesntsypsganvvetategqwtywviqevpgtpynvnvphplhLHGHDFFilGtgtdtwdakfsqnlnynnpprrdtamlpsngwivlawqtdnpgawLFhCHiAwhsdeglgvqfleaadqmqsvnpigtefdqqcqawhdfypvhasylkhdsgv\n",
    "```\n",
    "\n",
    "**Accession** This is the protein ID and any other data that was provided on the description line for the protein in its source FASTA file. If the protein is retrieved from the genomic assembly a gene name may have already been assigned to the protein, and this gene name may appear in CAZy for another protein. For example, the GenBank protein [XP_001545605.1](https://www.ncbi.nlm.nih.gov/protein/XP_001545605.1), is annotated as a hypothetical protein. CUPP has labeled it as BCIN_01g00260 and in [UniProt](https://www.uniprot.org/uniprot/A0A384J3W2) this is the name of the gene and in CAZy this is the name of a [AA4 CAZyme](http://www.cazy.org/AA4_all.html) with the GenBank accession number ATZ45204.1.\n",
    "\n",
    "**Raw function** This is produced when \"consider[ing] all EC functions of a CUPP equally\", as was described in the comment of the Python script file. This is written as the predicated CAZy family ':' predicated EC number\n",
    "\n",
    "**Best function** This is the \"most occuring EC function of the CUPP group\",as was described in the comment of the Python script file. It is inferred from the raw code that the 'CUPP group' refers to the CAZy family k-mer cluster for which the query protein scored most highly and has thus been assigned. -- This will taken as the predicated CAZy family for the query protein. In the fasta file, this is written as the predicated CAZy family ':' predicated EC number.\n",
    "\n",
    "**Additional information** This includes several pieces of information, each of which are only included if they were calculated by CUPP. Therefore, it can be difficult to determine which specific pieces of inforamtion have been provided. This information is presented in the following order, and as defined in the comments of the Python script:\n",
    "- \"Ratio of maximum frequencies\"\n",
    "- \"Ratio of maximum frequencies by coverage\"\n",
    "- \"Centre for mass for domain\"\n",
    "- \"Domain range\" -- this is identifable by the presense of '...' between two numbers and could potentially be useful when looking for regions of positive selection down the road\n",
    "- \"Number of important positions in domain\" -- very ambgious of meaning\n",
    "\n",
    "**Subfam** This is the predicated CAZy subfamily for the query protein\n",
    "\n",
    "**Seq** This is the query protein sequence -- have not had a change to check why some letters are lower case and some are upper\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Output log file\n",
    "The log file presents the same data as the fasta file, but instead of separating the data by '|' it instead separates the data by tabulated white space '\\t'.\n",
    "\n",
    "For example:\n",
    "QIX00022.1 AMS68_005539\tAA1\tAA1:6.1\t20.3\t11.61\t256\t124..357\t115\tAA1:1.10.3.2\tAA1:1.10.3.2\tAA1_3\n",
    "\n",
    "The data is presented in a slightly different order as well:\n",
    "- Predicated CAZy family (equivalent to the FASTA file best function)\n",
    "- CUPP (k-mer cluster) group which query protein scored highest for, and the associated score\n",
    "- Additional information\n",
    "- Raw_function (same as the fasta file (cazy_fam:ec#))\n",
    "- Best_function (same as the fasta file (cazy_fam:ec#))\n",
    "- Subfam\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      ">XP_001545605.1 BCIN_01g00260|&|AA4:1.1.3.38|AA4:1.1.3.38|AA4:1.1(13.9,122..568)|\n",
      "maasivttdtplalrvpnaattpaandplalppgtssekfaefishaeqttgaenviiirkkeeltkelytdpskahdmyhlfdkdyfvvsatiaprnveevqaivrlcndyeiplwpfsigrnvGYGgAAPRVPGSigldmgknmnkvlevnsegayallEPGvTFFSLydylkekklddqlwidtpdlgggsvvGNTIERGvGYTPYGDHFMMHCGLeVVlPTGELIRTGMGALPDPTQptveggsldqqpgnkcwqlfpygfgpYnDGIfSQSNlGIVTKMGIWLMpNPGGyQpylitfpkdtdlpkivdiirplrlqmviqnvpsirhilldaavmgtkksyydvdrplneeeldaiakelnlgrwnfygAlYgPKPvrdvlwqvvkdafgtiegakfflpedikekcvlhiraktlqgiptidelswvdwlpngahlffspiskisgedaskqyaltqkmtleagfdfigtftigmremhhiiclvfdrededqkrrahklireliqvcadngwgeYrTHIALMdqiaetygwndnaqmklnekiknaLDpKGILaPGKNGVWpAsYdrsawrldansertrtrp\n",
      ">XP_024545922.1 BCIN_01g00340|&|GT2:2.4.1.117|GT2:2.4.1.117|GT2:23.1(20.2,116..300)|\n",
      "mdgllelpvqilgglwevvrdtpvhvlgviligllglglgliyallllvapvprppypsektyitttpsgttetkpltcwydswvahreasvsksdscpantirtgaiepatlemslvvPaYNEEERLIgMLeealsfldttygrvargtgtgtgyeillvNDgSRDRTVEIAlDFSrknglhdvlrivtleenRGkGGAVTHGMRHVRGEYAVfADADGASRFADLGkLvkgvrevvdeegrgvavGSRaWLVGseAvvkRSALRNtlMhsfhlllrlltppatsrirdtQcGFKLFTRAAlPhiipymhaegwifdvemlmlaesapgvedaekgkgnggkakgikvseqpiawqevggsklnvmwdslgmawglavlrggwgmgvwrrr\n",
      ">XP_001547235.1 BCIN_01g00640|&|AA3:Unknown|AA3:Unknown|AA3:3.1(64.0,26..553)|AA3_1\n",
      "mkvsyqfslvvglvtgvwgqadnawiqeqWDAIIVGSGPAGIVVASRLSEAGLKTLLLEGGNQsYgvTGGDLDARRpEWLSgTNlSRVDVPGLYKSIFADGGglvcgaevnayGGCTIGGSSAINAGLFFEPPASDWDLYFPaEWNSTNMDaSIQRlYDTQTSTNlTSqDGIRSLQTgYDaARKWLVDGLGYKDVDINAQADdkTDVFGyPIFDYnNgQrGGPVTTYLQSALQRSNFRLQSGVRVVRVERDGDTATGVTALINGVeTIISTtPTgRVVLSAgAIQSPSLLMYSGIgDAETLsKLSAAGKlSPNLTSAEWInNTAIGAGLFDNPNTFIELQGDSiESYTySYDSPPaDDKalyldSRSgPYSFASETSVFWTTIAHADgSIAGlqGTiDSSGYSDFttnntiTLNVYGTSGLLSSGKVILdDNFIPGPSDDvYysNPRDaKDIAsFIHTIFSaLPnSGLtSMNIpqNATESeIetYITTASAYARGMVNHwSSSCRLGSCVDVDTTVIGmQNVHvVdASILAPVTTNPQFAVMAAGERASELILGVAGlecsfcasddecere\n",
      ">XP_001547254.1 BCIN_01g00800|&|AA1:1.10.3.2|AA1:1.10.3.2|AA1:6.1(10.5,202..455)|AA1_3\n",
      "mrlhillalclsfcqihasplfgwsfnpiatlqadllkiinqaeclvatkvsiipsiltfgnsnpvtsnkcsttsgkttssaaiqtslsslqsvasskwsslqsllstttslppilttsftattakptttssvkststtantvsstakpssitstasstasstsktsssavasgtpcagntaadrskwcnyststdyynevpdtgVtREYWFNVQdgvaspdgisryvqtingsiPGpTIIADWgDNVVVhvtnslttngssIHfHGIhQKNTNqndgvpSVTQcPIAYGDTyTyrwratqyGSSWYhSHVGLQawEgvtggiiInGPATANYDEDKGTLmlsdwghetvdelyqkvqtvgpqsmttglinGTNVFGAdgasnqtgsrfstsvqsgksyRFrLVNSAIDsqfkfsvdshtmtvmamdfvpIVPYQtEIlniaigQRYDVVITAnQASVAsdffiraipqsscsendnsdnirgvlhygsstglptttgytftdecvdepvaslvpylpktvssaskspeeavtvaqnsqklfkwylnnstflsewedptllmvqnnvtafgtpdnlvevpnanewiylvihsalpvphpihLHGHDFFVvSQqsttfdattavstynlnnpprrdvatlpgggylvlafetdnpgAWLMhCHIGWHtsmgfamqfleraseipklldttqmldtctawnghaiiqedsgv\n",
      ">XP_024546029.1 BCIN_01g01760|&|GH18:Unknown|GH18:Unknown|GH18:12.1(11.4,116..258)+PL7:0.1(1.1,989..1040)|PL7_2\n",
      "mypsksftvaaatfaliapaaalydasspanlalywgsgpsqtnlsyyceqstvdiiplafmnvfpaqgdgypaenfgnacygqpiftpgpgyplgdvntskdqlyvqcpgiqegipycQSLGKKiLLSLGGASNTYqltgaadgeYfADFLWGSYgpfkqswldaggirpmdggyygtdssvhididgfdFDIEFGPTdysegyiamiNRLRehFaenpskkyfISGAPqCPLPEPnMGAMIAgaQfdllWIqFYnNAAAQctarqwadnyaltgqedsaeftydqwlstinngasagasiylgllgstlagtasdyispleaqsliesyhnkpqfggvmiweatysqentdeelkgdsyhgfiksclasyapppptstssivhsttissssstsevsitssvpastessaatsstvisisststsvsessststltssessttsssslletsttstststsvsetstvtsssslessaaqsstysevlptqassgytipgyshkatsslthsqvhsskpagsytftghsskptsiatgytskshsakpsskpssgytsvyssnsttsgsvhssrpystksvvtgtgvsskpyhtkssvigtgastkpyptkssivgtgqsskprptkslgtgvkscksrsktrtrtktktktssagpsgypastssvvisstapygnstyshsettpassatssvegitvssgtatetgpgssatdstvvgpstsgtselptgsatnsgsvtsetgsasasgsissgtgsitsaiqyttstayvttvytitscaatvtdcpgrigsvttetissyttvcpvtatatstssaegisvsskpssapgvtasqtpsssaqgsatlpitsvvqyttstayvtsvytitscaatvtdcpgrigsvttetissyttvcpvtetsgnsgivsstsspaggpvptttvtsygtstvyqtsvytitscapsvtdcparsgqvttetissytsvypitvstapinvptsgASsAAGIetsvpgvkvsstasiftsapaAGAssIAGVssavptkgASSAAGIvtsvpvigasssaatpeaasptslyttlistlinsqtttisaivapsiitvvpvaiaspspasspytsiappygagngtlvatgtassitrvpkpssyvvkaegestsaksgsvttqipvatpsavtfeggaeklvkglgvwaavgmmglvlfi\n"
     ]
    }
   ],
   "source": [
    "# reading the fasta file\n",
    "with open(\"gca_143535_cupp.fasta\", \"r\") as fh:\n",
    "    fasta_file = fh.read().splitlines()\n",
    "for line in fasta_file[:10]:\n",
    "    print(line)\n",
    "\n",
    "# data is separated by '|'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "'>XP_001545605.1 BCIN_01g00260|&|AA4:1.1.3.38|AA4:1.1.3.38|AA4:1.1(13.9,122..568)|'\n",
      "'maasivttdtplalrvpnaattpaandplalppgtssekfaefishaeqttgaenviiirkkeeltkelytdpskahdmyhlfdkdyfvvsatiaprnveevqaivrlcndyeiplwpfsigrnvGYGgAAPRVPGSigldmgknmnkvlevnsegayallEPGvTFFSLydylkekklddqlwidtpdlgggsvvGNTIERGvGYTPYGDHFMMHCGLeVVlPTGELIRTGMGALPDPTQptveggsldqqpgnkcwqlfpygfgpYnDGIfSQSNlGIVTKMGIWLMpNPGGyQpylitfpkdtdlpkivdiirplrlqmviqnvpsirhilldaavmgtkksyydvdrplneeeldaiakelnlgrwnfygAlYgPKPvrdvlwqvvkdafgtiegakfflpedikekcvlhiraktlqgiptidelswvdwlpngahlffspiskisgedaskqyaltqkmtleagfdfigtftigmremhhiiclvfdrededqkrrahklireliqvcadngwgeYrTHIALMdqiaetygwndnaqmklnekiknaLDpKGILaPGKNGVWpAsYdrsawrldansertrtrp'\n",
      "'>XP_024545922.1 BCIN_01g00340|&|GT2:2.4.1.117|GT2:2.4.1.117|GT2:23.1(20.2,116..300)|'\n",
      "'mdgllelpvqilgglwevvrdtpvhvlgviligllglglgliyallllvapvprppypsektyitttpsgttetkpltcwydswvahreasvsksdscpantirtgaiepatlemslvvPaYNEEERLIgMLeealsfldttygrvargtgtgtgyeillvNDgSRDRTVEIAlDFSrknglhdvlrivtleenRGkGGAVTHGMRHVRGEYAVfADADGASRFADLGkLvkgvrevvdeegrgvavGSRaWLVGseAvvkRSALRNtlMhsfhlllrlltppatsrirdtQcGFKLFTRAAlPhiipymhaegwifdvemlmlaesapgvedaekgkgnggkakgikvseqpiawqevggsklnvmwdslgmawglavlrggwgmgvwrrr'\n",
      "'>XP_001547235.1 BCIN_01g00640|&|AA3:Unknown|AA3:Unknown|AA3:3.1(64.0,26..553)|AA3_1'\n",
      "'mkvsyqfslvvglvtgvwgqadnawiqeqWDAIIVGSGPAGIVVASRLSEAGLKTLLLEGGNQsYgvTGGDLDARRpEWLSgTNlSRVDVPGLYKSIFADGGglvcgaevnayGGCTIGGSSAINAGLFFEPPASDWDLYFPaEWNSTNMDaSIQRlYDTQTSTNlTSqDGIRSLQTgYDaARKWLVDGLGYKDVDINAQADdkTDVFGyPIFDYnNgQrGGPVTTYLQSALQRSNFRLQSGVRVVRVERDGDTATGVTALINGVeTIISTtPTgRVVLSAgAIQSPSLLMYSGIgDAETLsKLSAAGKlSPNLTSAEWInNTAIGAGLFDNPNTFIELQGDSiESYTySYDSPPaDDKalyldSRSgPYSFASETSVFWTTIAHADgSIAGlqGTiDSSGYSDFttnntiTLNVYGTSGLLSSGKVILdDNFIPGPSDDvYysNPRDaKDIAsFIHTIFSaLPnSGLtSMNIpqNATESeIetYITTASAYARGMVNHwSSSCRLGSCVDVDTTVIGmQNVHvVdASILAPVTTNPQFAVMAAGERASELILGVAGlecsfcasddecere'\n",
      "'>XP_001547254.1 BCIN_01g00800|&|AA1:1.10.3.2|AA1:1.10.3.2|AA1:6.1(10.5,202..455)|AA1_3'\n",
      "'mrlhillalclsfcqihasplfgwsfnpiatlqadllkiinqaeclvatkvsiipsiltfgnsnpvtsnkcsttsgkttssaaiqtslsslqsvasskwsslqsllstttslppilttsftattakptttssvkststtantvsstakpssitstasstasstsktsssavasgtpcagntaadrskwcnyststdyynevpdtgVtREYWFNVQdgvaspdgisryvqtingsiPGpTIIADWgDNVVVhvtnslttngssIHfHGIhQKNTNqndgvpSVTQcPIAYGDTyTyrwratqyGSSWYhSHVGLQawEgvtggiiInGPATANYDEDKGTLmlsdwghetvdelyqkvqtvgpqsmttglinGTNVFGAdgasnqtgsrfstsvqsgksyRFrLVNSAIDsqfkfsvdshtmtvmamdfvpIVPYQtEIlniaigQRYDVVITAnQASVAsdffiraipqsscsendnsdnirgvlhygsstglptttgytftdecvdepvaslvpylpktvssaskspeeavtvaqnsqklfkwylnnstflsewedptllmvqnnvtafgtpdnlvevpnanewiylvihsalpvphpihLHGHDFFVvSQqsttfdattavstynlnnpprrdvatlpgggylvlafetdnpgAWLMhCHIGWHtsmgfamqfleraseipklldttqmldtctawnghaiiqedsgv'\n",
      "'>XP_024546029.1 BCIN_01g01760|&|GH18:Unknown|GH18:Unknown|GH18:12.1(11.4,116..258)+PL7:0.1(1.1,989..1040)|PL7_2'\n",
      "'mypsksftvaaatfaliapaaalydasspanlalywgsgpsqtnlsyyceqstvdiiplafmnvfpaqgdgypaenfgnacygqpiftpgpgyplgdvntskdqlyvqcpgiqegipycQSLGKKiLLSLGGASNTYqltgaadgeYfADFLWGSYgpfkqswldaggirpmdggyygtdssvhididgfdFDIEFGPTdysegyiamiNRLRehFaenpskkyfISGAPqCPLPEPnMGAMIAgaQfdllWIqFYnNAAAQctarqwadnyaltgqedsaeftydqwlstinngasagasiylgllgstlagtasdyispleaqsliesyhnkpqfggvmiweatysqentdeelkgdsyhgfiksclasyapppptstssivhsttissssstsevsitssvpastessaatsstvisisststsvsessststltssessttsssslletsttstststsvsetstvtsssslessaaqsstysevlptqassgytipgyshkatsslthsqvhsskpagsytftghsskptsiatgytskshsakpsskpssgytsvyssnsttsgsvhssrpystksvvtgtgvsskpyhtkssvigtgastkpyptkssivgtgqsskprptkslgtgvkscksrsktrtrtktktktssagpsgypastssvvisstapygnstyshsettpassatssvegitvssgtatetgpgssatdstvvgpstsgtselptgsatnsgsvtsetgsasasgsissgtgsitsaiqyttstayvttvytitscaatvtdcpgrigsvttetissyttvcpvtatatstssaegisvsskpssapgvtasqtpsssaqgsatlpitsvvqyttstayvtsvytitscaatvtdcpgrigsvttetissyttvcpvtetsgnsgivsstsspaggpvptttvtsygtstvyqtsvytitscapsvtdcparsgqvttetissytsvypitvstapinvptsgASsAAGIetsvpgvkvsstasiftsapaAGAssIAGVssavptkgASSAAGIvtsvpvigasssaatpeaasptslyttlistlinsqtttisaivapsiitvvpvaiaspspasspytsiappygagngtlvatgtassitrvpkpssyvvkaegestsaksgsvttqipvatpsavtfeggaeklvkglgvwaavgmmglvlfi'\n"
     ]
    }
   ],
   "source": [
    "# read the raw text in the fasta file\n",
    "for line in fasta_file[:10]:\n",
    "    print(repr(line))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "XP_001545605.1 BCIN_01g00260\tAA4\tAA4:1.1\t13.9\t7.39\t268\t122..568\t107\tAA4:1.1.3.38\tAA4:1.1.3.38\t\n",
      "XP_024545922.1 BCIN_01g00340\tGT2\tGT2:23.1\t20.2\t7.74\t211\t116..300\t77\tGT2:2.4.1.117\tGT2:2.4.1.117\t\n",
      "XP_001547235.1 BCIN_01g00640\tAA3\tAA3:3.1\t64.0\t144.16\t274\t26..553\t453\t\t\tAA3_1\n",
      "XP_001547254.1 BCIN_01g00800\tAA1\tAA1:6.1\t10.5\t5.13\t329\t202..455\t98\tAA1:1.10.3.2\tAA1:1.10.3.2\tAA1_3\n",
      "XP_024546029.1 BCIN_01g01760\tGH18\tGH18:12.1\t11.4\t2.38\t182\t116..258\t42\t\t\t\n",
      "XP_024546029.1 BCIN_01g01760\tPL7\tPL7:0.1\t1.1\t0.11\t1023\t989..1040\t20\t\t\tPL7:2\n",
      "XP_001548518.1 BCIN_01g02310\tGT39\tGT39:2.1\t71.3\t44.37\t114\t62..264\t125\tGT39:2.4.1.109\tGT39:2.4.1.109\t\n",
      "XP_001548545.1 BCIN_01g02520\tGT2\tGT2:11.1\t33.2\t16.05\t415\t213..452\t97\tGT2:2.4.1.16\tGT2:2.4.1.16\t\n",
      "XP_001549934.1 BCIN_01g02970\tAA11\tAA11:1.1\t6.6\t1.08\t104\t55..160\t33\tAA11:1.*.*.*\tAA11:1.*.*.*\t\n",
      "XP_024546135.1 BCIN_01g03210\tGH71\tGH71:1.2\t7.0\t2.08\t181\t71..381\t60\tGH71:3.2.1.59\tGH71:3.2.1.59\t\n"
     ]
    }
   ],
   "source": [
    "# read the log file\n",
    "with open(\"gca_143535_cupp.fasta.log\", \"r\") as lfh:\n",
    "    log_file = lfh.read().splitlines()\n",
    "\n",
    "for line in log_file[:10]:\n",
    "    print(line)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "'XP_001545605.1 BCIN_01g00260\\tAA4\\tAA4:1.1\\t13.9\\t7.39\\t268\\t122..568\\t107\\tAA4:1.1.3.38\\tAA4:1.1.3.38\\t'\n",
      "'XP_024545922.1 BCIN_01g00340\\tGT2\\tGT2:23.1\\t20.2\\t7.74\\t211\\t116..300\\t77\\tGT2:2.4.1.117\\tGT2:2.4.1.117\\t'\n",
      "'XP_001547235.1 BCIN_01g00640\\tAA3\\tAA3:3.1\\t64.0\\t144.16\\t274\\t26..553\\t453\\t\\t\\tAA3_1'\n",
      "'XP_001547254.1 BCIN_01g00800\\tAA1\\tAA1:6.1\\t10.5\\t5.13\\t329\\t202..455\\t98\\tAA1:1.10.3.2\\tAA1:1.10.3.2\\tAA1_3'\n",
      "'XP_024546029.1 BCIN_01g01760\\tGH18\\tGH18:12.1\\t11.4\\t2.38\\t182\\t116..258\\t42\\t\\t\\t'\n",
      "'XP_024546029.1 BCIN_01g01760\\tPL7\\tPL7:0.1\\t1.1\\t0.11\\t1023\\t989..1040\\t20\\t\\t\\tPL7:2'\n",
      "'XP_001548518.1 BCIN_01g02310\\tGT39\\tGT39:2.1\\t71.3\\t44.37\\t114\\t62..264\\t125\\tGT39:2.4.1.109\\tGT39:2.4.1.109\\t'\n",
      "'XP_001548545.1 BCIN_01g02520\\tGT2\\tGT2:11.1\\t33.2\\t16.05\\t415\\t213..452\\t97\\tGT2:2.4.1.16\\tGT2:2.4.1.16\\t'\n",
      "'XP_001549934.1 BCIN_01g02970\\tAA11\\tAA11:1.1\\t6.6\\t1.08\\t104\\t55..160\\t33\\tAA11:1.*.*.*\\tAA11:1.*.*.*\\t'\n",
      "'XP_024546135.1 BCIN_01g03210\\tGH71\\tGH71:1.2\\t7.0\\t2.08\\t181\\t71..381\\t60\\tGH71:3.2.1.59\\tGH71:3.2.1.59\\t'\n"
     ]
    }
   ],
   "source": [
    "# read the raw text in the log file\n",
    "for line in log_file[:10]:\n",
    "    print(repr(line))\n",
    "# data is separated by '\\t'"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "A user or database may chose to put a '|' within the name of their protein or species. Therefore, using '|' to identify separations between individual items of data may be a brittle approach.\n",
    "\n",
    "It is very unlikely a database or user would use `\\t` within the name of their protein, especially because CUPP will be reading FASTA files written by `pyrewton` which does not put `\\t` within names. Additionally, a note can be made in the README to tell users if they create their own fasta files do not include `\\t` in the description line. The same could be said for `|` but the issue is dealing with databases and they cannot be told how to present their data.\n",
    "\n",
    "Each line in the log file contains a unique protein. The first item is the description retrieved from the FASTA file, the first item of which is the protein accession number retrieved from UniProt of the genomic assembly. The second item is the gene name."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['XP_001545605.1 BCIN_01g00260', 'AA4', 'AA4:1.1', '13.9', '7.39', '268', '122..568', '107', 'AA4:1.1.3.38', 'AA4:1.1.3.38', '']\n",
      "['XP_024545922.1 BCIN_01g00340', 'GT2', 'GT2:23.1', '20.2', '7.74', '211', '116..300', '77', 'GT2:2.4.1.117', 'GT2:2.4.1.117', '']\n",
      "['XP_001547235.1 BCIN_01g00640', 'AA3', 'AA3:3.1', '64.0', '144.16', '274', '26..553', '453', '', '', 'AA3_1']\n",
      "['XP_001547254.1 BCIN_01g00800', 'AA1', 'AA1:6.1', '10.5', '5.13', '329', '202..455', '98', 'AA1:1.10.3.2', 'AA1:1.10.3.2', 'AA1_3']\n"
     ]
    }
   ],
   "source": [
    "# first split the line into the individual pieces of data\n",
    "for line in log_file[:4]:\n",
    "    prediction_output = line.split(\"\\t\")\n",
    "    print(prediction_output)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'protein_accession': 'XP_001545605.1 BCIN_01g00260', 'cazy_family': 'AA4', 'cazy_subfamily': '', 'ec_number': 'AA4:1.1.3.38', 'domain_range': None}\n",
      "{'protein_accession': 'XP_024545922.1 BCIN_01g00340', 'cazy_family': 'GT2', 'cazy_subfamily': '', 'ec_number': 'GT2:2.4.1.117', 'domain_range': None}\n",
      "{'protein_accession': 'XP_001547235.1 BCIN_01g00640', 'cazy_family': 'AA3', 'cazy_subfamily': 'AA3_1', 'ec_number': '', 'domain_range': None}\n",
      "{'protein_accession': 'XP_001547254.1 BCIN_01g00800', 'cazy_family': 'AA1', 'cazy_subfamily': 'AA1_3', 'ec_number': 'AA1:1.10.3.2', 'domain_range': None}\n"
     ]
    }
   ],
   "source": [
    "for line in log_file[:4]:\n",
    "    prediction_output = line.split(\"\\t\")\n",
    "    for item in prediction_output:\n",
    "        if item.find(\"...\") != -1:\n",
    "            domain_range = item\n",
    "            break\n",
    "        else:\n",
    "            domain_range = None\n",
    "            pass\n",
    "    prediction = {\n",
    "        \"protein_accession\": prediction_output[0],\n",
    "        \"cazy_family\": prediction_output[1],\n",
    "        \"cazy_subfamily\": prediction_output[-1],\n",
    "        \"ec_number\": prediction_output[-2],\n",
    "        \"domain_range\": domain_range,\n",
    "    }\n",
    "    print(prediction)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "This generates dataframes that can easily be added to a Pandas dataframe.\n",
    "\n",
    "Check the meaning of PL7:2 in subfamily. There is not PL7_2 subfam, is this a CUPP suggested subfam."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "*PL7:2\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>protein_accession</th>\n",
       "      <th>cazy_family</th>\n",
       "      <th>cazy_subfamily</th>\n",
       "      <th>ec_number</th>\n",
       "      <th>domain_range</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>XP_001545605.1 BCIN_01g00260</td>\n",
       "      <td>AA4</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.1.3.38</td>\n",
       "      <td>122..568</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>XP_024545922.1 BCIN_01g00340</td>\n",
       "      <td>GT2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2.4.1.117</td>\n",
       "      <td>116..300</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>XP_001547235.1 BCIN_01g00640</td>\n",
       "      <td>AA3</td>\n",
       "      <td>AA3_1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>26..553</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>XP_001547254.1 BCIN_01g00800</td>\n",
       "      <td>AA1</td>\n",
       "      <td>AA1_3</td>\n",
       "      <td>1.10.3.2</td>\n",
       "      <td>202..455</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>XP_024546029.1 BCIN_01g01760</td>\n",
       "      <td>GH18</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>116..258</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "              protein_accession cazy_family cazy_subfamily  ec_number  \\\n",
       "0  XP_001545605.1 BCIN_01g00260         AA4            NaN   1.1.3.38   \n",
       "0  XP_024545922.1 BCIN_01g00340         GT2            NaN  2.4.1.117   \n",
       "0  XP_001547235.1 BCIN_01g00640         AA3          AA3_1        NaN   \n",
       "0  XP_001547254.1 BCIN_01g00800         AA1          AA1_3   1.10.3.2   \n",
       "0  XP_024546029.1 BCIN_01g01760        GH18            NaN        NaN   \n",
       "\n",
       "  domain_range  \n",
       "0     122..568  \n",
       "0     116..300  \n",
       "0      26..553  \n",
       "0     202..455  \n",
       "0     116..258  "
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# build an empty dataframe\n",
    "cupp_df = pd.DataFrame(columns=[\"protein_accession\",\"cazy_family\",\"cazy_subfamily\",\"ec_number\",\"domain_range\"])\n",
    "\n",
    "# add in predictions from log file to dataframe\n",
    "for line in log_file[:10]:\n",
    "    prediction_output = line.split(\"\\t\")\n",
    "    \n",
    "    # retrieve domain range if given\n",
    "    for item in prediction_output:\n",
    "        if item.find(\"..\") != -1:\n",
    "            domain_range = item\n",
    "            break\n",
    "        else:\n",
    "            domain_range = np.nan\n",
    "\n",
    "    try:\n",
    "        ec_number = prediction_output[-2].split(\":\")[1]\n",
    "    except IndexError:\n",
    "        ec_number = np.nan\n",
    "\n",
    "    # build dict to enable easy building of df\n",
    "    prediction = {\n",
    "        \"protein_accession\": [prediction_output[0]],\n",
    "        \"cazy_family\": [prediction_output[1]],\n",
    "        \"cazy_subfamily\": [prediction_output[-1]],\n",
    "        \"ec_number\": ec_number,\n",
    "        \"domain_range\": [domain_range],\n",
    "    }\n",
    "\n",
    "    # change empty string to null value\n",
    "    if len(prediction[\"cazy_subfamily\"][0]) == 0:\n",
    "        prediction[\"cazy_subfamily\"][0] = np.nan\n",
    "    # If CUPP suggested subfamily, i.e. a subfamily no in CAZy, mark this with '*'\n",
    "    else:\n",
    "        if prediction[\"cazy_subfamily\"][0].find(\":\") != -1:\n",
    "            prediction[\"cazy_subfamily\"][0] = \"*\" + prediction[\"cazy_subfamily\"][0]\n",
    "            print(prediction[\"cazy_subfamily\"][0])\n",
    "\n",
    "    prediction_df = pd.DataFrame(prediction)\n",
    "    cupp_df = cupp_df.append(prediction_df)\n",
    "\n",
    "cupp_df.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "As a single function:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "*PL7:2\n",
      "*AA1:2\n",
      "*GH13:31+GH13:40\n",
      "*PL1:4\n",
      "*AA3:2\n",
      "*AA1:3\n",
      "*AA3:2\n",
      "*AA1:3\n",
      "*AA1:3\n",
      "*AA1:3\n",
      "*GH5:49\n",
      "*AA3:1\n",
      "*PL1:4\n",
      "*AA3:2\n",
      "*AA3:2\n",
      "*AA3:2\n",
      "*PL1:4\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>protein_accession</th>\n",
       "      <th>cazy_family</th>\n",
       "      <th>cazy_subfamily</th>\n",
       "      <th>ec_number</th>\n",
       "      <th>domain_range</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>XP_001545605.1 BCIN_01g00260</td>\n",
       "      <td>AA4</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.1.3.38</td>\n",
       "      <td>122..568</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>XP_024545922.1 BCIN_01g00340</td>\n",
       "      <td>GT2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2.4.1.117</td>\n",
       "      <td>116..300</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>XP_001547235.1 BCIN_01g00640</td>\n",
       "      <td>AA3</td>\n",
       "      <td>AA3_1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>26..553</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>XP_001547254.1 BCIN_01g00800</td>\n",
       "      <td>AA1</td>\n",
       "      <td>AA1_3</td>\n",
       "      <td>1.10.3.2</td>\n",
       "      <td>202..455</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>XP_024546029.1 BCIN_01g01760</td>\n",
       "      <td>GH18</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>116..258</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>XP_001547801.1 BCIN_16g04060</td>\n",
       "      <td>GH16</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3.2.1.6-GH16</td>\n",
       "      <td>74..281</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>XP_024553953.1 BCIN_16g04060</td>\n",
       "      <td>GH16</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>74..134</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>XP_024553962.1 BCIN_16g04160</td>\n",
       "      <td>GH47</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3.2.1.113</td>\n",
       "      <td>129..565</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>XP_024553961.1 BCIN_16g04160</td>\n",
       "      <td>GH47</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3.2.1.113</td>\n",
       "      <td>134..570</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>XP_024554066.1 BCIN_16g05200</td>\n",
       "      <td>GH32</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3.2.1.7</td>\n",
       "      <td>353..665</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>482 rows × 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "               protein_accession cazy_family cazy_subfamily     ec_number  \\\n",
       "0   XP_001545605.1 BCIN_01g00260         AA4            NaN      1.1.3.38   \n",
       "0   XP_024545922.1 BCIN_01g00340         GT2            NaN     2.4.1.117   \n",
       "0   XP_001547235.1 BCIN_01g00640         AA3          AA3_1           NaN   \n",
       "0   XP_001547254.1 BCIN_01g00800         AA1          AA1_3      1.10.3.2   \n",
       "0   XP_024546029.1 BCIN_01g01760        GH18            NaN           NaN   \n",
       "..                           ...         ...            ...           ...   \n",
       "0   XP_001547801.1 BCIN_16g04060        GH16            NaN  3.2.1.6-GH16   \n",
       "0   XP_024553953.1 BCIN_16g04060        GH16            NaN           NaN   \n",
       "0   XP_024553962.1 BCIN_16g04160        GH47            NaN     3.2.1.113   \n",
       "0   XP_024553961.1 BCIN_16g04160        GH47            NaN     3.2.1.113   \n",
       "0   XP_024554066.1 BCIN_16g05200        GH32            NaN       3.2.1.7   \n",
       "\n",
       "   domain_range  \n",
       "0      122..568  \n",
       "0      116..300  \n",
       "0       26..553  \n",
       "0      202..455  \n",
       "0      116..258  \n",
       "..          ...  \n",
       "0       74..281  \n",
       "0       74..134  \n",
       "0      129..565  \n",
       "0      134..570  \n",
       "0      353..665  \n",
       "\n",
       "[482 rows x 5 columns]"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def parse_cupp_output(log_file_path):\n",
    "    \"\"\"Parse the output from the output log file from CUPP and write out data to a dataframe.\n",
    "    \n",
    "    Retrieves the protein accession/name/identifier, predicated CAZy family, predicated CAZy subfamily,\n",
    "    predicated EC number and predicated range of domain within the protein sequence (the index of the\n",
    "    first and last residues of the domain).\n",
    "    \n",
    "    :param log_file_path: path, path to the output log file\n",
    "    \n",
    "    Return Pandas dataframe containing CUPP output\n",
    "    \"\"\"\n",
    "    \n",
    "    with open(log_file_path, \"r\") as lfh:\n",
    "        log_file = lfh.read().splitlines()\n",
    "    \n",
    "    # build an empty dataframe to add predication outputs to\n",
    "    cupp_df = pd.DataFrame(columns=[\"protein_accession\",\"cazy_family\",\"cazy_subfamily\",\"ec_number\",\"domain_range\"])\n",
    "\n",
    "    # add in predictions from log file to dataframe\n",
    "    for line in log_file:\n",
    "        prediction_output = line.split(\"\\t\")\n",
    "\n",
    "        # retrieve domain range if given\n",
    "        for item in prediction_output:\n",
    "            if item.find(\"..\") != -1:\n",
    "                domain_range = item\n",
    "                break\n",
    "            else:\n",
    "                domain_range = np.nan\n",
    "        \n",
    "        try:\n",
    "            ec_number = prediction_output[-2].split(\":\")[1]\n",
    "        except IndexError:\n",
    "            ec_number = np.nan\n",
    "\n",
    "        # build dict to enable easy building of df\n",
    "        prediction = {\n",
    "            \"protein_accession\": [prediction_output[0]],\n",
    "            \"cazy_family\": [prediction_output[1]],\n",
    "            \"cazy_subfamily\": [prediction_output[-1]],\n",
    "            \"ec_number\": [ec_number],\n",
    "            \"domain_range\": [domain_range],\n",
    "        }\n",
    "\n",
    "        # change empty string to null value\n",
    "        if len(prediction[\"cazy_subfamily\"][0]) == 0:\n",
    "            prediction[\"cazy_subfamily\"][0] = np.nan\n",
    "        # If CUPP suggested subfamily, i.e. a subfamily no in CAZy, mark this with '*'\n",
    "        else:\n",
    "            if prediction[\"cazy_subfamily\"][0].find(\":\") != -1:\n",
    "                prediction[\"cazy_subfamily\"][0] = \"*\" + prediction[\"cazy_subfamily\"][0]\n",
    "                print(prediction[\"cazy_subfamily\"][0])\n",
    "\n",
    "        prediction_df = pd.DataFrame(prediction)\n",
    "        cupp_df = cupp_df.append(prediction_df)\n",
    "\n",
    "    return cupp_df\n",
    "\n",
    "parsed_cupp_result = parse_cupp_output(\"gca_143535_cupp.fasta.log\")\n",
    "parsed_cupp_result\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Parsing the output from eCAMI\n",
    "\n",
    "### Understanding the output from eCAMI\n",
    "\n",
    "eCAMI produces a single output file, a plain text file.\n",
    "\n",
    "Each protein is given two lines, and a new protein is indicated by the first line starting with '>'.\n",
    "\n",
    "For example:\n",
    "```\n",
    ">XP_001545605.1 BCIN_01g00260 \tAA4:3\tAA4:24|1.1.3.38:1\n",
    "EIPLWPFS(113),IPLWPFSI(114),PLWPFSIG(115),LWPFSIGR(116),WPFSIGRN(117),PFSIGRNV(118),FSIGRNVG(119),SIGRNVGY(120),IGRNVGYG(121),GRNVGYGG(122),RNVGYGGA(123),NVGYGGAA(124),VGYGGAAP(125),GYGGAAPR(126),YGGAAPRV(127),GGAAPRVP(128),GAAPRVPG(129),AAPRVPGS(130),APRVPGSI(131),PRVPGSIG(132),RVPGSIGL(133),VPGSIGLD(134),DTPDLGGG(186),TPDLGGGS(187),PDLGGGSV(188),ERGVGYTP(201),RGVGYTPY(202),GVGYTPYG(203),VGYTPYGD(204),GYTPYGDH(205),YTPYGDHF(206),TPYGDHFM(207),PYGDHFMM(208),YGDHFMMH(209),GDHFMMHC(210),DHFMMHCG(211),GLEVVLPT(218),LEVVLPTG(219),EVVLPTGE(220),IRTGMGAL(229),RTGMGALP(230),TGMGALPD(231),GMGALPDP(232),LFPYGFGP(260),FPYGFGPY(261),YGFGPYND(263),GFGPYNDG(264),FGPYNDGI(265),GPYNDGIF(266),DGIFSQSN(270),FSQSNLGI(273),QSNLGIVT(275),SNLGIVTK(276),WLMPNPGG(287),LMPNPGGY(288),MPNPGGYQ(289),IRHILLDA(330),RHILLDAA(331),HILLDAAV(332),LLDAAVMG(334),LNLGRWNF(365),NLGRWNFY(366),LGRWNFYG(367),GRWNFYGA(368),RWNFYGAL(369),WNFYGALY(370),NFYGALYG(371),FYGALYGP(372),DWLPNGAH(434),WLPNGAHL(435),LPNGAHLF(436),PNGAHLFF(437),NGAHLFFS(438),GAHLFFSP(439),AHLFFSPI(440),EAGFDFIG(468),AGFDFIGT(469),GFDFIGTF(470),GMREMHHI(480),NGWGEYRT(517),GWGEYRTH(518),IKNALDPK(550),KNALDPKG(551),NALDPKGI(552),ALDPKGIL(553),LDPKGILA(554),DPKGILAP(555),PKGILAPG(556),KGILAPGK(557),GILAPGKN(558),ILAPGKNG(559),\n",
    "```\n",
    "\n",
    "The first line contains:\n",
    "- Protein accession, taken straight from the description/identifier line from the input fasta file\n",
    "- Predicated CAZy family : group number # believe this i the k-mer cluster/group number\n",
    "- \"Subname of group\": \"subfam_name_count\"\n",
    "\n",
    "Each of these items are separated by what appears to be a tab ('\\t') white space. - This needs to be checked.\n",
    "\n",
    "There is only one protein accession given, and only one predicated CAZy family. But multiple subgroups can be named, each separated by '|'. This appears to additional domains/modules eCAMI has found, such as other catalytic domains and CBMs.\n",
    "\n",
    "The predicated CAZy family is the 'best' or highest scoring result.\n",
    "\n",
    "The second line contains each of the k-mers that were identified in the query sequences and the number of the parent k-mer cluster in parenthesises.\n",
    "\n",
    "Sometimes many many other domains can be found. For example:\n",
    "```\n",
    ">XP_024550696.1 BCIN_09g01110 \tCBM1:16\tCBM1:138|GH7:117|3.2.1.176:32|GH12:3|3.2.1.132:2|GH45:2|CE16:1|AA9:1|GH5_5:1|CE1:1|3.1.1.-:1|3.2.1.4:1|GH16:1|AA8:1|PL4_1:1|GH30_7:1|CBM18:1|GH43_6:1\n",
    "```\n",
    "This seems to frequently occur when the 'best' result is a CBM domain. This would appear to be an issue with identifying CBM domains. It might be that the reason a CBM domain appears to have multiple difference CAZy class functions is becuase the binding of the CBMs associated substrate is shared by many other catalytic domains.\n",
    "\n",
    "If the predicated subfamily is wanted to be retrieved, the final section of other domains needs to retrieved and filtered to retrieve the item that matches the predicated CAZy family with the addition of \"\\_\" followed by numbers.\n",
    "\n",
    "Sometimes an EC number is also predicated. This appears in the final section of the first line, but it's position changes. Therefore, if it is to retrieved the final section will need to be iterated over to retrieve the EC number. The EcC number is presented followed by the ':' and the group number.\n",
    "\n",
    "**Data to collect:**\n",
    "- Protein accession\n",
    "- Predicated CAZy family\n",
    "- Predicated CAZy subfamily\n",
    "- Predicated EC number\n",
    "- List of additionally domains - will be good to see if it can pick up multi-domain proteins and the individual functions of the enzyme\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      ">XP_001545605.1 BCIN_01g00260 \tAA4:3\tAA4:24|1.1.3.38:1\n",
      "EIPLWPFS(113),IPLWPFSI(114),PLWPFSIG(115),LWPFSIGR(116),WPFSIGRN(117),PFSIGRNV(118),FSIGRNVG(119),SIGRNVGY(120),IGRNVGYG(121),GRNVGYGG(122),RNVGYGGA(123),NVGYGGAA(124),VGYGGAAP(125),GYGGAAPR(126),YGGAAPRV(127),GGAAPRVP(128),GAAPRVPG(129),AAPRVPGS(130),APRVPGSI(131),PRVPGSIG(132),RVPGSIGL(133),VPGSIGLD(134),DTPDLGGG(186),TPDLGGGS(187),PDLGGGSV(188),ERGVGYTP(201),RGVGYTPY(202),GVGYTPYG(203),VGYTPYGD(204),GYTPYGDH(205),YTPYGDHF(206),TPYGDHFM(207),PYGDHFMM(208),YGDHFMMH(209),GDHFMMHC(210),DHFMMHCG(211),GLEVVLPT(218),LEVVLPTG(219),EVVLPTGE(220),IRTGMGAL(229),RTGMGALP(230),TGMGALPD(231),GMGALPDP(232),LFPYGFGP(260),FPYGFGPY(261),YGFGPYND(263),GFGPYNDG(264),FGPYNDGI(265),GPYNDGIF(266),DGIFSQSN(270),FSQSNLGI(273),QSNLGIVT(275),SNLGIVTK(276),WLMPNPGG(287),LMPNPGGY(288),MPNPGGYQ(289),IRHILLDA(330),RHILLDAA(331),HILLDAAV(332),LLDAAVMG(334),LNLGRWNF(365),NLGRWNFY(366),LGRWNFYG(367),GRWNFYGA(368),RWNFYGAL(369),WNFYGALY(370),NFYGALYG(371),FYGALYGP(372),DWLPNGAH(434),WLPNGAHL(435),LPNGAHLF(436),PNGAHLFF(437),NGAHLFFS(438),GAHLFFSP(439),AHLFFSPI(440),EAGFDFIG(468),AGFDFIGT(469),GFDFIGTF(470),GMREMHHI(480),NGWGEYRT(517),GWGEYRTH(518),IKNALDPK(550),KNALDPKG(551),NALDPKGI(552),ALDPKGIL(553),LDPKGILA(554),DPKGILAP(555),PKGILAPG(556),KGILAPGK(557),GILAPGKN(558),ILAPGKNG(559),\n",
      ">XP_024545922.1 BCIN_01g00340 \tGT2:1526\tGT2:37\n",
      "VVPAYNEE(118),NRGKGGAV(194),RGKGGAVT(195),GKGGAVTH(196),KGGAVTHG(197),GGAVTHGM(198),GAVTHGMR(199),AVTHGMRH(200),VTHGMRHV(201),THGMRHVR(202),HGMRHVRG(203),VFADADGA(214),FADADGAS(215),ADADGASR(216),DADGASRF(217),LVGSEAVV(253),VGSEAVVK(254),GSEAVVKR(255),SEAVVKRS(256),EAVVKRSA(257),LTPPATSR(280),TPPATSRI(281),PPATSRIR(282),PATSRIRD(283),ATSRIRDT(284),TSRIRDTQ(285),SRIRDTQC(286),RIRDTQCG(287),IRDTQCGF(288),RDTQCGFK(289),DTQCGFKL(290),TQCGFKLF(291),QCGFKLFT(292),CGFKLFTR(293),EGWIFDVE(313),GWIFDVEM(314),WIFDVEML(315),IFDVEMLM(316),FDVEMLML(317),DVEMLMLA(318),VEMLMLAE(319),EMLMLAES(320),MLMLAESA(321),LMLAESAP(322),EVGGSKLN(358),VGGSKLNV(359),\n"
     ]
    }
   ],
   "source": [
    "# Read the output file from eCAMI\n",
    "with open(\"143535_output.txt\", \"r\") as fh:\n",
    "    ecami_file = fh.read().splitlines()\n",
    "\n",
    "for line in ecami_file[:4]:\n",
    "    print(line)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "'>XP_001545605.1 BCIN_01g00260 \\tAA4:3\\tAA4:24|1.1.3.38:1'\n",
      "'EIPLWPFS(113),IPLWPFSI(114),PLWPFSIG(115),LWPFSIGR(116),WPFSIGRN(117),PFSIGRNV(118),FSIGRNVG(119),SIGRNVGY(120),IGRNVGYG(121),GRNVGYGG(122),RNVGYGGA(123),NVGYGGAA(124),VGYGGAAP(125),GYGGAAPR(126),YGGAAPRV(127),GGAAPRVP(128),GAAPRVPG(129),AAPRVPGS(130),APRVPGSI(131),PRVPGSIG(132),RVPGSIGL(133),VPGSIGLD(134),DTPDLGGG(186),TPDLGGGS(187),PDLGGGSV(188),ERGVGYTP(201),RGVGYTPY(202),GVGYTPYG(203),VGYTPYGD(204),GYTPYGDH(205),YTPYGDHF(206),TPYGDHFM(207),PYGDHFMM(208),YGDHFMMH(209),GDHFMMHC(210),DHFMMHCG(211),GLEVVLPT(218),LEVVLPTG(219),EVVLPTGE(220),IRTGMGAL(229),RTGMGALP(230),TGMGALPD(231),GMGALPDP(232),LFPYGFGP(260),FPYGFGPY(261),YGFGPYND(263),GFGPYNDG(264),FGPYNDGI(265),GPYNDGIF(266),DGIFSQSN(270),FSQSNLGI(273),QSNLGIVT(275),SNLGIVTK(276),WLMPNPGG(287),LMPNPGGY(288),MPNPGGYQ(289),IRHILLDA(330),RHILLDAA(331),HILLDAAV(332),LLDAAVMG(334),LNLGRWNF(365),NLGRWNFY(366),LGRWNFYG(367),GRWNFYGA(368),RWNFYGAL(369),WNFYGALY(370),NFYGALYG(371),FYGALYGP(372),DWLPNGAH(434),WLPNGAHL(435),LPNGAHLF(436),PNGAHLFF(437),NGAHLFFS(438),GAHLFFSP(439),AHLFFSPI(440),EAGFDFIG(468),AGFDFIGT(469),GFDFIGTF(470),GMREMHHI(480),NGWGEYRT(517),GWGEYRTH(518),IKNALDPK(550),KNALDPKG(551),NALDPKGI(552),ALDPKGIL(553),LDPKGILA(554),DPKGILAP(555),PKGILAPG(556),KGILAPGK(557),GILAPGKN(558),ILAPGKNG(559),'\n",
      "'>XP_024545922.1 BCIN_01g00340 \\tGT2:1526\\tGT2:37'\n",
      "'VVPAYNEE(118),NRGKGGAV(194),RGKGGAVT(195),GKGGAVTH(196),KGGAVTHG(197),GGAVTHGM(198),GAVTHGMR(199),AVTHGMRH(200),VTHGMRHV(201),THGMRHVR(202),HGMRHVRG(203),VFADADGA(214),FADADGAS(215),ADADGASR(216),DADGASRF(217),LVGSEAVV(253),VGSEAVVK(254),GSEAVVKR(255),SEAVVKRS(256),EAVVKRSA(257),LTPPATSR(280),TPPATSRI(281),PPATSRIR(282),PATSRIRD(283),ATSRIRDT(284),TSRIRDTQ(285),SRIRDTQC(286),RIRDTQCG(287),IRDTQCGF(288),RDTQCGFK(289),DTQCGFKL(290),TQCGFKLF(291),QCGFKLFT(292),CGFKLFTR(293),EGWIFDVE(313),GWIFDVEM(314),WIFDVEML(315),IFDVEMLM(316),FDVEMLML(317),DVEMLMLA(318),VEMLMLAE(319),EMLMLAES(320),MLMLAESA(321),LMLAESAP(322),EVGGSKLN(358),VGGSKLNV(359),'\n"
     ]
    }
   ],
   "source": [
    "# Check how the data is separated\n",
    "# Reading the code shows it could be tabs\n",
    "for line in ecami_file[:4]:\n",
    "    print(repr(line))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['>XP_001545605.1 BCIN_01g00260 ', 'AA4:3', 'AA4:24|1.1.3.38:1']\n",
      "['>XP_024545922.1 BCIN_01g00340 ', 'GT2:1526', 'GT2:37']\n",
      "['>XP_001547235.1 BCIN_01g00640 ', 'AA3:9', 'AA3_1:9']\n"
     ]
    }
   ],
   "source": [
    "# separate out the data for each protein\n",
    "for line in ecami_file[:6]:\n",
    "    if line.startswith(\">\"):  # identifies new protein\n",
    "        prediction_output = line.split(\"\\t\")\n",
    "        print(prediction_output)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "additional_domains= ['AA4']\n",
      "subfam= nan\n",
      "ecnum= 1.1.3.38\n",
      "additional_domains= ['GT2']\n",
      "subfam= nan\n",
      "ecnum= nan\n",
      "additional_domains= []\n",
      "subfam= AA3_1\n",
      "ecnum= nan\n",
      "additional_domains= ['CBM18']\n",
      "subfam= AA1_3\n",
      "ecnum= 1.10.3.2\n",
      "additional_domains= ['GH18', 'CBM50', 'CBM18']\n",
      "subfam= nan\n",
      "ecnum= nan\n"
     ]
    }
   ],
   "source": [
    "# parse the outputs so in format suitable for final dataframe\n",
    "for line in ecami_file[:10]:\n",
    "    if line.startswith(\">\"):  # identifies new protein\n",
    "        prediction_output = line.split(\"\\t\")\n",
    "        cazy_fam = prediction_output[1].split(\":\")[0]\n",
    "        \n",
    "        cazy_subfam = np.nan\n",
    "        ec_number = np.nan\n",
    "        \n",
    "        # check for additional domains\n",
    "        other_domains = prediction_output[2].split(\"|\")\n",
    "        additional_domains = []\n",
    "        for domain in other_domains:\n",
    "            domain = domain.split(\":\")[0]  # drop the eCAMI group number\n",
    "            # check if subfamily is predicated for the main CAZy family\n",
    "            if len(domain.split(\"_\")) != 1:\n",
    "                cazy_subfam = domain\n",
    "            # check if predicated EC number\n",
    "            elif domain.find(\".\") != -1:\n",
    "                ec_number = domain\n",
    "            else:\n",
    "                additional_domains.append(domain)\n",
    "        \n",
    "        # check parsing is correct\n",
    "        print(\"additional_domains=\", additional_domains)\n",
    "        print(\"subfam=\", cazy_subfam)\n",
    "        print(\"ecnum=\", ec_number)\n",
    "        "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'protein_accession': ['>XP_024546029.1 BCIN_01g01760 '], 'cazy_family': ['GH18'], 'cazy_subfamily': [nan], 'ec_number': [nan], 'additional_domains': [['GH18', 'CBM50', 'CBM18']]}\n"
     ]
    }
   ],
   "source": [
    "        # build dict to enable easy building of df\n",
    "        prediction = {\n",
    "            \"protein_accession\": [prediction_output[0]],\n",
    "            \"cazy_family\": [prediction_output[1].split(\":\")[0]],\n",
    "            \"cazy_subfamily\": [cazy_subfam ],\n",
    "            \"ec_number\": [ec_number],\n",
    "            \"additional_domains\": [additional_domains],\n",
    "        }\n",
    "        \n",
    "        print(prediction)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "As a single function including building a dataframe and adding the predictions to it:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>protein_accession</th>\n",
       "      <th>cazy_family</th>\n",
       "      <th>cazy_subfamily</th>\n",
       "      <th>ec_number</th>\n",
       "      <th>domain_range</th>\n",
       "      <th>additional_domains</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>&gt;XP_001545605.1 BCIN_01g00260</td>\n",
       "      <td>AA4</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.1.3.38</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[AA4]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>&gt;XP_024545922.1 BCIN_01g00340</td>\n",
       "      <td>GT2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[GT2]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>&gt;XP_001547235.1 BCIN_01g00640</td>\n",
       "      <td>AA3</td>\n",
       "      <td>AA3_1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>&gt;XP_001547254.1 BCIN_01g00800</td>\n",
       "      <td>AA1</td>\n",
       "      <td>AA1_3</td>\n",
       "      <td>1.10.3.2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[CBM18]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>&gt;XP_024546029.1 BCIN_01g01760</td>\n",
       "      <td>GH18</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[GH18, CBM50, CBM18]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>&gt;XP_024553954.1 BCIN_16g04060</td>\n",
       "      <td>GH16</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[GH16]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>&gt;XP_024553952.1 BCIN_16g04060</td>\n",
       "      <td>GH16</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[GH16]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>&gt;XP_024553962.1 BCIN_16g04160</td>\n",
       "      <td>GH47</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3.2.1.113</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[GH47]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>&gt;XP_024553961.1 BCIN_16g04160</td>\n",
       "      <td>GH47</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3.2.1.113</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[GH47]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>&gt;XP_024554066.1 BCIN_16g05200</td>\n",
       "      <td>GH32</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3.2.1.7</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[GH32]</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>585 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                 protein_accession cazy_family cazy_subfamily  ec_number  \\\n",
       "0   >XP_001545605.1 BCIN_01g00260          AA4            NaN   1.1.3.38   \n",
       "0   >XP_024545922.1 BCIN_01g00340          GT2            NaN        NaN   \n",
       "0   >XP_001547235.1 BCIN_01g00640          AA3          AA3_1        NaN   \n",
       "0   >XP_001547254.1 BCIN_01g00800          AA1          AA1_3   1.10.3.2   \n",
       "0   >XP_024546029.1 BCIN_01g01760         GH18            NaN        NaN   \n",
       "..                             ...         ...            ...        ...   \n",
       "0   >XP_024553954.1 BCIN_16g04060         GH16            NaN        NaN   \n",
       "0   >XP_024553952.1 BCIN_16g04060         GH16            NaN        NaN   \n",
       "0   >XP_024553962.1 BCIN_16g04160         GH47            NaN  3.2.1.113   \n",
       "0   >XP_024553961.1 BCIN_16g04160         GH47            NaN  3.2.1.113   \n",
       "0   >XP_024554066.1 BCIN_16g05200         GH32            NaN    3.2.1.7   \n",
       "\n",
       "   domain_range    additional_domains  \n",
       "0           NaN                 [AA4]  \n",
       "0           NaN                 [GT2]  \n",
       "0           NaN                   NaN  \n",
       "0           NaN               [CBM18]  \n",
       "0           NaN  [GH18, CBM50, CBM18]  \n",
       "..          ...                   ...  \n",
       "0           NaN                [GH16]  \n",
       "0           NaN                [GH16]  \n",
       "0           NaN                [GH47]  \n",
       "0           NaN                [GH47]  \n",
       "0           NaN                [GH32]  \n",
       "\n",
       "[585 rows x 6 columns]"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def parse_ecami_output(txt_file_path):\n",
    "    \"\"\"Parse the output from the output text file from eCAMI and write out data to a dataframe.\n",
    "    \n",
    "    Retrieves the protein accession/name/identifier, predicated CAZy family, predicated CAZy subfamily,\n",
    "    predicated EC number and additional/other domains predicated to also be within the protein sequence,\n",
    "    indicating prediciton of a multiple module enzymes.\n",
    "    \n",
    "    :param text_file_path: path, path to the output text file\n",
    "    \n",
    "    Return Pandas dataframe containing eCAMI output\n",
    "    \"\"\"\n",
    "    with open(txt_file_path, \"r\") as fh:\n",
    "        ecami_file = fh.read().splitlines()\n",
    "    \n",
    "    # build an empty dataframe to add predication outputs to\n",
    "    ecami_df = pd.DataFrame(columns=[\"protein_accession\",\"cazy_family\",\"cazy_subfamily\",\"ec_number\",\"domain_range\"])\n",
    "        \n",
    "    # parse the outputs so in format suitable for final dataframe\n",
    "    for line in ecami_file:\n",
    "        if line.startswith(\">\"):  # identifies new protein\n",
    "            prediction_output = line.split(\"\\t\")\n",
    "            cazy_fam = prediction_output[1].split(\":\")[0]\n",
    "\n",
    "            cazy_subfam = np.nan\n",
    "            ec_number = np.nan\n",
    "\n",
    "            # check for additional domains\n",
    "            other_domains = prediction_output[2].split(\"|\")\n",
    "            additional_domains = []\n",
    "            for domain in other_domains:\n",
    "                domain = domain.split(\":\")[0]  # drop the eCAMI group number\n",
    "                # check if subfamily is predicated for the main CAZy family\n",
    "                if len(domain.split(\"_\")) != 1:\n",
    "                    cazy_subfam = domain\n",
    "                # check if predicated EC number\n",
    "                elif domain.find(\".\") != -1:\n",
    "                    ec_number = domain\n",
    "                else:\n",
    "                    additional_domains.append(domain)\n",
    "            \n",
    "            if len(additional_domains) == 0:\n",
    "                additional_domains = np.nan\n",
    "\n",
    "            # build dict to enable easy building of df\n",
    "            prediction = {\n",
    "                \"protein_accession\": [prediction_output[0]],\n",
    "                \"cazy_family\": [prediction_output[1].split(\":\")[0]],\n",
    "                \"cazy_subfamily\": [cazy_subfam ],\n",
    "                \"ec_number\": [ec_number],\n",
    "                \"additional_domains\": [additional_domains],\n",
    "            }\n",
    "            \n",
    "            prediction_df = pd.DataFrame(prediction)\n",
    "            ecami_df = ecami_df.append(prediction_df)\n",
    "    \n",
    "    return ecami_df\n",
    "\n",
    "parsed_ecami_output = parse_ecami_output(\"143535_output.txt\")\n",
    "parsed_ecami_output\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Parse output from dbCAN\n",
    "\n",
    "dbCAN produces a directory of output files. This is becuase dbCAN includes the output from the three prediction tools:\n",
    "- DIAMOND\n",
    "- HMMER\n",
    "- Hotpep\n",
    "\n",
    "Each prediction tools has its own specific output, which includes additional predictions other than only the predicated CAZy family. A summary output file is also generated which summarises the output from each of the tools.\n",
    "\n",
    "A summary dataframe, containing the consensus result of the three predictions tools, and a dataframe per prediction tool will be produced for dbCAN.\n",
    "\n",
    "### The data in the summary dbCAN output file\n",
    "\n",
    "This data is kept in the overview.txt file, and contains:\n",
    "- Protein accession ('Gene ID')\n",
    "- Domains predicated by HMMER, separated by '+' and including domain ranges (the index of the first and last residue in the protein sequence)\n",
    "- Domains predicated by Hotpep, separated by '+' and includes the k-mer cluster/group to which the CAZy family belongs in parenthesises\n",
    "- Predicated CAZy family\n",
    "- The number of tools which predicated the query protein sequence was a CAZyme\n",
    "\n",
    "If no prediction is made for a given tool, this is indicated by '-'\n",
    "The data appears to be separated by `\\t`, but need to check the script and when parsing the file.\n",
    "\n",
    "Each protein is must on a separate line.\n",
    "\n",
    "### The data in the DIAMOND output file\n",
    "\n",
    "Data is stored in the file `diamond.out`, and data appears to be separated by `\\t` but this is a .out file so will need to check. This file may be readable as a dataframe.\n",
    "\n",
    "Contains:\n",
    "- Protein accession (\"Gene ID\")\n",
    "- Predicated CAZy family\n",
    "- Percentage identity between query sequence and model of CAZy family\n",
    "- Length - length of protein?\n",
    "- Mismatches - number of mismatched residues?\n",
    "- Gap open\n",
    "- Gene start and End\n",
    "- CAZY start and end, but these numbers can be floats, therefore not fully sure what this represents\n",
    "- E value and Bit score of the alignment\n",
    "\n",
    "Data to retrieve is the protein accession and predicated CAZy family. Need to check what is meant by the CAZy start and end and if this will be helpful (by looking at the Python script). But most of the data is easier to retrieved from the txt file.\n",
    "\n",
    "Multiple predicted domains are not as common as the other two tools (HMMER and Hotpep), and are separated by '+'.\n",
    "\n",
    "Each protein is a separate line.\n",
    "\n",
    "### The data in the HMMER output file\n",
    "\n",
    "Data appears to be separated by `\\t`, and is in the file `hmmer.out`. Again this is .out file and may be readable as a dataframe.\n",
    "\n",
    "Contains the:\n",
    "- HMM profile which matched\n",
    "- \"profile length\"?\n",
    "- protein accession (\"Gene ID\")\n",
    "- gene length\n",
    "- E value of result\n",
    "- The first residue that matched the CAZy family HMM profile\n",
    "- The last residue that matches the CAZy family HMM profile\n",
    "- Gene start and end\n",
    "- Coverage of the protein sequence by the HMM profiles\n",
    "\n",
    "Each unique protein is put on a separate line.\n",
    "\n",
    "Data to retrieve is the protein accession, the CAZy family and the domain range but this will be easier to retrieved from the 'overview.txt' file.\n",
    "\n",
    "### The data in the Hotpep output file\n",
    "\n",
    "The data is stored in the `Hotpep.out` file. It includes:\n",
    "- The predicated CAZy family\n",
    "- The number of the PPR family associated with the CAZy family\n",
    "- Protein accession (\"Gene ID\")\n",
    "- Frequency - ?\n",
    "- The number of matched PPR k-mers\n",
    "- The k-mers that were found in the query sequence\n",
    "- Signature peptides - ?\n",
    "- EC number\n",
    "\n",
    "Data to retrieve includes the protein accession, predicated CAZy family and the EC number. The EC number can be retrieved here but the CAZy family and accesion may be easier to retrieve from the overview, becuase it appears that the other domains are not listed in Hotpep.\n",
    "\n",
    "Each protein is on a unique line."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Gene ID\tHMMER\tHotpep\tDIAMOND\t#ofTools\n",
      "XP_001545265.2\tGH53(20-346)\tGH53(7)\tGH53\t3\n",
      "XP_001545514.1\tGT2_Chitin_synth_2(1213-1720)\tGT2(161)\tGT2\t3\n",
      "XP_001545605.1\tAA4(28-577)\tAA4(1)\tAA4\t3\n",
      "XP_001545700.1\tCBM21(353-459)\tCBM21(4)\tCBM21\t3\n",
      "'Gene ID\\tHMMER\\tHotpep\\tDIAMOND\\t#ofTools'\n",
      "'XP_001545265.2\\tGH53(20-346)\\tGH53(7)\\tGH53\\t3'\n",
      "'XP_001545514.1\\tGT2_Chitin_synth_2(1213-1720)\\tGT2(161)\\tGT2\\t3'\n",
      "'XP_001545605.1\\tAA4(28-577)\\tAA4(1)\\tAA4\\t3'\n",
      "'XP_001545700.1\\tCBM21(353-459)\\tCBM21(4)\\tCBM21\\t3'\n"
     ]
    }
   ],
   "source": [
    "# Check how text is separated in overview.txt\n",
    "with open (\"overview.txt\", \"r\") as ofh:\n",
    "    overview_file = ofh.read().splitlines()\n",
    "\n",
    "for line in overview_file[:5]:\n",
    "    print(line)\n",
    "\n",
    "for line in overview_file[:5]:\n",
    "    print(repr(line))\n",
    "# the data is separated by tabs `\\t`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene ID\\tCAZy ID\\t% Identical\\tLength\\tMismatches\\tGap Open\\tGene Start\\tGene End\\tCAZy Start\\tCAZy End\\tE Value\\tBit Score</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>XP_001545605.1\\tXP_001545605.1|AA4\\t100.0\\t590...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>XP_024545922.1\\tCCD55608.1|GT2\\t100.0\\t392\\t0\\...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>XP_001547235.1\\tATZ45248.1|AA3_1\\t100.0\\t572\\t...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>XP_001547254.1\\tATZ45267.1|AA1_3\\t100.0\\t710\\t...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>XP_024546029.1\\tATZ45383.1|GH18\\t100.0\\t1198\\t...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>594</th>\n",
       "      <td>XP_024553945.1\\tATZ58688.1|GH1\\t100.0\\t576\\t0\\...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>595</th>\n",
       "      <td>XP_001547801.1\\tATZ58699.1|GH16\\t100.0\\t315\\t0...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>596</th>\n",
       "      <td>XP_024553962.1\\tATZ58714.1|GH47\\t100.0\\t568\\t0...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>597</th>\n",
       "      <td>XP_024553961.1\\tCCD55952.1|GH47\\t100.0\\t573\\t0...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>598</th>\n",
       "      <td>XP_024554066.1\\tATZ58863.1|GH32\\t100.0\\t836\\t0...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>599 rows × 1 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    Gene ID\\tCAZy ID\\t% Identical\\tLength\\tMismatches\\tGap Open\\tGene Start\\tGene End\\tCAZy Start\\tCAZy End\\tE Value\\tBit Score\n",
       "0    XP_001545605.1\\tXP_001545605.1|AA4\\t100.0\\t590...                                                                         \n",
       "1    XP_024545922.1\\tCCD55608.1|GT2\\t100.0\\t392\\t0\\...                                                                         \n",
       "2    XP_001547235.1\\tATZ45248.1|AA3_1\\t100.0\\t572\\t...                                                                         \n",
       "3    XP_001547254.1\\tATZ45267.1|AA1_3\\t100.0\\t710\\t...                                                                         \n",
       "4    XP_024546029.1\\tATZ45383.1|GH18\\t100.0\\t1198\\t...                                                                         \n",
       "..                                                 ...                                                                         \n",
       "594  XP_024553945.1\\tATZ58688.1|GH1\\t100.0\\t576\\t0\\...                                                                         \n",
       "595  XP_001547801.1\\tATZ58699.1|GH16\\t100.0\\t315\\t0...                                                                         \n",
       "596  XP_024553962.1\\tATZ58714.1|GH47\\t100.0\\t568\\t0...                                                                         \n",
       "597  XP_024553961.1\\tCCD55952.1|GH47\\t100.0\\t573\\t0...                                                                         \n",
       "598  XP_024554066.1\\tATZ58863.1|GH32\\t100.0\\t836\\t0...                                                                         \n",
       "\n",
       "[599 rows x 1 columns]"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Check if pandas can easily handle .out files\n",
    "diamond_df = pd.read_csv(\"diamond.out\")\n",
    "diamond_df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Pandas cannot pass pass the output file becuase the data is separated by tabs (`\\t`), and setting deliminator for regex expression is not advisable (as written in the Pandas documentation).\n",
    "\n",
    "Therefore, all the dbCAN output files must be parsed in a similar manner to the output from CUPP and eCAMI.\n",
    "\n",
    "Start with parsing the data from the overview.txt file to retrieve all the predicated domains for each protein.\n",
    "\n",
    "\n",
    "I need to also determine what is meanr by CAZy start and CAZy end in the output from DIAMOND. This data is taken straight from the output of DIAMOND when looking at the run_dbcan.py script."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['Gene ID', 'CAZy ID', '% Identical', 'Length', 'Mismatches', 'Gap Open', 'Gene Start', 'Gene End', 'CAZy Start', 'CAZy End', 'E Value', 'Bit Score']\n",
      "['XP_001545605.1', 'XP_001545605.1|AA4', '100.0', '590', '0', '0', '1', '590', '1', '590', '0.0e+00', '1215.7']\n",
      "['XP_024545922.1', 'CCD55608.1|GT2', '100.0', '392', '0', '0', '1', '392', '1', '392', '2.2e-222', '773.9']\n",
      "['XP_001547235.1', 'ATZ45248.1|AA3_1', '100.0', '572', '0', '0', '1', '572', '1', '572', '0.0e+00', '1135.2']\n",
      "['XP_001547254.1', 'ATZ45267.1|AA1_3', '100.0', '710', '0', '0', '1', '710', '1', '710', '0.0e+00', '1236.9']\n"
     ]
    }
   ],
   "source": [
    "with open(\"diamond.out\", \"r\") as fh:\n",
    "    diamond = fh.read().splitlines()\n",
    "\n",
    "for line in diamond[:5]:\n",
    "    line = line.split(\"\\t\")\n",
    "    print(line)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "length= Length\n",
      "cazy start= Gene Start\n",
      "cazy end= Gene End\n",
      "length= 590\n",
      "cazy start= 1\n",
      "cazy end= 590\n",
      "length= 392\n",
      "cazy start= 1\n",
      "cazy end= 392\n",
      "length= 572\n",
      "cazy start= 1\n",
      "cazy end= 572\n",
      "length= 710\n",
      "cazy start= 1\n",
      "cazy end= 710\n"
     ]
    }
   ],
   "source": [
    "for line in diamond[:5]:\n",
    "    line = line.split(\"\\t\")\n",
    "    print(\"length=\", line[3])\n",
    "    print(\"cazy start=\", line[6])\n",
    "    print(\"cazy end=\", line[7])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene ID</th>\n",
       "      <th>CAZy ID</th>\n",
       "      <th>% Identical</th>\n",
       "      <th>Length</th>\n",
       "      <th>Mismatches</th>\n",
       "      <th>Gap Open</th>\n",
       "      <th>Gene Start</th>\n",
       "      <th>Gene End</th>\n",
       "      <th>CAZy Start</th>\n",
       "      <th>CAZy End</th>\n",
       "      <th>E Value</th>\n",
       "      <th>Bit Score</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>XP_001545605.1</td>\n",
       "      <td>XP_001545605.1|AA4</td>\n",
       "      <td>100.0</td>\n",
       "      <td>590</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>590</td>\n",
       "      <td>1</td>\n",
       "      <td>590</td>\n",
       "      <td>0.0e+00</td>\n",
       "      <td>1215.7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>XP_024545922.1</td>\n",
       "      <td>CCD55608.1|GT2</td>\n",
       "      <td>100.0</td>\n",
       "      <td>392</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>392</td>\n",
       "      <td>1</td>\n",
       "      <td>392</td>\n",
       "      <td>2.2e-222</td>\n",
       "      <td>773.9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>XP_001547235.1</td>\n",
       "      <td>ATZ45248.1|AA3_1</td>\n",
       "      <td>100.0</td>\n",
       "      <td>572</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>572</td>\n",
       "      <td>1</td>\n",
       "      <td>572</td>\n",
       "      <td>0.0e+00</td>\n",
       "      <td>1135.2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>XP_001547254.1</td>\n",
       "      <td>ATZ45267.1|AA1_3</td>\n",
       "      <td>100.0</td>\n",
       "      <td>710</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>710</td>\n",
       "      <td>1</td>\n",
       "      <td>710</td>\n",
       "      <td>0.0e+00</td>\n",
       "      <td>1236.9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>XP_024546029.1</td>\n",
       "      <td>ATZ45383.1|GH18</td>\n",
       "      <td>100.0</td>\n",
       "      <td>1198</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1198</td>\n",
       "      <td>1</td>\n",
       "      <td>1198</td>\n",
       "      <td>0.0e+00</td>\n",
       "      <td>1496.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>594</th>\n",
       "      <td>XP_024553945.1</td>\n",
       "      <td>ATZ58688.1|GH1</td>\n",
       "      <td>100.0</td>\n",
       "      <td>576</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>576</td>\n",
       "      <td>1</td>\n",
       "      <td>576</td>\n",
       "      <td>0.0e+00</td>\n",
       "      <td>1191.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>595</th>\n",
       "      <td>XP_001547801.1</td>\n",
       "      <td>ATZ58699.1|GH16</td>\n",
       "      <td>100.0</td>\n",
       "      <td>315</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>315</td>\n",
       "      <td>1</td>\n",
       "      <td>315</td>\n",
       "      <td>2.0e-181</td>\n",
       "      <td>637.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>596</th>\n",
       "      <td>XP_024553962.1</td>\n",
       "      <td>ATZ58714.1|GH47</td>\n",
       "      <td>100.0</td>\n",
       "      <td>568</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>568</td>\n",
       "      <td>1</td>\n",
       "      <td>568</td>\n",
       "      <td>0.0e+00</td>\n",
       "      <td>1171.8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>597</th>\n",
       "      <td>XP_024553961.1</td>\n",
       "      <td>CCD55952.1|GH47</td>\n",
       "      <td>100.0</td>\n",
       "      <td>573</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>573</td>\n",
       "      <td>1</td>\n",
       "      <td>573</td>\n",
       "      <td>0.0e+00</td>\n",
       "      <td>1181.4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>598</th>\n",
       "      <td>XP_024554066.1</td>\n",
       "      <td>ATZ58863.1|GH32</td>\n",
       "      <td>100.0</td>\n",
       "      <td>836</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>836</td>\n",
       "      <td>1</td>\n",
       "      <td>836</td>\n",
       "      <td>0.0e+00</td>\n",
       "      <td>1585.5</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>599 rows × 12 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "            Gene ID             CAZy ID % Identical Length Mismatches  \\\n",
       "0    XP_001545605.1  XP_001545605.1|AA4       100.0    590          0   \n",
       "1    XP_024545922.1      CCD55608.1|GT2       100.0    392          0   \n",
       "2    XP_001547235.1    ATZ45248.1|AA3_1       100.0    572          0   \n",
       "3    XP_001547254.1    ATZ45267.1|AA1_3       100.0    710          0   \n",
       "4    XP_024546029.1     ATZ45383.1|GH18       100.0   1198          0   \n",
       "..              ...                 ...         ...    ...        ...   \n",
       "594  XP_024553945.1      ATZ58688.1|GH1       100.0    576          0   \n",
       "595  XP_001547801.1     ATZ58699.1|GH16       100.0    315          0   \n",
       "596  XP_024553962.1     ATZ58714.1|GH47       100.0    568          0   \n",
       "597  XP_024553961.1     CCD55952.1|GH47       100.0    573          0   \n",
       "598  XP_024554066.1     ATZ58863.1|GH32       100.0    836          0   \n",
       "\n",
       "    Gap Open Gene Start Gene End CAZy Start CAZy End   E Value Bit Score  \n",
       "0          0          1      590          1      590   0.0e+00    1215.7  \n",
       "1          0          1      392          1      392  2.2e-222     773.9  \n",
       "2          0          1      572          1      572   0.0e+00    1135.2  \n",
       "3          0          1      710          1      710   0.0e+00    1236.9  \n",
       "4          0          1     1198          1     1198   0.0e+00    1496.1  \n",
       "..       ...        ...      ...        ...      ...       ...       ...  \n",
       "594        0          1      576          1      576   0.0e+00    1191.0  \n",
       "595        0          1      315          1      315  2.0e-181     637.5  \n",
       "596        0          1      568          1      568   0.0e+00    1171.8  \n",
       "597        0          1      573          1      573   0.0e+00    1181.4  \n",
       "598        0          1      836          1      836   0.0e+00    1585.5  \n",
       "\n",
       "[599 rows x 12 columns]"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# appears that it should be predicating where the protein starts and end\n",
    "# this maybe a feature for when a DNA and not an AA sequence is given\n",
    "# check if the CAZy end != length\n",
    "\n",
    "# first build a df to help with vistualisation\n",
    "df_data = []\n",
    "for line in diamond[1:]:\n",
    "    line = line.split(\"\\t\")\n",
    "    df_data.append(line)\n",
    "    \n",
    "diamond_df = pd.DataFrame(df_data, columns=['Gene ID', 'CAZy ID', '% Identical', 'Length', 'Mismatches', 'Gap Open', 'Gene Start', 'Gene End', 'CAZy Start', 'CAZy End', 'E Value', 'Bit Score'])\n",
    "diamond_df\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Gene ID         XP_024546124.1\n",
      "CAZy ID        EAA31326.1|GH71\n",
      "% Identical               72.7\n",
      "Length                     600\n",
      "Mismatches                 152\n",
      "Gap Open                     4\n",
      "Gene Start                  94\n",
      "Gene End                   688\n",
      "CAZy Start                 623\n",
      "CAZy End                  1215\n",
      "E Value               5.6e-253\n",
      "Bit Score                876.3\n",
      "Name: 9, dtype: object \n",
      "\n",
      "Gene ID         XP_024546154.1\n",
      "CAZy ID        AWP04987.1|GT13\n",
      "% Identical               42.6\n",
      "Length                     967\n",
      "Mismatches                 522\n",
      "Gap Open                    14\n",
      "Gene Start                 119\n",
      "Gene End                  1065\n",
      "CAZy Start                  80\n",
      "CAZy End                  1033\n",
      "E Value               6.4e-219\n",
      "Bit Score                763.8\n",
      "Name: 14, dtype: object \n",
      "\n",
      "Gene ID               XP_001547281.1\n",
      "CAZy ID        EAA64118.1|CBM20|GH15\n",
      "% Identical                     47.3\n",
      "Length                           579\n",
      "Mismatches                       261\n",
      "Gap Open                          11\n",
      "Gene Start                         2\n",
      "Gene End                         571\n",
      "CAZy Start                       693\n",
      "CAZy End                        1236\n",
      "E Value                     1.6e-152\n",
      "Bit Score                      542.3\n",
      "Name: 18, dtype: object \n",
      "\n",
      "Gene ID         XP_024546394.1\n",
      "CAZy ID        CCA39958.1|GT15\n",
      "% Identical               33.6\n",
      "Length                     906\n",
      "Mismatches                 463\n",
      "Gap Open                    23\n",
      "Gene Start                   1\n",
      "Gene End                   845\n",
      "CAZy Start                 447\n",
      "CAZy End                  1274\n",
      "E Value               1.4e-119\n",
      "Bit Score                433.7\n",
      "Name: 27, dtype: object \n",
      "\n",
      "Gene ID         XP_024546395.1\n",
      "CAZy ID        CCA39958.1|GT15\n",
      "% Identical               33.6\n",
      "Length                     906\n",
      "Mismatches                 463\n",
      "Gap Open                    23\n",
      "Gene Start                   1\n",
      "Gene End                   845\n",
      "CAZy Start                 447\n",
      "CAZy End                  1274\n",
      "E Value               1.4e-119\n",
      "Bit Score                433.7\n",
      "Name: 28, dtype: object \n",
      "\n",
      "Gene ID         XP_024546458.1\n",
      "CAZy ID        AWP04987.1|GT13\n",
      "% Identical               44.4\n",
      "Length                     846\n",
      "Mismatches                 449\n",
      "Gap Open                    13\n",
      "Gene Start                  69\n",
      "Gene End                   905\n",
      "CAZy Start                  84\n",
      "CAZy End                   917\n",
      "E Value               8.5e-205\n",
      "Bit Score                716.8\n",
      "Name: 37, dtype: object \n",
      "\n",
      "Gene ID         XP_024546457.1\n",
      "CAZy ID        AWP04987.1|GT13\n",
      "% Identical               44.7\n",
      "Length                     758\n",
      "Mismatches                 403\n",
      "Gap Open                    11\n",
      "Gene Start                   1\n",
      "Gene End                   750\n",
      "CAZy Start                 168\n",
      "CAZy End                   917\n",
      "E Value               3.5e-183\n",
      "Bit Score                644.8\n",
      "Name: 38, dtype: object \n",
      "\n",
      "Gene ID        XP_001561110.1\n",
      "CAZy ID        ANH22675.1|CE4\n",
      "% Identical              57.9\n",
      "Length                    399\n",
      "Mismatches                150\n",
      "Gap Open                    3\n",
      "Gene Start                 19\n",
      "Gene End                  416\n",
      "CAZy Start                  7\n",
      "CAZy End                  388\n",
      "E Value              4.0e-129\n",
      "Bit Score               464.2\n",
      "Name: 39, dtype: object \n",
      "\n",
      "Gene ID                  XP_024546573.1\n",
      "CAZy ID        EAA65758.1|GH13_25|GH133\n",
      "% Identical                        49.6\n",
      "Length                              794\n",
      "Mismatches                          360\n",
      "Gap Open                              9\n",
      "Gene Start                           30\n",
      "Gene End                            809\n",
      "CAZy Start                         1585\n",
      "CAZy End                           2352\n",
      "E Value                        1.3e-216\n",
      "Bit Score                         755.7\n",
      "Name: 46, dtype: object \n",
      "\n",
      "Gene ID               XP_024546619.1\n",
      "CAZy ID        AWP09643.1|CBM13|GT27\n",
      "% Identical                     48.6\n",
      "Length                           414\n",
      "Mismatches                       202\n",
      "Gap Open                           6\n",
      "Gene Start                        32\n",
      "Gene End                         441\n",
      "CAZy Start                       574\n",
      "CAZy End                         980\n",
      "E Value                     1.4e-107\n",
      "Bit Score                      392.9\n",
      "Name: 52, dtype: object \n",
      "\n",
      "Gene ID         XP_024546782.1\n",
      "CAZy ID        BAE64464.1|GH28\n",
      "% Identical               54.4\n",
      "Length                     666\n",
      "Mismatches                 274\n",
      "Gap Open                    12\n",
      "Gene Start                  24\n",
      "Gene End                   673\n",
      "CAZy Start                 598\n",
      "CAZy End                  1249\n",
      "E Value               3.9e-189\n",
      "Bit Score                664.5\n",
      "Name: 65, dtype: object \n",
      "\n",
      "Gene ID         XP_024546791.1\n",
      "CAZy ID        EAA58990.1|GH71\n",
      "% Identical               58.0\n",
      "Length                     493\n",
      "Mismatches                 192\n",
      "Gap Open                     5\n",
      "Gene Start                 130\n",
      "Gene End                   619\n",
      "CAZy Start                 500\n",
      "CAZy End                   980\n",
      "E Value               1.9e-159\n",
      "Bit Score                565.5\n",
      "Name: 69, dtype: object \n",
      "\n",
      "Gene ID         XP_024547084.1\n",
      "CAZy ID        EAA65371.1|GT22\n",
      "% Identical               68.5\n",
      "Length                     355\n",
      "Mismatches                  98\n",
      "Gap Open                     3\n",
      "Gene Start                  87\n",
      "Gene End                   429\n",
      "CAZy Start                 608\n",
      "CAZy End                   960\n",
      "E Value               4.2e-142\n",
      "Bit Score                507.3\n",
      "Name: 98, dtype: object \n",
      "\n",
      "Gene ID          XP_024547088.1\n",
      "CAZy ID        SCD20010.1|GH146\n",
      "% Identical                40.4\n",
      "Length                      804\n",
      "Mismatches                  450\n",
      "Gap Open                      7\n",
      "Gene Start                   29\n",
      "Gene End                    814\n",
      "CAZy Start                   81\n",
      "CAZy End                    873\n",
      "E Value                2.0e-167\n",
      "Bit Score                 592.4\n",
      "Name: 99, dtype: object \n",
      "\n",
      "Gene ID              XP_024547162.1\n",
      "CAZy ID        EAA57680.1|GH78|GH79\n",
      "% Identical                    51.0\n",
      "Length                          465\n",
      "Mismatches                      216\n",
      "Gap Open                          5\n",
      "Gene Start                       87\n",
      "Gene End                        549\n",
      "CAZy Start                     1334\n",
      "CAZy End                       1788\n",
      "E Value                    7.4e-126\n",
      "Bit Score                     453.8\n",
      "Name: 101, dtype: object \n",
      "\n",
      "Gene ID        XP_024547374.1\n",
      "CAZy ID        BAZ00088.1|GT2\n",
      "% Identical              31.8\n",
      "Length                    942\n",
      "Mismatches                562\n",
      "Gap Open                   23\n",
      "Gene Start                384\n",
      "Gene End                 1291\n",
      "CAZy Start               1214\n",
      "CAZy End                 2109\n",
      "E Value              1.5e-110\n",
      "Bit Score               404.8\n",
      "Name: 119, dtype: object \n",
      "\n",
      "Gene ID        XP_024547375.1\n",
      "CAZy ID        BAZ00088.1|GT2\n",
      "% Identical              31.8\n",
      "Length                    942\n",
      "Mismatches                562\n",
      "Gap Open                   23\n",
      "Gene Start                384\n",
      "Gene End                 1291\n",
      "CAZy Start               1214\n",
      "CAZy End                 2109\n",
      "E Value              1.5e-110\n",
      "Bit Score               404.8\n",
      "Name: 120, dtype: object \n",
      "\n",
      "Gene ID         XP_001559355.1\n",
      "CAZy ID        AIT18914.1|GH18\n",
      "% Identical               32.4\n",
      "Length                    1104\n",
      "Mismatches                 657\n",
      "Gap Open                    34\n",
      "Gene Start                  24\n",
      "Gene End                  1091\n",
      "CAZy Start                  15\n",
      "CAZy End                  1065\n",
      "E Value               4.1e-133\n",
      "Bit Score                478.8\n",
      "Name: 130, dtype: object \n",
      "\n",
      "Gene ID               XP_024547556.1\n",
      "CAZy ID        CEI39478.1|CBM18|GH18\n",
      "% Identical                     34.9\n",
      "Length                           893\n",
      "Mismatches                       430\n",
      "Gap Open                          25\n",
      "Gene Start                         1\n",
      "Gene End                         801\n",
      "CAZy Start                       826\n",
      "CAZy End                        1659\n",
      "E Value                     1.0e-117\n",
      "Bit Score                      427.2\n",
      "Name: 135, dtype: object \n",
      "\n",
      "Gene ID        XP_024547690.1\n",
      "CAZy ID        AHX26706.1|GT2\n",
      "% Identical              34.2\n",
      "Length                    734\n",
      "Mismatches                434\n",
      "Gap Open                   15\n",
      "Gene Start                173\n",
      "Gene End                  868\n",
      "CAZy Start                 98\n",
      "CAZy End                  820\n",
      "E Value              3.9e-104\n",
      "Bit Score               383.6\n",
      "Name: 143, dtype: object \n",
      "\n",
      "Gene ID         XP_024548313.1\n",
      "CAZy ID        SMR56934.1|GT22\n",
      "% Identical               61.3\n",
      "Length                     640\n",
      "Mismatches                 219\n",
      "Gap Open                     5\n",
      "Gene Start                   4\n",
      "Gene End                   630\n",
      "CAZy Start                 513\n",
      "CAZy End                  1136\n",
      "E Value               3.6e-227\n",
      "Bit Score                790.4\n",
      "Name: 172, dtype: object \n",
      "\n",
      "Gene ID            XP_024548343.1\n",
      "CAZy ID        AIE96983.1|GT0|GT7\n",
      "% Identical                  30.2\n",
      "Length                       1245\n",
      "Mismatches                    683\n",
      "Gap Open                       32\n",
      "Gene Start                     90\n",
      "Gene End                     1203\n",
      "CAZy Start                      7\n",
      "CAZy End                     1196\n",
      "E Value                  3.9e-137\n",
      "Bit Score                   492.3\n",
      "Name: 173, dtype: object \n",
      "\n",
      "Gene ID         XP_024548384.1\n",
      "CAZy ID        AIT18868.1|GH18\n",
      "% Identical               38.4\n",
      "Length                     683\n",
      "Mismatches                 333\n",
      "Gap Open                    18\n",
      "Gene Start                 120\n",
      "Gene End                   792\n",
      "CAZy Start                 376\n",
      "CAZy End                   980\n",
      "E Value               2.9e-104\n",
      "Bit Score                382.5\n",
      "Name: 176, dtype: object \n",
      "\n",
      "Gene ID            XP_001549316.1\n",
      "CAZy ID        SMY27695.1|GH43_21\n",
      "% Identical                  59.8\n",
      "Length                        503\n",
      "Mismatches                    200\n",
      "Gap Open                        2\n",
      "Gene Start                     36\n",
      "Gene End                      538\n",
      "CAZy Start                     44\n",
      "CAZy End                      544\n",
      "E Value                  2.2e-172\n",
      "Bit Score                   608.2\n",
      "Name: 185, dtype: object \n",
      "\n",
      "Gene ID        XP_001550386.1\n",
      "CAZy ID        EAA63395.1|GH2\n",
      "% Identical              52.0\n",
      "Length                    425\n",
      "Mismatches                192\n",
      "Gap Open                    3\n",
      "Gene Start                  3\n",
      "Gene End                  427\n",
      "CAZy Start                854\n",
      "CAZy End                 1266\n",
      "E Value              3.2e-118\n",
      "Bit Score               427.9\n",
      "Name: 189, dtype: object \n",
      "\n",
      "Gene ID          XP_001550354.1\n",
      "CAZy ID        ACU75807.1|CBM13\n",
      "% Identical                37.2\n",
      "Length                      701\n",
      "Mismatches                  397\n",
      "Gap Open                     11\n",
      "Gene Start                   19\n",
      "Gene End                    707\n",
      "CAZy Start                   84\n",
      "CAZy End                    753\n",
      "E Value                8.0e-122\n",
      "Bit Score                 440.7\n",
      "Name: 191, dtype: object \n",
      "\n",
      "Gene ID        XP_001560033.1\n",
      "CAZy ID        AIT83212.1|GT4\n",
      "% Identical              70.2\n",
      "Length                    302\n",
      "Mismatches                 88\n",
      "Gap Open                    2\n",
      "Gene Start                  1\n",
      "Gene End                  301\n",
      "CAZy Start                 14\n",
      "CAZy End                  314\n",
      "E Value              1.6e-120\n",
      "Bit Score               435.3\n",
      "Name: 202, dtype: object \n",
      "\n",
      "Gene ID         XP_024548909.1\n",
      "CAZy ID        ANH56445.1|GH36\n",
      "% Identical               56.9\n",
      "Length                     350\n",
      "Mismatches                 149\n",
      "Gap Open                     2\n",
      "Gene Start                   1\n",
      "Gene End                   350\n",
      "CAZy Start                   1\n",
      "CAZy End                   348\n",
      "E Value               4.1e-111\n",
      "Bit Score                404.1\n",
      "Name: 216, dtype: object \n",
      "\n",
      "Gene ID        XP_001547166.1\n",
      "CAZy ID        EAA63395.1|GH2\n",
      "% Identical              49.3\n",
      "Length                    428\n",
      "Mismatches                206\n",
      "Gap Open                    5\n",
      "Gene Start                  3\n",
      "Gene End                  430\n",
      "CAZy Start                854\n",
      "CAZy End                 1270\n",
      "E Value              2.3e-108\n",
      "Bit Score               395.2\n",
      "Name: 255, dtype: object \n",
      "\n",
      "Gene ID         XP_001555974.1\n",
      "CAZy ID        AWO98658.1|GH37\n",
      "% Identical               66.0\n",
      "Length                     430\n",
      "Mismatches                 134\n",
      "Gap Open                     2\n",
      "Gene Start                   5\n",
      "Gene End                   422\n",
      "CAZy Start                  42\n",
      "CAZy End                   471\n",
      "E Value               7.5e-165\n",
      "Bit Score                583.2\n",
      "Name: 281, dtype: object \n",
      "\n",
      "Gene ID             XP_001556039.1\n",
      "CAZy ID        EAA63677.1|GH1|GH95\n",
      "% Identical                   52.8\n",
      "Length                         479\n",
      "Mismatches                     197\n",
      "Gap Open                        10\n",
      "Gene Start                      38\n",
      "Gene End                       503\n",
      "CAZy Start                    1215\n",
      "CAZy End                      1677\n",
      "E Value                   4.3e-122\n",
      "Bit Score                    441.0\n",
      "Name: 284, dtype: object \n",
      "\n",
      "Gene ID         XP_024549602.1\n",
      "CAZy ID        ATZ51462.1|GT20\n",
      "% Identical              100.0\n",
      "Length                     911\n",
      "Mismatches                   0\n",
      "Gap Open                     0\n",
      "Gene Start                   1\n",
      "Gene End                   911\n",
      "CAZy Start                  86\n",
      "CAZy End                   996\n",
      "E Value                0.0e+00\n",
      "Bit Score               1845.5\n",
      "Name: 288, dtype: object \n",
      "\n",
      "Gene ID         XP_024549601.1\n",
      "CAZy ID        ATZ51462.1|GT20\n",
      "% Identical              100.0\n",
      "Length                     911\n",
      "Mismatches                   0\n",
      "Gap Open                     0\n",
      "Gene Start                   1\n",
      "Gene End                   911\n",
      "CAZy Start                  86\n",
      "CAZy End                   996\n",
      "E Value                0.0e+00\n",
      "Bit Score               1845.5\n",
      "Name: 289, dtype: object \n",
      "\n",
      "Gene ID         XP_001558652.1\n",
      "CAZy ID        AIT18866.1|GH18\n",
      "% Identical               31.1\n",
      "Length                    1053\n",
      "Mismatches                 659\n",
      "Gap Open                    22\n",
      "Gene Start                  14\n",
      "Gene End                  1023\n",
      "CAZy Start                  89\n",
      "CAZy End                  1118\n",
      "E Value               9.5e-148\n",
      "Bit Score                527.3\n",
      "Name: 302, dtype: object \n",
      "\n",
      "Gene ID        XP_024549743.1\n",
      "CAZy ID        BAZ00088.1|GT2\n",
      "% Identical              32.5\n",
      "Length                    844\n",
      "Mismatches                504\n",
      "Gap Open                   20\n",
      "Gene Start                 71\n",
      "Gene End                  885\n",
      "CAZy Start               1214\n",
      "CAZy End                 2020\n",
      "E Value              2.4e-108\n",
      "Bit Score               397.5\n",
      "Name: 304, dtype: object \n",
      "\n",
      "Gene ID         XP_024549896.1\n",
      "CAZy ID        EAA58983.1|GH18\n",
      "% Identical               46.2\n",
      "Length                     517\n",
      "Mismatches                 242\n",
      "Gap Open                    10\n",
      "Gene Start                 118\n",
      "Gene End                   613\n",
      "CAZy Start                 361\n",
      "CAZy End                   862\n",
      "E Value               5.7e-116\n",
      "Bit Score                421.0\n",
      "Name: 320, dtype: object \n",
      "\n",
      "Gene ID         XP_024549897.1\n",
      "CAZy ID        EAA58983.1|GH18\n",
      "% Identical               46.2\n",
      "Length                     517\n",
      "Mismatches                 242\n",
      "Gap Open                    10\n",
      "Gene Start                 118\n",
      "Gene End                   613\n",
      "CAZy Start                 361\n",
      "CAZy End                   862\n",
      "E Value               5.7e-116\n",
      "Bit Score                421.0\n",
      "Name: 321, dtype: object \n",
      "\n",
      "Gene ID         XP_024549895.1\n",
      "CAZy ID        EAA58983.1|GH18\n",
      "% Identical               46.9\n",
      "Length                     465\n",
      "Mismatches                 212\n",
      "Gap Open                     9\n",
      "Gene Start                  10\n",
      "Gene End                   453\n",
      "CAZy Start                 412\n",
      "CAZy End                   862\n",
      "E Value               4.3e-108\n",
      "Bit Score                394.4\n",
      "Name: 322, dtype: object \n",
      "\n",
      "Gene ID         XP_024549940.1\n",
      "CAZy ID        ATZ52025.1|GH47\n",
      "% Identical              100.0\n",
      "Length                     483\n",
      "Mismatches                   0\n",
      "Gap Open                     0\n",
      "Gene Start                   1\n",
      "Gene End                   483\n",
      "CAZy Start                 151\n",
      "CAZy End                   633\n",
      "E Value               3.5e-286\n",
      "Bit Score                986.1\n",
      "Name: 325, dtype: object \n",
      "\n",
      "Gene ID         XP_001549629.1\n",
      "CAZy ID        CUI24525.1|GH19\n",
      "% Identical               59.9\n",
      "Length                     322\n",
      "Mismatches                 123\n",
      "Gap Open                     2\n",
      "Gene Start                  37\n",
      "Gene End                   352\n",
      "CAZy Start                   7\n",
      "CAZy End                   328\n",
      "E Value               1.9e-104\n",
      "Bit Score                382.9\n",
      "Name: 326, dtype: object \n",
      "\n",
      "Gene ID          XP_024550245.1\n",
      "CAZy ID        ATZ52510.1|AA3_2\n",
      "% Identical               100.0\n",
      "Length                      222\n",
      "Mismatches                    0\n",
      "Gap Open                      0\n",
      "Gene Start                    1\n",
      "Gene End                    222\n",
      "CAZy Start                  334\n",
      "CAZy End                    555\n",
      "E Value                1.5e-122\n",
      "Bit Score                 441.4\n",
      "Name: 346, dtype: object \n",
      "\n",
      "Gene ID         XP_024550334.1\n",
      "CAZy ID        CDS04937.1|GH18\n",
      "% Identical               34.0\n",
      "Length                    2379\n",
      "Mismatches                1330\n",
      "Gap Open                    49\n",
      "Gene Start                 292\n",
      "Gene End                  2605\n",
      "CAZy Start                 868\n",
      "CAZy End                  3071\n",
      "E Value                0.0e+00\n",
      "Bit Score               1216.1\n",
      "Name: 352, dtype: object \n",
      "\n",
      "Gene ID         XP_001553964.2\n",
      "CAZy ID        CED83340.1|GT31\n",
      "% Identical               43.4\n",
      "Length                    1071\n",
      "Mismatches                 455\n",
      "Gap Open                    25\n",
      "Gene Start                 243\n",
      "Gene End                  1179\n",
      "CAZy Start                 859\n",
      "CAZy End                  1912\n",
      "E Value               2.5e-216\n",
      "Bit Score                755.4\n",
      "Name: 353, dtype: object \n",
      "\n",
      "Gene ID            XP_001552323.1\n",
      "CAZy ID        AOY72300.1|GH43_17\n",
      "% Identical                  34.3\n",
      "Length                        753\n",
      "Mismatches                    392\n",
      "Gap Open                       21\n",
      "Gene Start                     20\n",
      "Gene End                      703\n",
      "CAZy Start                    303\n",
      "CAZy End                     1021\n",
      "E Value                  1.7e-108\n",
      "Bit Score                   396.4\n",
      "Name: 359, dtype: object \n",
      "\n",
      "Gene ID         XP_024550615.1\n",
      "CAZy ID        AAM94807.1|GH18\n",
      "% Identical              100.0\n",
      "Length                     304\n",
      "Mismatches                   0\n",
      "Gap Open                     0\n",
      "Gene Start                   1\n",
      "Gene End                   304\n",
      "CAZy Start                 126\n",
      "CAZy End                   429\n",
      "E Value               4.2e-173\n",
      "Bit Score                609.8\n",
      "Name: 369, dtype: object \n",
      "\n",
      "Gene ID         XP_024550754.1\n",
      "CAZy ID        AIT18900.1|GH18\n",
      "% Identical               30.3\n",
      "Length                    1089\n",
      "Mismatches                 636\n",
      "Gap Open                    33\n",
      "Gene Start                  17\n",
      "Gene End                  1047\n",
      "CAZy Start                  15\n",
      "CAZy End                  1038\n",
      "E Value               1.0e-123\n",
      "Bit Score                447.6\n",
      "Name: 381, dtype: object \n",
      "\n",
      "Gene ID         XP_024550762.1\n",
      "CAZy ID        CAN80640.1|GT47\n",
      "% Identical               47.5\n",
      "Length                     432\n",
      "Mismatches                 213\n",
      "Gap Open                     7\n",
      "Gene Start                   3\n",
      "Gene End                   430\n",
      "CAZy Start                 942\n",
      "CAZy End                  1363\n",
      "E Value               8.0e-109\n",
      "Bit Score                396.7\n",
      "Name: 384, dtype: object \n",
      "\n",
      "Gene ID          XP_024550885.1\n",
      "CAZy ID        SMY25446.1|CBM18\n",
      "% Identical                38.7\n",
      "Length                      592\n",
      "Mismatches                  244\n",
      "Gap Open                     13\n",
      "Gene Start                   15\n",
      "Gene End                    518\n",
      "CAZy Start                    9\n",
      "CAZy End                    569\n",
      "E Value                4.3e-104\n",
      "Bit Score                 381.3\n",
      "Name: 390, dtype: object \n",
      "\n",
      "Gene ID          XP_001551566.1\n",
      "CAZy ID        SMY25446.1|CBM18\n",
      "% Identical                38.7\n",
      "Length                      592\n",
      "Mismatches                  244\n",
      "Gap Open                     13\n",
      "Gene Start                   15\n",
      "Gene End                    518\n",
      "CAZy Start                    9\n",
      "CAZy End                    569\n",
      "E Value                4.3e-104\n",
      "Bit Score                 381.3\n",
      "Name: 391, dtype: object \n",
      "\n",
      "Gene ID        XP_024551132.1\n",
      "CAZy ID        ATZ53949.1|GH3\n",
      "% Identical             100.0\n",
      "Length                    759\n",
      "Mismatches                  0\n",
      "Gap Open                    0\n",
      "Gene Start                  1\n",
      "Gene End                  759\n",
      "CAZy Start                 50\n",
      "CAZy End                  808\n",
      "E Value               0.0e+00\n",
      "Bit Score              1527.3\n",
      "Name: 412, dtype: object \n",
      "\n",
      "Gene ID        XP_024551133.1\n",
      "CAZy ID        ATZ53950.1|GH3\n",
      "% Identical             100.0\n",
      "Length                    741\n",
      "Mismatches                  0\n",
      "Gap Open                    0\n",
      "Gene Start                  1\n",
      "Gene End                  741\n",
      "CAZy Start                 50\n",
      "CAZy End                  790\n",
      "E Value               0.0e+00\n",
      "Bit Score              1488.4\n",
      "Name: 413, dtype: object \n",
      "\n",
      "Gene ID            XP_024551147.1\n",
      "CAZy ID        ATZ53965.1|GH43_26\n",
      "% Identical                 100.0\n",
      "Length                        357\n",
      "Mismatches                      0\n",
      "Gap Open                        0\n",
      "Gene Start                      1\n",
      "Gene End                      357\n",
      "CAZy Start                     28\n",
      "CAZy End                      384\n",
      "E Value                  8.7e-218\n",
      "Bit Score                   758.4\n",
      "Name: 416, dtype: object \n",
      "\n",
      "Gene ID        XP_001546108.1\n",
      "CAZy ID        AWK40335.1|GH6\n",
      "% Identical              36.7\n",
      "Length                    700\n",
      "Mismatches                409\n",
      "Gap Open                    9\n",
      "Gene Start                184\n",
      "Gene End                  875\n",
      "CAZy Start                180\n",
      "CAZy End                  853\n",
      "E Value              1.3e-116\n",
      "Bit Score               423.7\n",
      "Name: 425, dtype: object \n",
      "\n",
      "Gene ID           XP_024551612.1\n",
      "CAZy ID        CBX92400.1|GH30_7\n",
      "% Identical                 41.6\n",
      "Length                      1002\n",
      "Mismatches                   531\n",
      "Gap Open                      18\n",
      "Gene Start                     2\n",
      "Gene End                     970\n",
      "CAZy Start                   643\n",
      "CAZy End                    1623\n",
      "E Value                 1.9e-206\n",
      "Bit Score                  722.2\n",
      "Name: 438, dtype: object \n",
      "\n",
      "Gene ID         XP_001556605.2\n",
      "CAZy ID        AIT18866.1|GH18\n",
      "% Identical               35.8\n",
      "Length                    1055\n",
      "Mismatches                 635\n",
      "Gap Open                    18\n",
      "Gene Start                  28\n",
      "Gene End                  1062\n",
      "CAZy Start                  90\n",
      "CAZy End                  1122\n",
      "E Value               7.2e-183\n",
      "Bit Score                644.0\n",
      "Name: 440, dtype: object \n",
      "\n",
      "Gene ID         XP_001556820.2\n",
      "CAZy ID        ATY59959.1|GT31\n",
      "% Identical               49.3\n",
      "Length                     678\n",
      "Mismatches                 279\n",
      "Gap Open                    14\n",
      "Gene Start                   1\n",
      "Gene End                   663\n",
      "CAZy Start                   1\n",
      "CAZy End                   628\n",
      "E Value               1.9e-149\n",
      "Bit Score                532.3\n",
      "Name: 447, dtype: object \n",
      "\n",
      "Gene ID        XP_001558489.2\n",
      "CAZy ID        AWK40335.1|GH6\n",
      "% Identical              36.5\n",
      "Length                    690\n",
      "Mismatches                413\n",
      "Gap Open                    8\n",
      "Gene Start                 70\n",
      "Gene End                  749\n",
      "CAZy Start                165\n",
      "CAZy End                  839\n",
      "E Value              5.2e-114\n",
      "Bit Score               414.8\n",
      "Name: 449, dtype: object \n",
      "\n",
      "Gene ID         XP_001557974.1\n",
      "CAZy ID        EAA29319.1|GH93\n",
      "% Identical               61.2\n",
      "Length                     456\n",
      "Mismatches                 164\n",
      "Gap Open                     5\n",
      "Gene Start                   1\n",
      "Gene End                   446\n",
      "CAZy Start                   1\n",
      "CAZy End                   453\n",
      "E Value               4.4e-158\n",
      "Bit Score                560.5\n",
      "Name: 458, dtype: object \n",
      "\n",
      "Gene ID            XP_024552087.1\n",
      "CAZy ID        XP_003653167.1|GT0\n",
      "% Identical                  43.1\n",
      "Length                        684\n",
      "Mismatches                    335\n",
      "Gap Open                       12\n",
      "Gene Start                      7\n",
      "Gene End                      656\n",
      "CAZy Start                      4\n",
      "CAZy End                      667\n",
      "E Value                  5.1e-145\n",
      "Bit Score                   517.7\n",
      "Name: 469, dtype: object \n",
      "\n",
      "Gene ID        XP_001558190.1\n",
      "CAZy ID        EAA65303.1|GT0\n",
      "% Identical              69.7\n",
      "Length                    914\n",
      "Mismatches                261\n",
      "Gap Open                    4\n",
      "Gene Start                  1\n",
      "Gene End                  907\n",
      "CAZy Start                677\n",
      "CAZy End                 1581\n",
      "E Value               0.0e+00\n",
      "Bit Score              1243.8\n",
      "Name: 476, dtype: object \n",
      "\n",
      "Gene ID         XP_024552437.1\n",
      "CAZy ID        CBX97514.1|GT57\n",
      "% Identical               34.6\n",
      "Length                    1118\n",
      "Mismatches                 541\n",
      "Gap Open                    23\n",
      "Gene Start                   2\n",
      "Gene End                   970\n",
      "CAZy Start                 673\n",
      "CAZy End                  1749\n",
      "E Value               2.4e-140\n",
      "Bit Score                502.7\n",
      "Name: 492, dtype: object \n",
      "\n",
      "Gene ID         XP_024552436.1\n",
      "CAZy ID        CBX97514.1|GT57\n",
      "% Identical               35.8\n",
      "Length                    1136\n",
      "Mismatches                 579\n",
      "Gap Open                    28\n",
      "Gene Start                   2\n",
      "Gene End                  1046\n",
      "CAZy Start                 673\n",
      "CAZy End                  1749\n",
      "E Value               1.5e-148\n",
      "Bit Score                530.0\n",
      "Name: 493, dtype: object \n",
      "\n",
      "Gene ID          XP_001555351.2\n",
      "CAZy ID        KFY84961.1|GH135\n",
      "% Identical                56.7\n",
      "Length                      383\n",
      "Mismatches                  156\n",
      "Gap Open                      4\n",
      "Gene Start                    1\n",
      "Gene End                    382\n",
      "CAZy Start                    1\n",
      "CAZy End                    374\n",
      "E Value                2.5e-117\n",
      "Bit Score                 424.9\n",
      "Name: 495, dtype: object \n",
      "\n",
      "Gene ID         XP_001554752.1\n",
      "CAZy ID        AVI60826.1|GT22\n",
      "% Identical               87.8\n",
      "Length                     426\n",
      "Mismatches                  52\n",
      "Gap Open                     0\n",
      "Gene Start                  11\n",
      "Gene End                   436\n",
      "CAZy Start                 583\n",
      "CAZy End                  1008\n",
      "E Value               8.7e-220\n",
      "Bit Score                765.4\n",
      "Name: 513, dtype: object \n",
      "\n",
      "Gene ID        XP_001552859.2\n",
      "CAZy ID        BAZ00088.1|GT2\n",
      "% Identical              22.2\n",
      "Length                   3257\n",
      "Mismatches               2019\n",
      "Gap Open                  114\n",
      "Gene Start                704\n",
      "Gene End                 3754\n",
      "CAZy Start                318\n",
      "CAZy End                 3264\n",
      "E Value              4.2e-120\n",
      "Bit Score               437.6\n",
      "Name: 529, dtype: object \n",
      "\n",
      "Gene ID        XP_024552826.1\n",
      "CAZy ID        BAZ00088.1|GT2\n",
      "% Identical              22.6\n",
      "Length                   3033\n",
      "Mismatches               1865\n",
      "Gap Open                  107\n",
      "Gene Start                  7\n",
      "Gene End                 2853\n",
      "CAZy Start                529\n",
      "CAZy End                 3264\n",
      "E Value              1.8e-118\n",
      "Bit Score               431.8\n",
      "Name: 530, dtype: object \n",
      "\n",
      "Gene ID            XP_024552845.1\n",
      "CAZy ID        AIE96983.1|GT0|GT7\n",
      "% Identical                  33.3\n",
      "Length                       1285\n",
      "Mismatches                    704\n",
      "Gap Open                       28\n",
      "Gene Start                     26\n",
      "Gene End                     1253\n",
      "CAZy Start                      7\n",
      "CAZy End                     1195\n",
      "E Value                  1.1e-185\n",
      "Bit Score                   653.7\n",
      "Name: 531, dtype: object \n",
      "\n",
      "Gene ID          XP_001551326.1\n",
      "CAZy ID        CBX98831.1|GH132\n",
      "% Identical                61.3\n",
      "Length                      775\n",
      "Mismatches                  246\n",
      "Gap Open                     10\n",
      "Gene Start                    3\n",
      "Gene End                    733\n",
      "CAZy Start                    2\n",
      "CAZy End                    766\n",
      "E Value                8.4e-260\n",
      "Bit Score                 899.0\n",
      "Name: 541, dtype: object \n",
      "\n",
      "Gene ID                   XP_024553156.1\n",
      "CAZy ID        XP_361931.1|CBM66|GH43_34\n",
      "% Identical                         55.4\n",
      "Length                               368\n",
      "Mismatches                           154\n",
      "Gap Open                               5\n",
      "Gene Start                             4\n",
      "Gene End                             366\n",
      "CAZy Start                           561\n",
      "CAZy End                             923\n",
      "E Value                         1.0e-107\n",
      "Bit Score                          392.9\n",
      "Name: 551, dtype: object \n",
      "\n",
      "Gene ID         XP_001553665.1\n",
      "CAZy ID        AIT18866.1|GH18\n",
      "% Identical               36.9\n",
      "Length                    1041\n",
      "Mismatches                 615\n",
      "Gap Open                    19\n",
      "Gene Start                  11\n",
      "Gene End                  1040\n",
      "CAZy Start                  82\n",
      "CAZy End                  1091\n",
      "E Value               2.3e-192\n",
      "Bit Score                675.6\n",
      "Name: 555, dtype: object \n",
      "\n",
      "Gene ID        XP_024553215.1\n",
      "CAZy ID        BBE14729.1|GT4\n",
      "% Identical              55.7\n",
      "Length                    510\n",
      "Mismatches                207\n",
      "Gap Open                    3\n",
      "Gene Start                 51\n",
      "Gene End                  549\n",
      "CAZy Start                  6\n",
      "CAZy End                  507\n",
      "E Value              1.1e-155\n",
      "Bit Score               552.7\n",
      "Name: 558, dtype: object \n",
      "\n",
      "Gene ID         XP_001546880.1\n",
      "CAZy ID        CBX99002.1|GH18\n",
      "% Identical               82.6\n",
      "Length                     304\n",
      "Mismatches                  49\n",
      "Gap Open                     2\n",
      "Gene Start                   1\n",
      "Gene End                   303\n",
      "CAZy Start                   1\n",
      "CAZy End                   301\n",
      "E Value               5.5e-142\n",
      "Bit Score                506.5\n",
      "Name: 582, dtype: object \n",
      "\n",
      "Gene ID           XP_024553622.1\n",
      "CAZy ID        CBX92400.1|GH30_7\n",
      "% Identical                 37.2\n",
      "Length                       991\n",
      "Mismatches                   476\n",
      "Gap Open                      21\n",
      "Gene Start                     2\n",
      "Gene End                     855\n",
      "CAZy Start                   642\n",
      "CAZy End                    1623\n",
      "E Value                 3.5e-164\n",
      "Bit Score                  581.6\n",
      "Name: 583, dtype: object \n",
      "\n",
      "Gene ID           XP_001546886.1\n",
      "CAZy ID        CBX92400.1|GH30_7\n",
      "% Identical                 37.2\n",
      "Length                       991\n",
      "Mismatches                   476\n",
      "Gap Open                      21\n",
      "Gene Start                     2\n",
      "Gene End                     855\n",
      "CAZy Start                   642\n",
      "CAZy End                    1623\n",
      "E Value                 3.5e-164\n",
      "Bit Score                  581.6\n",
      "Name: 584, dtype: object \n",
      "\n",
      "Gene ID          XP_024553815.1\n",
      "CAZy ID        AGT85755.1|CBM13\n",
      "% Identical                37.8\n",
      "Length                      711\n",
      "Mismatches                  382\n",
      "Gap Open                     16\n",
      "Gene Start                   18\n",
      "Gene End                    705\n",
      "CAZy Start                   44\n",
      "CAZy End                    717\n",
      "E Value                4.7e-117\n",
      "Bit Score                 424.9\n",
      "Name: 590, dtype: object \n",
      "\n",
      "Gene ID          XP_024553814.1\n",
      "CAZy ID        AGT85755.1|CBM13\n",
      "% Identical                37.9\n",
      "Length                      715\n",
      "Mismatches                  380\n",
      "Gap Open                     16\n",
      "Gene Start                   18\n",
      "Gene End                    709\n",
      "CAZy Start                   44\n",
      "CAZy End                    717\n",
      "E Value                3.9e-119\n",
      "Bit Score                 431.8\n",
      "Name: 591, dtype: object \n",
      "\n"
     ]
    }
   ],
   "source": [
    "# check if CAZy end ever differs to length\n",
    "index = 0\n",
    "for index in range(len(diamond_df[\"Gene ID\"])):\n",
    "    if int(diamond_df[\"Length\"].iloc[index]) != int(diamond_df[\"CAZy End\"].iloc[index]):\n",
    "        print(diamond_df.iloc[index], \"\\n\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Using XP_024546124.1 as an example. The protien sequence is 695 residues long. Therefore, waht is meant by the CAZy start and end being 623 and 1215 respectively?\n",
    "\n",
    "Therefore, the CAZy start and end do not appear to be informative when protein sequences are the input sequence type for dbCAN.\n",
    "\n",
    "Consequently, data for Diamond need only be retrieved from the overview.txt file."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**HMMER**\n",
    "\n",
    "First, retrieve the output from HMMER. The desired data is all predicated CAZy familes/domains and their respective domain ranges.\n",
    "\n",
    "The text in the overview.txt file is separated by tabs (`\\t`) thus allowing the data for each prediction tool to be easily separated out. The data is presented in the following order:\n",
    "0. Gene ID (protein name)\n",
    "1. HMMER\n",
    "2. Hotpep\n",
    "3. Diamond\n",
    "4. #ofTools (number of tools that predicated the protein sequence contained at least one CAZy domain)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['XP_001545265.2', 'GH53(20-346)', 'GH53(7)', 'GH53', '3']\n",
      "['XP_001545514.1', 'GT2_Chitin_synth_2(1213-1720)', 'GT2(161)', 'GT2', '3']\n",
      "['XP_001545605.1', 'AA4(28-577)', 'AA4(1)', 'AA4', '3']\n",
      "['XP_001545700.1', 'CBM21(353-459)', 'CBM21(4)', 'CBM21', '3']\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene ID</th>\n",
       "      <th>HMMER</th>\n",
       "      <th>Hotpep</th>\n",
       "      <th>DIAMOND</th>\n",
       "      <th>#ofTools</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>XP_001545265.2</td>\n",
       "      <td>GH53(20-346)</td>\n",
       "      <td>GH53(7)</td>\n",
       "      <td>GH53</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>XP_001545514.1</td>\n",
       "      <td>GT2_Chitin_synth_2(1213-1720)</td>\n",
       "      <td>GT2(161)</td>\n",
       "      <td>GT2</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>XP_001545605.1</td>\n",
       "      <td>AA4(28-577)</td>\n",
       "      <td>AA4(1)</td>\n",
       "      <td>AA4</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>XP_001545700.1</td>\n",
       "      <td>CBM21(353-459)</td>\n",
       "      <td>CBM21(4)</td>\n",
       "      <td>CBM21</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          Gene ID                          HMMER    Hotpep DIAMOND #ofTools\n",
       "0  XP_001545265.2                   GH53(20-346)   GH53(7)    GH53        3\n",
       "1  XP_001545514.1  GT2_Chitin_synth_2(1213-1720)  GT2(161)     GT2        3\n",
       "2  XP_001545605.1                    AA4(28-577)    AA4(1)     AA4        3\n",
       "3  XP_001545700.1                 CBM21(353-459)  CBM21(4)   CBM21        3"
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Check how text is separated in overview.txt\n",
    "with open (\"overview.txt\", \"r\") as ofh:\n",
    "    overview_file = ofh.read().splitlines()\n",
    "\n",
    "data_ov = []\n",
    "for line in overview_file[1:5]:  # skip the first line becuase this is the head titles\n",
    "    line = line.split(\"\\t\")\n",
    "    print(line)\n",
    "    data_ov.append(line)\n",
    "\n",
    "# or better presented as a df\n",
    "df = pd.DataFrame(data_ov, columns=[\"Gene ID\", \"HMMER\", \"Hotpep\", \"DIAMOND\", \"#ofTools\"])\n",
    "df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "When multiple domains are predicated by HMMER the domains are separated by '+' and the CAZy family is immediately followed by its domain range within parenthesises.\n",
    "\n",
    "Sometimes proteins predicated as containing GT2 domains are given the CAZy family \"GT2_Chitin_synth_2\". Why this is I am not sure, but there is not GT2_2 subfamily. Therefore, the names of the predicted CAZy families need to be standardised to GT2.\n",
    "\n",
    "Additionally, when no CAZy domains are predicated to be within the protein sequence a result of '-' is provided. It will be better for the final dataframe if this is converted to a try null value.\n",
    "\n",
    "Lastly, sometimes HMMER predicates the CAZy subfamily. Unlike with eCAMI and CUPP the subfamily is not passed to a separate column, although this is helpful when it comes to evaluating the tools performances. Therefore, when HMMER predicates the CAZy subfamily this result needs to be passed to retrieve the CAZy family and then store the CAZy family and subfamily separately."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "hmmer_prediction= ['GH53(20-346)'] familes= ['GH53'] subfams [nan] domain_ranges= ['20-346']\n",
      "hmmer_prediction= ['GT2_Chitin_synth_2(1213-1720)'] familes= ['GT2'] subfams [nan] domain_ranges= ['1213-1720']\n",
      "hmmer_prediction= ['AA4(28-577)'] familes= ['AA4'] subfams [nan] domain_ranges= ['28-577']\n",
      "hmmer_prediction= ['CBM21(353-459)'] familes= ['CBM21'] subfams [nan] domain_ranges= ['353-459']\n",
      "hmmer_prediction= ['GH132(57-318)'] familes= ['GH132'] subfams [nan] domain_ranges= ['57-318']\n",
      "hmmer_prediction= ['GT32(85-165)'] familes= ['GT32'] subfams [nan] domain_ranges= ['85-165']\n",
      "hmmer_prediction= ['GT25(58-285)'] familes= ['GT25'] subfams [nan] domain_ranges= ['58-285']\n",
      "hmmer_prediction= ['GH17(442-672)'] familes= ['GH17'] subfams [nan] domain_ranges= ['442-672']\n",
      "hmmer_prediction= ['GT71(136-358)'] familes= ['GT71'] subfams [nan] domain_ranges= ['136-358']\n",
      "hmmer_prediction= ['GH16(87-284)'] familes= ['GH16'] subfams [nan] domain_ranges= ['87-284']\n",
      "hmmer_prediction= ['AA2(84-274)'] familes= ['AA2'] subfams [nan] domain_ranges= ['84-274']\n",
      "hmmer_prediction= ['GH18(31-309)'] familes= ['GH18'] subfams [nan] domain_ranges= ['31-309']\n",
      "hmmer_prediction= ['CE10(54-319)'] familes= ['CE10'] subfams [nan] domain_ranges= ['54-319']\n",
      "hmmer_prediction= ['CE4(61-165)'] familes= ['CE4'] subfams [nan] domain_ranges= ['61-165']\n",
      "hmmer_prediction= ['GH131(22-260)'] familes= ['GH131'] subfams [nan] domain_ranges= ['22-260']\n",
      "hmmer_prediction= ['CE10(46-321)'] familes= ['CE10'] subfams [nan] domain_ranges= ['46-321']\n",
      "hmmer_prediction= ['CE5(41-239)'] familes= ['CE5'] subfams [nan] domain_ranges= ['41-239']\n",
      "hmmer_prediction= ['GT90(140-405)'] familes= ['GT90'] subfams [nan] domain_ranges= ['140-405']\n",
      "hmmer_prediction= ['GH105(43-384)'] familes= ['GH105'] subfams [nan] domain_ranges= ['43-384']\n",
      "hmmer_prediction= ['PL1_4(116-301)'] familes= ['PL1'] subfams ['PL1_4'] domain_ranges= ['116-301']\n",
      "hmmer_prediction= ['AA3_1(26-554)'] familes= ['AA3'] subfams ['AA3_1'] domain_ranges= ['26-554']\n",
      "hmmer_prediction= ['AA1_3(195-505)'] familes= ['AA1'] subfams ['AA1_3'] domain_ranges= ['195-505']\n",
      "hmmer_prediction= ['CE10(36-344)'] familes= ['CE10'] subfams [nan] domain_ranges= ['36-344']\n",
      "hmmer_prediction= ['AA7(35-262)'] familes= ['AA7'] subfams [nan] domain_ranges= ['35-262']\n",
      "hmmer_prediction= ['CE12(21-233)'] familes= ['CE12'] subfams [nan] domain_ranges= ['21-233']\n",
      "hmmer_prediction= ['AA3_2(45-669)'] familes= ['AA3'] subfams ['AA3_2'] domain_ranges= ['45-669']\n",
      "hmmer_prediction= ['GH72(25-331)', 'CBM43(382-468)'] familes= ['GH72', 'CBM43'] subfams [nan, nan] domain_ranges= ['25-331', '382-468']\n",
      "hmmer_prediction= ['GH28(62-412)'] familes= ['GH28'] subfams [nan] domain_ranges= ['62-412']\n",
      "hmmer_prediction= ['GH93(44-356)'] familes= ['GH93'] subfams [nan] domain_ranges= ['44-356']\n",
      "hmmer_prediction= ['GT15(74-352)'] familes= ['GT15'] subfams [nan] domain_ranges= ['74-352']\n",
      "hmmer_prediction= ['GH16(81-329)'] familes= ['GH16'] subfams [nan] domain_ranges= ['81-329']\n",
      "hmmer_prediction= ['GH5_23(158-436)'] familes= ['GH5'] subfams ['GH5_23'] domain_ranges= ['158-436']\n",
      "hmmer_prediction= ['GH45(26-230)'] familes= ['GH45'] subfams [nan] domain_ranges= ['26-230']\n",
      "hmmer_prediction= ['CE5(48-239)'] familes= ['CE5'] subfams [nan] domain_ranges= ['48-239']\n",
      "hmmer_prediction= ['GH152(177-285)'] familes= ['GH152'] subfams [nan] domain_ranges= ['177-285']\n",
      "hmmer_prediction= ['GH16(71-314)'] familes= ['GH16'] subfams [nan] domain_ranges= ['71-314']\n",
      "hmmer_prediction= ['CBM67(151-322)', 'GH78(351-870)'] familes= ['CBM67', 'GH78'] subfams [nan, nan] domain_ranges= ['151-322', '351-870']\n",
      "hmmer_prediction= ['GH11(57-218)'] familes= ['GH11'] subfams [nan] domain_ranges= ['57-218']\n",
      "hmmer_prediction= ['GH16(83-297)'] familes= ['GH16'] subfams [nan] domain_ranges= ['83-297']\n",
      "hmmer_prediction= ['GT2_Glyco_tranf_2_3(49-284)'] familes= ['GT2'] subfams ['GT2_Glyco_tranf_2_3'] domain_ranges= ['49-284']\n",
      "hmmer_prediction= ['GH64(48-409)'] familes= ['GH64'] subfams [nan] domain_ranges= ['48-409']\n",
      "hmmer_prediction= ['AA4(122-310)'] familes= ['AA4'] subfams [nan] domain_ranges= ['122-310']\n",
      "hmmer_prediction= ['GT1(352-660)'] familes= ['GT1'] subfams [nan] domain_ranges= ['352-660']\n",
      "hmmer_prediction= ['AA3_2(27-612)'] familes= ['AA3'] subfams ['AA3_2'] domain_ranges= ['27-612']\n",
      "hmmer_prediction= ['GH71(258-659)'] familes= ['GH71'] subfams [nan] domain_ranges= ['258-659']\n",
      "hmmer_prediction= ['GH31(312-867)'] familes= ['GH31'] subfams [nan] domain_ranges= ['312-867']\n",
      "hmmer_prediction= ['GH145(21-349)'] familes= ['GH145'] subfams [nan] domain_ranges= ['21-349']\n",
      "hmmer_prediction= ['CE12(30-222)'] familes= ['CE12'] subfams [nan] domain_ranges= ['30-222']\n",
      "hmmer_prediction= ['GH28(86-438)'] familes= ['GH28'] subfams [nan] domain_ranges= ['86-438']\n",
      "hmmer_prediction= ['AA9(7-209)'] familes= ['AA9'] subfams [nan] domain_ranges= ['7-209']\n",
      "hmmer_prediction= ['GH115(191-812)'] familes= ['GH115'] subfams [nan] domain_ranges= ['191-812']\n",
      "hmmer_prediction= ['GT39(55-300)'] familes= ['GT39'] subfams [nan] domain_ranges= ['55-300']\n",
      "hmmer_prediction= ['GT2_Chitin_synth_1(215-385)'] familes= ['GT2'] subfams [nan] domain_ranges= ['215-385']\n",
      "hmmer_prediction= ['AA3_2(8-571)'] familes= ['AA3'] subfams ['AA3_2'] domain_ranges= ['8-571']\n",
      "hmmer_prediction= ['GT25(82-284)'] familes= ['GT25'] subfams [nan] domain_ranges= ['82-284']\n",
      "hmmer_prediction= ['AA7(37-463)'] familes= ['AA7'] subfams [nan] domain_ranges= ['37-463']\n",
      "hmmer_prediction= ['PL1_4(116-300)'] familes= ['PL1'] subfams ['PL1_4'] domain_ranges= ['116-300']\n",
      "hmmer_prediction= ['AA7(61-265)'] familes= ['AA7'] subfams [nan] domain_ranges= ['61-265']\n",
      "hmmer_prediction= ['GH16(124-449)'] familes= ['GH16'] subfams [nan] domain_ranges= ['124-449']\n",
      "hmmer_prediction= ['GH5_9(86-386)'] familes= ['GH5'] subfams ['GH5_9'] domain_ranges= ['86-386']\n",
      "hmmer_prediction= ['GH132(159-469)'] familes= ['GH132'] subfams [nan] domain_ranges= ['159-469']\n",
      "hmmer_prediction= ['GH17(55-294)'] familes= ['GH17'] subfams [nan] domain_ranges= ['55-294']\n",
      "hmmer_prediction= ['GH28(58-401)'] familes= ['GH28'] subfams [nan] domain_ranges= ['58-401']\n",
      "hmmer_prediction= ['GH152(81-363)'] familes= ['GH152'] subfams [nan] domain_ranges= ['81-363']\n",
      "hmmer_prediction= ['CE10(163-305)'] familes= ['CE10'] subfams [nan] domain_ranges= ['163-305']\n",
      "hmmer_prediction= ['PL1_4(118-310)'] familes= ['PL1'] subfams ['PL1_4'] domain_ranges= ['118-310']\n",
      "hmmer_prediction= ['AA2(21-265)'] familes= ['AA2'] subfams [nan] domain_ranges= ['21-265']\n",
      "hmmer_prediction= ['GT8(157-386)'] familes= ['GT8'] subfams [nan] domain_ranges= ['157-386']\n",
      "hmmer_prediction= ['GH1(13-444)'] familes= ['GH1'] subfams [nan] domain_ranges= ['13-444']\n",
      "hmmer_prediction= ['AA1_3(53-371)'] familes= ['AA1'] subfams ['AA1_3'] domain_ranges= ['53-371']\n",
      "hmmer_prediction= ['PL1_4(109-296)'] familes= ['PL1'] subfams ['PL1_4'] domain_ranges= ['109-296']\n",
      "hmmer_prediction= ['AA11(21-213)'] familes= ['AA11'] subfams [nan] domain_ranges= ['21-213']\n",
      "hmmer_prediction= ['GH28(60-370)'] familes= ['GH28'] subfams [nan] domain_ranges= ['60-370']\n",
      "hmmer_prediction= ['CE8(31-304)'] familes= ['CE8'] subfams [nan] domain_ranges= ['31-304']\n",
      "hmmer_prediction= ['CE10(127-351)'] familes= ['CE10'] subfams [nan] domain_ranges= ['127-351']\n",
      "hmmer_prediction= ['GH76(27-385)'] familes= ['GH76'] subfams [nan] domain_ranges= ['27-385']\n",
      "hmmer_prediction= ['GH62(20-290)'] familes= ['GH62'] subfams [nan] domain_ranges= ['20-290']\n",
      "hmmer_prediction= ['AA1_3(1-124)'] familes= ['AA1'] subfams ['AA1_3'] domain_ranges= ['1-124']\n",
      "hmmer_prediction= ['GH125(92-524)'] familes= ['GH125'] subfams [nan] domain_ranges= ['92-524']\n",
      "hmmer_prediction= ['GT2_Chitin_synth_1(479-641)', 'GT2_Chitin_synth_2(618-883)'] familes= ['GT2', 'GT2'] subfams [nan, nan] domain_ranges= ['479-641', '618-883']\n",
      "hmmer_prediction= ['CE10(122-358)'] familes= ['CE10'] subfams [nan] domain_ranges= ['122-358']\n",
      "hmmer_prediction= ['GH47(57-598)'] familes= ['GH47'] subfams [nan] domain_ranges= ['57-598']\n",
      "hmmer_prediction= ['GH5_49(114-417)'] familes= ['GH5'] subfams ['GH5_49'] domain_ranges= ['114-417']\n",
      "hmmer_prediction= ['GH16(138-465)'] familes= ['GH16'] subfams [nan] domain_ranges= ['138-465']\n",
      "hmmer_prediction= ['GH72(33-331)'] familes= ['GH72'] subfams [nan] domain_ranges= ['33-331']\n",
      "hmmer_prediction= ['GH7(18-449)'] familes= ['GH7'] subfams [nan] domain_ranges= ['18-449']\n",
      "hmmer_prediction= ['GH154(24-373)'] familes= ['GH154'] subfams [nan] domain_ranges= ['24-373']\n",
      "hmmer_prediction= ['AA3(104-694)'] familes= ['AA3'] subfams [nan] domain_ranges= ['104-694']\n",
      "hmmer_prediction= ['GH71(23-411)'] familes= ['GH71'] subfams [nan] domain_ranges= ['23-411']\n",
      "hmmer_prediction= ['GH43_6(28-317)'] familes= ['GH43'] subfams ['GH43_6'] domain_ranges= ['28-317']\n",
      "hmmer_prediction= ['AA1_3(129-438)'] familes= ['AA1'] subfams ['AA1_3'] domain_ranges= ['129-438']\n",
      "hmmer_prediction= ['CBM35(28-144)', 'GH26(162-462)'] familes= ['CBM35', 'GH26'] subfams [nan, nan] domain_ranges= ['28-144', '162-462']\n",
      "hmmer_prediction= ['GH51(14-502)'] familes= ['GH51'] subfams [nan] domain_ranges= ['14-502']\n",
      "hmmer_prediction= ['GH3(106-327)'] familes= ['GH3'] subfams [nan] domain_ranges= ['106-327']\n",
      "hmmer_prediction= ['PL3_2(133-323)'] familes= ['PL3'] subfams ['PL3_2'] domain_ranges= ['133-323']\n",
      "hmmer_prediction= ['GH3(110-324)'] familes= ['GH3'] subfams [nan] domain_ranges= ['110-324']\n",
      "hmmer_prediction= ['CE2(171-402)'] familes= ['CE2'] subfams [nan] domain_ranges= ['171-402']\n",
      "hmmer_prediction= ['CE10(73-354)'] familes= ['CE10'] subfams [nan] domain_ranges= ['73-354']\n",
      "hmmer_prediction= ['GH16(64-335)'] familes= ['GH16'] subfams [nan] domain_ranges= ['64-335']\n",
      "hmmer_prediction= ['GH3(169-381)'] familes= ['GH3'] subfams [nan] domain_ranges= ['169-381']\n",
      "hmmer_prediction= ['GH5_5(23-327)'] familes= ['GH5'] subfams ['GH5_5'] domain_ranges= ['23-327']\n",
      "hmmer_prediction= ['CE10(67-282)'] familes= ['CE10'] subfams [nan] domain_ranges= ['67-282']\n",
      "hmmer_prediction= ['GH2(12-770)'] familes= ['GH2'] subfams [nan] domain_ranges= ['12-770']\n",
      "hmmer_prediction= ['GH47(106-630)'] familes= ['GH47'] subfams [nan] domain_ranges= ['106-630']\n",
      "hmmer_prediction= ['CBM67(11-191)', 'GH78(265-767)'] familes= ['CBM67', 'GH78'] subfams [nan, nan] domain_ranges= ['11-191', '265-767']\n",
      "hmmer_prediction= ['AA3_2(10-616)'] familes= ['AA3'] subfams ['AA3_2'] domain_ranges= ['10-616']\n",
      "hmmer_prediction= ['CE10(161-428)'] familes= ['CE10'] subfams [nan] domain_ranges= ['161-428']\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "hmmer_prediction= ['GH1(166-661)'] familes= ['GH1'] subfams [nan] domain_ranges= ['166-661']\n",
      "hmmer_prediction= ['GH17(24-303)'] familes= ['GH17'] subfams [nan] domain_ranges= ['24-303']\n",
      "hmmer_prediction= ['GH51(182-653)'] familes= ['GH51'] subfams [nan] domain_ranges= ['182-653']\n",
      "hmmer_prediction= ['GT90(349-621)'] familes= ['GT90'] subfams [nan] domain_ranges= ['349-621']\n",
      "hmmer_prediction= ['GH17(82-309)'] familes= ['GH17'] subfams [nan] domain_ranges= ['82-309']\n",
      "hmmer_prediction= ['GH28(21-361)'] familes= ['GH28'] subfams [nan] domain_ranges= ['21-361']\n",
      "hmmer_prediction= ['GH13_31(38-394)'] familes= ['GH13'] subfams ['GH13_31'] domain_ranges= ['38-394']\n",
      "hmmer_prediction= ['GT24(1186-1433)'] familes= ['GT24'] subfams [nan] domain_ranges= ['1186-1433']\n",
      "hmmer_prediction= ['GH6(125-418)'] familes= ['GH6'] subfams [nan] domain_ranges= ['125-418']\n",
      "hmmer_prediction= ['GH5_5(89-384)'] familes= ['GH5'] subfams ['GH5_5'] domain_ranges= ['89-384']\n",
      "hmmer_prediction= ['PL7_4(71-286)'] familes= ['PL7'] subfams ['PL7_4'] domain_ranges= ['71-286']\n",
      "hmmer_prediction= ['GT15(113-422)'] familes= ['GT15'] subfams [nan] domain_ranges= ['113-422']\n",
      "hmmer_prediction= ['GH16(40-284)'] familes= ['GH16'] subfams [nan] domain_ranges= ['40-284']\n",
      "hmmer_prediction= ['GH95(22-784)'] familes= ['GH95'] subfams [nan] domain_ranges= ['22-784']\n",
      "hmmer_prediction= ['AA1_3(56-369)'] familes= ['AA1'] subfams ['AA1_3'] domain_ranges= ['56-369']\n",
      "hmmer_prediction= ['CE10(94-248)'] familes= ['CE10'] subfams [nan] domain_ranges= ['94-248']\n",
      "hmmer_prediction= ['AA9(7-230)'] familes= ['AA9'] subfams [nan] domain_ranges= ['7-230']\n",
      "hmmer_prediction= ['AA9(11-237)'] familes= ['AA9'] subfams [nan] domain_ranges= ['11-237']\n",
      "hmmer_prediction= ['GH63(596-807)'] familes= ['GH63'] subfams [nan] domain_ranges= ['596-807']\n",
      "hmmer_prediction= ['GH13_22(99-497)', 'GH13_22(1174-1637)'] familes= ['GH13', 'GH13'] subfams ['GH13_22', 'GH13_22'] domain_ranges= ['99-497', '1174-1637']\n",
      "hmmer_prediction= ['GH13_1(72-377)'] familes= ['GH13'] subfams ['GH13_1'] domain_ranges= ['72-377']\n",
      "hmmer_prediction= ['AA3_2(8-546)'] familes= ['AA3'] subfams ['AA3_2'] domain_ranges= ['8-546']\n",
      "hmmer_prediction= ['CE5(33-233)'] familes= ['CE5'] subfams [nan] domain_ranges= ['33-233']\n",
      "hmmer_prediction= ['CE4(106-233)'] familes= ['CE4'] subfams [nan] domain_ranges= ['106-233']\n",
      "hmmer_prediction= ['GH106(50-741)'] familes= ['GH106'] subfams [nan] domain_ranges= ['50-741']\n",
      "hmmer_prediction= ['GH128(244-465)'] familes= ['GH128'] subfams [nan] domain_ranges= ['244-465']\n",
      "hmmer_prediction= ['CE5(63-231)'] familes= ['CE5'] subfams [nan] domain_ranges= ['63-231']\n",
      "hmmer_prediction= ['GH47(218-993)'] familes= ['GH47'] subfams [nan] domain_ranges= ['218-993']\n",
      "hmmer_prediction= ['GT71(150-385)'] familes= ['GT71'] subfams [nan] domain_ranges= ['150-385']\n",
      "hmmer_prediction= ['AA8(11-844)'] familes= ['AA8'] subfams [nan] domain_ranges= ['11-844']\n",
      "hmmer_prediction= ['GT17(49-340)'] familes= ['GT17'] subfams [nan] domain_ranges= ['49-340']\n",
      "hmmer_prediction= ['GH16(60-297)'] familes= ['GH16'] subfams [nan] domain_ranges= ['60-297']\n",
      "hmmer_prediction= ['GT34(162-423)'] familes= ['GT34'] subfams [nan] domain_ranges= ['162-423']\n",
      "hmmer_prediction= ['GH5_22(49-336)'] familes= ['GH5'] subfams ['GH5_22'] domain_ranges= ['49-336']\n",
      "hmmer_prediction= ['GH128(52-278)'] familes= ['GH128'] subfams [nan] domain_ranges= ['52-278']\n",
      "hmmer_prediction= ['AA7(61-277)'] familes= ['AA7'] subfams [nan] domain_ranges= ['61-277']\n",
      "hmmer_prediction= ['GH28(62-425)'] familes= ['GH28'] subfams [nan] domain_ranges= ['62-425']\n",
      "hmmer_prediction= ['PL1_4(125-306)'] familes= ['PL1'] subfams ['PL1_4'] domain_ranges= ['125-306']\n",
      "hmmer_prediction= ['GT32(127-208)'] familes= ['GT32'] subfams [nan] domain_ranges= ['127-208']\n",
      "hmmer_prediction= ['AA3_2(13-541)'] familes= ['AA3'] subfams ['AA3_2'] domain_ranges= ['13-541']\n",
      "hmmer_prediction= ['GH20(203-548)'] familes= ['GH20'] subfams [nan] domain_ranges= ['203-548']\n",
      "hmmer_prediction= ['GT1(126-447)'] familes= ['GT1'] subfams [nan] domain_ranges= ['126-447']\n",
      "hmmer_prediction= ['GH55(32-772)'] familes= ['GH55'] subfams [nan] domain_ranges= ['32-772']\n",
      "hmmer_prediction= ['CE10(10-295)'] familes= ['CE10'] subfams [nan] domain_ranges= ['10-295']\n",
      "hmmer_prediction= ['GT8(19-259)'] familes= ['GT8'] subfams [nan] domain_ranges= ['19-259']\n",
      "hmmer_prediction= ['GH2(29-937)'] familes= ['GH2'] subfams [nan] domain_ranges= ['29-937']\n",
      "hmmer_prediction= ['GH71(47-435)'] familes= ['GH71'] subfams [nan] domain_ranges= ['47-435']\n",
      "hmmer_prediction= ['PL3_2(31-221)'] familes= ['PL3'] subfams ['PL3_2'] domain_ranges= ['31-221']\n",
      "hmmer_prediction= ['GH78(393-790)'] familes= ['GH78'] subfams [nan] domain_ranges= ['393-790']\n",
      "hmmer_prediction= ['GH3(81-306)'] familes= ['GH3'] subfams [nan] domain_ranges= ['81-306']\n",
      "hmmer_prediction= ['GH76(31-396)'] familes= ['GH76'] subfams [nan] domain_ranges= ['31-396']\n",
      "hmmer_prediction= ['GT90(169-402)'] familes= ['GT90'] subfams [nan] domain_ranges= ['169-402']\n",
      "hmmer_prediction= ['CE5(83-253)'] familes= ['CE5'] subfams [nan] domain_ranges= ['83-253']\n",
      "hmmer_prediction= ['AA5_2(76-680)'] familes= ['AA5'] subfams ['AA5_2'] domain_ranges= ['76-680']\n",
      "hmmer_prediction= ['GT71(131-352)'] familes= ['GT71'] subfams [nan] domain_ranges= ['131-352']\n",
      "hmmer_prediction= ['GH13_1(54-364)'] familes= ['GH13'] subfams ['GH13_1'] domain_ranges= ['54-364']\n",
      "hmmer_prediction= ['CE4(145-269)'] familes= ['CE4'] subfams [nan] domain_ranges= ['145-269']\n",
      "hmmer_prediction= ['GT1(199-481)'] familes= ['GT1'] subfams [nan] domain_ranges= ['199-481']\n",
      "hmmer_prediction= ['AA7(77-505)'] familes= ['AA7'] subfams [nan] domain_ranges= ['77-505']\n",
      "hmmer_prediction= ['GH7(19-452)'] familes= ['GH7'] subfams [nan] domain_ranges= ['19-452']\n",
      "hmmer_prediction= ['GH43_37(53-289)'] familes= ['GH43'] subfams ['GH43_37'] domain_ranges= ['53-289']\n",
      "hmmer_prediction= ['GH28(163-445)'] familes= ['GH28'] subfams [nan] domain_ranges= ['163-445']\n",
      "hmmer_prediction= ['AA7(155-351)'] familes= ['AA7'] subfams [nan] domain_ranges= ['155-351']\n",
      "hmmer_prediction= ['GH16(79-300)'] familes= ['GH16'] subfams [nan] domain_ranges= ['79-300']\n",
      "hmmer_prediction= ['GH76(129-404)'] familes= ['GH76'] subfams [nan] domain_ranges= ['129-404']\n",
      "hmmer_prediction= ['CE10(16-193)'] familes= ['CE10'] subfams [nan] domain_ranges= ['16-193']\n",
      "hmmer_prediction= ['GH18(11-234)'] familes= ['GH18'] subfams [nan] domain_ranges= ['11-234']\n",
      "hmmer_prediction= ['GT21(54-369)'] familes= ['GT21'] subfams [nan] domain_ranges= ['54-369']\n",
      "hmmer_prediction= ['GT90(334-624)'] familes= ['GT90'] subfams [nan] domain_ranges= ['334-624']\n",
      "hmmer_prediction= ['CE16(33-291)'] familes= ['CE16'] subfams [nan] domain_ranges= ['33-291']\n",
      "hmmer_prediction= ['GT2_Glyco_trans_2_3(273-467)'] familes= ['GT2'] subfams ['GT2_Glyco_trans_2_3'] domain_ranges= ['273-467']\n",
      "hmmer_prediction= ['GH76(33-388)'] familes= ['GH76'] subfams [nan] domain_ranges= ['33-388']\n",
      "hmmer_prediction= ['GT62(76-339)'] familes= ['GT62'] subfams [nan] domain_ranges= ['76-339']\n",
      "hmmer_prediction= ['GH12(139-272)'] familes= ['GH12'] subfams [nan] domain_ranges= ['139-272']\n",
      "hmmer_prediction= ['GH16(106-269)'] familes= ['GH16'] subfams [nan] domain_ranges= ['106-269']\n",
      "hmmer_prediction= ['AA2(61-243)'] familes= ['AA2'] subfams [nan] domain_ranges= ['61-243']\n",
      "hmmer_prediction= ['GH54(28-343)', 'CBM42(361-501)'] familes= ['GH54', 'CBM42'] subfams [nan, nan] domain_ranges= ['28-343', '361-501']\n",
      "hmmer_prediction= ['CE5(34-201)'] familes= ['CE5'] subfams [nan] domain_ranges= ['34-201']\n",
      "hmmer_prediction= ['GH13_25(1082-1535)'] familes= ['GH13'] subfams ['GH13_25'] domain_ranges= ['1082-1535']\n",
      "hmmer_prediction= ['GH53(30-360)'] familes= ['GH53'] subfams [nan] domain_ranges= ['30-360']\n",
      "hmmer_prediction= ['GT2_Glycos_transf_2(11-184)'] familes= ['GT2'] subfams ['GT2_Glycos_transf_2'] domain_ranges= ['11-184']\n",
      "hmmer_prediction= ['AA3_2(38-608)'] familes= ['AA3'] subfams ['AA3_2'] domain_ranges= ['38-608']\n",
      "hmmer_prediction= ['AA7(137-567)'] familes= ['AA7'] subfams [nan] domain_ranges= ['137-567']\n",
      "hmmer_prediction= ['CE10(93-232)'] familes= ['CE10'] subfams [nan] domain_ranges= ['93-232']\n",
      "hmmer_prediction= ['GT2_Chitin_synth_1(218-386)'] familes= ['GT2'] subfams [nan] domain_ranges= ['218-386']\n",
      "hmmer_prediction= ['GH31(211-665)'] familes= ['GH31'] subfams [nan] domain_ranges= ['211-665']\n",
      "hmmer_prediction= ['CE16(203-456)'] familes= ['CE16'] subfams [nan] domain_ranges= ['203-456']\n",
      "hmmer_prediction= ['GH28(71-376)'] familes= ['GH28'] subfams [nan] domain_ranges= ['71-376']\n",
      "hmmer_prediction= ['AA5(7-502)'] familes= ['AA5'] subfams [nan] domain_ranges= ['7-502']\n",
      "hmmer_prediction= ['GH15(81-493)', 'CBM20(573-663)'] familes= ['GH15', 'CBM20'] subfams [nan, nan] domain_ranges= ['81-493', '573-663']\n",
      "hmmer_prediction= ['GH76(60-391)'] familes= ['GH76'] subfams [nan] domain_ranges= ['60-391']\n",
      "hmmer_prediction= ['CE10(100-272)'] familes= ['CE10'] subfams [nan] domain_ranges= ['100-272']\n",
      "hmmer_prediction= ['GT32(180-257)'] familes= ['GT32'] subfams [nan] domain_ranges= ['180-257']\n",
      "hmmer_prediction= ['GH28(68-376)'] familes= ['GH28'] subfams [nan] domain_ranges= ['68-376']\n",
      "hmmer_prediction= ['GH5_7(74-355)'] familes= ['GH5'] subfams ['GH5_7'] domain_ranges= ['74-355']\n",
      "hmmer_prediction= ['GH35(49-384)'] familes= ['GH35'] subfams [nan] domain_ranges= ['49-384']\n",
      "hmmer_prediction= ['GH11(43-218)'] familes= ['GH11'] subfams [nan] domain_ranges= ['43-218']\n",
      "hmmer_prediction= ['AA3_2(11-575)'] familes= ['AA3'] subfams ['AA3_2'] domain_ranges= ['11-575']\n",
      "hmmer_prediction= ['GH3(109-307)'] familes= ['GH3'] subfams [nan] domain_ranges= ['109-307']\n",
      "hmmer_prediction= ['CBM24(177-252)'] familes= ['CBM24'] subfams [nan] domain_ranges= ['177-252']\n",
      "hmmer_prediction= ['AA8(2-780)'] familes= ['AA8'] subfams [nan] domain_ranges= ['2-780']\n",
      "hmmer_prediction= ['GT32(97-174)'] familes= ['GT32'] subfams [nan] domain_ranges= ['97-174']\n",
      "hmmer_prediction= ['GH16(61-226)'] familes= ['GH16'] subfams [nan] domain_ranges= ['61-226']\n",
      "hmmer_prediction= ['CE5(30-198)'] familes= ['CE5'] subfams [nan] domain_ranges= ['30-198']\n",
      "hmmer_prediction= ['AA1_2(183-544)'] familes= ['AA1'] subfams ['AA1_2'] domain_ranges= ['183-544']\n",
      "hmmer_prediction= ['AA7(35-467)'] familes= ['AA7'] subfams [nan] domain_ranges= ['35-467']\n",
      "hmmer_prediction= ['GT1(273-478)'] familes= ['GT1'] subfams [nan] domain_ranges= ['273-478']\n",
      "hmmer_prediction= ['GT69(119-393)'] familes= ['GT69'] subfams [nan] domain_ranges= ['119-393']\n",
      "hmmer_prediction= ['GH105(47-369)'] familes= ['GH105'] subfams [nan] domain_ranges= ['47-369']\n",
      "hmmer_prediction= ['GH5_16(40-374)'] familes= ['GH5'] subfams ['GH5_16'] domain_ranges= ['40-374']\n",
      "hmmer_prediction= ['AA3_2(20-600)'] familes= ['AA3'] subfams ['AA3_2'] domain_ranges= ['20-600']\n",
      "hmmer_prediction= ['AA1_2(45-381)'] familes= ['AA1'] subfams ['AA1_2'] domain_ranges= ['45-381']\n",
      "hmmer_prediction= ['CE10(101-257)'] familes= ['CE10'] subfams [nan] domain_ranges= ['101-257']\n",
      "hmmer_prediction= ['GH128(192-406)'] familes= ['GH128'] subfams [nan] domain_ranges= ['192-406']\n",
      "hmmer_prediction= ['GT66(23-675)'] familes= ['GT66'] subfams [nan] domain_ranges= ['23-675']\n",
      "hmmer_prediction= ['AA9(7-231)'] familes= ['AA9'] subfams [nan] domain_ranges= ['7-231']\n",
      "hmmer_prediction= ['GH43_1(6-319)'] familes= ['GH43'] subfams ['GH43_1'] domain_ranges= ['6-319']\n",
      "hmmer_prediction= ['GH45(27-231)'] familes= ['GH45'] subfams [nan] domain_ranges= ['27-231']\n",
      "hmmer_prediction= ['GH27(118-365)'] familes= ['GH27'] subfams [nan] domain_ranges= ['118-365']\n",
      "hmmer_prediction= ['GT71(155-384)'] familes= ['GT71'] subfams [nan] domain_ranges= ['155-384']\n",
      "hmmer_prediction= ['GH64(32-405)'] familes= ['GH64'] subfams [nan] domain_ranges= ['32-405']\n",
      "hmmer_prediction= ['AA13(25-254)', 'CBM20(303-393)'] familes= ['AA13', 'CBM20'] subfams [nan, nan] domain_ranges= ['25-254', '303-393']\n",
      "hmmer_prediction= ['GH3(110-339)'] familes= ['GH3'] subfams [nan] domain_ranges= ['110-339']\n",
      "hmmer_prediction= ['GH5_15(93-392)'] familes= ['GH5'] subfams ['GH5_15'] domain_ranges= ['93-392']\n",
      "hmmer_prediction= ['CE1(52-229)'] familes= ['CE1'] subfams [nan] domain_ranges= ['52-229']\n",
      "hmmer_prediction= ['AA3_2(14-617)'] familes= ['AA3'] subfams ['AA3_2'] domain_ranges= ['14-617']\n",
      "hmmer_prediction= ['AA3_2(49-626)'] familes= ['AA3'] subfams ['AA3_2'] domain_ranges= ['49-626']\n",
      "hmmer_prediction= ['AA3_3(5-639)'] familes= ['AA3'] subfams ['AA3_3'] domain_ranges= ['5-639']\n",
      "hmmer_prediction= ['GH13_5(42-383)'] familes= ['GH13'] subfams ['GH13_5'] domain_ranges= ['42-383']\n",
      "hmmer_prediction= ['AA16(20-188)'] familes= ['AA16'] subfams [nan] domain_ranges= ['20-188']\n",
      "hmmer_prediction= ['AA8(3-784)'] familes= ['AA8'] subfams [nan] domain_ranges= ['3-784']\n",
      "hmmer_prediction= ['GH10(20-332)'] familes= ['GH10'] subfams [nan] domain_ranges= ['20-332']\n",
      "hmmer_prediction= ['AA3_2(32-610)'] familes= ['AA3'] subfams ['AA3_2'] domain_ranges= ['32-610']\n",
      "hmmer_prediction= ['AA7(68-518)'] familes= ['AA7'] subfams [nan] domain_ranges= ['68-518']\n",
      "hmmer_prediction= ['CE10(31-280)'] familes= ['CE10'] subfams [nan] domain_ranges= ['31-280']\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "hmmer_prediction= ['GH5_9(83-382)'] familes= ['GH5'] subfams ['GH5_9'] domain_ranges= ['83-382']\n",
      "hmmer_prediction= ['AA3_2(49-672)'] familes= ['AA3'] subfams ['AA3_2'] domain_ranges= ['49-672']\n",
      "hmmer_prediction= ['GH12(116-258)'] familes= ['GH12'] subfams [nan] domain_ranges= ['116-258']\n",
      "hmmer_prediction= ['GH55(52-808)'] familes= ['GH55'] subfams [nan] domain_ranges= ['52-808']\n",
      "hmmer_prediction= ['GH135(7-235)'] familes= ['GH135'] subfams [nan] domain_ranges= ['7-235']\n",
      "hmmer_prediction= ['GT17(70-378)'] familes= ['GT17'] subfams [nan] domain_ranges= ['70-378']\n",
      "hmmer_prediction= ['GH79(41-414)'] familes= ['GH79'] subfams [nan] domain_ranges= ['41-414']\n",
      "hmmer_prediction= ['GH92(288-793)'] familes= ['GH92'] subfams [nan] domain_ranges= ['288-793']\n",
      "hmmer_prediction= ['GH12(100-246)'] familes= ['GH12'] subfams [nan] domain_ranges= ['100-246']\n",
      "hmmer_prediction= ['GT2_Glyco_tranf_2_3(219-467)'] familes= ['GT2'] subfams ['GT2_Glyco_tranf_2_3'] domain_ranges= ['219-467']\n",
      "hmmer_prediction= ['CE8(49-312)'] familes= ['CE8'] subfams [nan] domain_ranges= ['49-312']\n",
      "hmmer_prediction= ['GH17(445-704)'] familes= ['GH17'] subfams [nan] domain_ranges= ['445-704']\n",
      "hmmer_prediction= ['GH5_5(121-401)'] familes= ['GH5'] subfams ['GH5_5'] domain_ranges= ['121-401']\n",
      "hmmer_prediction= ['GT32(74-155)'] familes= ['GT32'] subfams [nan] domain_ranges= ['74-155']\n",
      "hmmer_prediction= ['CE4(239-366)'] familes= ['CE4'] subfams [nan] domain_ranges= ['239-366']\n",
      "hmmer_prediction= ['GH13_1(63-364)'] familes= ['GH13'] subfams ['GH13_1'] domain_ranges= ['63-364']\n",
      "hmmer_prediction= ['PL1_7(82-264)'] familes= ['PL1'] subfams ['PL1_7'] domain_ranges= ['82-264']\n",
      "hmmer_prediction= ['GH135(7-232)'] familes= ['GH135'] subfams [nan] domain_ranges= ['7-232']\n",
      "hmmer_prediction= ['GH27(116-327)'] familes= ['GH27'] subfams [nan] domain_ranges= ['116-327']\n",
      "hmmer_prediction= ['GH125(74-498)'] familes= ['GH125'] subfams [nan] domain_ranges= ['74-498']\n",
      "hmmer_prediction= ['GH43_13(7-307)'] familes= ['GH43'] subfams ['GH43_13'] domain_ranges= ['7-307']\n",
      "hmmer_prediction= ['GH28(89-438)'] familes= ['GH28'] subfams [nan] domain_ranges= ['89-438']\n",
      "hmmer_prediction= ['GH92(249-760)'] familes= ['GH92'] subfams [nan] domain_ranges= ['249-760']\n",
      "hmmer_prediction= ['CE10(110-309)'] familes= ['CE10'] subfams [nan] domain_ranges= ['110-309']\n",
      "hmmer_prediction= ['CE12(92-301)'] familes= ['CE12'] subfams [nan] domain_ranges= ['92-301']\n",
      "hmmer_prediction= ['GH47(193-686)'] familes= ['GH47'] subfams [nan] domain_ranges= ['193-686']\n",
      "hmmer_prediction= ['GH16(81-233)'] familes= ['GH16'] subfams [nan] domain_ranges= ['81-233']\n",
      "hmmer_prediction= ['CE10(18-299)'] familes= ['CE10'] subfams [nan] domain_ranges= ['18-299']\n",
      "hmmer_prediction= ['GH114(103-298)'] familes= ['GH114'] subfams [nan] domain_ranges= ['103-298']\n",
      "hmmer_prediction= ['GT2_Glycos_transf_2(116-250)'] familes= ['GT2'] subfams ['GT2_Glycos_transf_2'] domain_ranges= ['116-250']\n",
      "hmmer_prediction= ['GH18(31-309)'] familes= ['GH18'] subfams [nan] domain_ranges= ['31-309']\n",
      "hmmer_prediction= ['AA4(108-389)'] familes= ['AA4'] subfams [nan] domain_ranges= ['108-389']\n",
      "hmmer_prediction= ['CBM63(136-206)'] familes= ['CBM63'] subfams [nan] domain_ranges= ['136-206']\n",
      "hmmer_prediction= ['GH71(44-432)'] familes= ['GH71'] subfams [nan] domain_ranges= ['44-432']\n",
      "hmmer_prediction= ['CE10(441-687)'] familes= ['CE10'] subfams [nan] domain_ranges= ['441-687']\n",
      "hmmer_prediction= ['GH38(303-560)'] familes= ['GH38'] subfams [nan] domain_ranges= ['303-560']\n",
      "hmmer_prediction= ['GH38(284-541)'] familes= ['GH38'] subfams [nan] domain_ranges= ['284-541']\n",
      "hmmer_prediction= ['GT2_Chitin_synth_2(661-1184)'] familes= ['GT2'] subfams [nan] domain_ranges= ['661-1184']\n",
      "hmmer_prediction= ['GT2_Chitin_synth_2(661-1184)'] familes= ['GT2'] subfams [nan] domain_ranges= ['661-1184']\n",
      "hmmer_prediction= ['CE10(92-272)'] familes= ['CE10'] subfams [nan] domain_ranges= ['92-272']\n",
      "hmmer_prediction= ['GH47(54-525)'] familes= ['GH47'] subfams [nan] domain_ranges= ['54-525']\n",
      "hmmer_prediction= ['GH135(16-256)'] familes= ['GH135'] subfams [nan] domain_ranges= ['16-256']\n",
      "hmmer_prediction= ['CE10(27-148)'] familes= ['CE10'] subfams [nan] domain_ranges= ['27-148']\n",
      "hmmer_prediction= ['GT4(228-344)'] familes= ['GT4'] subfams [nan] domain_ranges= ['228-344']\n",
      "hmmer_prediction= ['AA1_3(66-383)'] familes= ['AA1'] subfams ['AA1_3'] domain_ranges= ['66-383']\n",
      "hmmer_prediction= ['GH28(55-371)'] familes= ['GH28'] subfams [nan] domain_ranges= ['55-371']\n",
      "hmmer_prediction= ['AA1_3(54-359)'] familes= ['AA1'] subfams ['AA1_3'] domain_ranges= ['54-359']\n",
      "hmmer_prediction= ['GT50(131-417)'] familes= ['GT50'] subfams [nan] domain_ranges= ['131-417']\n",
      "hmmer_prediction= ['CE16(49-330)'] familes= ['CE16'] subfams [nan] domain_ranges= ['49-330']\n",
      "hmmer_prediction= ['GT20(180-630)'] familes= ['GT20'] subfams [nan] domain_ranges= ['180-630']\n",
      "hmmer_prediction= ['AA2(102-354)'] familes= ['AA2'] subfams [nan] domain_ranges= ['102-354']\n",
      "hmmer_prediction= ['GH72(1-237)'] familes= ['GH72'] subfams [nan] domain_ranges= ['1-237']\n",
      "hmmer_prediction= ['GH3(78-303)'] familes= ['GH3'] subfams [nan] domain_ranges= ['78-303']\n",
      "hmmer_prediction= ['CE9(17-429)'] familes= ['CE9'] subfams [nan] domain_ranges= ['17-429']\n",
      "hmmer_prediction= ['CE10(125-389)'] familes= ['CE10'] subfams [nan] domain_ranges= ['125-389']\n",
      "hmmer_prediction= ['GH78(106-681)'] familes= ['GH78'] subfams [nan] domain_ranges= ['106-681']\n",
      "hmmer_prediction= ['CE8(50-308)'] familes= ['CE8'] subfams [nan] domain_ranges= ['50-308']\n",
      "hmmer_prediction= ['GH76(21-396)'] familes= ['GH76'] subfams [nan] domain_ranges= ['21-396']\n",
      "hmmer_prediction= ['AA3_2(32-591)'] familes= ['AA3'] subfams ['AA3_2'] domain_ranges= ['32-591']\n",
      "hmmer_prediction= ['GH88(95-467)'] familes= ['GH88'] subfams [nan] domain_ranges= ['95-467']\n",
      "hmmer_prediction= ['GH28(163-445)'] familes= ['GH28'] subfams [nan] domain_ranges= ['163-445']\n",
      "hmmer_prediction= ['AA7(135-319)'] familes= ['AA7'] subfams [nan] domain_ranges= ['135-319']\n",
      "hmmer_prediction= ['GH3(67-319)'] familes= ['GH3'] subfams [nan] domain_ranges= ['67-319']\n",
      "hmmer_prediction= ['GH71(43-433)', 'CBM24(552-628)', 'CBM24(676-751)'] familes= ['GH71', 'CBM24', 'CBM24'] subfams [nan, nan, nan] domain_ranges= ['43-433', '552-628', '676-751']\n",
      "hmmer_prediction= ['GH71(43-433)', 'CBM24(551-627)', 'CBM24(675-750)'] familes= ['GH71', 'CBM24', 'CBM24'] subfams [nan, nan, nan] domain_ranges= ['43-433', '551-627', '675-750']\n",
      "hmmer_prediction= ['GT8(105-358)'] familes= ['GT8'] subfams [nan] domain_ranges= ['105-358']\n",
      "hmmer_prediction= ['GH128(40-264)'] familes= ['GH128'] subfams [nan] domain_ranges= ['40-264']\n",
      "hmmer_prediction= ['GH128(40-202)'] familes= ['GH128'] subfams [nan] domain_ranges= ['40-202']\n",
      "hmmer_prediction= ['GH31(371-815)'] familes= ['GH31'] subfams [nan] domain_ranges= ['371-815']\n",
      "hmmer_prediction= ['GH13_1(59-362)'] familes= ['GH13'] subfams ['GH13_1'] domain_ranges= ['59-362']\n",
      "hmmer_prediction= ['GT39(55-300)'] familes= ['GT39'] subfams [nan] domain_ranges= ['55-300']\n",
      "hmmer_prediction= ['CE10(151-420)'] familes= ['CE10'] subfams [nan] domain_ranges= ['151-420']\n",
      "hmmer_prediction= ['GH1(104-580)'] familes= ['GH1'] subfams [nan] domain_ranges= ['104-580']\n",
      "hmmer_prediction= ['CE10(100-361)'] familes= ['CE10'] subfams [nan] domain_ranges= ['100-361']\n",
      "hmmer_prediction= ['CE10(47-318)'] familes= ['CE10'] subfams [nan] domain_ranges= ['47-318']\n",
      "hmmer_prediction= ['CE10(228-379)'] familes= ['CE10'] subfams [nan] domain_ranges= ['228-379']\n",
      "hmmer_prediction= ['GT22(40-467)'] familes= ['GT22'] subfams [nan] domain_ranges= ['40-467']\n",
      "hmmer_prediction= ['GH17(209-454)'] familes= ['GH17'] subfams [nan] domain_ranges= ['209-454']\n",
      "hmmer_prediction= ['GH17(209-454)'] familes= ['GH17'] subfams [nan] domain_ranges= ['209-454']\n",
      "hmmer_prediction= ['GT48(861-1638)'] familes= ['GT48'] subfams [nan] domain_ranges= ['861-1638']\n",
      "hmmer_prediction= ['GH16(83-297)'] familes= ['GH16'] subfams [nan] domain_ranges= ['83-297']\n",
      "hmmer_prediction= ['GH76(33-393)'] familes= ['GH76'] subfams [nan] domain_ranges= ['33-393']\n",
      "hmmer_prediction= ['GH5_12(15-548)'] familes= ['GH5'] subfams ['GH5_12'] domain_ranges= ['15-548']\n",
      "hmmer_prediction= ['GT20(10-477)'] familes= ['GT20'] subfams [nan] domain_ranges= ['10-477']\n",
      "hmmer_prediction= ['CE10(2197-2530)'] familes= ['CE10'] subfams [nan] domain_ranges= ['2197-2530']\n",
      "hmmer_prediction= ['GT4(215-352)'] familes= ['GT4'] subfams [nan] domain_ranges= ['215-352']\n",
      "hmmer_prediction= ['GH11(48-225)'] familes= ['GH11'] subfams [nan] domain_ranges= ['48-225']\n",
      "hmmer_prediction= ['AA7(315-522)'] familes= ['AA7'] subfams [nan] domain_ranges= ['315-522']\n",
      "hmmer_prediction= ['CE5(31-147)'] familes= ['CE5'] subfams [nan] domain_ranges= ['31-147']\n",
      "hmmer_prediction= ['GH5_7(133-404)'] familes= ['GH5'] subfams ['GH5_7'] domain_ranges= ['133-404']\n",
      "hmmer_prediction= ['GH26(173-308)'] familes= ['GH26'] subfams [nan] domain_ranges= ['173-308']\n",
      "hmmer_prediction= ['GH28(70-390)'] familes= ['GH28'] subfams [nan] domain_ranges= ['70-390']\n",
      "hmmer_prediction= ['GT22(8-399)'] familes= ['GT22'] subfams [nan] domain_ranges= ['8-399']\n",
      "hmmer_prediction= ['GH27(20-354)'] familes= ['GH27'] subfams [nan] domain_ranges= ['20-354']\n",
      "hmmer_prediction= ['GH18(125-462)'] familes= ['GH18'] subfams [nan] domain_ranges= ['125-462']\n",
      "hmmer_prediction= ['GH10(93-386)'] familes= ['GH10'] subfams [nan] domain_ranges= ['93-386']\n",
      "hmmer_prediction= ['CE8(55-350)'] familes= ['CE8'] subfams [nan] domain_ranges= ['55-350']\n",
      "hmmer_prediction= ['GH81(179-857)'] familes= ['GH81'] subfams [nan] domain_ranges= ['179-857']\n",
      "hmmer_prediction= ['AA7(29-454)'] familes= ['AA7'] subfams [nan] domain_ranges= ['29-454']\n",
      "hmmer_prediction= ['AA9(9-213)'] familes= ['AA9'] subfams [nan] domain_ranges= ['9-213']\n",
      "hmmer_prediction= ['GT76(17-449)'] familes= ['GT76'] subfams [nan] domain_ranges= ['17-449']\n",
      "hmmer_prediction= ['GT8(35-304)'] familes= ['GT8'] subfams [nan] domain_ranges= ['35-304']\n",
      "hmmer_prediction= ['GT90(373-643)'] familes= ['GT90'] subfams [nan] domain_ranges= ['373-643']\n",
      "hmmer_prediction= ['CE10(94-328)'] familes= ['CE10'] subfams [nan] domain_ranges= ['94-328']\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "hmmer_prediction= ['CE10(68-314)'] familes= ['CE10'] subfams [nan] domain_ranges= ['68-314']\n",
      "hmmer_prediction= ['CE10(68-314)'] familes= ['CE10'] subfams [nan] domain_ranges= ['68-314']\n",
      "hmmer_prediction= ['GH3(83-295)'] familes= ['GH3'] subfams [nan] domain_ranges= ['83-295']\n",
      "hmmer_prediction= ['GH15(50-458)'] familes= ['GH15'] subfams [nan] domain_ranges= ['50-458']\n",
      "hmmer_prediction= ['GH71(21-408)', 'CBM24(459-539)', 'CBM24(558-634)'] familes= ['GH71', 'CBM24', 'CBM24'] subfams [nan, nan, nan] domain_ranges= ['21-408', '459-539', '558-634']\n",
      "hmmer_prediction= ['GT62(100-362)'] familes= ['GT62'] subfams [nan] domain_ranges= ['100-362']\n",
      "hmmer_prediction= ['CE10(175-582)'] familes= ['CE10'] subfams [nan] domain_ranges= ['175-582']\n",
      "hmmer_prediction= ['GH15(81-493)', 'CBM20(573-663)'] familes= ['GH15', 'CBM20'] subfams [nan, nan] domain_ranges= ['81-493', '573-663']\n",
      "hmmer_prediction= ['GH28(183-502)'] familes= ['GH28'] subfams [nan] domain_ranges= ['183-502']\n",
      "hmmer_prediction= ['GH35(35-369)'] familes= ['GH35'] subfams [nan] domain_ranges= ['35-369']\n",
      "hmmer_prediction= ['GT2_Glyco_trans_2_3(355-539)'] familes= ['GT2'] subfams ['GT2_Glyco_trans_2_3'] domain_ranges= ['355-539']\n",
      "hmmer_prediction= ['GH13_1(61-361)'] familes= ['GH13'] subfams ['GH13_1'] domain_ranges= ['61-361']\n",
      "hmmer_prediction= ['GT4(358-514)'] familes= ['GT4'] subfams [nan] domain_ranges= ['358-514']\n",
      "hmmer_prediction= ['AA7(126-311)'] familes= ['AA7'] subfams [nan] domain_ranges= ['126-311']\n",
      "hmmer_prediction= ['CE10(78-229)'] familes= ['CE10'] subfams [nan] domain_ranges= ['78-229']\n",
      "hmmer_prediction= ['GH43_6(34-318)'] familes= ['GH43'] subfams ['GH43_6'] domain_ranges= ['34-318']\n",
      "hmmer_prediction= ['GT2_Glyco_trans_2_3(481-685)'] familes= ['GT2'] subfams ['GT2_Glyco_trans_2_3'] domain_ranges= ['481-685']\n",
      "hmmer_prediction= ['GT1(272-612)'] familes= ['GT1'] subfams [nan] domain_ranges= ['272-612']\n",
      "hmmer_prediction= ['GH28(86-442)'] familes= ['GH28'] subfams [nan] domain_ranges= ['86-442']\n",
      "hmmer_prediction= ['GH18(36-360)'] familes= ['GH18'] subfams [nan] domain_ranges= ['36-360']\n",
      "hmmer_prediction= ['GT8(71-302)'] familes= ['GT8'] subfams [nan] domain_ranges= ['71-302']\n",
      "hmmer_prediction= ['GT69(60-318)'] familes= ['GT69'] subfams [nan] domain_ranges= ['60-318']\n",
      "hmmer_prediction= ['AA1_3(59-371)'] familes= ['AA1'] subfams ['AA1_3'] domain_ranges= ['59-371']\n",
      "hmmer_prediction= ['GH28(135-417)'] familes= ['GH28'] subfams [nan] domain_ranges= ['135-417']\n",
      "hmmer_prediction= ['GH72(21-298)'] familes= ['GH72'] subfams [nan] domain_ranges= ['21-298']\n",
      "hmmer_prediction= ['GT2_Glyco_trans_2_3(197-391)'] familes= ['GT2'] subfams ['GT2_Glyco_trans_2_3'] domain_ranges= ['197-391']\n",
      "hmmer_prediction= ['GH3(82-295)'] familes= ['GH3'] subfams [nan] domain_ranges= ['82-295']\n",
      "hmmer_prediction= ['GH76(107-496)'] familes= ['GH76'] subfams [nan] domain_ranges= ['107-496']\n",
      "hmmer_prediction= ['AA7(31-224)'] familes= ['AA7'] subfams [nan] domain_ranges= ['31-224']\n",
      "hmmer_prediction= ['AA7(168-360)'] familes= ['AA7'] subfams [nan] domain_ranges= ['168-360']\n",
      "hmmer_prediction= ['GH89(73-743)'] familes= ['GH89'] subfams [nan] domain_ranges= ['73-743']\n",
      "hmmer_prediction= ['GT71(157-423)'] familes= ['GT71'] subfams [nan] domain_ranges= ['157-423']\n",
      "hmmer_prediction= ['GT71(157-423)'] familes= ['GT71'] subfams [nan] domain_ranges= ['157-423']\n",
      "hmmer_prediction= ['GT1(106-500)'] familes= ['GT1'] subfams [nan] domain_ranges= ['106-500']\n",
      "hmmer_prediction= ['AA9(11-227)'] familes= ['AA9'] subfams [nan] domain_ranges= ['11-227']\n",
      "hmmer_prediction= ['GH92(293-809)'] familes= ['GH92'] subfams [nan] domain_ranges= ['293-809']\n",
      "hmmer_prediction= ['AA7(60-263)'] familes= ['AA7'] subfams [nan] domain_ranges= ['60-263']\n",
      "hmmer_prediction= ['GH135(7-237)'] familes= ['GH135'] subfams [nan] domain_ranges= ['7-237']\n",
      "hmmer_prediction= ['GH43_24(44-275)'] familes= ['GH43'] subfams ['GH43_24'] domain_ranges= ['44-275']\n",
      "hmmer_prediction= ['GH79(38-513)'] familes= ['GH79'] subfams [nan] domain_ranges= ['38-513']\n",
      "hmmer_prediction= ['AA5(5-817)'] familes= ['AA5'] subfams [nan] domain_ranges= ['5-817']\n",
      "hmmer_prediction= ['GH28(53-402)'] familes= ['GH28'] subfams [nan] domain_ranges= ['53-402']\n",
      "hmmer_prediction= ['GH125(178-647)'] familes= ['GH125'] subfams [nan] domain_ranges= ['178-647']\n",
      "hmmer_prediction= ['GH35(45-390)'] familes= ['GH35'] subfams [nan] domain_ranges= ['45-390']\n",
      "hmmer_prediction= ['GH35(45-390)'] familes= ['GH35'] subfams [nan] domain_ranges= ['45-390']\n",
      "hmmer_prediction= ['GH47(105-590)'] familes= ['GH47'] subfams [nan] domain_ranges= ['105-590']\n",
      "hmmer_prediction= ['GH31(258-699)'] familes= ['GH31'] subfams [nan] domain_ranges= ['258-699']\n",
      "hmmer_prediction= ['GT15(167-434)'] familes= ['GT15'] subfams [nan] domain_ranges= ['167-434']\n",
      "hmmer_prediction= ['GT15(167-283)'] familes= ['GT15'] subfams [nan] domain_ranges= ['167-283']\n",
      "hmmer_prediction= ['GH47(102-562)'] familes= ['GH47'] subfams [nan] domain_ranges= ['102-562']\n",
      "hmmer_prediction= ['GH132(159-469)'] familes= ['GH132'] subfams [nan] domain_ranges= ['159-469']\n",
      "hmmer_prediction= ['AA6(4-199)'] familes= ['AA6'] subfams [nan] domain_ranges= ['4-199']\n",
      "hmmer_prediction= ['CE16(103-357)'] familes= ['CE16'] subfams [nan] domain_ranges= ['103-357']\n",
      "hmmer_prediction= ['GH78(243-635)'] familes= ['GH78'] subfams [nan] domain_ranges= ['243-635']\n",
      "hmmer_prediction= ['GT1(199-481)'] familes= ['GT1'] subfams [nan] domain_ranges= ['199-481']\n",
      "hmmer_prediction= ['CE5(33-232)'] familes= ['CE5'] subfams [nan] domain_ranges= ['33-232']\n",
      "hmmer_prediction= ['AA9(10-230)'] familes= ['AA9'] subfams [nan] domain_ranges= ['10-230']\n",
      "hmmer_prediction= ['AA7(78-508)'] familes= ['AA7'] subfams [nan] domain_ranges= ['78-508']\n",
      "hmmer_prediction= ['AA7(77-253)'] familes= ['AA7'] subfams [nan] domain_ranges= ['77-253']\n",
      "hmmer_prediction= ['GT33(47-456)'] familes= ['GT33'] subfams [nan] domain_ranges= ['47-456']\n",
      "hmmer_prediction= ['GH71(22-407)', 'CBM24(449-528)', 'CBM24(555-625)'] familes= ['GH71', 'CBM24', 'CBM24'] subfams [nan, nan, nan] domain_ranges= ['22-407', '449-528', '555-625']\n",
      "hmmer_prediction= ['GH18(7-359)'] familes= ['GH18'] subfams [nan] domain_ranges= ['7-359']\n",
      "hmmer_prediction= ['GT22(55-587)'] familes= ['GT22'] subfams [nan] domain_ranges= ['55-587']\n",
      "hmmer_prediction= ['GH18(30-282)'] familes= ['GH18'] subfams [nan] domain_ranges= ['30-282']\n",
      "hmmer_prediction= ['GT20(68-588)'] familes= ['GT20'] subfams [nan] domain_ranges= ['68-588']\n",
      "hmmer_prediction= ['GT20(68-588)'] familes= ['GT20'] subfams [nan] domain_ranges= ['68-588']\n",
      "hmmer_prediction= ['GT20(153-673)'] familes= ['GT20'] subfams [nan] domain_ranges= ['153-673']\n",
      "hmmer_prediction= ['CE5(86-255)'] familes= ['CE5'] subfams [nan] domain_ranges= ['86-255']\n",
      "hmmer_prediction= ['GT2_Chitin_synth_2(255-545)'] familes= ['GT2'] subfams [nan] domain_ranges= ['255-545']\n",
      "hmmer_prediction= ['GH92(246-749)'] familes= ['GH92'] subfams [nan] domain_ranges= ['246-749']\n",
      "hmmer_prediction= ['CE10(457-698)'] familes= ['CE10'] subfams [nan] domain_ranges= ['457-698']\n",
      "hmmer_prediction= ['GH15(285-662)'] familes= ['GH15'] subfams [nan] domain_ranges= ['285-662']\n",
      "hmmer_prediction= ['GH15(303-680)'] familes= ['GH15'] subfams [nan] domain_ranges= ['303-680']\n",
      "hmmer_prediction= ['GH71(36-404)', 'CBM24(522-597)'] familes= ['GH71', 'CBM24'] subfams [nan, nan] domain_ranges= ['36-404', '522-597']\n",
      "hmmer_prediction= ['GH55(20-781)'] familes= ['GH55'] subfams [nan] domain_ranges= ['20-781']\n",
      "hmmer_prediction= ['GT69(119-393)'] familes= ['GT69'] subfams [nan] domain_ranges= ['119-393']\n",
      "hmmer_prediction= ['GT69(119-393)'] familes= ['GT69'] subfams [nan] domain_ranges= ['119-393']\n",
      "hmmer_prediction= ['AA3_3(8-600)'] familes= ['AA3'] subfams ['AA3_3'] domain_ranges= ['8-600']\n",
      "hmmer_prediction= ['CE16(81-388)'] familes= ['CE16'] subfams [nan] domain_ranges= ['81-388']\n",
      "hmmer_prediction= ['AA11(19-224)'] familes= ['AA11'] subfams [nan] domain_ranges= ['19-224']\n",
      "hmmer_prediction= ['GH92(265-775)'] familes= ['GH92'] subfams [nan] domain_ranges= ['265-775']\n",
      "hmmer_prediction= ['GT22(8-393)'] familes= ['GT22'] subfams [nan] domain_ranges= ['8-393']\n",
      "hmmer_prediction= ['GH20(183-528)'] familes= ['GH20'] subfams [nan] domain_ranges= ['183-528']\n",
      "hmmer_prediction= ['GH16(76-241)'] familes= ['GH16'] subfams [nan] domain_ranges= ['76-241']\n",
      "hmmer_prediction= ['GH16(110-336)'] familes= ['GH16'] subfams [nan] domain_ranges= ['110-336']\n",
      "hmmer_prediction= ['GH47(1-480)'] familes= ['GH47'] subfams [nan] domain_ranges= ['1-480']\n",
      "hmmer_prediction= ['AA2(6-233)'] familes= ['AA2'] subfams [nan] domain_ranges= ['6-233']\n",
      "hmmer_prediction= ['GH16(393-717)'] familes= ['GH16'] subfams [nan] domain_ranges= ['393-717']\n",
      "hmmer_prediction= ['GH1(13-482)'] familes= ['GH1'] subfams [nan] domain_ranges= ['13-482']\n",
      "hmmer_prediction= ['GH95(8-735)'] familes= ['GH95'] subfams [nan] domain_ranges= ['8-735']\n",
      "hmmer_prediction= ['GH43_14(62-343)'] familes= ['GH43'] subfams ['GH43_14'] domain_ranges= ['62-343']\n",
      "hmmer_prediction= ['GT1(208-534)'] familes= ['GT1'] subfams [nan] domain_ranges= ['208-534']\n",
      "hmmer_prediction= ['CE10(1788-2115)'] familes= ['CE10'] subfams [nan] domain_ranges= ['1788-2115']\n",
      "hmmer_prediction= ['CE16(36-294)'] familes= ['CE16'] subfams [nan] domain_ranges= ['36-294']\n",
      "hmmer_prediction= ['GT1(185-518)'] familes= ['GT1'] subfams [nan] domain_ranges= ['185-518']\n",
      "hmmer_prediction= ['GH5_4(144-412)'] familes= ['GH5'] subfams ['GH5_4'] domain_ranges= ['144-412']\n",
      "hmmer_prediction= ['CE8(40-314)'] familes= ['CE8'] subfams [nan] domain_ranges= ['40-314']\n",
      "hmmer_prediction= ['GH16(60-286)'] familes= ['GH16'] subfams [nan] domain_ranges= ['60-286']\n",
      "hmmer_prediction= ['GH16(60-286)'] familes= ['GH16'] subfams [nan] domain_ranges= ['60-286']\n",
      "hmmer_prediction= ['GH3(84-271)'] familes= ['GH3'] subfams [nan] domain_ranges= ['84-271']\n",
      "hmmer_prediction= ['GH18(43-395)'] familes= ['GH18'] subfams [nan] domain_ranges= ['43-395']\n",
      "hmmer_prediction= ['GH18(2-270)'] familes= ['GH18'] subfams [nan] domain_ranges= ['2-270']\n",
      "hmmer_prediction= ['GH76(32-408)'] familes= ['GH76'] subfams [nan] domain_ranges= ['32-408']\n",
      "hmmer_prediction= ['GH7(20-446)'] familes= ['GH7'] subfams [nan] domain_ranges= ['20-446']\n",
      "hmmer_prediction= ['GH131(22-260)'] familes= ['GH131'] subfams [nan] domain_ranges= ['22-260']\n",
      "hmmer_prediction= ['GT2_Chitin_synth_1(270-432)'] familes= ['GT2'] subfams [nan] domain_ranges= ['270-432']\n",
      "hmmer_prediction= ['GT2_Glyco_trans_2_3(508-720)'] familes= ['GT2'] subfams ['GT2_Glyco_trans_2_3'] domain_ranges= ['508-720']\n",
      "hmmer_prediction= ['GH37(146-703)'] familes= ['GH37'] subfams [nan] domain_ranges= ['146-703']\n",
      "hmmer_prediction= ['CE3(51-240)'] familes= ['CE3'] subfams [nan] domain_ranges= ['51-240']\n",
      "hmmer_prediction= ['GT2_Glyco_trans_2_3(493-699)'] familes= ['GT2'] subfams ['GT2_Glyco_trans_2_3'] domain_ranges= ['493-699']\n",
      "hmmer_prediction= ['AA1_3(63-377)'] familes= ['AA1'] subfams ['AA1_3'] domain_ranges= ['63-377']\n",
      "hmmer_prediction= ['GT1(165-509)'] familes= ['GT1'] subfams [nan] domain_ranges= ['165-509']\n",
      "hmmer_prediction= ['AA5_2(31-634)'] familes= ['AA5'] subfams ['AA5_2'] domain_ranges= ['31-634']\n",
      "hmmer_prediction= ['AA5_2(31-634)'] familes= ['AA5'] subfams ['AA5_2'] domain_ranges= ['31-634']\n",
      "hmmer_prediction= ['AA5(610-1112)'] familes= ['AA5'] subfams [nan] domain_ranges= ['610-1112']\n",
      "hmmer_prediction= ['GH65(393-745)'] familes= ['GH65'] subfams [nan] domain_ranges= ['393-745']\n",
      "hmmer_prediction= ['AA3_2(38-618)'] familes= ['AA3'] subfams ['AA3_2'] domain_ranges= ['38-618']\n",
      "hmmer_prediction= ['GH16(40-324)'] familes= ['GH16'] subfams [nan] domain_ranges= ['40-324']\n",
      "hmmer_prediction= ['AA1_3(50-355)'] familes= ['AA1'] subfams ['AA1_3'] domain_ranges= ['50-355']\n",
      "hmmer_prediction= ['AA1_3(36-341)'] familes= ['AA1'] subfams ['AA1_3'] domain_ranges= ['36-341']\n",
      "hmmer_prediction= ['AA1_3(52-357)'] familes= ['AA1'] subfams ['AA1_3'] domain_ranges= ['52-357']\n",
      "hmmer_prediction= ['GH3(110-328)'] familes= ['GH3'] subfams [nan] domain_ranges= ['110-328']\n",
      "hmmer_prediction= ['GT32(104-188)'] familes= ['GT32'] subfams [nan] domain_ranges= ['104-188']\n",
      "hmmer_prediction= ['AA7(26-209)'] familes= ['AA7'] subfams [nan] domain_ranges= ['26-209']\n",
      "hmmer_prediction= ['AA7(26-209)'] familes= ['AA7'] subfams [nan] domain_ranges= ['26-209']\n",
      "hmmer_prediction= ['GH3(46-260)'] familes= ['GH3'] subfams [nan] domain_ranges= ['46-260']\n",
      "hmmer_prediction= ['GH3(46-260)'] familes= ['GH3'] subfams [nan] domain_ranges= ['46-260']\n",
      "hmmer_prediction= ['GH3(95-309)'] familes= ['GH3'] subfams [nan] domain_ranges= ['95-309']\n",
      "hmmer_prediction= ['GH3(95-309)'] familes= ['GH3'] subfams [nan] domain_ranges= ['95-309']\n",
      "hmmer_prediction= ['GH43_26(47-338)'] familes= ['GH43'] subfams ['GH43_26'] domain_ranges= ['47-338']\n",
      "hmmer_prediction= ['GH43_26(67-358)'] familes= ['GH43'] subfams ['GH43_26'] domain_ranges= ['67-358']\n",
      "hmmer_prediction= ['GH43_26(20-311)'] familes= ['GH43'] subfams ['GH43_26'] domain_ranges= ['20-311']\n",
      "hmmer_prediction= ['GH43_26(40-331)'] familes= ['GH43'] subfams ['GH43_26'] domain_ranges= ['40-331']\n",
      "hmmer_prediction= ['GH55(138-878)'] familes= ['GH55'] subfams [nan] domain_ranges= ['138-878']\n",
      "hmmer_prediction= ['AA3_1(222-738)'] familes= ['AA3'] subfams ['AA3_1'] domain_ranges= ['222-738']\n",
      "hmmer_prediction= ['AA7(67-265)'] familes= ['AA7'] subfams [nan] domain_ranges= ['67-265']\n",
      "hmmer_prediction= ['AA7(142-350)'] familes= ['AA7'] subfams [nan] domain_ranges= ['142-350']\n",
      "hmmer_prediction= ['AA3_2(38-638)'] familes= ['AA3'] subfams ['AA3_2'] domain_ranges= ['38-638']\n",
      "hmmer_prediction= ['GT8(9-221)'] familes= ['GT8'] subfams [nan] domain_ranges= ['9-221']\n",
      "hmmer_prediction= ['CE1(25-281)'] familes= ['CE1'] subfams [nan] domain_ranges= ['25-281']\n",
      "hmmer_prediction= ['GH1(4-471)'] familes= ['GH1'] subfams [nan] domain_ranges= ['4-471']\n",
      "hmmer_prediction= ['GH127(16-566)'] familes= ['GH127'] subfams [nan] domain_ranges= ['16-566']\n",
      "hmmer_prediction= ['GH3(111-332)'] familes= ['GH3'] subfams [nan] domain_ranges= ['111-332']\n",
      "hmmer_prediction= ['GH32(56-366)'] familes= ['GH32'] subfams [nan] domain_ranges= ['56-366']\n",
      "hmmer_prediction= ['CE10(34-203)'] familes= ['CE10'] subfams [nan] domain_ranges= ['34-203']\n",
      "hmmer_prediction= ['AA7(125-328)'] familes= ['AA7'] subfams [nan] domain_ranges= ['125-328']\n",
      "hmmer_prediction= ['AA3_2(8-568)'] familes= ['AA3'] subfams ['AA3_2'] domain_ranges= ['8-568']\n",
      "hmmer_prediction= ['GH55(64-813)'] familes= ['GH55'] subfams [nan] domain_ranges= ['64-813']\n",
      "hmmer_prediction= ['GH35(51-384)'] familes= ['GH35'] subfams [nan] domain_ranges= ['51-384']\n",
      "hmmer_prediction= ['CE4(244-373)'] familes= ['CE4'] subfams [nan] domain_ranges= ['244-373']\n",
      "hmmer_prediction= ['AA9(9-236)'] familes= ['AA9'] subfams [nan] domain_ranges= ['9-236']\n",
      "hmmer_prediction= ['AA7(123-311)'] familes= ['AA7'] subfams [nan] domain_ranges= ['123-311']\n",
      "hmmer_prediction= ['GH106(42-560)'] familes= ['GH106'] subfams [nan] domain_ranges= ['42-560']\n",
      "hmmer_prediction= ['GT15(84-354)'] familes= ['GT15'] subfams [nan] domain_ranges= ['84-354']\n",
      "hmmer_prediction= ['GT4(241-396)'] familes= ['GT4'] subfams [nan] domain_ranges= ['241-396']\n",
      "hmmer_prediction= ['GT2_Chitin_synth_1(370-532)', 'GT2_Chitin_synth_2(509-774)'] familes= ['GT2', 'GT2'] subfams [nan, nan] domain_ranges= ['370-532', '509-774']\n",
      "hmmer_prediction= ['AA5(505-1022)'] familes= ['AA5'] subfams [nan] domain_ranges= ['505-1022']\n",
      "hmmer_prediction= ['GH16(138-476)'] familes= ['GH16'] subfams [nan] domain_ranges= ['138-476']\n",
      "hmmer_prediction= ['GT2_Glyco_trans_2_3(565-773)'] familes= ['GT2'] subfams ['GT2_Glyco_trans_2_3'] domain_ranges= ['565-773']\n",
      "hmmer_prediction= ['GT62(91-354)'] familes= ['GT62'] subfams [nan] domain_ranges= ['91-354']\n",
      "hmmer_prediction= ['AA3_2(37-626)'] familes= ['AA3'] subfams ['AA3_2'] domain_ranges= ['37-626']\n",
      "hmmer_prediction= ['GH15(64-456)', 'CBM20(506-595)'] familes= ['GH15', 'CBM20'] subfams [nan, nan] domain_ranges= ['64-456', '506-595']\n",
      "hmmer_prediction= ['AA9(10-230)'] familes= ['AA9'] subfams [nan] domain_ranges= ['10-230']\n",
      "hmmer_prediction= ['GT58(43-405)'] familes= ['GT58'] subfams [nan] domain_ranges= ['43-405']\n",
      "hmmer_prediction= ['GT2_Chitin_synth_2(1063-1571)'] familes= ['GT2'] subfams [nan] domain_ranges= ['1063-1571']\n",
      "hmmer_prediction= ['GH131(27-285)'] familes= ['GH131'] subfams [nan] domain_ranges= ['27-285']\n",
      "hmmer_prediction= ['CE10(111-271)'] familes= ['CE10'] subfams [nan] domain_ranges= ['111-271']\n",
      "hmmer_prediction= ['AA1_3(119-388)'] familes= ['AA1'] subfams ['AA1_3'] domain_ranges= ['119-388']\n",
      "hmmer_prediction= ['GT57(15-496)'] familes= ['GT57'] subfams [nan] domain_ranges= ['15-496']\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "hmmer_prediction= ['GH12(105-251)'] familes= ['GH12'] subfams [nan] domain_ranges= ['105-251']\n",
      "hmmer_prediction= ['GH72(20-335)'] familes= ['GH72'] subfams [nan] domain_ranges= ['20-335']\n",
      "hmmer_prediction= ['AA7(3-335)'] familes= ['AA7'] subfams [nan] domain_ranges= ['3-335']\n",
      "hmmer_prediction= ['GH28(64-373)'] familes= ['GH28'] subfams [nan] domain_ranges= ['64-373']\n",
      "hmmer_prediction= ['CE10(137-433)'] familes= ['CE10'] subfams [nan] domain_ranges= ['137-433']\n",
      "hmmer_prediction= ['CE10(137-426)'] familes= ['CE10'] subfams [nan] domain_ranges= ['137-426']\n",
      "hmmer_prediction= ['GH5_9(473-796)'] familes= ['GH5'] subfams ['GH5_9'] domain_ranges= ['473-796']\n",
      "hmmer_prediction= ['CE10(120-380)'] familes= ['CE10'] subfams [nan] domain_ranges= ['120-380']\n",
      "hmmer_prediction= ['GH76(23-386)'] familes= ['GH76'] subfams [nan] domain_ranges= ['23-386']\n",
      "hmmer_prediction= ['GT4(221-389)'] familes= ['GT4'] subfams [nan] domain_ranges= ['221-389']\n",
      "hmmer_prediction= ['GH78(419-833)'] familes= ['GH78'] subfams [nan] domain_ranges= ['419-833']\n",
      "hmmer_prediction= ['GH78(82-638)'] familes= ['GH78'] subfams [nan] domain_ranges= ['82-638']\n",
      "hmmer_prediction= ['AA3_2(6-545)'] familes= ['AA3'] subfams ['AA3_2'] domain_ranges= ['6-545']\n",
      "hmmer_prediction= ['AA3_2(54-617)'] familes= ['AA3'] subfams ['AA3_2'] domain_ranges= ['54-617']\n",
      "hmmer_prediction= ['GH3(32-246)'] familes= ['GH3'] subfams [nan] domain_ranges= ['32-246']\n",
      "hmmer_prediction= ['GH28(72-402)'] familes= ['GH28'] subfams [nan] domain_ranges= ['72-402']\n",
      "hmmer_prediction= ['GH28(81-411)'] familes= ['GH28'] subfams [nan] domain_ranges= ['81-411']\n",
      "hmmer_prediction= ['GH28(44-374)'] familes= ['GH28'] subfams [nan] domain_ranges= ['44-374']\n",
      "hmmer_prediction= ['GH28(44-374)'] familes= ['GH28'] subfams [nan] domain_ranges= ['44-374']\n",
      "hmmer_prediction= ['GH28(44-374)'] familes= ['GH28'] subfams [nan] domain_ranges= ['44-374']\n",
      "hmmer_prediction= ['GH31(258-813)'] familes= ['GH31'] subfams [nan] domain_ranges= ['258-813']\n",
      "hmmer_prediction= ['PL1_7(87-267)'] familes= ['PL1'] subfams ['PL1_7'] domain_ranges= ['87-267']\n",
      "hmmer_prediction= ['GT8(7-278)'] familes= ['GT8'] subfams [nan] domain_ranges= ['7-278']\n",
      "hmmer_prediction= ['AA3_2(4-515)'] familes= ['AA3'] subfams ['AA3_2'] domain_ranges= ['4-515']\n",
      "hmmer_prediction= ['AA3_2(4-515)'] familes= ['AA3'] subfams ['AA3_2'] domain_ranges= ['4-515']\n",
      "hmmer_prediction= ['AA7(45-237)'] familes= ['AA7'] subfams [nan] domain_ranges= ['45-237']\n",
      "hmmer_prediction= ['CE10(105-250)'] familes= ['CE10'] subfams [nan] domain_ranges= ['105-250']\n",
      "hmmer_prediction= ['GT34(64-292)'] familes= ['GT34'] subfams [nan] domain_ranges= ['64-292']\n",
      "hmmer_prediction= ['CE16(36-332)'] familes= ['CE16'] subfams [nan] domain_ranges= ['36-332']\n",
      "hmmer_prediction= ['CE16(36-323)'] familes= ['CE16'] subfams [nan] domain_ranges= ['36-323']\n",
      "hmmer_prediction= ['CE9(23-389)'] familes= ['CE9'] subfams [nan] domain_ranges= ['23-389']\n",
      "hmmer_prediction= ['AA3_2(23-589)'] familes= ['AA3'] subfams ['AA3_2'] domain_ranges= ['23-589']\n",
      "hmmer_prediction= ['CE5(31-235)'] familes= ['CE5'] subfams [nan] domain_ranges= ['31-235']\n",
      "hmmer_prediction= ['GT1(951-1361)'] familes= ['GT1'] subfams [nan] domain_ranges= ['951-1361']\n",
      "hmmer_prediction= ['CE1(43-271)'] familes= ['CE1'] subfams [nan] domain_ranges= ['43-271']\n",
      "hmmer_prediction= ['GH71(52-432)', 'CBM24(475-548)', 'CBM24(568-615)'] familes= ['GH71', 'CBM24', 'CBM24'] subfams [nan, nan, nan] domain_ranges= ['52-432', '475-548', '568-615']\n",
      "hmmer_prediction= ['GT57(74-304)', 'GT57(343-574)'] familes= ['GT57', 'GT57'] subfams [nan, nan] domain_ranges= ['74-304', '343-574']\n",
      "hmmer_prediction= ['GH13_8(247-540)'] familes= ['GH13'] subfams ['GH13_8'] domain_ranges= ['247-540']\n",
      "hmmer_prediction= ['GH13_8(247-540)'] familes= ['GH13'] subfams ['GH13_8'] domain_ranges= ['247-540']\n",
      "hmmer_prediction= ['GT3(16-652)'] familes= ['GT3'] subfams [nan] domain_ranges= ['16-652']\n",
      "hmmer_prediction= ['GH3(62-271)'] familes= ['GH3'] subfams [nan] domain_ranges= ['62-271']\n",
      "hmmer_prediction= ['AA7(129-548)'] familes= ['AA7'] subfams [nan] domain_ranges= ['129-548']\n",
      "hmmer_prediction= ['AA9(11-212)'] familes= ['AA9'] subfams [nan] domain_ranges= ['11-212']\n",
      "hmmer_prediction= ['GT39(66-304)'] familes= ['GT39'] subfams [nan] domain_ranges= ['66-304']\n",
      "hmmer_prediction= ['GT35(159-874)'] familes= ['GT35'] subfams [nan] domain_ranges= ['159-874']\n",
      "hmmer_prediction= ['GT59(53-542)'] familes= ['GT59'] subfams [nan] domain_ranges= ['53-542']\n",
      "hmmer_prediction= ['GH28(47-393)'] familes= ['GH28'] subfams [nan] domain_ranges= ['47-393']\n",
      "hmmer_prediction= ['GH72(35-335)'] familes= ['GH72'] subfams [nan] domain_ranges= ['35-335']\n",
      "hmmer_prediction= ['GH76(30-392)'] familes= ['GH76'] subfams [nan] domain_ranges= ['30-392']\n",
      "hmmer_prediction= ['AA7(180-379)'] familes= ['AA7'] subfams [nan] domain_ranges= ['180-379']\n",
      "hmmer_prediction= ['AA7(180-379)'] familes= ['AA7'] subfams [nan] domain_ranges= ['180-379']\n",
      "hmmer_prediction= ['GT1(134-503)'] familes= ['GT1'] subfams [nan] domain_ranges= ['134-503']\n",
      "hmmer_prediction= ['AA7(127-332)'] familes= ['AA7'] subfams [nan] domain_ranges= ['127-332']\n",
      "hmmer_prediction= ['AA7(127-296)'] familes= ['AA7'] subfams [nan] domain_ranges= ['127-296']\n",
      "hmmer_prediction= ['GT69(93-365)'] familes= ['GT69'] subfams [nan] domain_ranges= ['93-365']\n",
      "hmmer_prediction= ['CE10(128-373)'] familes= ['CE10'] subfams [nan] domain_ranges= ['128-373']\n",
      "hmmer_prediction= ['AA7(179-378)'] familes= ['AA7'] subfams [nan] domain_ranges= ['179-378']\n",
      "hmmer_prediction= ['GH1(97-561)'] familes= ['GH1'] subfams [nan] domain_ranges= ['97-561']\n",
      "hmmer_prediction= ['GH152(177-285)'] familes= ['GH152'] subfams [nan] domain_ranges= ['177-285']\n",
      "hmmer_prediction= ['GH47(105-571)'] familes= ['GH47'] subfams [nan] domain_ranges= ['105-571']\n",
      "hmmer_prediction= ['GH47(100-566)'] familes= ['GH47'] subfams [nan] domain_ranges= ['100-566']\n",
      "hmmer_prediction= ['CE10(2225-2533)'] familes= ['CE10'] subfams [nan] domain_ranges= ['2225-2533']\n",
      "hmmer_prediction= ['CE10(2247-2555)'] familes= ['CE10'] subfams [nan] domain_ranges= ['2247-2555']\n",
      "hmmer_prediction= ['AA7(57-278)'] familes= ['AA7'] subfams [nan] domain_ranges= ['57-278']\n",
      "hmmer_prediction= ['GH32(353-670)'] familes= ['GH32'] subfams [nan] domain_ranges= ['353-670']\n",
      "hmmer_prediction= ['-'] familes= [nan] subfams [nan] domain_ranges= [nan]\n",
      "hmmer_prediction= ['-'] familes= [nan] subfams [nan] domain_ranges= [nan]\n",
      "hmmer_prediction= ['-'] familes= [nan] subfams [nan] domain_ranges= [nan]\n",
      "hmmer_prediction= ['-'] familes= [nan] subfams [nan] domain_ranges= [nan]\n",
      "hmmer_prediction= ['-'] familes= [nan] subfams [nan] domain_ranges= [nan]\n",
      "hmmer_prediction= ['-'] familes= [nan] subfams [nan] domain_ranges= [nan]\n",
      "hmmer_prediction= ['-'] familes= [nan] subfams [nan] domain_ranges= [nan]\n",
      "hmmer_prediction= ['-'] familes= [nan] subfams [nan] domain_ranges= [nan]\n",
      "hmmer_prediction= ['-'] familes= [nan] subfams [nan] domain_ranges= [nan]\n",
      "hmmer_prediction= ['-'] familes= [nan] subfams [nan] domain_ranges= [nan]\n",
      "hmmer_prediction= ['-'] familes= [nan] subfams [nan] domain_ranges= [nan]\n",
      "hmmer_prediction= ['-'] familes= [nan] subfams [nan] domain_ranges= [nan]\n",
      "hmmer_prediction= ['-'] familes= [nan] subfams [nan] domain_ranges= [nan]\n",
      "hmmer_prediction= ['-'] familes= [nan] subfams [nan] domain_ranges= [nan]\n",
      "hmmer_prediction= ['-'] familes= [nan] subfams [nan] domain_ranges= [nan]\n",
      "hmmer_prediction= ['-'] familes= [nan] subfams [nan] domain_ranges= [nan]\n",
      "hmmer_prediction= ['-'] familes= [nan] subfams [nan] domain_ranges= [nan]\n",
      "hmmer_prediction= ['-'] familes= [nan] subfams [nan] domain_ranges= [nan]\n",
      "hmmer_prediction= ['-'] familes= [nan] subfams [nan] domain_ranges= [nan]\n",
      "hmmer_prediction= ['-'] familes= [nan] subfams [nan] domain_ranges= [nan]\n",
      "hmmer_prediction= ['-'] familes= [nan] subfams [nan] domain_ranges= [nan]\n",
      "hmmer_prediction= ['-'] familes= [nan] subfams [nan] domain_ranges= [nan]\n",
      "hmmer_prediction= ['-'] familes= [nan] subfams [nan] domain_ranges= [nan]\n",
      "hmmer_prediction= ['-'] familes= [nan] subfams [nan] domain_ranges= [nan]\n",
      "hmmer_prediction= ['-'] familes= [nan] subfams [nan] domain_ranges= [nan]\n",
      "hmmer_prediction= ['-'] familes= [nan] subfams [nan] domain_ranges= [nan]\n",
      "hmmer_prediction= ['-'] familes= [nan] subfams [nan] domain_ranges= [nan]\n",
      "hmmer_prediction= ['-'] familes= [nan] subfams [nan] domain_ranges= [nan]\n",
      "hmmer_prediction= ['-'] familes= [nan] subfams [nan] domain_ranges= [nan]\n",
      "hmmer_prediction= ['-'] familes= [nan] subfams [nan] domain_ranges= [nan]\n",
      "hmmer_prediction= ['-'] familes= [nan] subfams [nan] domain_ranges= [nan]\n",
      "hmmer_prediction= ['-'] familes= [nan] subfams [nan] domain_ranges= [nan]\n",
      "hmmer_prediction= ['-'] familes= [nan] subfams [nan] domain_ranges= [nan]\n",
      "hmmer_prediction= ['-'] familes= [nan] subfams [nan] domain_ranges= [nan]\n",
      "hmmer_prediction= ['-'] familes= [nan] subfams [nan] domain_ranges= [nan]\n",
      "hmmer_prediction= ['-'] familes= [nan] subfams [nan] domain_ranges= [nan]\n",
      "hmmer_prediction= ['-'] familes= [nan] subfams [nan] domain_ranges= [nan]\n",
      "hmmer_prediction= ['-'] familes= [nan] subfams [nan] domain_ranges= [nan]\n",
      "hmmer_prediction= ['-'] familes= [nan] subfams [nan] domain_ranges= [nan]\n",
      "hmmer_prediction= ['-'] familes= [nan] subfams [nan] domain_ranges= [nan]\n",
      "hmmer_prediction= ['-'] familes= [nan] subfams [nan] domain_ranges= [nan]\n",
      "hmmer_prediction= ['-'] familes= [nan] subfams [nan] domain_ranges= [nan]\n",
      "hmmer_prediction= ['-'] familes= [nan] subfams [nan] domain_ranges= [nan]\n",
      "hmmer_prediction= ['-'] familes= [nan] subfams [nan] domain_ranges= [nan]\n",
      "hmmer_prediction= ['-'] familes= [nan] subfams [nan] domain_ranges= [nan]\n",
      "hmmer_prediction= ['-'] familes= [nan] subfams [nan] domain_ranges= [nan]\n",
      "hmmer_prediction= ['-'] familes= [nan] subfams [nan] domain_ranges= [nan]\n",
      "hmmer_prediction= ['-'] familes= [nan] subfams [nan] domain_ranges= [nan]\n",
      "hmmer_prediction= ['-'] familes= [nan] subfams [nan] domain_ranges= [nan]\n",
      "hmmer_prediction= ['-'] familes= [nan] subfams [nan] domain_ranges= [nan]\n",
      "hmmer_prediction= ['-'] familes= [nan] subfams [nan] domain_ranges= [nan]\n",
      "hmmer_prediction= ['-'] familes= [nan] subfams [nan] domain_ranges= [nan]\n",
      "hmmer_prediction= ['-'] familes= [nan] subfams [nan] domain_ranges= [nan]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "hmmer_prediction= ['-'] familes= [nan] subfams [nan] domain_ranges= [nan]\n",
      "hmmer_prediction= ['-'] familes= [nan] subfams [nan] domain_ranges= [nan]\n",
      "hmmer_prediction= ['-'] familes= [nan] subfams [nan] domain_ranges= [nan]\n",
      "hmmer_prediction= ['-'] familes= [nan] subfams [nan] domain_ranges= [nan]\n",
      "hmmer_prediction= ['-'] familes= [nan] subfams [nan] domain_ranges= [nan]\n",
      "hmmer_prediction= ['-'] familes= [nan] subfams [nan] domain_ranges= [nan]\n",
      "hmmer_prediction= ['-'] familes= [nan] subfams [nan] domain_ranges= [nan]\n",
      "hmmer_prediction= ['-'] familes= [nan] subfams [nan] domain_ranges= [nan]\n",
      "hmmer_prediction= ['-'] familes= [nan] subfams [nan] domain_ranges= [nan]\n",
      "hmmer_prediction= ['-'] familes= [nan] subfams [nan] domain_ranges= [nan]\n",
      "hmmer_prediction= ['-'] familes= [nan] subfams [nan] domain_ranges= [nan]\n",
      "hmmer_prediction= ['-'] familes= [nan] subfams [nan] domain_ranges= [nan]\n",
      "hmmer_prediction= ['-'] familes= [nan] subfams [nan] domain_ranges= [nan]\n",
      "hmmer_prediction= ['-'] familes= [nan] subfams [nan] domain_ranges= [nan]\n",
      "hmmer_prediction= ['-'] familes= [nan] subfams [nan] domain_ranges= [nan]\n",
      "hmmer_prediction= ['-'] familes= [nan] subfams [nan] domain_ranges= [nan]\n",
      "hmmer_prediction= ['-'] familes= [nan] subfams [nan] domain_ranges= [nan]\n",
      "hmmer_prediction= ['-'] familes= [nan] subfams [nan] domain_ranges= [nan]\n",
      "hmmer_prediction= ['-'] familes= [nan] subfams [nan] domain_ranges= [nan]\n",
      "hmmer_prediction= ['-'] familes= [nan] subfams [nan] domain_ranges= [nan]\n",
      "hmmer_prediction= ['-'] familes= [nan] subfams [nan] domain_ranges= [nan]\n",
      "hmmer_prediction= ['-'] familes= [nan] subfams [nan] domain_ranges= [nan]\n",
      "hmmer_prediction= ['-'] familes= [nan] subfams [nan] domain_ranges= [nan]\n",
      "hmmer_prediction= ['-'] familes= [nan] subfams [nan] domain_ranges= [nan]\n",
      "hmmer_prediction= ['-'] familes= [nan] subfams [nan] domain_ranges= [nan]\n",
      "hmmer_prediction= ['-'] familes= [nan] subfams [nan] domain_ranges= [nan]\n",
      "hmmer_prediction= ['-'] familes= [nan] subfams [nan] domain_ranges= [nan]\n",
      "hmmer_prediction= ['-'] familes= [nan] subfams [nan] domain_ranges= [nan]\n",
      "hmmer_prediction= ['-'] familes= [nan] subfams [nan] domain_ranges= [nan]\n",
      "hmmer_prediction= ['-'] familes= [nan] subfams [nan] domain_ranges= [nan]\n",
      "hmmer_prediction= ['-'] familes= [nan] subfams [nan] domain_ranges= [nan]\n",
      "hmmer_prediction= ['-'] familes= [nan] subfams [nan] domain_ranges= [nan]\n",
      "hmmer_prediction= ['-'] familes= [nan] subfams [nan] domain_ranges= [nan]\n",
      "hmmer_prediction= ['-'] familes= [nan] subfams [nan] domain_ranges= [nan]\n",
      "hmmer_prediction= ['-'] familes= [nan] subfams [nan] domain_ranges= [nan]\n",
      "hmmer_prediction= ['-'] familes= [nan] subfams [nan] domain_ranges= [nan]\n",
      "hmmer_prediction= ['-'] familes= [nan] subfams [nan] domain_ranges= [nan]\n",
      "hmmer_prediction= ['-'] familes= [nan] subfams [nan] domain_ranges= [nan]\n",
      "hmmer_prediction= ['-'] familes= [nan] subfams [nan] domain_ranges= [nan]\n",
      "hmmer_prediction= ['-'] familes= [nan] subfams [nan] domain_ranges= [nan]\n",
      "hmmer_prediction= ['-'] familes= [nan] subfams [nan] domain_ranges= [nan]\n",
      "hmmer_prediction= ['-'] familes= [nan] subfams [nan] domain_ranges= [nan]\n",
      "hmmer_prediction= ['-'] familes= [nan] subfams [nan] domain_ranges= [nan]\n",
      "hmmer_prediction= ['-'] familes= [nan] subfams [nan] domain_ranges= [nan]\n",
      "hmmer_prediction= ['-'] familes= [nan] subfams [nan] domain_ranges= [nan]\n",
      "hmmer_prediction= ['-'] familes= [nan] subfams [nan] domain_ranges= [nan]\n",
      "hmmer_prediction= ['-'] familes= [nan] subfams [nan] domain_ranges= [nan]\n",
      "hmmer_prediction= ['-'] familes= [nan] subfams [nan] domain_ranges= [nan]\n",
      "hmmer_prediction= ['-'] familes= [nan] subfams [nan] domain_ranges= [nan]\n",
      "hmmer_prediction= ['-'] familes= [nan] subfams [nan] domain_ranges= [nan]\n",
      "hmmer_prediction= ['-'] familes= [nan] subfams [nan] domain_ranges= [nan]\n",
      "hmmer_prediction= ['-'] familes= [nan] subfams [nan] domain_ranges= [nan]\n",
      "hmmer_prediction= ['-'] familes= [nan] subfams [nan] domain_ranges= [nan]\n",
      "hmmer_prediction= ['-'] familes= [nan] subfams [nan] domain_ranges= [nan]\n",
      "hmmer_prediction= ['-'] familes= [nan] subfams [nan] domain_ranges= [nan]\n",
      "hmmer_prediction= ['-'] familes= [nan] subfams [nan] domain_ranges= [nan]\n",
      "hmmer_prediction= ['-'] familes= [nan] subfams [nan] domain_ranges= [nan]\n",
      "hmmer_prediction= ['-'] familes= [nan] subfams [nan] domain_ranges= [nan]\n",
      "hmmer_prediction= ['-'] familes= [nan] subfams [nan] domain_ranges= [nan]\n",
      "hmmer_prediction= ['-'] familes= [nan] subfams [nan] domain_ranges= [nan]\n",
      "hmmer_prediction= ['-'] familes= [nan] subfams [nan] domain_ranges= [nan]\n",
      "hmmer_prediction= ['-'] familes= [nan] subfams [nan] domain_ranges= [nan]\n",
      "hmmer_prediction= ['-'] familes= [nan] subfams [nan] domain_ranges= [nan]\n",
      "hmmer_prediction= ['-'] familes= [nan] subfams [nan] domain_ranges= [nan]\n",
      "hmmer_prediction= ['-'] familes= [nan] subfams [nan] domain_ranges= [nan]\n",
      "hmmer_prediction= ['-'] familes= [nan] subfams [nan] domain_ranges= [nan]\n",
      "hmmer_prediction= ['-'] familes= [nan] subfams [nan] domain_ranges= [nan]\n",
      "hmmer_prediction= ['-'] familes= [nan] subfams [nan] domain_ranges= [nan]\n",
      "hmmer_prediction= ['-'] familes= [nan] subfams [nan] domain_ranges= [nan]\n",
      "hmmer_prediction= ['-'] familes= [nan] subfams [nan] domain_ranges= [nan]\n",
      "hmmer_prediction= ['-'] familes= [nan] subfams [nan] domain_ranges= [nan]\n",
      "hmmer_prediction= ['-'] familes= [nan] subfams [nan] domain_ranges= [nan]\n",
      "hmmer_prediction= ['-'] familes= [nan] subfams [nan] domain_ranges= [nan]\n"
     ]
    }
   ],
   "source": [
    "for line in overview_file[1:]:  # skip the first line becuase this is the head titles\n",
    "    line = line.split(\"\\t\")\n",
    "    \n",
    "    protein_id = line[0]\n",
    "\n",
    "    # Retrieve predictions from HMMER (predicated domains and domain ranges)\n",
    "    # separate out each of the domain predications\n",
    "    hmmer_prediction = line[1].split(\"+\")\n",
    "    \n",
    "    # create empty lists to store predicated domains and ranges\n",
    "    cazy_fam = []\n",
    "    cazy_subfam = []\n",
    "    domain_ranges = []\n",
    "\n",
    "    # separate out the name of the predicated domain (the CAZy family) and the domain's AA range\n",
    "    for domain in hmmer_prediction:\n",
    "        domain = domain.split(\"(\")\n",
    "        # standardise name if necessary, retrieve subfam if preciated, and produce null values\n",
    "        domain_name = domain[0]\n",
    "        if domain_name.startswith(\"GT2_Chitin_\"):\n",
    "            cazy_fam.append(\"GT2\")\n",
    "            cazy_subfam.append(np.nan)\n",
    "\n",
    "        elif domain_name == '-':\n",
    "            cazy_fam.append(np.nan)\n",
    "            cazy_subfam.append(np.nan)\n",
    "\n",
    "        elif domain_name.find(\"_\") != -1:\n",
    "            cutoff = domain_name.find(\"_\")\n",
    "            cazy_fam.append(domain_name[:cutoff])\n",
    "            cazy_subfam.append(domain_name)\n",
    "\n",
    "        else:\n",
    "            cazy_fam.append(domain_name)\n",
    "            cazy_subfam.append(np.nan)\n",
    "\n",
    "        try:\n",
    "            if domain[1] == \"-\":\n",
    "                domain_ranges.append(np.nan)\n",
    "            else:\n",
    "                domain_ranges.append(domain[1][:-1])  # exlude the final \")\"\n",
    "        except IndexError: # no domain predicated so no domain ranges\n",
    "            domain_ranges.append(np.nan)\n",
    "\n",
    "    print(\"hmmer_prediction=\", hmmer_prediction, \"familes=\", cazy_fam, \"subfams\", cazy_subfam, \"domain_ranges=\", domain_ranges)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**HOTPEP**\n",
    "\n",
    "Next is to retrieve the data from the Hotpep predications. This data is layed out in the same manner as for HMMER, when looking at the output from both tools in the overview.txt file. However, instead of providing domain ranges within the parenthesises, the number of the CAZy family's associated k-mer cluster/group within Hotpep is given and is thus not wanted/needed."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "hotpep_prediction= ['GH53(7)'] familes= ['GH53'] subfams [nan]\n",
      "hotpep_prediction= ['GT2(161)'] familes= ['GT2'] subfams [nan]\n",
      "hotpep_prediction= ['AA4(1)'] familes= ['AA4'] subfams [nan]\n",
      "hotpep_prediction= ['CBM21(4)'] familes= ['CBM21'] subfams [nan]\n",
      "hotpep_prediction= ['GH132(2)'] familes= ['GH132'] subfams [nan]\n",
      "hotpep_prediction= ['GT32(54)'] familes= ['GT32'] subfams [nan]\n",
      "hotpep_prediction= ['GT25(35)'] familes= ['GT25'] subfams [nan]\n",
      "hotpep_prediction= ['GH17(29)'] familes= ['GH17'] subfams [nan]\n",
      "hotpep_prediction= ['GT71(7)'] familes= ['GT71'] subfams [nan]\n",
      "hotpep_prediction= ['-'] familes= [nan] subfams [nan]\n",
      "hotpep_prediction= ['-'] familes= [nan] subfams [nan]\n",
      "hotpep_prediction= ['GH18(92)', 'CBM1(13)', 'CBM19(1)'] familes= ['GH18', 'CBM1', 'CBM19'] subfams [nan, nan, nan]\n",
      "hotpep_prediction= ['-'] familes= [nan] subfams [nan]\n",
      "hotpep_prediction= ['-'] familes= [nan] subfams [nan]\n"
     ]
    }
   ],
   "source": [
    "for line in overview_file[1:15]:  # skip the first line becuase this is the head titles\n",
    "    line = line.split(\"\\t\")\n",
    "\n",
    "    # Retrieve predictions from Hotpep\n",
    "    hotpep_prediction = line[2].split(\"+\")\n",
    "    # separate the predicated CAZy fam and subfam\n",
    "    cazy_fams = []\n",
    "    cazy_subfams = []\n",
    "    # remove the k-mer cluster number\n",
    "    for prediction in hotpep_prediction:\n",
    "        prediction = prediction.split(\"(\")[0]\n",
    "        if prediction == \"-\":\n",
    "            cazy_fams.append(np.nan)\n",
    "            cazy_subfams.append(np.nan)\n",
    "        elif prediction.find(\"_\") != -1:\n",
    "            cutoff = prediction.find(\"_\")\n",
    "            cazy_fams.append(predictions[:cutoff])\n",
    "            cazy_subfams.append(prediction)\n",
    "        else:\n",
    "            cazy_fams.append(prediction)\n",
    "            cazy_subfams.append(np.nan)\n",
    "    print(\"hotpep_prediction=\", hotpep_prediction, \"familes=\", cazy_fams, \"subfams\", cazy_subfams)\n",
    "\n",
    "# checked more rows to make sure handling of when no domains is predicated '-' is ok\n",
    "# there are no subfamily predictions in the output file use by Hotpep can predict CAZy subfamiles"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The predicated EC number from Hotpep would also be good but this has to be retrieved from the specific Hotpep output file therefore, to keep focusing on the overview.txt file, lets look at retrieving the output from DIAMOND.\n",
    "\n",
    "**DIAMOND**\n",
    "\n",
    "DIAMOND can also predicated multiple domains and these are separated by a '+' and no additional data is given within the overview.txt file, and none of the additional data within the DIAMOND output file is wanted. Therefore, a similar approach to the previous two can be used for retrieving the output for DIAMOND.\n",
    "\n",
    "Again, if not CAZy domains are predicated from a query protein a '-' is given, which will be converted to a null value.\n",
    "\n",
    "Lastly, DIAMOND can predict CAZy subfamiles, and like HMMER it does not produce a separate CAZy family and subfamily output, therefore, one needs to be created mannually."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "diamond_predictions= ['GH53'] fams= ['GH53'] subfams= [nan]\n",
      "diamond_predictions= ['GT2'] fams= ['GT2'] subfams= [nan]\n",
      "diamond_predictions= ['AA4'] fams= ['AA4'] subfams= [nan]\n",
      "diamond_predictions= ['CBM21'] fams= ['CBM21'] subfams= [nan]\n",
      "diamond_predictions= ['GH132'] fams= ['GH132'] subfams= [nan]\n",
      "diamond_predictions= ['GT32'] fams= ['GT32'] subfams= [nan]\n",
      "diamond_predictions= ['GT25'] fams= ['GT25'] subfams= [nan]\n",
      "diamond_predictions= ['GH17'] fams= ['GH17'] subfams= [nan]\n",
      "diamond_predictions= ['GT71'] fams= ['GT71'] subfams= [nan]\n",
      "diamond_predictions= ['GH16'] fams= ['GH16'] subfams= [nan]\n",
      "diamond_predictions= ['-'] fams= [nan] subfams= [nan]\n",
      "diamond_predictions= ['CBM1', 'GH18'] fams= ['CBM1', 'GH18'] subfams= [nan, nan]\n",
      "diamond_predictions= ['-'] fams= [nan] subfams= [nan]\n",
      "diamond_predictions= ['-'] fams= [nan] subfams= [nan]\n",
      "diamond_predictions= ['CBM1', 'GH131'] fams= ['CBM1', 'GH131'] subfams= [nan, nan]\n",
      "diamond_predictions= ['-'] fams= [nan] subfams= [nan]\n",
      "diamond_predictions= ['CE5'] fams= ['CE5'] subfams= [nan]\n",
      "diamond_predictions= ['-'] fams= [nan] subfams= [nan]\n",
      "diamond_predictions= ['GH105'] fams= ['GH105'] subfams= [nan]\n",
      "diamond_predictions= ['PL1_4'] fams= ['PL1'] subfams= ['PL1_4']\n",
      "diamond_predictions= ['AA3_1'] fams= ['AA3'] subfams= ['AA3_1']\n",
      "diamond_predictions= ['AA1_3'] fams= ['AA1'] subfams= ['AA1_3']\n",
      "diamond_predictions= ['-'] fams= [nan] subfams= [nan]\n",
      "diamond_predictions= ['-'] fams= [nan] subfams= [nan]\n"
     ]
    }
   ],
   "source": [
    "for line in overview_file[1:25]:  # skip the first line becuase this is the head titles\n",
    "    line = line.split(\"\\t\")\n",
    "\n",
    "    # Retrieve domaind predications from DIAMOND\n",
    "    diamond_prediction = line[3].split('+')\n",
    "    cazy_fams = []\n",
    "    cazy_subfams = []\n",
    "    for pred in diamond_prediction:\n",
    "        if pred == '-':\n",
    "            cazy_fams.append(np.nan)\n",
    "            cazy_subfams.append(np.nan)\n",
    "        elif pred.find(\"_\") != -1:\n",
    "            cutoff = pred.find(\"_\")\n",
    "            cazy_fams.append(pred[:cutoff])\n",
    "            cazy_subfams.append(pred)\n",
    "        else:\n",
    "            cazy_fams.append(pred)\n",
    "            cazy_subfams.append(np.nan)\n",
    "    print(\"diamond_predictions=\", diamond_prediction, \"fams=\", cazy_fams, \"subfams=\", cazy_subfams)\n",
    "# extended number of lines in trial to make sure subfam capture worked"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**HOTPEP AGAIN**\n",
    "It could be useful to retrieve the predicated EC number from Hotpep.out (the output file for Hotpep).\n",
    "\n",
    "Like the overview.txt file, each protein is on a separate line and the data is separated by tables (`\\t`). The EC numbers are stored at the end of each line."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['CE0', '9', 'XP_001551543.2', '23.2', '48', 'HPTDVG,GHIKVA,VGHIKV,YKGRWD,DSKYVS,VAITFG,IKVASH,DIVVIN,VSWWSA,VLVGYR,KGRWDS,DTLYWN,SYKGRW,AGLGNT,GRWDSK,SWWSAP,DVGHIK,WWSAPG,GLGNTE,SKYVSW,THLLVS,HNDINP,ADIVVI,GNTEYS,KVASHL,NLGTND,HLLVSP,ATHLLV,RWDSKY,YSITAY,RVTNWA,TFELRV,YVSWWS,EYSITA,KYVSWW,TLYWND,YFNTTG,PTDVGH,LGNTEY,ITAYPG,GAGLGN,WDSKYV,YTSDTS,TDVGHI,HIKVAS,SITAYP,LLVSPE,IGDSLS', '3.1.1.-:17']\n",
      "['CE0', '16', 'XP_024553067.1', '52.7', '70', 'LGTSAG,PNWVEY,VEYLTS,LFMNLP,GGPNWV,FMNLPP,TSAGGP,QNYSFI,MDPSKT,GTSAGG,VQNQLG,DTDSYT,PPLQRT,AGGPNW,HNYTVS,AFGDSY,QNQLGT,DSYTFP,SLENQI,NWVEYL,GDTDSY,QLWDFA,WDFAFA,NYSFIG,DFAFAG,GINDIG,NFSYTP,FPSHNS,VAIWIG,AAIETI,NLPPLQ,FGDSYT,WVEYLT,TFPSHN,NQLGTS,LHHNYT,GLPSKC,IGINDI,LTSCFS,INDIGD,LWDFAF,GDSYTY,NDIGDT,AGSDVS,YSFIGD,AFAGSD,PSKTLV,AIWIGI,MNLPPL,GPNWVE,KQLWDF,FAFAGS,DSYTYV,LVAIWI,YTFPSH,FAAIET,IWIGIN,FSYTPA,SGLPSK,SFIGDL,GSDVST,SYTFPS,SAGGPN,LYTNII,QLGTSA,TSCFSG,DIGDTD,WIGIND,AIETIY,SYTYVQ', 'NA']\n",
      "['CE0', '16', 'XP_024553068.1', '52.7', '70', 'LGTSAG,PNWVEY,VEYLTS,LFMNLP,GGPNWV,FMNLPP,TSAGGP,QNYSFI,MDPSKT,GTSAGG,VQNQLG,DTDSYT,PPLQRT,AGGPNW,HNYTVS,AFGDSY,QNQLGT,DSYTFP,SLENQI,NWVEYL,GDTDSY,QLWDFA,WDFAFA,NYSFIG,DFAFAG,GINDIG,NFSYTP,FPSHNS,VAIWIG,AAIETI,NLPPLQ,FGDSYT,WVEYLT,TFPSHN,NQLGTS,LHHNYT,GLPSKC,IGINDI,LTSCFS,INDIGD,LWDFAF,GDSYTY,NDIGDT,AGSDVS,YSFIGD,AFAGSD,PSKTLV,AIWIGI,MNLPPL,GPNWVE,KQLWDF,FAFAGS,DSYTYV,LVAIWI,YTFPSH,FAAIET,IWIGIN,FSYTPA,SGLPSK,SFIGDL,GSDVST,SYTFPS,SAGGPN,LYTNII,QLGTSA,TSCFSG,DIGDTD,WIGIND,AIETIY,SYTYVQ', 'NA']\n",
      "['CE1', '14', 'XP_024553208.1', '6.13', '10', 'AMMTNV,TTLYPQ,DTTLYP,GTSSGA,SGAMMT,TLYPQN,GAMMTN,SSGAMM,TSSGAM,CWDVSS', '3.1.1.72:347, 3.1.1.-:53']\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CAZy Family</th>\n",
       "      <th>PPR Subfamily</th>\n",
       "      <th>Gene ID</th>\n",
       "      <th>Frequency</th>\n",
       "      <th>Hits</th>\n",
       "      <th>Signature Peptides</th>\n",
       "      <th>EC number</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CE0</td>\n",
       "      <td>9</td>\n",
       "      <td>XP_001551543.2</td>\n",
       "      <td>23.2</td>\n",
       "      <td>48</td>\n",
       "      <td>HPTDVG,GHIKVA,VGHIKV,YKGRWD,DSKYVS,VAITFG,IKVA...</td>\n",
       "      <td>3.1.1.-:17</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>CE0</td>\n",
       "      <td>16</td>\n",
       "      <td>XP_024553067.1</td>\n",
       "      <td>52.7</td>\n",
       "      <td>70</td>\n",
       "      <td>LGTSAG,PNWVEY,VEYLTS,LFMNLP,GGPNWV,FMNLPP,TSAG...</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>CE0</td>\n",
       "      <td>16</td>\n",
       "      <td>XP_024553068.1</td>\n",
       "      <td>52.7</td>\n",
       "      <td>70</td>\n",
       "      <td>LGTSAG,PNWVEY,VEYLTS,LFMNLP,GGPNWV,FMNLPP,TSAG...</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>CE1</td>\n",
       "      <td>14</td>\n",
       "      <td>XP_024553208.1</td>\n",
       "      <td>6.13</td>\n",
       "      <td>10</td>\n",
       "      <td>AMMTNV,TTLYPQ,DTTLYP,GTSSGA,SGAMMT,TLYPQN,GAMM...</td>\n",
       "      <td>3.1.1.72:347, 3.1.1.-:53</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  CAZy Family PPR Subfamily         Gene ID Frequency Hits  \\\n",
       "0         CE0             9  XP_001551543.2      23.2   48   \n",
       "1         CE0            16  XP_024553067.1      52.7   70   \n",
       "2         CE0            16  XP_024553068.1      52.7   70   \n",
       "3         CE1            14  XP_024553208.1      6.13   10   \n",
       "\n",
       "                                  Signature Peptides                 EC number  \n",
       "0  HPTDVG,GHIKVA,VGHIKV,YKGRWD,DSKYVS,VAITFG,IKVA...                3.1.1.-:17  \n",
       "1  LGTSAG,PNWVEY,VEYLTS,LFMNLP,GGPNWV,FMNLPP,TSAG...                        NA  \n",
       "2  LGTSAG,PNWVEY,VEYLTS,LFMNLP,GGPNWV,FMNLPP,TSAG...                        NA  \n",
       "3  AMMTNV,TTLYPQ,DTTLYP,GTSSGA,SGAMMT,TLYPQN,GAMM...  3.1.1.72:347, 3.1.1.-:53  "
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "with open(\"Hotpep.out\", \"r\") as hf:\n",
    "    hotpep_file = hf.read().splitlines()\n",
    "\n",
    "df_data = []\n",
    "column_names = hotpep_file[0].split(\"\\t\")\n",
    "for line in hotpep_file[1:5]:\n",
    "    line = line.split(\"\\t\")\n",
    "    df_data.append(line)\n",
    "    print(line)\n",
    "\n",
    "df = pd.DataFrame(df_data, columns=column_names)\n",
    "df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Looking at the output, the Gene ID is in the third column (`line[2]`) and the EC number in the last (`line[-1]`). When no EC number is predicated the string \"NA\" is given but this should be changed to a null value.\n",
    "\n",
    "Additionally, the EC number needs formating, becuase each predicted EC number is immediately followed by \":_score_\" which is not wanted.\n",
    "\n",
    "Arguably, it would be easier to retrieve all the data from the hotpep.out file, BUT, file only includes the highest scoring CAZy family and does not show the multi-domain prediction.\n",
    "\n",
    "Therefore, we need to retrieve the EC number and the \"Gene ID\" (protein name/id) of the associated protein, and make sure this data is linked."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CE0\n",
      "CE0\n",
      "CE0\n",
      "CE1\n",
      "CE12\n",
      "CE12\n",
      "CE12\n",
      "CE16\n",
      "CE16\n",
      "CE16\n",
      "CE16\n",
      "CE16\n",
      "CE2\n",
      "CE4\n",
      "CE4\n",
      "CE4\n",
      "CE4\n",
      "CE5\n",
      "CE5\n",
      "CE5\n",
      "CE5\n",
      "CE5\n",
      "CE5\n",
      "CE5\n",
      "CE5\n",
      "CE5\n",
      "CE5\n",
      "CE8\n",
      "CE8\n",
      "CE8\n",
      "CE8\n",
      "CE9\n",
      "CE9\n",
      "GH0\n",
      "GH0\n",
      "GH0\n",
      "GH0\n",
      "GH0\n",
      "GH0\n",
      "GH0\n",
      "GH1\n",
      "GH1\n",
      "GH1\n",
      "GH10\n",
      "GH10\n",
      "GH105\n",
      "GH105\n",
      "GH106\n",
      "GH106\n",
      "GH11\n",
      "GH11\n",
      "GH11\n",
      "GH114\n",
      "GH115\n",
      "GH12\n",
      "GH12\n",
      "GH12\n",
      "GH125\n",
      "GH125\n",
      "GH125\n",
      "GH127\n",
      "GH128\n",
      "GH128\n",
      "GH13\n",
      "GH13\n",
      "GH13\n",
      "GH13\n",
      "GH131\n",
      "GH131\n",
      "GH131\n",
      "GH132\n",
      "GH132\n",
      "GH132\n",
      "GH133\n",
      "GH135\n",
      "GH135\n",
      "GH145\n",
      "GH15\n",
      "GH15\n",
      "GH15\n",
      "GH15\n",
      "GH15\n",
      "GH15\n",
      "GH152\n",
      "GH154\n",
      "GH16\n",
      "GH16\n",
      "GH16\n",
      "GH16\n",
      "GH16\n",
      "GH16\n",
      "GH16\n",
      "GH16\n",
      "GH16\n",
      "GH16\n",
      "GH16\n",
      "GH16\n",
      "GH16\n",
      "GH16\n",
      "GH16\n",
      "GH16\n",
      "GH16\n",
      "GH17\n",
      "GH17\n",
      "GH17\n",
      "GH17\n",
      "GH17\n",
      "GH18\n",
      "GH18\n",
      "GH18\n",
      "GH18\n",
      "GH18\n",
      "GH18\n",
      "GH2\n",
      "GH2\n",
      "GH20\n",
      "GH20\n",
      "GH26\n",
      "GH27\n",
      "GH27\n",
      "GH27\n",
      "GH27\n",
      "GH28\n",
      "GH28\n",
      "GH28\n",
      "GH28\n",
      "GH28\n",
      "GH28\n",
      "GH28\n",
      "GH28\n",
      "GH28\n",
      "GH28\n",
      "GH28\n",
      "GH28\n",
      "GH28\n",
      "GH28\n",
      "GH28\n",
      "GH28\n",
      "GH3\n",
      "GH3\n",
      "GH3\n",
      "GH3\n",
      "GH3\n",
      "GH3\n",
      "GH3\n",
      "GH3\n",
      "GH3\n",
      "GH3\n",
      "GH3\n",
      "GH3\n",
      "GH3\n",
      "GH3\n",
      "GH3\n",
      "GH3\n",
      "GH3\n",
      "GH3\n",
      "GH31\n",
      "GH31\n",
      "GH31\n",
      "GH31\n",
      "GH31\n",
      "GH32\n",
      "GH32\n",
      "GH35\n",
      "GH35\n",
      "GH35\n",
      "GH35\n",
      "GH35\n",
      "GH36\n",
      "GH36\n",
      "GH37\n",
      "GH38\n",
      "GH38\n",
      "GH39\n",
      "GH43\n",
      "GH43\n",
      "GH43\n",
      "GH43\n",
      "GH43\n",
      "GH43\n",
      "GH43\n",
      "GH43\n",
      "GH45\n",
      "GH45\n",
      "GH47\n",
      "GH47\n",
      "GH47\n",
      "GH47\n",
      "GH47\n",
      "GH47\n",
      "GH47\n",
      "GH47\n",
      "GH47\n",
      "GH47\n",
      "GH5\n",
      "GH5\n",
      "GH5\n",
      "GH5\n",
      "GH5\n",
      "GH5\n",
      "GH5\n",
      "GH5\n",
      "GH5\n",
      "GH5\n",
      "GH5\n",
      "GH5\n",
      "GH5\n",
      "GH51\n",
      "GH51\n",
      "GH51\n",
      "GH53\n",
      "GH53\n",
      "GH54\n",
      "GH55\n",
      "GH55\n",
      "GH55\n",
      "GH55\n",
      "GH55\n",
      "GH6\n",
      "GH62\n",
      "GH63\n",
      "GH64\n",
      "GH64\n",
      "GH65\n",
      "GH7\n",
      "GH7\n",
      "GH7\n",
      "GH71\n",
      "GH71\n",
      "GH71\n",
      "GH71\n",
      "GH71\n",
      "GH71\n",
      "GH71\n",
      "GH71\n",
      "GH72\n",
      "GH72\n",
      "GH72\n",
      "GH72\n",
      "GH72\n",
      "GH76\n",
      "GH76\n",
      "GH76\n",
      "GH76\n",
      "GH76\n",
      "GH76\n",
      "GH76\n",
      "GH76\n",
      "GH76\n",
      "GH76\n",
      "GH78\n",
      "GH78\n",
      "GH78\n",
      "GH78\n",
      "GH78\n",
      "GH78\n",
      "GH79\n",
      "GH79\n",
      "GH81\n",
      "GH88\n",
      "GH89\n",
      "GH92\n",
      "GH92\n",
      "GH92\n",
      "GH92\n",
      "GH92\n",
      "GH93\n",
      "GH95\n",
      "GH95\n",
      "AA0\n",
      "AA1\n",
      "AA1\n",
      "AA1\n",
      "AA1\n",
      "AA1\n",
      "AA1\n",
      "AA1\n",
      "AA1\n",
      "AA1\n",
      "AA1\n",
      "AA1\n",
      "AA1\n",
      "AA1\n",
      "AA1\n",
      "AA1\n",
      "AA1\n",
      "AA11\n",
      "AA13\n",
      "AA16\n",
      "AA3\n",
      "AA3\n",
      "AA3\n",
      "AA3\n",
      "AA3\n",
      "AA3\n",
      "AA3\n",
      "AA3\n",
      "AA3\n",
      "AA3\n",
      "AA3\n",
      "AA3\n",
      "AA3\n",
      "AA3\n",
      "AA3\n",
      "AA3\n",
      "AA3\n",
      "AA3\n",
      "AA3\n",
      "AA3\n",
      "AA3\n",
      "AA3\n",
      "AA3\n",
      "AA3\n",
      "AA3\n",
      "AA3\n",
      "AA3\n",
      "AA4\n",
      "AA5\n",
      "AA5\n",
      "AA5\n",
      "AA5\n",
      "AA5\n",
      "AA5\n",
      "AA6\n",
      "AA7\n",
      "AA8\n",
      "AA8\n",
      "AA8\n",
      "AA9\n",
      "AA9\n",
      "AA9\n",
      "AA9\n",
      "AA9\n",
      "AA9\n",
      "AA9\n",
      "AA9\n",
      "AA9\n",
      "AA9\n",
      "PL1\n",
      "PL1\n",
      "PL1\n",
      "PL1\n",
      "PL1\n",
      "PL1\n",
      "PL3\n",
      "PL3\n",
      "PL7\n",
      "GT0\n",
      "GT0\n",
      "GT0\n",
      "GT0\n",
      "GT1\n",
      "GT1\n",
      "GT1\n",
      "GT1\n",
      "GT1\n",
      "GT1\n",
      "GT1\n",
      "GT1\n",
      "GT15\n",
      "GT15\n",
      "GT15\n",
      "GT15\n",
      "GT17\n",
      "GT17\n",
      "GT2\n",
      "GT2\n",
      "GT2\n",
      "GT2\n",
      "GT2\n",
      "GT2\n",
      "GT2\n",
      "GT2\n",
      "GT2\n",
      "GT2\n",
      "GT20\n",
      "GT20\n",
      "GT20\n",
      "GT20\n",
      "GT20\n",
      "GT21\n",
      "GT22\n",
      "GT22\n",
      "GT22\n",
      "GT22\n",
      "GT24\n",
      "GT25\n",
      "GT25\n",
      "GT3\n",
      "GT31\n",
      "GT31\n",
      "GT31\n",
      "GT31\n",
      "GT31\n",
      "GT32\n",
      "GT32\n",
      "GT32\n",
      "GT32\n",
      "GT32\n",
      "GT33\n",
      "GT34\n",
      "GT34\n",
      "GT35\n",
      "GT39\n",
      "GT39\n",
      "GT39\n",
      "GT4\n",
      "GT4\n",
      "GT4\n",
      "GT41\n",
      "GT48\n",
      "GT5\n",
      "GT50\n",
      "GT57\n",
      "GT57\n",
      "GT58\n",
      "GT59\n",
      "GT62\n",
      "GT62\n",
      "GT62\n",
      "GT66\n",
      "GT69\n",
      "GT69\n",
      "GT69\n",
      "GT69\n",
      "GT69\n",
      "GT71\n",
      "GT71\n",
      "GT71\n",
      "GT76\n",
      "GT8\n",
      "GT8\n",
      "GT8\n",
      "GT8\n",
      "GT90\n",
      "GT90\n",
      "GT90\n",
      "CBM1\n",
      "CBM1\n",
      "CBM1\n",
      "CBM1\n",
      "CBM1\n",
      "CBM1\n",
      "CBM1\n",
      "CBM1\n",
      "CBM1\n",
      "CBM1\n",
      "CBM1\n",
      "CBM1\n",
      "CBM1\n",
      "CBM1\n",
      "CBM1\n",
      "CBM1\n",
      "CBM1\n",
      "CBM1\n",
      "CBM1\n",
      "CBM1\n",
      "CBM1\n",
      "CBM1\n",
      "CBM1\n",
      "CBM1\n",
      "CBM1\n",
      "CBM1\n",
      "CBM1\n",
      "CBM1\n",
      "CBM1\n",
      "CBM1\n",
      "CBM1\n",
      "CBM1\n",
      "CBM1\n",
      "CBM1\n",
      "CBM1\n",
      "CBM1\n",
      "CBM1\n",
      "CBM1\n",
      "CBM1\n",
      "CBM1\n",
      "CBM1\n",
      "CBM1\n",
      "CBM1\n",
      "CBM1\n",
      "CBM1\n",
      "CBM1\n",
      "CBM1\n",
      "CBM1\n",
      "CBM10\n",
      "CBM10\n",
      "CBM10\n",
      "CBM12\n",
      "CBM12\n",
      "CBM12\n",
      "CBM13\n",
      "CBM13\n",
      "CBM13\n",
      "CBM13\n",
      "CBM13\n",
      "CBM16\n",
      "CBM16\n",
      "CBM16\n",
      "CBM18\n",
      "CBM18\n",
      "CBM18\n",
      "CBM18\n",
      "CBM18\n",
      "CBM18\n",
      "CBM18\n",
      "CBM18\n",
      "CBM18\n",
      "CBM19\n",
      "CBM19\n",
      "CBM19\n",
      "CBM2\n",
      "CBM2\n",
      "CBM2\n",
      "CBM2\n",
      "CBM20\n",
      "CBM20\n",
      "CBM20\n",
      "CBM20\n",
      "CBM20\n",
      "CBM20\n",
      "CBM20\n",
      "CBM20\n",
      "CBM20\n",
      "CBM20\n",
      "CBM20\n",
      "CBM20\n",
      "CBM20\n",
      "CBM21\n",
      "CBM22\n",
      "CBM24\n",
      "CBM24\n",
      "CBM24\n",
      "CBM24\n",
      "CBM24\n",
      "CBM24\n",
      "CBM24\n",
      "CBM27\n",
      "CBM3\n",
      "CBM32\n",
      "CBM32\n",
      "CBM32\n",
      "CBM32\n",
      "CBM32\n",
      "CBM35\n",
      "CBM35\n",
      "CBM35\n",
      "CBM35\n",
      "CBM36\n",
      "CBM38\n",
      "CBM39\n",
      "CBM42\n",
      "CBM43\n",
      "CBM43\n",
      "CBM46\n",
      "CBM48\n",
      "CBM48\n",
      "CBM48\n",
      "CBM5\n",
      "CBM5\n",
      "CBM5\n",
      "CBM5\n",
      "CBM51\n",
      "CBM51\n",
      "CBM51\n",
      "CBM54\n",
      "CBM55\n",
      "CBM55\n",
      "CBM57\n",
      "CBM57\n",
      "CBM57\n",
      "CBM57\n",
      "CBM57\n",
      "CBM57\n",
      "CBM57\n",
      "CBM57\n",
      "CBM6\n",
      "CBM6\n",
      "CBM6\n",
      "CBM6\n",
      "CBM6\n",
      "CBM6\n",
      "CBM6\n",
      "CBM60\n",
      "CBM63\n",
      "CBM66\n",
      "CBM67\n",
      "CBM9\n"
     ]
    }
   ],
   "source": [
    "for line in hotpep_file[1:]:\n",
    "    line = line.split(\"\\t\")\n",
    "    print(line[0])\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'XP_001551543.2': ['3.1.1.-'],\n",
       " 'XP_024553067.1': [nan],\n",
       " 'XP_024553068.1': [nan],\n",
       " 'XP_024553208.1': ['3.1.1.72', ' 3.1.1.-']}"
      ]
     },
     "execution_count": 29,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "hotpep_ec_predictions = {}\n",
    "for line in hotpep_file[1:5]:\n",
    "    line = line.split(\"\\t\")\n",
    "    ec_numbers = line[-1]\n",
    "    ec_numbers = ec_numbers.split(\",\")  # multiple EC numbers may be predicted\n",
    "    # formate EC numbers\n",
    "    index = 0\n",
    "    for index in range(len(ec_numbers)):\n",
    "        ec = ec_numbers[index].split(\":\")[0] # remove the (\":score\") from the EC number\n",
    "        if ec == \"NA\":  # change represented null value to true null value\n",
    "            ec_numbers[index] = np.nan\n",
    "        else:\n",
    "            ec_numbers[index] = ec\n",
    "    gene_id = line[2]\n",
    "    hotpep_ec_predictions[gene_id] = ec_numbers\n",
    "\n",
    "hotpep_ec_predictions"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "I do not want to rely on the order proteins are presented in the Hotpep.out file being the exactly same order they are presented in the overview.txt file, hence the importance of associating the predicated EC numbers and gene ID. These gene IDs can be used to make sure that the correct EC number predicitons are given to the correct proteins. To do this a single dataframe containing the Hotpep data is needed, and a column (either empty or pre-populated with null values) added to have the predicated EC numbers added/inserted into.\n",
    "\n",
    "Therefore, the next task is to build a dataframe each for the HMMER, DIAMOND and Hotpep from the overview.txt file.\n",
    "\n",
    "What needs to be considered at this stage is that for eCAMI the output produces the main (or heighest scoring) CAZy prediction then additional predictions. However, the run_dbcan.py script shows that the predicted CAZy families are taken 'as is' straight from the prediction tool ouput for HMMER, DIAMOND and Hotpep, meaning they do not appear to be order with the heighest scoring family listed first. -- NEED TO CHECK THE PAPERS THAT DISCUSS THE TOOLS INDIVIDUALLY!!!\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "done\n"
     ]
    }
   ],
   "source": [
    "def parse_dbcan_output(overview_file_path):\n",
    "    \"\"\"Parse the output from the run_dbCAN overview.txt file, writing a dataframe from each prediciotn tool.\n",
    "    \n",
    "    :param overview_file_path: path, path to the output 'overview.txt' file\n",
    "    \n",
    "    Return 3 dataframes, one each for HMMER, Hotpep and DIAMOND\n",
    "    \"\"\"\n",
    "    with open(overview_file_path, \"r\") as fh:\n",
    "        overview_file = fh.read().splitlines()\n",
    "        \n",
    "    # build empty dataframes to add the repsective prediction tools output to\n",
    "    hmmer_df = pd.DataFrame({}, columns=[\"protein_accession\", \"cazy_family\", \"cazy_subfamily\", \"domain_range\"])\n",
    "    hotpep_df = pd.DataFrame({}, columns=[\"protein_accession\", \"cazy_family\", \"cazy_subfamily\"]) # add ec later\n",
    "    diamond_df = pd.DataFrame({}, columns=[\"protein_accession\", \"cazy_family\", \"cazy_subfamily\"])\n",
    "\n",
    "    for line in overview_file[1:]: # skip the first line becuase this is the head titles\n",
    "        line = line.split(\"\\t\")\n",
    "        \n",
    "        # retrieve the data for HMMER\n",
    "        hmmer_df = hmmer_df.append(parse_hmmer_output(line), ignore_index=True)\n",
    "        \n",
    "        # retrieve the data for Hotpep\n",
    "        hotpep_df = hotpep_df.append(parse_hotpep_output(line), ignore_index=True)\n",
    "        \n",
    "        # retrieve the data for DIAMOND\n",
    "        diamond_df = diamond_df.append(parse_diamond_output(line), ignore_index=True)\n",
    "\n",
    "    return hmmer_df, hotpep_df, diamond_df\n",
    "\n",
    "\n",
    "def parse_hmmer_output(line):\n",
    "    \"\"\"Parse the output from HMMER from the overview.txt file.\n",
    "    \n",
    "    Retrieve the protein accession, predicated CAZy families, accomanying subfamiles and domain ranges.\n",
    "    Multiple domains can be predicated, and the families, subfamiles and ranges will be collected as\n",
    "    as lists, with the index for each item in each list being associated with the matching items (by index)\n",
    "    in the other lists. For example, the first item in every list will contain the prediciton for the first\n",
    "    predicated domain, the second item the second prediction etc.\n",
    "    \n",
    "    :param line: str, line from the dbcan overview.txt file\n",
    "    \n",
    "    Return pandas dataframe.\"\"\"\n",
    "    # Retrieve predictions from HMMER (predicated domains and domain ranges)\n",
    "    # separate out each of the domain predications\n",
    "    hmmer_prediction = line[1].split(\"+\")\n",
    "    \n",
    "    # create empty lists to store predicated domains and ranges\n",
    "    cazy_fam = []\n",
    "    cazy_subfam = []\n",
    "    domain_ranges = []\n",
    "\n",
    "    # separate out the name of the predicated domain (the CAZy family) and the domain's AA range\n",
    "    for domain in hmmer_prediction:\n",
    "        domain = domain.split(\"(\")\n",
    "        # standardise name if necessary, retrieve subfam if preciated, and produce null values\n",
    "        domain_name = domain[0]\n",
    "        if domain_name.startswith(\"GT2_Chitin_\"):\n",
    "            cazy_fam.append(\"GT2\")\n",
    "            cazy_subfam.append(np.nan)\n",
    "\n",
    "        elif domain_name == '-':\n",
    "            cazy_fam.append(np.nan)\n",
    "            cazy_subfam.append(np.nan)\n",
    "\n",
    "        elif domain_name.find(\"_\") != -1:\n",
    "            cutoff = domain_name.find(\"_\")\n",
    "            cazy_fam.append(domain_name[:cutoff])\n",
    "            cazy_subfam.append(domain_name)\n",
    "\n",
    "        else:\n",
    "            cazy_fam.append(domain_name)\n",
    "            cazy_subfam.append(np.nan)\n",
    "\n",
    "        try:\n",
    "            if domain[1] == \"-\":\n",
    "                domain_ranges.append(np.nan)\n",
    "            else:\n",
    "                domain_ranges.append(domain[1][:-1])  # exlude the final \")\"\n",
    "        except IndexError: # no domain predicated so no domain ranges\n",
    "            domain_ranges.append(np.nan)\n",
    "\n",
    "        \n",
    "    prediction_dict = {\n",
    "        \"protein_accession\": [line[0]],\n",
    "        \"cazy_family\": [cazy_fam],\n",
    "        \"cazy_subfamily\": [cazy_subfam],\n",
    "        \"domain_range\": [domain_ranges],\n",
    "    }\n",
    "\n",
    "    prediction_df = pd.DataFrame(prediction_dict)\n",
    "    \n",
    "    return prediction_df\n",
    "\n",
    "\n",
    "def parse_hotpep_output(line):\n",
    "    \"\"\"Parse the output from Hotpep from the overview.txt file.\n",
    "    \n",
    "    Retrieve the protein accession, predicated CAZy families, accomanying subfamiles.\n",
    "    Multiple domains can be predicated, and the families and subfamiles will be collected as\n",
    "    as lists, with the index for each item in each list being associated with the matching items (by index)\n",
    "    in the other lists. For example, the first item in every list will contain the prediciton for the first\n",
    "    predicated domain, the second item the second prediction etc.\n",
    "    \n",
    "    :param line: str, line from the dbcan overview.txt file\n",
    "    \n",
    "    Return pandas dataframe.\"\"\"\n",
    "    # Retrieve predictions from Hotpep\n",
    "    hotpep_prediction = line[2].split(\"+\")\n",
    "\n",
    "    # create empty lists to separate the predicated CAZy fam and subfam\n",
    "    cazy_fams = []\n",
    "    cazy_subfams = []\n",
    "\n",
    "    # remove the k-mer cluster number\n",
    "    for prediction in hotpep_prediction:\n",
    "        prediction = prediction.split(\"(\")[0]\n",
    "\n",
    "        if prediction == \"-\":\n",
    "            cazy_fams.append(np.nan)\n",
    "            cazy_subfams.append(np.nan)\n",
    "\n",
    "        elif prediction.find(\"_\") != -1:\n",
    "            cutoff = prediction.find(\"_\")\n",
    "            cazy_fams.append(predictions[:cutoff])\n",
    "            cazy_subfams.append(prediction)\n",
    "\n",
    "        else:\n",
    "            cazy_fams.append(prediction)\n",
    "            cazy_subfams.append(np.nan)\n",
    "\n",
    "    prediction_dict = {\n",
    "        \"protein_accession\": [line[0]],\n",
    "        \"cazy_family\": [cazy_fams],\n",
    "        \"cazy_subfamily\": [cazy_subfams],\n",
    "    }\n",
    "    \n",
    "    prediction_df = pd.DataFrame(prediction_dict)\n",
    "    \n",
    "    return prediction_df\n",
    "\n",
    "\n",
    "def parse_diamond_output(line):\n",
    "    \"\"\"Parse the output from DIAMOND from the overview.txt file.\n",
    "    \n",
    "    Retrieve the protein accession, predicated CAZy families, accomanying subfamiles.\n",
    "    Multiple domains can be predicated, and the families and subfamiles will be collected as\n",
    "    as lists, with the index for each item in each list being associated with the matching items (by index)\n",
    "    in the other lists. For example, the first item in every list will contain the prediciton for the first\n",
    "    predicated domain, the second item the second prediction etc.\n",
    "    \n",
    "    :param line: str, line from the dbcan overview.txt file\n",
    "    \n",
    "    Return pandas dataframe.\"\"\"\n",
    "    \n",
    "    # Retrieve domaind predications from DIAMOND\n",
    "    diamond_prediction = line[3].split('+')\n",
    "\n",
    "    # create empty lists to separate the CAZy family and subfamiles\n",
    "    cazy_fams = []\n",
    "    cazy_subfams = []\n",
    "\n",
    "    for pred in diamond_prediction:\n",
    "        if pred == '-':\n",
    "            cazy_fams.append(np.nan)\n",
    "            cazy_subfams.append(np.nan)\n",
    "\n",
    "        elif pred.find(\"_\") != -1:\n",
    "            cutoff = pred.find(\"_\")\n",
    "            cazy_fams.append(pred[:cutoff])\n",
    "            cazy_subfams.append(pred)\n",
    "\n",
    "        else:\n",
    "            cazy_fams.append(pred)\n",
    "            cazy_subfams.append(np.nan)\n",
    "\n",
    "    prediction_dict = {\n",
    "        \"protein_accession\": [line[0]],\n",
    "        \"cazy_family\": [cazy_fams],\n",
    "        \"cazy_subfamily\": [cazy_subfams],\n",
    "    }\n",
    "    \n",
    "    prediction_df = pd.DataFrame(prediction_dict)\n",
    "    \n",
    "    return prediction_df\n",
    "\n",
    "\n",
    "hmmer_df, hotpep_df, diamond_df = parse_dbcan_output(\"overview.txt\")\n",
    "print(\"done\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>protein_accession</th>\n",
       "      <th>cazy_family</th>\n",
       "      <th>cazy_subfamily</th>\n",
       "      <th>domain_range</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>XP_001545265.2</td>\n",
       "      <td>[GH53]</td>\n",
       "      <td>[nan]</td>\n",
       "      <td>[20-346]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>XP_001545514.1</td>\n",
       "      <td>[GT2]</td>\n",
       "      <td>[nan]</td>\n",
       "      <td>[1213-1720]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>XP_001545605.1</td>\n",
       "      <td>[AA4]</td>\n",
       "      <td>[nan]</td>\n",
       "      <td>[28-577]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>XP_001545700.1</td>\n",
       "      <td>[CBM21]</td>\n",
       "      <td>[nan]</td>\n",
       "      <td>[353-459]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>XP_001545706.2</td>\n",
       "      <td>[GH132]</td>\n",
       "      <td>[nan]</td>\n",
       "      <td>[57-318]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>701</th>\n",
       "      <td>XP_024553622.1</td>\n",
       "      <td>[nan]</td>\n",
       "      <td>[nan]</td>\n",
       "      <td>[nan]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>702</th>\n",
       "      <td>XP_001546886.1</td>\n",
       "      <td>[nan]</td>\n",
       "      <td>[nan]</td>\n",
       "      <td>[nan]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>703</th>\n",
       "      <td>XP_024553655.1</td>\n",
       "      <td>[nan]</td>\n",
       "      <td>[nan]</td>\n",
       "      <td>[nan]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>704</th>\n",
       "      <td>XP_024553815.1</td>\n",
       "      <td>[nan]</td>\n",
       "      <td>[nan]</td>\n",
       "      <td>[nan]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>705</th>\n",
       "      <td>XP_024553814.1</td>\n",
       "      <td>[nan]</td>\n",
       "      <td>[nan]</td>\n",
       "      <td>[nan]</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>706 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    protein_accession cazy_family cazy_subfamily domain_range\n",
       "0      XP_001545265.2      [GH53]          [nan]     [20-346]\n",
       "1      XP_001545514.1       [GT2]          [nan]  [1213-1720]\n",
       "2      XP_001545605.1       [AA4]          [nan]     [28-577]\n",
       "3      XP_001545700.1     [CBM21]          [nan]    [353-459]\n",
       "4      XP_001545706.2     [GH132]          [nan]     [57-318]\n",
       "..                ...         ...            ...          ...\n",
       "701    XP_024553622.1       [nan]          [nan]        [nan]\n",
       "702    XP_001546886.1       [nan]          [nan]        [nan]\n",
       "703    XP_024553655.1       [nan]          [nan]        [nan]\n",
       "704    XP_024553815.1       [nan]          [nan]        [nan]\n",
       "705    XP_024553814.1       [nan]          [nan]        [nan]\n",
       "\n",
       "[706 rows x 4 columns]"
      ]
     },
     "execution_count": 31,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "hmmer_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>protein_accession</th>\n",
       "      <th>cazy_family</th>\n",
       "      <th>cazy_subfamily</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>XP_001545265.2</td>\n",
       "      <td>[GH53]</td>\n",
       "      <td>[nan]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>XP_001545514.1</td>\n",
       "      <td>[GT2]</td>\n",
       "      <td>[nan]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>XP_001545605.1</td>\n",
       "      <td>[AA4]</td>\n",
       "      <td>[nan]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>XP_001545700.1</td>\n",
       "      <td>[CBM21]</td>\n",
       "      <td>[nan]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>XP_001545706.2</td>\n",
       "      <td>[GH132]</td>\n",
       "      <td>[nan]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>701</th>\n",
       "      <td>XP_024553622.1</td>\n",
       "      <td>[nan]</td>\n",
       "      <td>[nan]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>702</th>\n",
       "      <td>XP_001546886.1</td>\n",
       "      <td>[nan]</td>\n",
       "      <td>[nan]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>703</th>\n",
       "      <td>XP_024553655.1</td>\n",
       "      <td>[nan]</td>\n",
       "      <td>[nan]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>704</th>\n",
       "      <td>XP_024553815.1</td>\n",
       "      <td>[nan]</td>\n",
       "      <td>[nan]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>705</th>\n",
       "      <td>XP_024553814.1</td>\n",
       "      <td>[nan]</td>\n",
       "      <td>[nan]</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>706 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    protein_accession cazy_family cazy_subfamily\n",
       "0      XP_001545265.2      [GH53]          [nan]\n",
       "1      XP_001545514.1       [GT2]          [nan]\n",
       "2      XP_001545605.1       [AA4]          [nan]\n",
       "3      XP_001545700.1     [CBM21]          [nan]\n",
       "4      XP_001545706.2     [GH132]          [nan]\n",
       "..                ...         ...            ...\n",
       "701    XP_024553622.1       [nan]          [nan]\n",
       "702    XP_001546886.1       [nan]          [nan]\n",
       "703    XP_024553655.1       [nan]          [nan]\n",
       "704    XP_024553815.1       [nan]          [nan]\n",
       "705    XP_024553814.1       [nan]          [nan]\n",
       "\n",
       "[706 rows x 3 columns]"
      ]
     },
     "execution_count": 32,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "hotpep_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>protein_accession</th>\n",
       "      <th>cazy_family</th>\n",
       "      <th>cazy_subfamily</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>XP_001545265.2</td>\n",
       "      <td>[GH53]</td>\n",
       "      <td>[nan]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>XP_001545514.1</td>\n",
       "      <td>[GT2]</td>\n",
       "      <td>[nan]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>XP_001545605.1</td>\n",
       "      <td>[AA4]</td>\n",
       "      <td>[nan]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>XP_001545700.1</td>\n",
       "      <td>[CBM21]</td>\n",
       "      <td>[nan]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>XP_001545706.2</td>\n",
       "      <td>[GH132]</td>\n",
       "      <td>[nan]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>701</th>\n",
       "      <td>XP_024553622.1</td>\n",
       "      <td>[GH30]</td>\n",
       "      <td>[GH30_7]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>702</th>\n",
       "      <td>XP_001546886.1</td>\n",
       "      <td>[GH30]</td>\n",
       "      <td>[GH30_7]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>703</th>\n",
       "      <td>XP_024553655.1</td>\n",
       "      <td>[CBM24, GH71]</td>\n",
       "      <td>[nan, nan]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>704</th>\n",
       "      <td>XP_024553815.1</td>\n",
       "      <td>[CBM13]</td>\n",
       "      <td>[nan]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>705</th>\n",
       "      <td>XP_024553814.1</td>\n",
       "      <td>[CBM13]</td>\n",
       "      <td>[nan]</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>706 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    protein_accession    cazy_family cazy_subfamily\n",
       "0      XP_001545265.2         [GH53]          [nan]\n",
       "1      XP_001545514.1          [GT2]          [nan]\n",
       "2      XP_001545605.1          [AA4]          [nan]\n",
       "3      XP_001545700.1        [CBM21]          [nan]\n",
       "4      XP_001545706.2        [GH132]          [nan]\n",
       "..                ...            ...            ...\n",
       "701    XP_024553622.1         [GH30]       [GH30_7]\n",
       "702    XP_001546886.1         [GH30]       [GH30_7]\n",
       "703    XP_024553655.1  [CBM24, GH71]     [nan, nan]\n",
       "704    XP_024553815.1        [CBM13]          [nan]\n",
       "705    XP_024553814.1        [CBM13]          [nan]\n",
       "\n",
       "[706 rows x 3 columns]"
      ]
     },
     "execution_count": 33,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "diamond_df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "This produces a dataframe each for DIAMOND, HMMER and Hotpep. But we want to add the predicted EC numbers from Hotpep.out to the Hotpep dataframe."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>protein_accession</th>\n",
       "      <th>cazy_family</th>\n",
       "      <th>cazy_subfamily</th>\n",
       "      <th>ec_number</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>XP_001545265.2</td>\n",
       "      <td>[GH53]</td>\n",
       "      <td>[nan]</td>\n",
       "      <td>3.2.1.89</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>XP_001545514.1</td>\n",
       "      <td>[GT2]</td>\n",
       "      <td>[nan]</td>\n",
       "      <td>2.4.1.16</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>XP_001545605.1</td>\n",
       "      <td>[AA4]</td>\n",
       "      <td>[nan]</td>\n",
       "      <td>1.1.3.38</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>XP_001545700.1</td>\n",
       "      <td>[CBM21]</td>\n",
       "      <td>[nan]</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>XP_001545706.2</td>\n",
       "      <td>[GH132]</td>\n",
       "      <td>[nan]</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>701</th>\n",
       "      <td>XP_024553622.1</td>\n",
       "      <td>[nan]</td>\n",
       "      <td>[nan]</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>702</th>\n",
       "      <td>XP_001546886.1</td>\n",
       "      <td>[nan]</td>\n",
       "      <td>[nan]</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>703</th>\n",
       "      <td>XP_024553655.1</td>\n",
       "      <td>[nan]</td>\n",
       "      <td>[nan]</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>704</th>\n",
       "      <td>XP_024553815.1</td>\n",
       "      <td>[nan]</td>\n",
       "      <td>[nan]</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>705</th>\n",
       "      <td>XP_024553814.1</td>\n",
       "      <td>[nan]</td>\n",
       "      <td>[nan]</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>706 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    protein_accession cazy_family cazy_subfamily ec_number\n",
       "0      XP_001545265.2      [GH53]          [nan]  3.2.1.89\n",
       "1      XP_001545514.1       [GT2]          [nan]  2.4.1.16\n",
       "2      XP_001545605.1       [AA4]          [nan]  1.1.3.38\n",
       "3      XP_001545700.1     [CBM21]          [nan]       NaN\n",
       "4      XP_001545706.2     [GH132]          [nan]       NaN\n",
       "..                ...         ...            ...       ...\n",
       "701    XP_024553622.1       [nan]          [nan]       NaN\n",
       "702    XP_001546886.1       [nan]          [nan]       NaN\n",
       "703    XP_024553655.1       [nan]          [nan]       NaN\n",
       "704    XP_024553815.1       [nan]          [nan]       NaN\n",
       "705    XP_024553814.1       [nan]          [nan]       NaN\n",
       "\n",
       "[706 rows x 4 columns]"
      ]
     },
     "execution_count": 34,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# add a column full of null values for EC numbers\n",
    "hotpep_df[\"ec_number\"] = np.nan\n",
    "\n",
    "# retrieve EC numbers from Hotpep.out and add to the corresponding protein in the Hotpep df\n",
    "\n",
    "\n",
    "def add_hotpep_ec_predictions(hotpep_output_file, hotpep_df):\n",
    "    \"\"\"Retrieve predicated EC numbers from Hotpep output file and add to the Hotpep dataframe.\n",
    "    \n",
    "    :param hotpep_output_file: path, path to Hotpep.out file\n",
    "    :param hotpep_df: pandas dataframe, containing Hotpep predicated CAZy families\n",
    "    \n",
    "    Return pandas dataframe (hotpep_df with EC number predictions)\n",
    "    \"\"\"\n",
    "    with open(hotpep_output_file, \"r\") as fh:\n",
    "        hotpep_file = fh.read().splitlines()\n",
    "        \n",
    "        ec_predictions = {}\n",
    "        \n",
    "        for line in hotpep_file[1:]:  # skip the first line which contains the titles\n",
    "            line = line.split(\"\\t\")\n",
    "            \n",
    "            ec_numbers = line[-1].split(\",\") # multiple EC numbers may be predicted\n",
    "            \n",
    "            # remove (\":score\") from each EC number, and convert \"NA\" to proper null value\n",
    "            index = 0\n",
    "            for index in range(len(ec_numbers)):\n",
    "                ec = ec_numbers[index].split(\":\")[0] # remove the (\":score\") from the EC number\n",
    "                \n",
    "                if ec == \"NA\":\n",
    "                    ec_numbers[index] = np.nan\n",
    "                \n",
    "                else:\n",
    "                    ec_numbers[index] = ec\n",
    "            \n",
    "            protein_accession = line[2]\n",
    "            ec_predictions[protein_accession] = ec_numbers\n",
    "        \n",
    "\n",
    "    for protein_accession in ec_predictions:  # dict {protein_accession: [predicated EC#s]}\n",
    "        # Check if there are EC numbers to add\n",
    "        if type(ec_predictions[protein_accession][0]) == float:  # EC numbers is not stored as null value\n",
    "            continue\n",
    "            \n",
    "        # find index of row in hotpep_df with the same protein accession\n",
    "        row_index = hotpep_df.index[hotpep_df[\"protein_accession\"] == protein_accession].tolist()[0]\n",
    "        # add EC numbers to this row\n",
    "        string = \"\"\n",
    "        for item in ec_predictions[protein_accession][:-1]:\n",
    "            string += f\"{item},\"\n",
    "        string += ec_predictions[protein_accession][((len(ec_predictions[protein_accession])) - 1)]\n",
    "        last_item_index = ((len(ec_predictions[protein_accession])) - 1)\n",
    "        hotpep_df.iloc[row_index, 3] = string\n",
    "    \n",
    "    return hotpep_df\n",
    "\n",
    "hotpep_df = add_hotpep_ec_predictions(\"Hotpep.out\", hotpep_df)\n",
    "# returns the dict here so can check values were added ok\n",
    "hotpep_df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**dbCAN consensus result**\n",
    "\n",
    "The final task is to create a consensus result for the dbCAN tool.\n",
    "\n",
    "A consensus result will be defined as the result of which 2 or more tools agree. For example, if only the #ofTools count for a protein is 1 then the dbCAN consensus is that the query protein is not a CAZyme - that's the easy one to determine.\n",
    "\n",
    "The difficulty arises when #ofTools => 2 but the tools predicated domains, or one predicated multiple domains and the other only one. This will involve checking if a domain is listed in the others predictions.\n",
    "\n",
    "Again this involves parsing the overview.txt file. At the end this will be added to the parse_dbcan_output() function but for now it will be developed in isolation.\n",
    "\n",
    "Another issue to deal with is that the dataframes containing the standardised output from tool should be in the same order, becusae we iterated through the lines of the overview file added the data from each line to the hmmer, hotpep and diamond dfs at the same time. Therefore, you could use the dataframes to get the standardised results for each prediciton tool and find the consensus, or build the dbCAN df at the same times as the others. The former method appears to be potentially brittle and thus the latter approach will be taken.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>protein_accession</th>\n",
       "      <th>cazy_family_3</th>\n",
       "      <th>cazy_subfamily_3</th>\n",
       "      <th>cazy_family_2</th>\n",
       "      <th>cazy_subfamily_2</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>XP_001545265.2</td>\n",
       "      <td>[GH53]</td>\n",
       "      <td>[nan]</td>\n",
       "      <td>[nan]</td>\n",
       "      <td>[nan]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>XP_001545514.1</td>\n",
       "      <td>[GT2]</td>\n",
       "      <td>[nan]</td>\n",
       "      <td>[nan]</td>\n",
       "      <td>[nan]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>XP_001545605.1</td>\n",
       "      <td>[AA4]</td>\n",
       "      <td>[nan]</td>\n",
       "      <td>[nan]</td>\n",
       "      <td>[nan]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>XP_001545700.1</td>\n",
       "      <td>[CBM21]</td>\n",
       "      <td>[nan]</td>\n",
       "      <td>[nan]</td>\n",
       "      <td>[nan]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>XP_001545706.2</td>\n",
       "      <td>[GH132]</td>\n",
       "      <td>[nan]</td>\n",
       "      <td>[nan]</td>\n",
       "      <td>[nan]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>504</th>\n",
       "      <td>XP_001557402.1</td>\n",
       "      <td>[nan]</td>\n",
       "      <td>[nan]</td>\n",
       "      <td>[GT41]</td>\n",
       "      <td>[nan]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>505</th>\n",
       "      <td>XP_024547082.1</td>\n",
       "      <td>[nan]</td>\n",
       "      <td>[nan]</td>\n",
       "      <td>[CBM1]</td>\n",
       "      <td>[nan]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>506</th>\n",
       "      <td>XP_024550885.1</td>\n",
       "      <td>[nan]</td>\n",
       "      <td>[nan]</td>\n",
       "      <td>[CBM18]</td>\n",
       "      <td>[nan]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>507</th>\n",
       "      <td>XP_001551566.1</td>\n",
       "      <td>[nan]</td>\n",
       "      <td>[nan]</td>\n",
       "      <td>[CBM18]</td>\n",
       "      <td>[nan]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>508</th>\n",
       "      <td>XP_001555374.1</td>\n",
       "      <td>[nan]</td>\n",
       "      <td>[nan]</td>\n",
       "      <td>[CBM48]</td>\n",
       "      <td>[nan]</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>509 rows × 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    protein_accession cazy_family_3 cazy_subfamily_3 cazy_family_2  \\\n",
       "0      XP_001545265.2        [GH53]            [nan]         [nan]   \n",
       "1      XP_001545514.1         [GT2]            [nan]         [nan]   \n",
       "2      XP_001545605.1         [AA4]            [nan]         [nan]   \n",
       "3      XP_001545700.1       [CBM21]            [nan]         [nan]   \n",
       "4      XP_001545706.2       [GH132]            [nan]         [nan]   \n",
       "..                ...           ...              ...           ...   \n",
       "504    XP_001557402.1         [nan]            [nan]        [GT41]   \n",
       "505    XP_024547082.1         [nan]            [nan]        [CBM1]   \n",
       "506    XP_024550885.1         [nan]            [nan]       [CBM18]   \n",
       "507    XP_001551566.1         [nan]            [nan]       [CBM18]   \n",
       "508    XP_001555374.1         [nan]            [nan]       [CBM48]   \n",
       "\n",
       "    cazy_subfamily_2  \n",
       "0              [nan]  \n",
       "1              [nan]  \n",
       "2              [nan]  \n",
       "3              [nan]  \n",
       "4              [nan]  \n",
       "..               ...  \n",
       "504            [nan]  \n",
       "505            [nan]  \n",
       "506            [nan]  \n",
       "507            [nan]  \n",
       "508            [nan]  \n",
       "\n",
       "[509 rows x 5 columns]"
      ]
     },
     "execution_count": 35,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def parse_dbcan_output(overview_file_path):\n",
    "    \"\"\"Parse the output from the run_dbCAN overview.txt file, writing a dataframe from each prediciotn tool.\n",
    "    \n",
    "    :param overview_file_path: path, path to the output 'overview.txt' file\n",
    "    \n",
    "    Return 4 dataframes, one each for consensus dbCAN result, HMMER, Hotpep and DIAMOND\n",
    "    \"\"\"\n",
    "    with open(overview_file_path, \"r\") as fh:\n",
    "        overview_file = fh.read().splitlines()\n",
    "        \n",
    "    # build empty dataframes to add the repsective prediction tools output to\n",
    "    hmmer_df = pd.DataFrame({}, columns=[\"protein_accession\", \"cazy_family\", \"cazy_subfamily\", \"domain_range\"])\n",
    "    hotpep_df = pd.DataFrame({}, columns=[\"protein_accession\", \"cazy_family\", \"cazy_subfamily\"]) # add ec later\n",
    "    diamond_df = pd.DataFrame({}, columns=[\"protein_accession\", \"cazy_family\", \"cazy_subfamily\"])\n",
    "    dbcan_df = pd.DataFrame({}, columns=[\"protein_accession\", \"cazy_family_3\", \"cazy_subfamily_3\", \"cazy_family_2\", \"cazy_subfamily_2\"])\n",
    "\n",
    "    for line in overview_file[1:]: # skip the first line becuase this is the head titles\n",
    "        line = line.split(\"\\t\")\n",
    "        \n",
    "        # retrieve the data for HMMER\n",
    "        hmmer_temp_df = parse_hmmer_output(line)\n",
    "        hmmer_df = hmmer_df.append(hmmer_temp_df, ignore_index=True)\n",
    "        \n",
    "        # retrieve the data for Hotpep\n",
    "        hotpep_temp_df = parse_hotpep_output(line)\n",
    "        hotpep_df = hotpep_df.append(hotpep_temp_df, ignore_index=True)\n",
    "        \n",
    "        # retrieve the data for DIAMOND\n",
    "        diamond_temp_df = parse_diamond_output(line)\n",
    "        diamond_df = diamond_df.append(diamond_temp_df, ignore_index=True)\n",
    "        \n",
    "        # retrieve consensus results for dbCAN\n",
    "        if line[-1] != \"1\":   #ofTools != 1, if ==1, same approach as other tools, do not include if non-CAZyme prediction\n",
    "            fam_consensus_3, fam_consensus_2 = get_dbcan_consensus(\n",
    "                hmmer_temp_df.iloc[0, 1],\n",
    "                hotpep_temp_df.iloc[0, 1],\n",
    "                diamond_temp_df.iloc[0, 1],\n",
    "            )\n",
    "            subfam_consensus_3, subfam_consensus_2 = get_dbcan_consensus(\n",
    "                hmmer_temp_df.iloc[0, 2],\n",
    "                hotpep_temp_df.iloc[0, 2],\n",
    "                diamond_temp_df.iloc[0, 2],\n",
    "            )\n",
    "\n",
    "            consensus_dict = {\n",
    "                \"protein_accession\": [line[0]],\n",
    "                \"cazy_family_3\": [fam_consensus_3],\n",
    "                \"cazy_subfamily_3\": [subfam_consensus_3],\n",
    "                \"cazy_family_2\": [fam_consensus_2],\n",
    "                \"cazy_subfamily_2\": [subfam_consensus_2],\n",
    "            }\n",
    "            \n",
    "            temp_consensus_df = pd.DataFrame(consensus_dict)\n",
    "            \n",
    "            dbcan_df = dbcan_df.append(temp_consensus_df, ignore_index=True)\n",
    "\n",
    "    return dbcan_df, hmmer_df, hotpep_df, diamond_df\n",
    "\n",
    "\n",
    "def parse_hmmer_output(line):\n",
    "    \"\"\"Parse the output from HMMER from the overview.txt file.\n",
    "    \n",
    "    Retrieve the protein accession, predicated CAZy families, accomanying subfamiles and domain ranges.\n",
    "    Multiple domains can be predicated, and the families, subfamiles and ranges will be collected as\n",
    "    as lists, with the index for each item in each list being associated with the matching items (by index)\n",
    "    in the other lists. For example, the first item in every list will contain the prediciton for the first\n",
    "    predicated domain, the second item the second prediction etc.\n",
    "    \n",
    "    :param line: str, line from the dbcan overview.txt file\n",
    "    \n",
    "    Return pandas dataframe.\"\"\"\n",
    "    # Retrieve predictions from HMMER (predicated domains and domain ranges)\n",
    "    # separate out each of the domain predications\n",
    "    hmmer_prediction = line[1].split(\"+\")\n",
    "    \n",
    "    # create empty lists to store predicated domains and ranges\n",
    "    cazy_fam = []\n",
    "    cazy_subfam = []\n",
    "    domain_ranges = []\n",
    "\n",
    "    # separate out the name of the predicated domain (the CAZy family) and the domain's AA range\n",
    "    for domain in hmmer_prediction:\n",
    "        domain = domain.split(\"(\")\n",
    "        # standardise name if necessary, retrieve subfam if preciated, and produce null values\n",
    "        domain_name = domain[0]\n",
    "        if domain_name.startswith(\"GT2_Chitin_\"):\n",
    "            cazy_fam.append(\"GT2\")\n",
    "            cazy_subfam.append(np.nan)\n",
    "\n",
    "        elif domain_name == '-':\n",
    "            cazy_fam.append(np.nan)\n",
    "            cazy_subfam.append(np.nan)\n",
    "\n",
    "        elif domain_name.find(\"_\") != -1:\n",
    "            cutoff = domain_name.find(\"_\")\n",
    "            cazy_fam.append(domain_name[:cutoff])\n",
    "            cazy_subfam.append(domain_name)\n",
    "\n",
    "        else:\n",
    "            cazy_fam.append(domain_name)\n",
    "            cazy_subfam.append(np.nan)\n",
    "\n",
    "        try:\n",
    "            if domain[1] == \"-\":\n",
    "                domain_ranges.append(np.nan)\n",
    "            else:\n",
    "                domain_ranges.append(domain[1][:-1])  # exlude the final \")\"\n",
    "        except IndexError: # no domain predicated so no domain ranges\n",
    "            domain_ranges.append(np.nan)\n",
    "\n",
    "        \n",
    "    prediction_dict = {\n",
    "        \"protein_accession\": [line[0]],\n",
    "        \"cazy_family\": [cazy_fam],\n",
    "        \"cazy_subfamily\": [cazy_subfam],\n",
    "        \"domain_range\": [domain_ranges],\n",
    "    }\n",
    "\n",
    "    prediction_df = pd.DataFrame(prediction_dict)\n",
    "    \n",
    "    return prediction_df\n",
    "\n",
    "\n",
    "def parse_hotpep_output(line):\n",
    "    \"\"\"Parse the output from Hotpep from the overview.txt file.\n",
    "    \n",
    "    Retrieve the protein accession, predicated CAZy families, accomanying subfamiles.\n",
    "    Multiple domains can be predicated, and the families and subfamiles will be collected as\n",
    "    as lists, with the index for each item in each list being associated with the matching items (by index)\n",
    "    in the other lists. For example, the first item in every list will contain the prediciton for the first\n",
    "    predicated domain, the second item the second prediction etc.\n",
    "    \n",
    "    :param line: str, line from the dbcan overview.txt file\n",
    "    \n",
    "    Return pandas dataframe.\"\"\"\n",
    "    # Retrieve predictions from Hotpep\n",
    "    hotpep_prediction = line[2].split(\"+\")\n",
    "\n",
    "    # create empty lists to separate the predicated CAZy fam and subfam\n",
    "    cazy_fams = []\n",
    "    cazy_subfams = []\n",
    "\n",
    "    # remove the k-mer cluster number\n",
    "    for prediction in hotpep_prediction:\n",
    "        prediction = prediction.split(\"(\")[0]\n",
    "\n",
    "        if prediction == \"-\":\n",
    "            cazy_fams.append(np.nan)\n",
    "            cazy_subfams.append(np.nan)\n",
    "\n",
    "        elif prediction.find(\"_\") != -1:\n",
    "            cutoff = prediction.find(\"_\")\n",
    "            cazy_fams.append(predictions[:cutoff])\n",
    "            cazy_subfams.append(prediction)\n",
    "\n",
    "        else:\n",
    "            cazy_fams.append(prediction)\n",
    "            cazy_subfams.append(np.nan)\n",
    "\n",
    "    prediction_dict = {\n",
    "        \"protein_accession\": [line[0]],\n",
    "        \"cazy_family\": [cazy_fams],\n",
    "        \"cazy_subfamily\": [cazy_subfams],\n",
    "    }\n",
    "    \n",
    "    prediction_df = pd.DataFrame(prediction_dict)\n",
    "    \n",
    "    return prediction_df\n",
    "\n",
    "\n",
    "def parse_diamond_output(line):\n",
    "    \"\"\"Parse the output from DIAMOND from the overview.txt file.\n",
    "    \n",
    "    Retrieve the protein accession, predicated CAZy families, accomanying subfamiles.\n",
    "    Multiple domains can be predicated, and the families and subfamiles will be collected as\n",
    "    as lists, with the index for each item in each list being associated with the matching items (by index)\n",
    "    in the other lists. For example, the first item in every list will contain the prediciton for the first\n",
    "    predicated domain, the second item the second prediction etc.\n",
    "    \n",
    "    :param line: str, line from the dbcan overview.txt file\n",
    "    \n",
    "    Return pandas dataframe.\"\"\"\n",
    "    \n",
    "    # Retrieve domaind predications from DIAMOND\n",
    "    diamond_prediction = line[3].split('+')\n",
    "\n",
    "    # create empty lists to separate the CAZy family and subfamiles\n",
    "    cazy_fams = []\n",
    "    cazy_subfams = []\n",
    "\n",
    "    for pred in diamond_prediction:\n",
    "        if pred == '-':\n",
    "            cazy_fams.append(np.nan)\n",
    "            cazy_subfams.append(np.nan)\n",
    "\n",
    "        elif pred.find(\"_\") != -1:\n",
    "            cutoff = pred.find(\"_\")\n",
    "            cazy_fams.append(pred[:cutoff])\n",
    "            cazy_subfams.append(pred)\n",
    "\n",
    "        else:\n",
    "            cazy_fams.append(pred)\n",
    "            cazy_subfams.append(np.nan)\n",
    "\n",
    "    prediction_dict = {\n",
    "        \"protein_accession\": [line[0]],\n",
    "        \"cazy_family\": [cazy_fams],\n",
    "        \"cazy_subfamily\": [cazy_subfams],\n",
    "    }\n",
    "    \n",
    "    prediction_df = pd.DataFrame(prediction_dict)\n",
    "    \n",
    "    return prediction_df\n",
    "\n",
    "\n",
    "def get_dbcan_consensus(hmmer, hotpep, diamond):\n",
    "    \"\"\"Get consensus results across HMMER, Hotpep and DIAMOND.\n",
    "    \n",
    "    Retrieves list of items common to all three tools, and another list of items common to two of the tools. Then\n",
    "    builds a dataframe storing the data.\n",
    "    \n",
    "    :param hmmer: list of predictions from HMMER\n",
    "    :param hotpep: list of predictions from Hotpep\n",
    "    :param diamond: list of predictions from DIAMOND\n",
    "    \n",
    "    Return two lists, items where all 3 tools agree, items where two 2 tools agree\n",
    "    \"\"\"\n",
    "    # Retrieve list of items predicated by all three tools\n",
    "    consensus_3 = list(set(hmmer) & set(hotpep) & set(diamond))\n",
    "    if len(consensus_3) == 0:\n",
    "        consensus_3 = [np.nan]\n",
    "    \n",
    "    # Retrieve list of items predicated by two of the tools\n",
    "    consensus_2 = list(set(hmmer) & set(hotpep))\n",
    "    consensus_2 += list(set(hmmer) & set(diamond))\n",
    "    consensus_2 += list(set(hotpep) & set(diamond))\n",
    "    \n",
    "    # remove duplicates and items in consensus for all 3 tools\n",
    "    consensus_2 = list(dict.fromkeys(consensus_2))\n",
    "    for item in consensus_2:\n",
    "        if item in consensus_3:\n",
    "            consensus_2.remove(item)\n",
    "    \n",
    "    if len(consensus_2) == 0:\n",
    "        consensus_2 = [np.nan]\n",
    "    \n",
    "    return consensus_3, consensus_2\n",
    "\n",
    "\n",
    "dbcan_df, hmmer_df, hotpep_df, diamond_df = parse_dbcan_output(\"overview.txt\")\n",
    "dbcan_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>protein_accession</th>\n",
       "      <th>cazy_family</th>\n",
       "      <th>cazy_subfamily</th>\n",
       "      <th>domain_range</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>XP_001545265.2</td>\n",
       "      <td>[GH53]</td>\n",
       "      <td>[nan]</td>\n",
       "      <td>[20-346]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>XP_001545514.1</td>\n",
       "      <td>[GT2]</td>\n",
       "      <td>[nan]</td>\n",
       "      <td>[1213-1720]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>XP_001545605.1</td>\n",
       "      <td>[AA4]</td>\n",
       "      <td>[nan]</td>\n",
       "      <td>[28-577]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>XP_001545700.1</td>\n",
       "      <td>[CBM21]</td>\n",
       "      <td>[nan]</td>\n",
       "      <td>[353-459]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>XP_001545706.2</td>\n",
       "      <td>[GH132]</td>\n",
       "      <td>[nan]</td>\n",
       "      <td>[57-318]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>701</th>\n",
       "      <td>XP_024553622.1</td>\n",
       "      <td>[nan]</td>\n",
       "      <td>[nan]</td>\n",
       "      <td>[nan]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>702</th>\n",
       "      <td>XP_001546886.1</td>\n",
       "      <td>[nan]</td>\n",
       "      <td>[nan]</td>\n",
       "      <td>[nan]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>703</th>\n",
       "      <td>XP_024553655.1</td>\n",
       "      <td>[nan]</td>\n",
       "      <td>[nan]</td>\n",
       "      <td>[nan]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>704</th>\n",
       "      <td>XP_024553815.1</td>\n",
       "      <td>[nan]</td>\n",
       "      <td>[nan]</td>\n",
       "      <td>[nan]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>705</th>\n",
       "      <td>XP_024553814.1</td>\n",
       "      <td>[nan]</td>\n",
       "      <td>[nan]</td>\n",
       "      <td>[nan]</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>706 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    protein_accession cazy_family cazy_subfamily domain_range\n",
       "0      XP_001545265.2      [GH53]          [nan]     [20-346]\n",
       "1      XP_001545514.1       [GT2]          [nan]  [1213-1720]\n",
       "2      XP_001545605.1       [AA4]          [nan]     [28-577]\n",
       "3      XP_001545700.1     [CBM21]          [nan]    [353-459]\n",
       "4      XP_001545706.2     [GH132]          [nan]     [57-318]\n",
       "..                ...         ...            ...          ...\n",
       "701    XP_024553622.1       [nan]          [nan]        [nan]\n",
       "702    XP_001546886.1       [nan]          [nan]        [nan]\n",
       "703    XP_024553655.1       [nan]          [nan]        [nan]\n",
       "704    XP_024553815.1       [nan]          [nan]        [nan]\n",
       "705    XP_024553814.1       [nan]          [nan]        [nan]\n",
       "\n",
       "[706 rows x 4 columns]"
      ]
     },
     "execution_count": 36,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "hmmer_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>protein_accession</th>\n",
       "      <th>cazy_family</th>\n",
       "      <th>cazy_subfamily</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>XP_001545265.2</td>\n",
       "      <td>[GH53]</td>\n",
       "      <td>[nan]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>XP_001545514.1</td>\n",
       "      <td>[GT2]</td>\n",
       "      <td>[nan]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>XP_001545605.1</td>\n",
       "      <td>[AA4]</td>\n",
       "      <td>[nan]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>XP_001545700.1</td>\n",
       "      <td>[CBM21]</td>\n",
       "      <td>[nan]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>XP_001545706.2</td>\n",
       "      <td>[GH132]</td>\n",
       "      <td>[nan]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>701</th>\n",
       "      <td>XP_024553622.1</td>\n",
       "      <td>[nan]</td>\n",
       "      <td>[nan]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>702</th>\n",
       "      <td>XP_001546886.1</td>\n",
       "      <td>[nan]</td>\n",
       "      <td>[nan]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>703</th>\n",
       "      <td>XP_024553655.1</td>\n",
       "      <td>[nan]</td>\n",
       "      <td>[nan]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>704</th>\n",
       "      <td>XP_024553815.1</td>\n",
       "      <td>[nan]</td>\n",
       "      <td>[nan]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>705</th>\n",
       "      <td>XP_024553814.1</td>\n",
       "      <td>[nan]</td>\n",
       "      <td>[nan]</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>706 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    protein_accession cazy_family cazy_subfamily\n",
       "0      XP_001545265.2      [GH53]          [nan]\n",
       "1      XP_001545514.1       [GT2]          [nan]\n",
       "2      XP_001545605.1       [AA4]          [nan]\n",
       "3      XP_001545700.1     [CBM21]          [nan]\n",
       "4      XP_001545706.2     [GH132]          [nan]\n",
       "..                ...         ...            ...\n",
       "701    XP_024553622.1       [nan]          [nan]\n",
       "702    XP_001546886.1       [nan]          [nan]\n",
       "703    XP_024553655.1       [nan]          [nan]\n",
       "704    XP_024553815.1       [nan]          [nan]\n",
       "705    XP_024553814.1       [nan]          [nan]\n",
       "\n",
       "[706 rows x 3 columns]"
      ]
     },
     "execution_count": 37,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "hotpep_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>protein_accession</th>\n",
       "      <th>cazy_family</th>\n",
       "      <th>cazy_subfamily</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>XP_001545265.2</td>\n",
       "      <td>[GH53]</td>\n",
       "      <td>[nan]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>XP_001545514.1</td>\n",
       "      <td>[GT2]</td>\n",
       "      <td>[nan]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>XP_001545605.1</td>\n",
       "      <td>[AA4]</td>\n",
       "      <td>[nan]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>XP_001545700.1</td>\n",
       "      <td>[CBM21]</td>\n",
       "      <td>[nan]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>XP_001545706.2</td>\n",
       "      <td>[GH132]</td>\n",
       "      <td>[nan]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>701</th>\n",
       "      <td>XP_024553622.1</td>\n",
       "      <td>[GH30]</td>\n",
       "      <td>[GH30_7]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>702</th>\n",
       "      <td>XP_001546886.1</td>\n",
       "      <td>[GH30]</td>\n",
       "      <td>[GH30_7]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>703</th>\n",
       "      <td>XP_024553655.1</td>\n",
       "      <td>[CBM24, GH71]</td>\n",
       "      <td>[nan, nan]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>704</th>\n",
       "      <td>XP_024553815.1</td>\n",
       "      <td>[CBM13]</td>\n",
       "      <td>[nan]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>705</th>\n",
       "      <td>XP_024553814.1</td>\n",
       "      <td>[CBM13]</td>\n",
       "      <td>[nan]</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>706 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    protein_accession    cazy_family cazy_subfamily\n",
       "0      XP_001545265.2         [GH53]          [nan]\n",
       "1      XP_001545514.1          [GT2]          [nan]\n",
       "2      XP_001545605.1          [AA4]          [nan]\n",
       "3      XP_001545700.1        [CBM21]          [nan]\n",
       "4      XP_001545706.2        [GH132]          [nan]\n",
       "..                ...            ...            ...\n",
       "701    XP_024553622.1         [GH30]       [GH30_7]\n",
       "702    XP_001546886.1         [GH30]       [GH30_7]\n",
       "703    XP_024553655.1  [CBM24, GH71]     [nan, nan]\n",
       "704    XP_024553815.1        [CBM13]          [nan]\n",
       "705    XP_024553814.1        [CBM13]          [nan]\n",
       "\n",
       "[706 rows x 3 columns]"
      ]
     },
     "execution_count": 38,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "diamond_df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Although it seems brilliant to separate out the domains that have been predicated by 3 tools, and those that have been predicated by 2, the dbCAN papers say that a consensus result is a result where at least 2 tools agree. Additionally, having a single columns of 'cazy_family' and 'cazy_subfamily' will make the statistical evaluation of the performance down the line much easier, becuase all the dataframes will have the same titled columns.\n",
    "\n",
    "Therefore, here is the final version of the parse_dbcan_output function group, where the dbCAN consensus result is defined as any domains that have been predicated by at least 2 tools within dbCAN.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>protein_accession</th>\n",
       "      <th>cazy_family</th>\n",
       "      <th>cazy_subfamily</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>XP_001545265.2</td>\n",
       "      <td>[GH53]</td>\n",
       "      <td>[nan]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>XP_001545514.1</td>\n",
       "      <td>[GT2]</td>\n",
       "      <td>[nan]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>XP_001545605.1</td>\n",
       "      <td>[AA4]</td>\n",
       "      <td>[nan]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>XP_001545700.1</td>\n",
       "      <td>[CBM21]</td>\n",
       "      <td>[nan]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>XP_001545706.2</td>\n",
       "      <td>[GH132]</td>\n",
       "      <td>[nan]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>504</th>\n",
       "      <td>XP_001557402.1</td>\n",
       "      <td>[GT41]</td>\n",
       "      <td>[nan]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>505</th>\n",
       "      <td>XP_024547082.1</td>\n",
       "      <td>[CBM1]</td>\n",
       "      <td>[nan]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>506</th>\n",
       "      <td>XP_024550885.1</td>\n",
       "      <td>[CBM18]</td>\n",
       "      <td>[nan]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>507</th>\n",
       "      <td>XP_001551566.1</td>\n",
       "      <td>[CBM18]</td>\n",
       "      <td>[nan]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>508</th>\n",
       "      <td>XP_001555374.1</td>\n",
       "      <td>[CBM48]</td>\n",
       "      <td>[nan]</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>509 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    protein_accession cazy_family cazy_subfamily\n",
       "0      XP_001545265.2      [GH53]          [nan]\n",
       "1      XP_001545514.1       [GT2]          [nan]\n",
       "2      XP_001545605.1       [AA4]          [nan]\n",
       "3      XP_001545700.1     [CBM21]          [nan]\n",
       "4      XP_001545706.2     [GH132]          [nan]\n",
       "..                ...         ...            ...\n",
       "504    XP_001557402.1      [GT41]          [nan]\n",
       "505    XP_024547082.1      [CBM1]          [nan]\n",
       "506    XP_024550885.1     [CBM18]          [nan]\n",
       "507    XP_001551566.1     [CBM18]          [nan]\n",
       "508    XP_001555374.1     [CBM48]          [nan]\n",
       "\n",
       "[509 rows x 3 columns]"
      ]
     },
     "execution_count": 39,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def parse_dbcan_output(overview_file_path):\n",
    "    \"\"\"Parse the output from the run_dbCAN overview.txt file, writing a dataframe from each prediciotn tool.\n",
    "    \n",
    "    :param overview_file_path: path, path to the output 'overview.txt' file\n",
    "    \n",
    "    Return 4 dataframes, one each for consensus dbCAN result, HMMER, Hotpep and DIAMOND\n",
    "    \"\"\"\n",
    "    with open(overview_file_path, \"r\") as fh:\n",
    "        overview_file = fh.read().splitlines()\n",
    "        \n",
    "    # build empty dataframes to add the repsective prediction tools output to\n",
    "    hmmer_df = pd.DataFrame({}, columns=[\"protein_accession\", \"cazy_family\", \"cazy_subfamily\", \"domain_range\"])\n",
    "    hotpep_df = pd.DataFrame({}, columns=[\"protein_accession\", \"cazy_family\", \"cazy_subfamily\"]) # add ec later\n",
    "    diamond_df = pd.DataFrame({}, columns=[\"protein_accession\", \"cazy_family\", \"cazy_subfamily\"])\n",
    "    dbcan_df = pd.DataFrame({}, columns=[\"protein_accession\", \"cazy_family\", \"cazy_subfamily\"])\n",
    "\n",
    "    for line in overview_file[1:]: # skip the first line becuase this is the head titles\n",
    "        line = line.split(\"\\t\")\n",
    "        \n",
    "        # retrieve the data for HMMER\n",
    "        hmmer_temp_df = parse_hmmer_output(line)\n",
    "        hmmer_df = hmmer_df.append(hmmer_temp_df, ignore_index=True)\n",
    "        \n",
    "        # retrieve the data for Hotpep\n",
    "        hotpep_temp_df = parse_hotpep_output(line)\n",
    "        hotpep_df = hotpep_df.append(hotpep_temp_df, ignore_index=True)\n",
    "        \n",
    "        # retrieve the data for DIAMOND\n",
    "        diamond_temp_df = parse_diamond_output(line)\n",
    "        diamond_df = diamond_df.append(diamond_temp_df, ignore_index=True)\n",
    "        \n",
    "        # retrieve consensus results for dbCAN\n",
    "        if line[-1] != \"1\":   #c heck #ofTools, same approach as other tools, don't include if non-CAZyme prediction\n",
    "            consensus_cazy_fam = get_dbcan_consensus(\n",
    "                hmmer_temp_df.iloc[0, 1],\n",
    "                hotpep_temp_df.iloc[0, 1],\n",
    "                diamond_temp_df.iloc[0, 1],\n",
    "            )\n",
    "            consensus_sub_fam = get_dbcan_consensus(\n",
    "                hmmer_temp_df.iloc[0, 2],\n",
    "                hotpep_temp_df.iloc[0, 2],\n",
    "                diamond_temp_df.iloc[0, 2],\n",
    "            )\n",
    "\n",
    "            consensus_dict = {\n",
    "                \"protein_accession\": [line[0]],\n",
    "                \"cazy_family\": [consensus_cazy_fam],\n",
    "                \"cazy_subfamily\": [consensus_sub_fam],\n",
    "            }\n",
    "            \n",
    "            temp_consensus_df = pd.DataFrame(consensus_dict)\n",
    "            \n",
    "            dbcan_df = dbcan_df.append(temp_consensus_df, ignore_index=True)\n",
    "\n",
    "    return dbcan_df, hmmer_df, hotpep_df, diamond_df\n",
    "\n",
    "\n",
    "def parse_hmmer_output(line):\n",
    "    \"\"\"Parse the output from HMMER from the overview.txt file.\n",
    "    \n",
    "    Retrieve the protein accession, predicated CAZy families, accomanying subfamiles and domain ranges.\n",
    "    Multiple domains can be predicated, and the families, subfamiles and ranges will be collected as\n",
    "    as lists, with the index for each item in each list being associated with the matching items (by index)\n",
    "    in the other lists. For example, the first item in every list will contain the prediciton for the first\n",
    "    predicated domain, the second item the second prediction etc.\n",
    "    \n",
    "    :param line: str, line from the dbcan overview.txt file\n",
    "    \n",
    "    Return pandas dataframe.\"\"\"\n",
    "    # Retrieve predictions from HMMER (predicated domains and domain ranges)\n",
    "    # separate out each of the domain predications\n",
    "    hmmer_prediction = line[1].split(\"+\")\n",
    "    \n",
    "    # create empty lists to store predicated domains and ranges\n",
    "    cazy_fam = []\n",
    "    cazy_subfam = []\n",
    "    domain_ranges = []\n",
    "\n",
    "    # separate out the name of the predicated domain (the CAZy family) and the domain's AA range\n",
    "    for domain in hmmer_prediction:\n",
    "        domain = domain.split(\"(\")\n",
    "        # standardise name if necessary, retrieve subfam if preciated, and produce null values\n",
    "        domain_name = domain[0]\n",
    "        if domain_name.startswith(\"GT2_Chitin_\"):\n",
    "            cazy_fam.append(\"GT2\")\n",
    "            cazy_subfam.append(np.nan)\n",
    "\n",
    "        elif domain_name == '-':\n",
    "            cazy_fam.append(np.nan)\n",
    "            cazy_subfam.append(np.nan)\n",
    "\n",
    "        elif domain_name.find(\"_\") != -1:\n",
    "            cutoff = domain_name.find(\"_\")\n",
    "            cazy_fam.append(domain_name[:cutoff])\n",
    "            cazy_subfam.append(domain_name)\n",
    "\n",
    "        else:\n",
    "            cazy_fam.append(domain_name)\n",
    "            cazy_subfam.append(np.nan)\n",
    "\n",
    "        try:\n",
    "            if domain[1] == \"-\":\n",
    "                domain_ranges.append(np.nan)\n",
    "            else:\n",
    "                domain_ranges.append(domain[1][:-1])  # exlude the final \")\"\n",
    "        except IndexError: # no domain predicated so no domain ranges\n",
    "            domain_ranges.append(np.nan)\n",
    "\n",
    "        \n",
    "    prediction_dict = {\n",
    "        \"protein_accession\": [line[0]],\n",
    "        \"cazy_family\": [cazy_fam],\n",
    "        \"cazy_subfamily\": [cazy_subfam],\n",
    "        \"domain_range\": [domain_ranges],\n",
    "    }\n",
    "\n",
    "    prediction_df = pd.DataFrame(prediction_dict)\n",
    "    \n",
    "    return prediction_df\n",
    "\n",
    "\n",
    "def parse_hotpep_output(line):\n",
    "    \"\"\"Parse the output from Hotpep from the overview.txt file.\n",
    "    \n",
    "    Retrieve the protein accession, predicated CAZy families, accomanying subfamiles.\n",
    "    Multiple domains can be predicated, and the families and subfamiles will be collected as\n",
    "    as lists, with the index for each item in each list being associated with the matching items (by index)\n",
    "    in the other lists. For example, the first item in every list will contain the prediciton for the first\n",
    "    predicated domain, the second item the second prediction etc.\n",
    "    \n",
    "    :param line: str, line from the dbcan overview.txt file\n",
    "    \n",
    "    Return pandas dataframe.\"\"\"\n",
    "    # Retrieve predictions from Hotpep\n",
    "    hotpep_prediction = line[2].split(\"+\")\n",
    "\n",
    "    # create empty lists to separate the predicated CAZy fam and subfam\n",
    "    cazy_fams = []\n",
    "    cazy_subfams = []\n",
    "\n",
    "    # remove the k-mer cluster number\n",
    "    for prediction in hotpep_prediction:\n",
    "        prediction = prediction.split(\"(\")[0]\n",
    "\n",
    "        if prediction == \"-\":\n",
    "            cazy_fams.append(np.nan)\n",
    "            cazy_subfams.append(np.nan)\n",
    "\n",
    "        elif prediction.find(\"_\") != -1:\n",
    "            cutoff = prediction.find(\"_\")\n",
    "            cazy_fams.append(predictions[:cutoff])\n",
    "            cazy_subfams.append(prediction)\n",
    "\n",
    "        else:\n",
    "            cazy_fams.append(prediction)\n",
    "            cazy_subfams.append(np.nan)\n",
    "\n",
    "    prediction_dict = {\n",
    "        \"protein_accession\": [line[0]],\n",
    "        \"cazy_family\": [cazy_fams],\n",
    "        \"cazy_subfamily\": [cazy_subfams],\n",
    "    }\n",
    "    \n",
    "    prediction_df = pd.DataFrame(prediction_dict)\n",
    "    \n",
    "    return prediction_df\n",
    "\n",
    "\n",
    "def parse_diamond_output(line):\n",
    "    \"\"\"Parse the output from DIAMOND from the overview.txt file.\n",
    "    \n",
    "    Retrieve the protein accession, predicated CAZy families, accomanying subfamiles.\n",
    "    Multiple domains can be predicated, and the families and subfamiles will be collected as\n",
    "    as lists, with the index for each item in each list being associated with the matching items (by index)\n",
    "    in the other lists. For example, the first item in every list will contain the prediciton for the first\n",
    "    predicated domain, the second item the second prediction etc.\n",
    "    \n",
    "    :param line: str, line from the dbcan overview.txt file\n",
    "    \n",
    "    Return pandas dataframe.\"\"\"\n",
    "    \n",
    "    # Retrieve domaind predications from DIAMOND\n",
    "    diamond_prediction = line[3].split('+')\n",
    "\n",
    "    # create empty lists to separate the CAZy family and subfamiles\n",
    "    cazy_fams = []\n",
    "    cazy_subfams = []\n",
    "\n",
    "    for pred in diamond_prediction:\n",
    "        if pred == '-':\n",
    "            cazy_fams.append(np.nan)\n",
    "            cazy_subfams.append(np.nan)\n",
    "\n",
    "        elif pred.find(\"_\") != -1:\n",
    "            cutoff = pred.find(\"_\")\n",
    "            cazy_fams.append(pred[:cutoff])\n",
    "            cazy_subfams.append(pred)\n",
    "\n",
    "        else:\n",
    "            cazy_fams.append(pred)\n",
    "            cazy_subfams.append(np.nan)\n",
    "\n",
    "    prediction_dict = {\n",
    "        \"protein_accession\": [line[0]],\n",
    "        \"cazy_family\": [cazy_fams],\n",
    "        \"cazy_subfamily\": [cazy_subfams],\n",
    "    }\n",
    "    \n",
    "    prediction_df = pd.DataFrame(prediction_dict)\n",
    "    \n",
    "    return prediction_df\n",
    "\n",
    "\n",
    "def get_dbcan_consensus(hmmer, hotpep, diamond):\n",
    "    \"\"\"Get consensus results across HMMER, Hotpep and DIAMOND.\n",
    "    \n",
    "    A consensus result is defined by a result at at least two of the tools in dbCAN have predicated. Stores the\n",
    "    consensus results in a list, with no duplicates.\n",
    "    \n",
    "    Retrieves list of items common to all three tools, and another list of items common to two of the tools. Then\n",
    "    builds a dataframe storing the data.\n",
    "    \n",
    "    :param hmmer: list of predictions from HMMER\n",
    "    :param hotpep: list of predictions from Hotpep\n",
    "    :param diamond: list of predictions from DIAMOND\n",
    "    \n",
    "    Returns a list.\n",
    "    \"\"\"\n",
    "    # Retrieve list of items predicated by all three tools\n",
    "    consensus = list(set(hmmer) & set(hotpep) & set(diamond))\n",
    "    \n",
    "    # add domains predicated by two of the tools\n",
    "    consensus += list(set(hmmer) & set(hotpep))\n",
    "    consensus += list(set(hmmer) & set(diamond))\n",
    "    consensus += list(set(hotpep) & set(diamond))\n",
    "    \n",
    "    # remove duplicates\n",
    "    consensus = list(dict.fromkeys(consensus))\n",
    "    \n",
    "    if len(consensus) == 0:\n",
    "        consensus = [np.nan]\n",
    "    \n",
    "    return consensus\n",
    "\n",
    "\n",
    "dbcan_df, hmmer_df, hotpep_df, diamond_df = parse_dbcan_output(\"overview.txt\")\n",
    "dbcan_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>protein_accession</th>\n",
       "      <th>cazy_family</th>\n",
       "      <th>cazy_subfamily</th>\n",
       "      <th>domain_range</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>XP_001545265.2</td>\n",
       "      <td>[GH53]</td>\n",
       "      <td>[nan]</td>\n",
       "      <td>[20-346]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>XP_001545514.1</td>\n",
       "      <td>[GT2]</td>\n",
       "      <td>[nan]</td>\n",
       "      <td>[1213-1720]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>XP_001545605.1</td>\n",
       "      <td>[AA4]</td>\n",
       "      <td>[nan]</td>\n",
       "      <td>[28-577]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>XP_001545700.1</td>\n",
       "      <td>[CBM21]</td>\n",
       "      <td>[nan]</td>\n",
       "      <td>[353-459]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>XP_001545706.2</td>\n",
       "      <td>[GH132]</td>\n",
       "      <td>[nan]</td>\n",
       "      <td>[57-318]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>701</th>\n",
       "      <td>XP_024553622.1</td>\n",
       "      <td>[nan]</td>\n",
       "      <td>[nan]</td>\n",
       "      <td>[nan]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>702</th>\n",
       "      <td>XP_001546886.1</td>\n",
       "      <td>[nan]</td>\n",
       "      <td>[nan]</td>\n",
       "      <td>[nan]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>703</th>\n",
       "      <td>XP_024553655.1</td>\n",
       "      <td>[nan]</td>\n",
       "      <td>[nan]</td>\n",
       "      <td>[nan]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>704</th>\n",
       "      <td>XP_024553815.1</td>\n",
       "      <td>[nan]</td>\n",
       "      <td>[nan]</td>\n",
       "      <td>[nan]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>705</th>\n",
       "      <td>XP_024553814.1</td>\n",
       "      <td>[nan]</td>\n",
       "      <td>[nan]</td>\n",
       "      <td>[nan]</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>706 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    protein_accession cazy_family cazy_subfamily domain_range\n",
       "0      XP_001545265.2      [GH53]          [nan]     [20-346]\n",
       "1      XP_001545514.1       [GT2]          [nan]  [1213-1720]\n",
       "2      XP_001545605.1       [AA4]          [nan]     [28-577]\n",
       "3      XP_001545700.1     [CBM21]          [nan]    [353-459]\n",
       "4      XP_001545706.2     [GH132]          [nan]     [57-318]\n",
       "..                ...         ...            ...          ...\n",
       "701    XP_024553622.1       [nan]          [nan]        [nan]\n",
       "702    XP_001546886.1       [nan]          [nan]        [nan]\n",
       "703    XP_024553655.1       [nan]          [nan]        [nan]\n",
       "704    XP_024553815.1       [nan]          [nan]        [nan]\n",
       "705    XP_024553814.1       [nan]          [nan]        [nan]\n",
       "\n",
       "[706 rows x 4 columns]"
      ]
     },
     "execution_count": 40,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "hmmer_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>protein_accession</th>\n",
       "      <th>cazy_family</th>\n",
       "      <th>cazy_subfamily</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>XP_001545265.2</td>\n",
       "      <td>[GH53]</td>\n",
       "      <td>[nan]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>XP_001545514.1</td>\n",
       "      <td>[GT2]</td>\n",
       "      <td>[nan]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>XP_001545605.1</td>\n",
       "      <td>[AA4]</td>\n",
       "      <td>[nan]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>XP_001545700.1</td>\n",
       "      <td>[CBM21]</td>\n",
       "      <td>[nan]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>XP_001545706.2</td>\n",
       "      <td>[GH132]</td>\n",
       "      <td>[nan]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>701</th>\n",
       "      <td>XP_024553622.1</td>\n",
       "      <td>[nan]</td>\n",
       "      <td>[nan]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>702</th>\n",
       "      <td>XP_001546886.1</td>\n",
       "      <td>[nan]</td>\n",
       "      <td>[nan]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>703</th>\n",
       "      <td>XP_024553655.1</td>\n",
       "      <td>[nan]</td>\n",
       "      <td>[nan]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>704</th>\n",
       "      <td>XP_024553815.1</td>\n",
       "      <td>[nan]</td>\n",
       "      <td>[nan]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>705</th>\n",
       "      <td>XP_024553814.1</td>\n",
       "      <td>[nan]</td>\n",
       "      <td>[nan]</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>706 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    protein_accession cazy_family cazy_subfamily\n",
       "0      XP_001545265.2      [GH53]          [nan]\n",
       "1      XP_001545514.1       [GT2]          [nan]\n",
       "2      XP_001545605.1       [AA4]          [nan]\n",
       "3      XP_001545700.1     [CBM21]          [nan]\n",
       "4      XP_001545706.2     [GH132]          [nan]\n",
       "..                ...         ...            ...\n",
       "701    XP_024553622.1       [nan]          [nan]\n",
       "702    XP_001546886.1       [nan]          [nan]\n",
       "703    XP_024553655.1       [nan]          [nan]\n",
       "704    XP_024553815.1       [nan]          [nan]\n",
       "705    XP_024553814.1       [nan]          [nan]\n",
       "\n",
       "[706 rows x 3 columns]"
      ]
     },
     "execution_count": 41,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "hotpep_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>protein_accession</th>\n",
       "      <th>cazy_family</th>\n",
       "      <th>cazy_subfamily</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>XP_001545265.2</td>\n",
       "      <td>[GH53]</td>\n",
       "      <td>[nan]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>XP_001545514.1</td>\n",
       "      <td>[GT2]</td>\n",
       "      <td>[nan]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>XP_001545605.1</td>\n",
       "      <td>[AA4]</td>\n",
       "      <td>[nan]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>XP_001545700.1</td>\n",
       "      <td>[CBM21]</td>\n",
       "      <td>[nan]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>XP_001545706.2</td>\n",
       "      <td>[GH132]</td>\n",
       "      <td>[nan]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>701</th>\n",
       "      <td>XP_024553622.1</td>\n",
       "      <td>[GH30]</td>\n",
       "      <td>[GH30_7]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>702</th>\n",
       "      <td>XP_001546886.1</td>\n",
       "      <td>[GH30]</td>\n",
       "      <td>[GH30_7]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>703</th>\n",
       "      <td>XP_024553655.1</td>\n",
       "      <td>[CBM24, GH71]</td>\n",
       "      <td>[nan, nan]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>704</th>\n",
       "      <td>XP_024553815.1</td>\n",
       "      <td>[CBM13]</td>\n",
       "      <td>[nan]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>705</th>\n",
       "      <td>XP_024553814.1</td>\n",
       "      <td>[CBM13]</td>\n",
       "      <td>[nan]</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>706 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    protein_accession    cazy_family cazy_subfamily\n",
       "0      XP_001545265.2         [GH53]          [nan]\n",
       "1      XP_001545514.1          [GT2]          [nan]\n",
       "2      XP_001545605.1          [AA4]          [nan]\n",
       "3      XP_001545700.1        [CBM21]          [nan]\n",
       "4      XP_001545706.2        [GH132]          [nan]\n",
       "..                ...            ...            ...\n",
       "701    XP_024553622.1         [GH30]       [GH30_7]\n",
       "702    XP_001546886.1         [GH30]       [GH30_7]\n",
       "703    XP_024553655.1  [CBM24, GH71]     [nan, nan]\n",
       "704    XP_024553815.1        [CBM13]          [nan]\n",
       "705    XP_024553814.1        [CBM13]          [nan]\n",
       "\n",
       "[706 rows x 3 columns]"
      ]
     },
     "execution_count": 42,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "diamond_df"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
