{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Parsing CUPP output\n",
    "\n",
    "The CAZyme prediction tools write out their output files in different formats, and do not present the data in a standardised manner across all tools. Additionally, the output from the prediction tools does not enable easy parsing of the data by the [Sklearn](https://scikit-learn.org/stable/index.html) library, which performs the statistical analysis of the tools preformances.\n",
    "\n",
    "This notebook explains the theorlogy and method used by the `pyrewton.cazymes.prediction.parse` submodule to parse the output from the CAZyme prediction tool CUPP, and write out Pandas dataframes.\n",
    "\n",
    "[CUPP](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC6489277/) is a k-mer based CAZyme prediction tool. The information defining the data included in each field of the output files came from the [README file](https://files.dtu.dk/userportal/#/shared/public/hLin6ni4p-SWuKfp/CUPP_program/) and the raw Python script.\n",
    "\n",
    "The following data is collected from the CUPP output:\n",
    "- protein_accession\n",
    "- cazy_family (all predicated CAZy families)\n",
    "- cazy_subfamily (all predicated CAZy subfamilies)\n",
    "- ec_number (all predicted EC numbers)\n",
    "- domain_range (the amino acid domain range of all predicted CAZy (sub)families)\n",
    "\n",
    "\n",
    "### Notebook structure\n",
    "\n",
    "A lot of the information about CUPP and its data is taken from its README, but predominately from the `CUPPprediction.py` Python script.\n",
    "\n",
    "The section titled 'exploring the output' shows the exploration of the tools output files. This is to demonstrate how the data is constructed and presented, and forms a foundation for explaining the developed method of parsing the data.\n",
    "\n",
    "The section 'parsing the output' shows the method of parsing the data from the CAZyme prediction tool.\n",
    "\n",
    "The final section shows any additional parsing that was performed to make the statistical analysis of the tool's performance easier. This typically includes removing duplicates, and combining rows that represent data for the same predicted CAZyme.\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Contents\n",
    "\n",
    "- [Imports](#notebook_imports)\n",
    "- [Exploring the CUPP FASTA file](#cupp_fasta)\n",
    "- [Exploring the CUPP log file](#cupp_log)\n",
    "- [Choosing the CUPP output file to path](#cupp_choice)\n",
    "- [Parsing the CUPP log file](#cupp_parse_log)\n",
    "    - [Creating a data model](#data_model)\n",
    "    - [Getting the predicted domain range](#cupp_domain_range)\n",
    "    - [Getting the predicted EC number](#cupp_ec)\n",
    "        - [Adding qualitying checking when getting the EC number](#cupp_ec_quality)\n",
    "    - [Getting the predicted subfamily](#cupp_subfam)\n",
    "        - [Standardising CAZy subfamilies (PL7:2)](#cupp_std_subfam)\n",
    "- [Combing functions and populating the data model](#cupp_functions)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<a id=\"notebook_imports\"></a>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Requirement already satisfied: pandas in /home/em/.local/lib/python3.8/site-packages (1.1.4)\n",
      "Requirement already satisfied: pytz>=2017.2 in /home/em/anaconda3/lib/python3.8/site-packages (from pandas) (2020.1)\n",
      "Requirement already satisfied: numpy>=1.15.4 in /home/em/.local/lib/python3.8/site-packages (from pandas) (1.19.4)\n",
      "Requirement already satisfied: python-dateutil>=2.7.3 in /home/em/anaconda3/lib/python3.8/site-packages (from pandas) (2.8.1)\n",
      "Requirement already satisfied: six>=1.5 in /home/em/anaconda3/lib/python3.8/site-packages (from python-dateutil>=2.7.3->pandas) (1.15.0)\n",
      "Requirement already satisfied: numpy in /home/em/.local/lib/python3.8/site-packages (1.19.4)\n"
     ]
    }
   ],
   "source": [
    "# Install necessary libraries\n",
    "!pip3 install pandas\n",
    "!pip3 install numpy\n",
    "\n",
    "import re\n",
    "\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "\n",
    "from tqdm.notebook import tqdm"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<a id=\"cupp_intro\"></a>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Understanding the output\n",
    "\n",
    "CUPP produces two output files (a log file and a FASTA file) that contain the same data, but present it slightly differently. \n",
    "\n",
    "#### FASTA file\n",
    "The format of the data in the output FASTA file for each protein is as follows:  \n",
    "```\n",
    "> accession | raw_function | best_function | additional_info | subfam \\n seq\n",
    "```\n",
    "\n",
    "For example:\n",
    "\n",
    "```\n",
    ">QIX00022.1 AMS68_005539|&|AA1:1.10.3.2|AA1:1.10.3.2|AA1:6.1(20.3,124..357)|AA1_3\n",
    "mvrsvnfplgvyfqykfikigtdttltweadpnreytvpsdcntspvlsgpwqyppvstsaavaiaatsttaqasavvtsqptycpntpttrncwadgfdintdfdnhwpttgrtvsytfeitnttlApDgvSRPvfaingqyPGPtIYADWgDMISVtvinklqdngttIHFHGVrQWHSNtqdgvpGLTECpIAPGTTRTyTfqatQfGTSwYHSHFSCQYGdGILGPiviqGpAtANYDIDVGpLpitdwyyptvnyhasitehtngsppvadnalINgTMTSSsggayaltklttgkrhRLrLINTSVDNHFMVSLdghqievitaDFvpIVPYNtTWimlgigqRYDVIITADQAPgnywfraevqdgcganannkniksmfaypghenetpistgsgytgrctdetqlvphwnsyvppkagfaeptffgtalnqstaadgtltlywqvngsslniqwdkptlqyvqesntsypsganvvetategqwtywviqevpgtpynvnvphplhLHGHDFFilGtgtdtwdakfsqnlnynnpprrdtamlpsngwivlawqtdnpgawLFhCHiAwhsdeglgvqfleaadqmqsvnpigtefdqqcqawhdfypvhasylkhdsgv\n",
    "```\n",
    "\n",
    "The following defines and the explains what data is included within each field of the output FASTA file.\n",
    "\n",
    "**Accession**\n",
    "This the data contained in the line starting with '>' in the input FASTA file containing the query sequences for CUPP. If the input FASTA file passed to CUPP was written output using `pyrewton` the first item in the accession is the GenBank protein accession number. This is the unique identifier for each protein.\n",
    "\n",
    "**Raw function** The following definition was taken from the raw Python script for `CUPP.CUPPprediction.py`: The 'raw function' is produced \"when consider[ing] all EC functions of a CUPP equally\". This is preseted as \"predicted_cazy_family : predicted_ec_number\".\n",
    "\n",
    "**Best function** The following definition was taken from the raw Python script for `CUPP.CUPPprediction.py`: the best function is the \"most occuring EC function of the CUPP group\". It was inferred from the Python script `CUPP.CUPPprediction.py` that the 'CUPP group' refers to the CAZy family k-mer cluster for which the query protein scored most highly. In the FASTA file this is presented as \"predicted_cazy_family : predicted_ec_number\". **This is the function that is taken from the output files as the CUPP CAZyme and CAZy family prediction.**\n",
    "\n",
    "**Additional information** This includes several pieces of information that are calculated by CUPP. Owing to the manner that CUPP presents this data in the FASTA file it can be difficult to determine exactly which pieces of data have been provided. The additional data is presented in the following order (and the definitions were taken from the script `CUPP.CUPPprediction.py`):\n",
    "- \"Ratio of maximum frequencies\"\n",
    "- \"Ratio of maximum frequencies by coverage\"\n",
    "- \"Centre for mass for domain\"\n",
    "- \"Domain range\" -- this is identifable by the presense of '...' between two numbers and could potentially be useful when looking for regions of positive selection down the road\n",
    "- \"Number of important positions in domain\" -- very ambgious of meaning\n",
    "\n",
    "**Subfam** This is the predicted CAZy subfamily for the query protein\n",
    "\n",
    "**Seq** This is the amino acid sequence of the query protein. It has not been determined why some letters up upper case and other lower case, although it is hypothesised that the upper case letters indicate the start of a k-mer hit found by CUPP.\n",
    "\n",
    "#### Log file\n",
    "The output log file for CUPP presents the same data as the FASTA, but presents the data in a different way. Instead of separating the data by '|' (which is used in the output FASTA file), the data is separated by tabulated white space ('\\t').\n",
    "\n",
    "Additionally, the data is presented in a different order to the output FASTA file:\n",
    "- Predicated CAZy family (equivalent to the FASTA file best function)\n",
    "- CUPP (k-mer cluster) group which the query protein scored highest for, and the associated score\n",
    "- Additional information\n",
    "- Raw_function (same as the fasta file (presented as \"cazy_fam:ec#\"))\n",
    "- Best_function (same as the fasta file (presented as \"cazy_fam:ec#\")) **This is the function that is taken from the output files as the CUPP CAZyme and CAZy family prediction.**\n",
    "- Subfam\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<a id=\"cupp_fasta\"></a>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Exploring the CUPP FASTA file\n",
    "\n",
    "The data in the CUPP FASTA file is separated by '|'.\n",
    "Each protein covers two lines of data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      ">XP_001545605.1 BCIN_01g00260|&|AA4:1.1.3.38|AA4:1.1.3.38|AA4:1.1(13.9,122..568)|\n",
      "maasivttdtplalrvpnaattpaandplalppgtssekfaefishaeqttgaenviiirkkeeltkelytdpskahdmyhlfdkdyfvvsatiaprnveevqaivrlcndyeiplwpfsigrnvGYGgAAPRVPGSigldmgknmnkvlevnsegayallEPGvTFFSLydylkekklddqlwidtpdlgggsvvGNTIERGvGYTPYGDHFMMHCGLeVVlPTGELIRTGMGALPDPTQptveggsldqqpgnkcwqlfpygfgpYnDGIfSQSNlGIVTKMGIWLMpNPGGyQpylitfpkdtdlpkivdiirplrlqmviqnvpsirhilldaavmgtkksyydvdrplneeeldaiakelnlgrwnfygAlYgPKPvrdvlwqvvkdafgtiegakfflpedikekcvlhiraktlqgiptidelswvdwlpngahlffspiskisgedaskqyaltqkmtleagfdfigtftigmremhhiiclvfdrededqkrrahklireliqvcadngwgeYrTHIALMdqiaetygwndnaqmklnekiknaLDpKGILaPGKNGVWpAsYdrsawrldansertrtrp\n",
      ">XP_024545922.1 BCIN_01g00340|&|GT2:2.4.1.117|GT2:2.4.1.117|GT2:23.1(20.2,116..300)|\n",
      "mdgllelpvqilgglwevvrdtpvhvlgviligllglglgliyallllvapvprppypsektyitttpsgttetkpltcwydswvahreasvsksdscpantirtgaiepatlemslvvPaYNEEERLIgMLeealsfldttygrvargtgtgtgyeillvNDgSRDRTVEIAlDFSrknglhdvlrivtleenRGkGGAVTHGMRHVRGEYAVfADADGASRFADLGkLvkgvrevvdeegrgvavGSRaWLVGseAvvkRSALRNtlMhsfhlllrlltppatsrirdtQcGFKLFTRAAlPhiipymhaegwifdvemlmlaesapgvedaekgkgnggkakgikvseqpiawqevggsklnvmwdslgmawglavlrggwgmgvwrrr\n"
     ]
    }
   ],
   "source": [
    "# Reading the CUPP output FASTA file\n",
    "with open(\"gca_143535_cupp.fasta\", \"r\") as fh:\n",
    "    fasta_file = fh.read().splitlines()\n",
    "for line in fasta_file[:4]:\n",
    "    print(line)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "'>XP_001545605.1 BCIN_01g00260|&|AA4:1.1.3.38|AA4:1.1.3.38|AA4:1.1(13.9,122..568)|'\n",
      "'maasivttdtplalrvpnaattpaandplalppgtssekfaefishaeqttgaenviiirkkeeltkelytdpskahdmyhlfdkdyfvvsatiaprnveevqaivrlcndyeiplwpfsigrnvGYGgAAPRVPGSigldmgknmnkvlevnsegayallEPGvTFFSLydylkekklddqlwidtpdlgggsvvGNTIERGvGYTPYGDHFMMHCGLeVVlPTGELIRTGMGALPDPTQptveggsldqqpgnkcwqlfpygfgpYnDGIfSQSNlGIVTKMGIWLMpNPGGyQpylitfpkdtdlpkivdiirplrlqmviqnvpsirhilldaavmgtkksyydvdrplneeeldaiakelnlgrwnfygAlYgPKPvrdvlwqvvkdafgtiegakfflpedikekcvlhiraktlqgiptidelswvdwlpngahlffspiskisgedaskqyaltqkmtleagfdfigtftigmremhhiiclvfdrededqkrrahklireliqvcadngwgeYrTHIALMdqiaetygwndnaqmklnekiknaLDpKGILaPGKNGVWpAsYdrsawrldansertrtrp'\n",
      "'>XP_024545922.1 BCIN_01g00340|&|GT2:2.4.1.117|GT2:2.4.1.117|GT2:23.1(20.2,116..300)|'\n",
      "'mdgllelpvqilgglwevvrdtpvhvlgviligllglglgliyallllvapvprppypsektyitttpsgttetkpltcwydswvahreasvsksdscpantirtgaiepatlemslvvPaYNEEERLIgMLeealsfldttygrvargtgtgtgyeillvNDgSRDRTVEIAlDFSrknglhdvlrivtleenRGkGGAVTHGMRHVRGEYAVfADADGASRFADLGkLvkgvrevvdeegrgvavGSRaWLVGseAvvkRSALRNtlMhsfhlllrlltppatsrirdtQcGFKLFTRAAlPhiipymhaegwifdvemlmlaesapgvedaekgkgnggkakgikvseqpiawqevggsklnvmwdslgmawglavlrggwgmgvwrrr'\n"
     ]
    }
   ],
   "source": [
    "# read the raw text in the fasta file\n",
    "for line in fasta_file[:4]:\n",
    "    print(repr(line))\n",
    "# the data is separated by '|'"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<a id=\"cupp_log\"></a>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Exploring the CUPP Log file\n",
    "\n",
    "The data in the CUPP log file is separated by '\\t' (tabulated white space). Each protein is contained on a single line"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "XP_001545605.1 BCIN_01g00260\tAA4\tAA4:1.1\t13.9\t7.39\t268\t122..568\t107\tAA4:1.1.3.38\tAA4:1.1.3.38\t\n",
      "XP_024545922.1 BCIN_01g00340\tGT2\tGT2:23.1\t20.2\t7.74\t211\t116..300\t77\tGT2:2.4.1.117\tGT2:2.4.1.117\t\n",
      "XP_001547235.1 BCIN_01g00640\tAA3\tAA3:3.1\t64.0\t144.16\t274\t26..553\t453\t\t\tAA3_1\n",
      "XP_001547254.1 BCIN_01g00800\tAA1\tAA1:6.1\t10.5\t5.13\t329\t202..455\t98\tAA1:1.10.3.2\tAA1:1.10.3.2\tAA1_3\n"
     ]
    }
   ],
   "source": [
    "# read the log file\n",
    "with open(\"gca_143535_cupp.fasta.log\", \"r\") as lfh:\n",
    "    log_file = lfh.read().splitlines()\n",
    "\n",
    "for line in log_file[:4]:\n",
    "    print(line)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "'XP_001545605.1 BCIN_01g00260\\tAA4\\tAA4:1.1\\t13.9\\t7.39\\t268\\t122..568\\t107\\tAA4:1.1.3.38\\tAA4:1.1.3.38\\t'\n",
      "'XP_024545922.1 BCIN_01g00340\\tGT2\\tGT2:23.1\\t20.2\\t7.74\\t211\\t116..300\\t77\\tGT2:2.4.1.117\\tGT2:2.4.1.117\\t'\n",
      "'XP_001547235.1 BCIN_01g00640\\tAA3\\tAA3:3.1\\t64.0\\t144.16\\t274\\t26..553\\t453\\t\\t\\tAA3_1'\n",
      "'XP_001547254.1 BCIN_01g00800\\tAA1\\tAA1:6.1\\t10.5\\t5.13\\t329\\t202..455\\t98\\tAA1:1.10.3.2\\tAA1:1.10.3.2\\tAA1_3'\n"
     ]
    }
   ],
   "source": [
    "# read the raw text in the log file\n",
    "for line in log_file[:4]:\n",
    "    print(repr(line))\n",
    "# data is separated by '\\t'"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<a id=\"cupp_choice\"></a>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Choosing the CUPP output file to path\n",
    "\n",
    "The output FASTA and log files of CUPP contain the same data, therefore, it is unnecessary to parse both output files.\n",
    "\n",
    "A parser of the CUPP FASTA file would separate the protein data by the '|' separator character. This would parse the individual items of the protein data into a list, in the same order as they appear in the output FASTA file. This would allow the parser to selective retrieve the predicted CAZy families, EC numbers, etc.\n",
    "\n",
    "However, '|' is a common data separator. A user or external database from which the input FASTA file containing the input query protein sequences for CUPP was retrieved, may include '|' characters in their FASTA files. For example, a protein data line may be written as \">XP_001545605.1|BCIN_01g00260|Botrytis cinerea B05.10|XM_001545555.2\" in the CUPP input FASTA file. The entire contents of this protein data line would be included in the CUPP output FASTA file under the 'Accession' field. If the 'Accession' field contains '|', the items withint the 'Accession' field would also be separated and thus disrupt the selective retrieval of data.\n",
    "\n",
    "The code blocks below demonstrate this theory."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['>XP_024545922.1 BCIN_01g00340',\n",
       " '&',\n",
       " 'GT2:2.4.1.117',\n",
       " 'GT2:2.4.1.117',\n",
       " 'GT2:23.1(20.2,116..300)',\n",
       " '']"
      ]
     },
     "execution_count": 43,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Example that would work\n",
    "protein_data_line = '>XP_024545922.1 BCIN_01g00340|&|GT2:2.4.1.117|GT2:2.4.1.117|GT2:23.1(20.2,116..300)|'\n",
    "protein_data = protein_data_line.split(\"|\")\n",
    "protein_data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "best function= GT2:2.4.1.117\n"
     ]
    }
   ],
   "source": [
    "print(\"best function=\", protein_data[3])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['>XP_001545605.1',\n",
       " 'BCIN_01g00260',\n",
       " 'Botrytis cinerea B05.10',\n",
       " 'XM_001545555.2',\n",
       " '&',\n",
       " 'GT2:2.4.1.117',\n",
       " 'GT2:2.4.1.117',\n",
       " 'GT2:23.1(20.2,116..300)',\n",
       " '']"
      ]
     },
     "execution_count": 45,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Example that would cause the incorrect data to be retrieved\n",
    "protein_data_line = '>XP_001545605.1|BCIN_01g00260|Botrytis cinerea B05.10|XM_001545555.2|&|GT2:2.4.1.117|GT2:2.4.1.117|GT2:23.1(20.2,116..300)|'\n",
    "protein_data = protein_data_line.split(\"|\")\n",
    "protein_data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "best function= XM_001545555.2\n"
     ]
    }
   ],
   "source": [
    "print(\"best function=\", protein_data[3])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Using '|' as a separator is common. Therefore, using '|' to identify separations between individual items of data may be a brittle approach when parsing the output from CUPP.\n",
    "\n",
    "The data in the output Log file is separated by tabs '\\t'. This is less common that '|' and was therefore, deemed the least brittle of the two files to parse. Consequently, a method to parse the CUPP output log file was developed."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<a id=\"cupp_parse_log\"></a>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Parsing the CUPP log file\n",
    "\n",
    "The first task is to separate the out the data fields created by CUPP into separate elements in a list."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<a id=\"data_model\"></a>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Creating a data model\n",
    "\n",
    "To store the data parsed from CUPP a class is created to store the data. This is to enable checking for multiple entries for the same protein in the CUPP log file, which does sometime occur. If there are multiple 'Best function' results for a given CAZyme, CUPP will write these out on separate lines in its log file. Each different 'Best function' prediction is interpreted as predicting a different/new CAZyme domain within the query protein. Additionally, sometimes a protein appears on multiple rows if multiple domain ranges are predicted for a given domain within the CAZyme.\n",
    "\n",
    "Therefore, each CAZyme predicted by CUPP can have multiple CAZyme domains (identified by their different predicted CAZy (sub)families). Additionally, each CAZyme domain has its own associated predicted amino acid domain ranges. To ensure the association between the domain ranges and their CAZyme domain are retained, a class for representing the CUPP predicted CAZyme domain was written. Each CAZyme domain instance represents one CAZyme domain predicted by CUPP, and these instances are stored within sets of their parent proteins respective `CUPPprediction` class instance, which represents the parent protein.\n",
    "\n",
    "Each CAZyme domain will be represented by a `CUPPCazymeDomain` instance, which stores the predicted:\n",
    "- CAZy family (str)\n",
    "- CAZy subfamily (str if given, else a null value)\n",
    "- EC numbers (list of strings, each string containing a unique EC number)\n",
    "- Amino acid domain range (list of strings, each string is different predicted domain range)\n",
    "\n",
    "Each protein from the input FASTA file containing all the query protein sequences that were parsed by CUPP will be represented by the `CUPPprediction` class, containing the:\n",
    "- protein accession (str)\n",
    "- CAZyme classificaiton (int, 1 if predicted to be a CAZyme, 0 if predicted to not be a CAZyme)\n",
    "- CAZyme domains (set of `CUPPCazymeDomain` instances)\n",
    "- Methods for returning dictionaries that enable the easy building of Pandas dataframes for down stream evaluation of the performance of CUPP\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "class CUPPCazymeDomain:\n",
    "    \"\"\"Single CAZyme domain in a protein, predicted by CUPP\"\"\"\n",
    "    \n",
    "    def __init__(self, protein_accession, cazy_family, cazy_subfamily=None, ec_numbers=None, domain_range=None):\n",
    "        \"\"\"Initiate instance\n",
    "        \n",
    "        :attr protein_accession: str\n",
    "        :attr cazy_family: str\n",
    "        :attr cazy_subfamily: str\n",
    "        :attr ec_numbers: list (list of str, each str contains a unique EC number)\n",
    "        :attr domain_range: list (list of str, each str contains a unique domain range)\n",
    "        \"\"\"\n",
    "        self.protein_accession = protein_accession\n",
    "        self.cazy_family = cazy_family\n",
    "        \n",
    "        # not all CAZyme domans are catalogued under a CAZy subfamily\n",
    "        if cazy_subfamily is None:\n",
    "            self.cazy_subfamiy = np.nan\n",
    "        else:\n",
    "            self.cazy_subfamily = cazy_subfamily\n",
    "        \n",
    "        # EC numbers are not predicted for all CAZyme domains\n",
    "        if ec_numbers is None:\n",
    "            self.ec_numbers = []  # enables adding in EC numbers included in another line of the output file\n",
    "        else:\n",
    "            self.ec_numbers = ec_numbers\n",
    "        \n",
    "        # A domain range may not be predicted\n",
    "        if domain_range is None:\n",
    "            self.domain_range = []  # enables adding domain range listed in another line of the ouput file\n",
    "        else:\n",
    "            self.domain_range = domain_range\n",
    "    \n",
    "    def __str__(self):\n",
    "        return f\"-CUPPCazymeDomain in {self.protein_accession}, fam={self.cazy_family}, subfam={self.cazy_subfamily}-\"\n",
    "    \n",
    "    def __repr__(self):\n",
    "        return f\"<CUPPCazymeDomain parent={self.protein_accession} fam={self.cazy_family}, subfam={self.cazy_subfamily}>\"\n",
    "\n",
    "\n",
    "class CUPPprediction:\n",
    "    \"\"\"Single protein and CUPP CAZyme/non-CAZyme classification and CAZyme domain predictions.\"\"\"\n",
    "    \n",
    "    cazyme_domains = set()  # store all CUPPCazymeDomain instances predicted to be in the protein\n",
    "    \n",
    "    def __init__(self, protein_accession, cazyme_classification):\n",
    "        \"\"\"Initate class instance.\n",
    "        \n",
    "        :attr protein_accession: str\n",
    "        :attr cazyme_classification: int, 1=CAZyme, 0=non-CAZyme\n",
    "        \"\"\"\n",
    "        self.protein_accession = protein_accession\n",
    "        self.cazyme_classification = cazyme_classification  # CAZyme=1, non-CAZyme=0\n",
    "    \n",
    "    def __str__(self):\n",
    "        if self.cazyme_classification == 0:\n",
    "            return f\"-CUPPprediction, protein={self.protein_accession}, non-CAZyme-\"\n",
    "        else:\n",
    "            return f\"-CUPPprediction, protein={self.protein_accession}, CAZyme domains={len(cazyme_domains)}-\"\n",
    "    \n",
    "    def __repr__(self):\n",
    "        return f\"<CUPPprediction, protein={self.protein_accession}>\"\n",
    "\n",
    "    def get_cazyme_classification_dicts(self):\n",
    "        \"\"\"Returns a dictionary that contains the protein accession and its CAZyme/non-CAZyme classification.\n",
    "        \n",
    "        This dictionary is used to create the cazyme_classification dataframe for evaluating the performance\n",
    "        of the CAZyme prediction tools in terms of differnetiating between CAZymes (identified with a score of 1)\n",
    "        and non-CAZymes (identified with a score of 0). The CAZy classes under which the protein is classifed are\n",
    "        retrieved as well to enable evaluating the CAZyme/non-CAZyme differneitation per CAZy class, and obsever\n",
    "        differences between this performance between classes. A score of 0 incidates a protein was not classified\n",
    "        under the class, a score of 1 indicates the protein as classified under the class.\n",
    "        \"\"\"\n",
    "        cazyme_classification_dict = {\n",
    "            \"protein_accession\": self.protein_accession,\n",
    "            \"cazyme_classification\": self.cazyme_classification,\n",
    "            \"GH\": 0,\n",
    "            \"GT\": 0,\n",
    "            \"CE\": 0,\n",
    "            \"PL\": 0,\n",
    "            \"AA\": 0,\n",
    "            \"CBM\": 0,\n",
    "        }\n",
    "        \n",
    "        # check which CAZy classes the predicted CAZyme domains were classified under\n",
    "        for domain in cazyme_domains:\n",
    "            if domain.cazy_family[:3] == \"CBM\":\n",
    "                cazyme_classification_dict[\"CBM\"] = 1\n",
    "            else:\n",
    "                cazyme_classification_dict[family[:2]] = 1\n",
    "\n",
    "        return cazyme_classification_dict\n",
    "\n",
    "    def get_cazy_family_dict(self):\n",
    "        \"\"\"Return a dictionary of the predicted CAZy families for a protein.\n",
    "        \n",
    "        This dictionary is used to create a dataframe of CAZy family predictions for evaluating the performance\n",
    "        of the prediction tools to predict the CAZy family of a CAZyme. The dictionary is keyed by the CAZy\n",
    "        classes, and valued by a list of the corresponding families predicted for the CAZyme. The is to enable\n",
    "        evaluating the CAZy family prediction performance per class.\"\"\"\n",
    "        cazy_family_dict = {\n",
    "            \"GH\": [],\n",
    "            \"GT\": [],\n",
    "            \"CE\": [],\n",
    "            \"PL\": [],\n",
    "            \"AA\": [],\n",
    "            \"CBM\": [],\n",
    "        }\n",
    "        \n",
    "        for domain in self.cazyme_domains:\n",
    "            if domain.cazy_family[:3] == \"CBM\":\n",
    "                cazy_family_dict[\"CBM\"].append(domain.cazy_family)\n",
    "            else:\n",
    "                cazy_family_dict[domain.cazy_family[:2].append(domain.cazy_family)]\n",
    "        \n",
    "        return cazy_family_dict\n",
    "    \n",
    "    def get_cazyme_domains(self):\n",
    "        \"\"\"Returns a list of all CAZyme domains dbCAN predicts is in the protein.\"\"\"\n",
    "        return self.cazyme_domains\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now lets split up the data in the CUPP log file to prepare it for adding it to the data model. Additionally, we create a dictionary, key by the protein accession and valued by its corresponding CUPPprediction instance. This is to enable quick checking if a CUPPprediction instance has already been created for a protein."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['XP_001545605.1 BCIN_01g00260', 'AA4', 'AA4:1.1', '13.9', '7.39', '268', '122..568', '107', 'AA4:1.1.3.38', 'AA4:1.1.3.38', '']\n",
      "['XP_024545922.1 BCIN_01g00340', 'GT2', 'GT2:23.1', '20.2', '7.74', '211', '116..300', '77', 'GT2:2.4.1.117', 'GT2:2.4.1.117', '']\n"
     ]
    }
   ],
   "source": [
    "cupp_predictions = {}  # stores proteins {protein_accession:CUPPprediction_instance}\n",
    "\n",
    "# first split the line into the individual pieces of data\n",
    "for line in log_file[:2]:\n",
    "    prediction_output = line.split(\"\\t\")\n",
    "    print(prediction_output)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<a id=\"cupp_domain_range\"></a>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Getting the predicted domain range\n",
    "\n",
    "The domain range is stored within the 'additional data' field by CUPP. However, if CUPP does not commute some of the fields within 'additional data' an empty object (such as an empty string) is _not_ written out to the output FASTA or log files.\n",
    "\n",
    "Reminder of the data included in the 'additional data' field by CUPP:\n",
    "- \"Ratio of maximum frequencies\"\n",
    "- \"Ratio of maximum frequencies by coverage\"\n",
    "- \"Centre for mass for domain\"\n",
    "- \"Domain range\" -- this is identifable by the presense of '...' between two numbers and could potentially be useful when looking for regions of positive selection down the road\n",
    "- \"Number of important positions in domain\" -- very ambgious of meaning\n",
    "\n",
    "Therefore, if only the domain range is calculated, only the data for the domain range will be included, producing the following output:\n",
    "`XP_001545605.1 BCIN_01g00260  AA4  AA4:1.1 15...250`\n",
    "If the 'Ratio of maximum frequencies' and domain range are calculated, then two elements of data will be included in the space allocated for 'additional data' within the output FASTA and log files. Also, these elements within the 'additional data' are separated by the same character used to separate the other data fields in the output file (such as '\\t' for output log file). Therefore, the following output would be produced:\n",
    "`XP_001545605.1 BCIN_01g00260  AA4  AA4:1.1  7.4  15...250` \n",
    "\n",
    "Therefore, retrieving the predicted domain range can not be achieved by calling for an element in the `prediction_output` list by its index. However, the domain range is easy to identify by its characteristic '..' between two digits. The remaining data can be retrieved based upon its index in the `prediction_output` list."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "protein_accession = XP_001547254.1 BCIN_01g00800\n",
      "CAZy family = AA1\n",
      "CAZy subfamily = AA1_3\n",
      "EC# = AA1:1.10.3.2\n",
      "Domain range = 202..455\n"
     ]
    }
   ],
   "source": [
    "for line in log_file[:4]:\n",
    "    # separate the data fields\n",
    "    prediction_output = line.split(\"\\t\")\n",
    "    \n",
    "    for item in prediction_output:\n",
    "        if item.find(\"..\") != -1:\n",
    "            domain_range = item\n",
    "            break\n",
    "        else:\n",
    "            domain_range = np.nan\n",
    "    \n",
    "    # store the prediction output in a dictionary\n",
    "    prediction = {\n",
    "        \"protein_accession\": prediction_output[0],\n",
    "        \"cazy_family\": prediction_output[1],\n",
    "        \"cazy_subfamily\": prediction_output[-1],\n",
    "        \"ec_number\": prediction_output[-2],\n",
    "        \"domain_range\": domain_range,\n",
    "    }\n",
    "\n",
    "print(\"protein_accession =\", prediction[\"protein_accession\"])\n",
    "print(\"CAZy family =\", prediction[\"cazy_family\"])\n",
    "print(\"CAZy subfamily =\", prediction[\"cazy_subfamily\"])\n",
    "print(\"EC# =\", prediction[\"ec_number\"])\n",
    "print(\"Domain range =\", prediction[\"domain_range\"])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "This can be written into a single function, and a 'quality' control to check the believed domain range is in the expected format (_digit_.._digit_), to increase the probability of correctly identifying the domain range."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "122..568\n",
      "116..300\n",
      "26..553\n",
      "202..455\n"
     ]
    }
   ],
   "source": [
    "def get_cupp_domain_range(prediction_output):\n",
    "    \"\"\"Retrieve the predicted amino acid range of predicted CAZyme domain from CUPP log file.\n",
    "    \n",
    "    :param prediction_output: list of items from log file line.\n",
    "    \n",
    "    Return string if domain range given, or null value if a domain range is not given.\n",
    "    \"\"\"\n",
    "    domain_range = []  # store as a string in case multiple domain ranges are given\n",
    "\n",
    "    for item in prediction_output:\n",
    "        if item.find(\"..\") != -1:\n",
    "            domain_range.append(item)\n",
    "\n",
    "    # check retrieved items are definetly the domain ranges\n",
    "    for item in domain_range:\n",
    "        try:\n",
    "            re.match(r\"\\d+\\.\\.\\d+\", item).group()\n",
    "        except AttributeError:\n",
    "            # write as logger in pyrewton\n",
    "            print(f\"RAISED ATTRIBUTE ERROR: {item} misidentified as domain range\")\n",
    "            domain_range.remove(item)\n",
    "\n",
    "    if len(domain_range) == 0:\n",
    "        domain_range = np.nan\n",
    "    \n",
    "    else:\n",
    "        domain_range = \", \".join(domain_range)\n",
    "    \n",
    "    return domain_range\n",
    "\n",
    "\n",
    "for line in log_file[:4]:\n",
    "    # separate the data fields\n",
    "    prediction_output = line.split(\"\\t\")\n",
    "    domain_range = get_cupp_domain_range(prediction_output)\n",
    "    print(domain_range)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<a id=\"cupp_ec\"></a>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Getting the predicted EC number\n",
    "\n",
    "The CAZy family and associated EC number are stored within the same element in the `prediction_output` list. The CAZy family is always listed first, followed by a colon and then the EC number, which allows the retrieval of only the EC number.\n",
    "\n",
    "The item at position `[-2]` in prediction_output is the 'Best function' predicted by CUPP, the item at position `[-3]` is the raw function. The item listed is the 'Best function' is taken as the predicted CAZy family and associated EC numbers for the protein."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "protein_accession = XP_001547254.1 BCIN_01g00800\n",
      "CAZy family = AA1\n",
      "CAZy subfamily = AA1_3\n",
      "EC# = 1.10.3.2\n",
      "Domain range = 202..455\n"
     ]
    }
   ],
   "source": [
    "for line in log_file[:4]:\n",
    "    # separate the data fields\n",
    "    prediction_output = line.split(\"\\t\")\n",
    "    \n",
    "    # retrieve the predicted domain range\n",
    "    for item in prediction_output:\n",
    "        if item.find(\"..\") != -1:\n",
    "            domain_range = item\n",
    "            break\n",
    "        else:\n",
    "            domain_range = np.nan\n",
    "    \n",
    "    # retrieve the predicted EC number\n",
    "    try:\n",
    "        ec_number = prediction_output[-2].split(\":\")[1]\n",
    "    except IndexError:\n",
    "        ec_number = np.nan\n",
    "    \n",
    "    # store the prediction output in a dictionary\n",
    "    prediction = {\n",
    "        \"protein_accession\": prediction_output[0],\n",
    "        \"cazy_family\": prediction_output[1],\n",
    "        \"cazy_subfamily\": prediction_output[-1],\n",
    "        \"ec_number\": ec_number,\n",
    "        \"domain_range\": domain_range,\n",
    "    }\n",
    "\n",
    "print(\"protein_accession =\", prediction[\"protein_accession\"])\n",
    "print(\"CAZy family =\", prediction[\"cazy_family\"])\n",
    "print(\"CAZy subfamily =\", prediction[\"cazy_subfamily\"])\n",
    "print(\"EC# =\", prediction[\"ec_number\"])\n",
    "print(\"Domain range =\", prediction[\"domain_range\"])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<a id=\"cupp_ec_quality\"></a>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Adding quality checks when retrieving EC numbers\n",
    "\n",
    "The retrieval of the predicted EC number appears susceptible to errors. The appear to result from not fully analysing all potential formats that CUPP may write out its predicted EC numbers.\n",
    "\n",
    "Therefore, first lets look at the EC numbers listed in the CUPP output log file."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "AA4:1.1.3.38 'AA4:1.1.3.38'\n",
      "GT2:2.4.1.117 'GT2:2.4.1.117'\n",
      " ''\n",
      "AA1:1.10.3.2 'AA1:1.10.3.2'\n",
      " ''\n",
      " ''\n",
      "GT39:2.4.1.109 'GT39:2.4.1.109'\n",
      "GT2:2.4.1.16 'GT2:2.4.1.16'\n",
      "AA11:1.*.*.* 'AA11:1.*.*.*'\n",
      "GH71:3.2.1.59 'GH71:3.2.1.59'\n",
      "GH71:3.2.1.59 'GH71:3.2.1.59'\n",
      "GH38:3.2.1.24 'GH38:3.2.1.24'\n",
      "GH38:3.2.1.24 'GH38:3.2.1.24'\n",
      "GT2:2.4.1.16 'GT2:2.4.1.16'\n",
      "GT2:2.4.1.16 'GT2:2.4.1.16'\n",
      " ''\n",
      "GH47:3.2.1.113 'GH47:3.2.1.113'\n",
      " ''\n",
      "GH114:3.2.1.109 'GH114:3.2.1.109'\n",
      "GH135:3.2.1.* 'GH135:3.2.1.*'\n",
      "GH16:2.4.1.*&3.2.1.*-GH16:3.2.1.39 'GH16:2.4.1.*&3.2.1.*-GH16:3.2.1.39'\n",
      "GH47:3.2.1.113 'GH47:3.2.1.113'\n",
      "CE12:3.1.1.* 'CE12:3.1.1.*'\n",
      "AA1:1.10.3.2 'AA1:1.10.3.2'\n",
      " ''\n",
      "GH28:3.2.1.67 'GH28:3.2.1.67'\n",
      " ''\n",
      "GH28:3.2.1.15 'GH28:3.2.1.15'\n",
      "GH125:3.2.1.* 'GH125:3.2.1.*'\n",
      "AA1:1.10.3.2 'AA1:1.10.3.2'\n",
      "GH27:3.2.1.49 'GH27:3.2.1.49'\n",
      "GH135:3.2.1.* 'GH135:3.2.1.*'\n",
      "GT50:2.4.1.* 'GT50:2.4.1.*'\n",
      "CE16:3.1.1.6 'CE16:3.1.1.6'\n",
      " ''\n",
      "GT20:2.4.1.15 'GT20:2.4.1.15'\n",
      " ''\n",
      "CE5:3.1.1.74 'CE5:3.1.1.74'\n",
      " ''\n",
      "CE9:3.5.1.25 'CE9:3.5.1.25'\n",
      "GH13:2.4.1.25&3.2.1.33 'GH13:2.4.1.25&3.2.1.33'\n",
      "GH13:2.4.1.25&3.2.1.33 'GH13:2.4.1.25&3.2.1.33'\n",
      " ''\n",
      "CE5:3.1.1.74 'CE5:3.1.1.74'\n",
      " ''\n",
      "CE8:3.1.1.11 'CE8:3.1.1.11'\n",
      "GH28:3.2.1.* 'GH28:3.2.1.*'\n",
      " ''\n",
      "GH76:3.2.1.101 'GH76:3.2.1.101'\n",
      " ''\n",
      "GH88:3.2.1.* 'GH88:3.2.1.*'\n",
      " ''\n",
      " ''\n",
      " ''\n",
      " ''\n",
      " ''\n",
      "GH71:3.2.1.59 'GH71:3.2.1.59'\n",
      "GH71:3.2.1.59 'GH71:3.2.1.59'\n",
      " ''\n",
      "GH18:3.2.1.96 'GH18:3.2.1.96'\n",
      "GT21:2.4.1.80 'GT21:2.4.1.80'\n",
      " ''\n",
      " ''\n",
      "GH27:3.2.1.22-GH27:3.2.1.49 'GH27:3.2.1.22-GH27:3.2.1.49'\n",
      "GH45:3.2.1.4 'GH45:3.2.1.4'\n",
      "GT39:2.4.1.109 'GT39:2.4.1.109'\n",
      " ''\n",
      "AA9:1.*.*.* 'AA9:1.*.*.*'\n",
      "GT66:2.4.99.18 'GT66:2.4.99.18'\n",
      " ''\n",
      "AA1:1.10.3.2 'AA1:1.10.3.2'\n",
      "AA1:1.10.3.2 'AA1:1.10.3.2'\n",
      "GH1:3.2.1.21&3.2.1.23 'GH1:3.2.1.21&3.2.1.23'\n",
      "GT15:2.4.1.131 'GT15:2.4.1.131'\n",
      " ''\n",
      "GH16:3.2.1.6-GH16:3.2.1.73 'GH16:3.2.1.6-GH16:3.2.1.73'\n",
      " ''\n",
      " ''\n",
      "GT22:2.4.1.259&2.4.1.261 'GT22:2.4.1.259&2.4.1.261'\n",
      "GT24:2.4.1.* 'GT24:2.4.1.*'\n",
      "GH13:3.2.1.20-GH13:5.4.99.11 'GH13:3.2.1.20-GH13:5.4.99.11'\n",
      "GH28:3.2.1.15 'GH28:3.2.1.15'\n",
      " ''\n",
      " ''\n",
      " ''\n",
      "GT48:2.4.1.34 'GT48:2.4.1.34'\n",
      "GH72:2.4.1.* 'GH72:2.4.1.*'\n",
      " ''\n",
      "CE12:3.1.1.* 'CE12:3.1.1.*'\n",
      "GH16:3.2.1.6-GH16:3.2.1.73 'GH16:3.2.1.6-GH16:3.2.1.73'\n",
      "GH16:3.2.1.6-GH16:3.2.1.73 'GH16:3.2.1.6-GH16:3.2.1.73'\n",
      " ''\n",
      "AA1:1.10.3.2 'AA1:1.10.3.2'\n",
      "GH43:3.2.1.99 'GH43:3.2.1.99'\n",
      "GH76:3.2.1.101 'GH76:3.2.1.101'\n",
      "GH5:3.2.1.45 'GH5:3.2.1.45'\n",
      "GT20:2.4.1.15 'GT20:2.4.1.15'\n",
      "GH93:3.2.1.* 'GH93:3.2.1.*'\n",
      "GH2:3.2.1.23 'GH2:3.2.1.23'\n",
      "GH71:3.2.1.59 'GH71:3.2.1.59'\n",
      "PL1:4.2.2.10 'PL1:4.2.2.10'\n",
      " ''\n",
      "GH11:3.2.1.8 'GH11:3.2.1.8'\n",
      " ''\n",
      "AA3:1.1.3.13 'AA3:1.1.3.13'\n",
      " ''\n",
      "GH5:3.2.1.78 'GH5:3.2.1.78'\n",
      " ''\n",
      " ''\n",
      "GH28:3.2.1.15 'GH28:3.2.1.15'\n",
      "GT22:2.4.1.* 'GT22:2.4.1.*'\n",
      " ''\n",
      "GH27:3.2.1.22 'GH27:3.2.1.22'\n",
      " ''\n",
      " ''\n",
      "GH10:3.2.1.8 'GH10:3.2.1.8'\n",
      " ''\n",
      "GH12:3.2.1.151 'GH12:3.2.1.151'\n",
      "CE8:3.1.1.11 'CE8:3.1.1.11'\n",
      " ''\n",
      "GH5:3.2.1.4 'GH5:3.2.1.4'\n",
      "GT32:2.4.1.* 'GT32:2.4.1.*'\n",
      " ''\n",
      "CE8:3.1.1.11 'CE8:3.1.1.11'\n",
      "GH81:3.2.1.39 'GH81:3.2.1.39'\n",
      " ''\n",
      " ''\n",
      "PL1:4.2.2.2 'PL1:4.2.2.2'\n",
      "AA9:1.*.*.* 'AA9:1.*.*.*'\n",
      "GT76:2.4.1.* 'GT76:2.4.1.*'\n",
      " ''\n",
      "GT90:2.4.1.*&2.4.2.* 'GT90:2.4.1.*&2.4.2.*'\n",
      " ''\n",
      " ''\n",
      " ''\n",
      "GH15:3.2.1.3 'GH15:3.2.1.3'\n",
      "GH71:3.2.1.59 'GH71:3.2.1.59'\n",
      "PL3:4.2.2.2 'PL3:4.2.2.2'\n",
      "GT62:2.4.1.* 'GT62:2.4.1.*'\n",
      " ''\n",
      " ''\n",
      "GT2:2.4.1.16 'GT2:2.4.1.16'\n",
      "GH15:3.2.1.3 'GH15:3.2.1.3'\n",
      "GH15:3.2.1.3 'GH15:3.2.1.3'\n",
      "GH28:3.2.1.15 'GH28:3.2.1.15'\n",
      "GH35:3.2.1.23 'GH35:3.2.1.23'\n",
      "GT2:2.4.1.* 'GT2:2.4.1.*'\n",
      " ''\n",
      " ''\n",
      "GT4:2.4.1.131 'GT4:2.4.1.131'\n",
      "GH43:3.2.1.99 'GH43:3.2.1.99'\n",
      "GH47:3.2.1.113 'GH47:3.2.1.113'\n",
      "GH2:3.2.1.25 'GH2:3.2.1.25'\n",
      "GH16:3.2.1.39 'GH16:3.2.1.39'\n",
      "GT1:2.4.1.173 'GT1:2.4.1.173'\n",
      " ''\n",
      " ''\n",
      " ''\n",
      "GH28:3.2.1.67 'GH28:3.2.1.67'\n",
      " ''\n",
      "GT69:2.4.1.* 'GT69:2.4.1.*'\n",
      " ''\n",
      "AA1:1.10.3.2 'AA1:1.10.3.2'\n",
      " ''\n",
      " ''\n",
      "GH72:2.4.1.* 'GH72:2.4.1.*'\n",
      "AA7:1.1.3.* 'AA7:1.1.3.*'\n",
      " ''\n",
      "GH10:3.2.1.8 'GH10:3.2.1.8'\n",
      "AA8:1.1.99.18 'AA8:1.1.99.18'\n",
      " ''\n",
      " ''\n",
      "GH64:3.2.1.39 'GH64:3.2.1.39'\n",
      "GH3:Unknown 'GH3:Unknown'\n",
      "GH76:3.2.1.101 'GH76:3.2.1.101'\n",
      "GH115:3.2.1.*-GH115:3.2.1.131 'GH115:3.2.1.*-GH115:3.2.1.131'\n",
      " ''\n",
      "GH28:3.2.1.67 'GH28:3.2.1.67'\n",
      " ''\n",
      "GH89:3.2.1.50 'GH89:3.2.1.50'\n",
      "GT71:2.4.1.* 'GT71:2.4.1.*'\n",
      "GT71:2.4.1.* 'GT71:2.4.1.*'\n",
      "GT1:2.4.1.173 'GT1:2.4.1.173'\n",
      "AA9:1.*.*.* 'AA9:1.*.*.*'\n",
      " ''\n",
      " ''\n",
      "GH55:3.2.1.58 'GH55:3.2.1.58'\n",
      "AA7:1.1.3.* 'AA7:1.1.3.*'\n",
      " ''\n",
      " ''\n",
      " ''\n",
      "GH43:3.2.1.145 'GH43:3.2.1.145'\n",
      "GH79:3.2.1.*&3.2.1.31 'GH79:3.2.1.*&3.2.1.31'\n",
      " ''\n",
      "GH28:3.2.1.40 'GH28:3.2.1.40'\n",
      "GT62:2.4.1.* 'GT62:2.4.1.*'\n",
      "GH125:3.2.1.* 'GH125:3.2.1.*'\n",
      "GH76:3.2.1.101 'GH76:3.2.1.101'\n",
      " ''\n",
      "CE16:3.1.1.6 'CE16:3.1.1.6'\n",
      " ''\n",
      " ''\n",
      " ''\n",
      "GH47:3.2.1.113 'GH47:3.2.1.113'\n",
      "GH5:3.2.1.75 'GH5:3.2.1.75'\n",
      "GH3:3.2.1.21 'GH3:3.2.1.21'\n",
      "GH31:3.2.1.177 'GH31:3.2.1.177'\n",
      "AA13:1.*.*.* 'AA13:1.*.*.*'\n",
      "GH64:3.2.1.39 'GH64:3.2.1.39'\n",
      "GT71:2.4.1.* 'GT71:2.4.1.*'\n",
      "GT15:2.4.1.131 'GT15:2.4.1.131'\n",
      "GH47:3.2.1.113 'GH47:3.2.1.113'\n",
      "GH132:3.2.1.* 'GH132:3.2.1.*'\n",
      "GH132:3.2.1.* 'GH132:3.2.1.*'\n",
      "AA6:1.6.5.6 'AA6:1.6.5.6'\n",
      "CE16:3.1.1.6 'CE16:3.1.1.6'\n",
      "GH78:3.2.1.40 'GH78:3.2.1.40'\n",
      " ''\n",
      " ''\n",
      " ''\n",
      "AA9:1.*.*.* 'AA9:1.*.*.*'\n",
      "GT33:2.4.1.142 'GT33:2.4.1.142'\n",
      "GH71:3.2.1.59 'GH71:3.2.1.59'\n",
      "GH18:3.2.1.14 'GH18:3.2.1.14'\n",
      "GT90:2.4.1.*&2.4.2.* 'GT90:2.4.1.*&2.4.2.*'\n",
      "GT22:2.4.1.* 'GT22:2.4.1.*'\n",
      "GH18:3.2.1.14 'GH18:3.2.1.14'\n",
      "GT20:3.1.3.12 'GT20:3.1.3.12'\n",
      "GT20:3.1.3.12 'GT20:3.1.3.12'\n",
      "GT20:3.1.3.12 'GT20:3.1.3.12'\n",
      "CE5:3.1.1.74 'CE5:3.1.1.74'\n",
      "GT2:2.4.1.16 'GT2:2.4.1.16'\n",
      " ''\n",
      " ''\n",
      "GH15:3.2.1.3 'GH15:3.2.1.3'\n",
      "GH15:3.2.1.3 'GH15:3.2.1.3'\n",
      "GH5:3.2.1.164 'GH5:3.2.1.164'\n",
      "GH71:3.2.1.59 'GH71:3.2.1.59'\n",
      "GH55:3.2.1.58 'GH55:3.2.1.58'\n",
      " ''\n",
      "GT69:2.4.1.* 'GT69:2.4.1.*'\n",
      "GT69:2.4.1.* 'GT69:2.4.1.*'\n",
      "GT69:2.4.1.* 'GT69:2.4.1.*'\n",
      "AA3:1.1.3.13 'AA3:1.1.3.13'\n",
      "CE16:3.1.1.6 'CE16:3.1.1.6'\n",
      "GH16:3.2.1.35 'GH16:3.2.1.35'\n",
      "GH16:3.2.1.35 'GH16:3.2.1.35'\n",
      "AA11:1.*.*.* 'AA11:1.*.*.*'\n",
      " ''\n",
      " ''\n",
      " ''\n",
      "GH20:3.2.1.52 'GH20:3.2.1.52'\n",
      "GH20:3.2.1.52 'GH20:3.2.1.52'\n",
      "GH16:2.4.1.*&3.2.1.*-GH16:3.2.1.39 'GH16:2.4.1.*&3.2.1.*-GH16:3.2.1.39'\n",
      " ''\n",
      "GH78:Unknown 'GH78:Unknown'\n",
      "GH47:3.2.1.113 'GH47:3.2.1.113'\n",
      "GH47:3.2.1.113 'GH47:3.2.1.113'\n",
      " ''\n",
      "GH132:3.2.1.* 'GH132:3.2.1.*'\n",
      "GH1:3.2.1.21 'GH1:3.2.1.21'\n",
      "GH1:3.2.1.21 'GH1:3.2.1.21'\n",
      "GH95:3.2.1.63 'GH95:3.2.1.63'\n",
      "AA1:1.10.3.2 'AA1:1.10.3.2'\n",
      " ''\n",
      "GH43:3.2.1.37 'GH43:3.2.1.37'\n",
      "GT1:2.4.1.173 'GT1:2.4.1.173'\n",
      "GT90:2.4.1.*&2.4.2.* 'GT90:2.4.1.*&2.4.2.*'\n",
      "GH51:3.2.1.55 'GH51:3.2.1.55'\n",
      "GH17:2.4.1.* 'GH17:2.4.1.*'\n",
      " ''\n",
      " ''\n",
      " ''\n",
      " ''\n",
      "AA3:Unknown 'AA3:Unknown'\n",
      "GH5:3.2.1.4 'GH5:3.2.1.4'\n",
      " ''\n",
      " ''\n",
      "GH13:2.4.1.* 'GH13:2.4.1.*'\n",
      "GH13:2.4.1.* 'GH13:2.4.1.*'\n",
      "CE8:3.1.1.11 'CE8:3.1.1.11'\n",
      "AA1:1.10.3.2 'AA1:1.10.3.2'\n",
      "PL1:4.2.2.10 'PL1:4.2.2.10'\n",
      "GT32:2.4.1.* 'GT32:2.4.1.*'\n",
      " ''\n",
      " ''\n",
      " ''\n",
      " ''\n",
      "GT15:2.4.1.131 'GT15:2.4.1.131'\n",
      "GH16:3.2.1.6 'GH16:3.2.1.6'\n",
      "GH95:3.2.1.63 'GH95:3.2.1.63'\n",
      "GT32:2.4.1.* 'GT32:2.4.1.*'\n",
      "GH76:3.2.1.101 'GH76:3.2.1.101'\n",
      "GH53:3.2.1.89 'GH53:3.2.1.89'\n",
      " ''\n",
      "GH18:3.2.1.14 'GH18:3.2.1.14'\n",
      "GH18:3.2.1.14 'GH18:3.2.1.14'\n",
      "GH76:3.2.1.101 'GH76:3.2.1.101'\n",
      " ''\n",
      " ''\n",
      "GH131:3.2.1.* 'GH131:3.2.1.*'\n",
      "GH131:3.2.1.* 'GH131:3.2.1.*'\n",
      "GT2:2.4.1.16 'GT2:2.4.1.16'\n",
      "GH37:3.2.1.28 'GH37:3.2.1.28'\n",
      " ''\n",
      " ''\n",
      "GH16:3.2.1.39 'GH16:3.2.1.39'\n",
      "AA1:1.10.3.2 'AA1:1.10.3.2'\n",
      "GH76:3.2.1.101 'GH76:3.2.1.101'\n",
      " ''\n",
      "GT1:2.4.1.173 'GT1:2.4.1.173'\n",
      " ''\n",
      "GH16:3.2.1.6-GH16:3.2.1.73 'GH16:3.2.1.6-GH16:3.2.1.73'\n",
      "AA5:1.1.3.9 'AA5:1.1.3.9'\n",
      "AA5:1.1.3.9 'AA5:1.1.3.9'\n",
      " ''\n",
      "GH51:3.2.1.55 'GH51:3.2.1.55'\n",
      "GH65:3.2.1.28 'GH65:3.2.1.28'\n",
      " ''\n",
      "GH16:3.2.1.6-GH16:3.2.1.73 'GH16:3.2.1.6-GH16:3.2.1.73'\n",
      "AA1:1.10.3.2 'AA1:1.10.3.2'\n",
      "AA1:1.10.3.2 'AA1:1.10.3.2'\n",
      "AA1:1.10.3.2 'AA1:1.10.3.2'\n",
      " ''\n",
      " ''\n",
      "GT32:2.4.1.* 'GT32:2.4.1.*'\n",
      "AA9:1.*.*.* 'AA9:1.*.*.*'\n",
      "AA9:1.*.*.* 'AA9:1.*.*.*'\n",
      " ''\n",
      " ''\n",
      " ''\n",
      " ''\n",
      " ''\n",
      " ''\n",
      " ''\n",
      " ''\n",
      "GH55:3.2.1.58 'GH55:3.2.1.58'\n",
      "GH55:3.2.1.58 'GH55:3.2.1.58'\n",
      " ''\n",
      " ''\n",
      "GT8:2.4.1.186 'GT8:2.4.1.186'\n",
      "GH1:3.2.1.21 'GH1:3.2.1.21'\n",
      " ''\n",
      "GH127:3.2.1.185 'GH127:3.2.1.185'\n",
      " ''\n",
      " ''\n",
      "GH28:3.2.1.171 'GH28:3.2.1.171'\n",
      " ''\n",
      " ''\n",
      "PL3:4.2.2.2 'PL3:4.2.2.2'\n",
      " ''\n",
      "GH32:3.2.1.26 'GH32:3.2.1.26'\n",
      "GH28:3.2.1.40 'GH28:3.2.1.40'\n",
      "GH1:3.2.1.21 'GH1:3.2.1.21'\n",
      " ''\n",
      "GH55:3.2.1.58 'GH55:3.2.1.58'\n",
      " ''\n",
      "GT2:2.4.1.83 'GT2:2.4.1.83'\n",
      " ''\n",
      "GH35:3.2.1.23 'GH35:3.2.1.23'\n",
      " ''\n",
      " ''\n",
      " ''\n",
      "GT1:2.4.2.40 'GT1:2.4.2.40'\n",
      "AA9:1.*.*.* 'AA9:1.*.*.*'\n",
      "GH31:3.2.1.20 'GH31:3.2.1.20'\n",
      " ''\n",
      "GH35:3.2.1.23 'GH35:3.2.1.23'\n",
      "GT15:2.4.1.131 'GT15:2.4.1.131'\n",
      "GH5:3.2.1.78 'GH5:3.2.1.78'\n",
      "GH28:3.2.1.171 'GH28:3.2.1.171'\n",
      " ''\n",
      "GT2:2.4.1.16 'GT2:2.4.1.16'\n",
      "GT2:2.4.1.16 'GT2:2.4.1.16'\n",
      " ''\n",
      "GH16:3.2.1.39 'GH16:3.2.1.39'\n",
      "GH125:3.2.1.* 'GH125:3.2.1.*'\n",
      " ''\n",
      "GT62:2.4.1.* 'GT62:2.4.1.*'\n",
      " ''\n",
      "AA3:Unknown 'AA3:Unknown'\n",
      "AA8:1.1.99.18 'AA8:1.1.99.18'\n",
      " ''\n",
      "GH15:3.2.1.3 'GH15:3.2.1.3'\n",
      " ''\n",
      "AA9:1.*.*.* 'AA9:1.*.*.*'\n",
      "GH28:3.2.1.173 'GH28:3.2.1.173'\n",
      "GT58:2.4.1.258 'GT58:2.4.1.258'\n",
      " ''\n",
      "GH5:3.2.1.4 'GH5:3.2.1.4'\n",
      "GT2:2.4.1.16 'GT2:2.4.1.16'\n",
      "GT2:2.4.1.16 'GT2:2.4.1.16'\n",
      " ''\n",
      " ''\n",
      "GH131:3.2.1.* 'GH131:3.2.1.*'\n",
      "GH16:3.2.1.6-GH16:3.2.1.73 'GH16:3.2.1.6-GH16:3.2.1.73'\n",
      "GH5:3.2.1.149 'GH5:3.2.1.149'\n",
      "GH45:3.2.1.4 'GH45:3.2.1.4'\n",
      "AA1:1.10.3.2 'AA1:1.10.3.2'\n",
      "GT57:2.4.1.* 'GT57:2.4.1.*'\n",
      " ''\n",
      "GH12:3.2.1.151 'GH12:3.2.1.151'\n",
      "GH72:2.4.1.* 'GH72:2.4.1.*'\n",
      "GH28:3.2.1.171 'GH28:3.2.1.171'\n",
      "GH5:3.2.1.58 'GH5:3.2.1.58'\n",
      "GH76:3.2.1.101 'GH76:3.2.1.101'\n",
      "GH16:2.4.1.*&3.2.1.*-GH16:3.2.1.39 'GH16:2.4.1.*&3.2.1.*-GH16:3.2.1.39'\n",
      "GT4:2.4.1.132&2.4.1.257 'GT4:2.4.1.132&2.4.1.257'\n",
      " ''\n",
      " ''\n",
      "GH78:3.2.1.174-GH78:3.2.1.40 'GH78:3.2.1.174-GH78:3.2.1.40'\n",
      "AA3:Unknown 'AA3:Unknown'\n",
      "GT1:2.4.1.173 'GT1:2.4.1.173'\n",
      "GH78:3.2.1.174 'GH78:3.2.1.174'\n",
      " ''\n",
      "GT71:2.4.1.* 'GT71:2.4.1.*'\n",
      "AA5:1.1.3.9 'AA5:1.1.3.9'\n",
      "CE5:3.1.1.74 'CE5:3.1.1.74'\n",
      "AA3:Unknown 'AA3:Unknown'\n",
      " ''\n",
      " ''\n",
      " ''\n",
      "GH28:3.2.1.15 'GH28:3.2.1.15'\n",
      "GH28:3.2.1.15 'GH28:3.2.1.15'\n",
      "GH28:3.2.1.15 'GH28:3.2.1.15'\n",
      "GH28:3.2.1.15 'GH28:3.2.1.15'\n",
      "GH28:3.2.1.15 'GH28:3.2.1.15'\n",
      " ''\n",
      "GH28:3.2.1.15 'GH28:3.2.1.15'\n",
      "CE8:3.1.1.11 'CE8:3.1.1.11'\n",
      "GT71:2.4.1.* 'GT71:2.4.1.*'\n",
      "PL1:4.2.2.2 'PL1:4.2.2.2'\n",
      "GH5:3.2.1.4 'GH5:3.2.1.4'\n",
      " ''\n",
      " ''\n",
      " ''\n",
      " ''\n",
      "AA1:1.10.3.2 'AA1:1.10.3.2'\n",
      "GH51:3.2.1.55 'GH51:3.2.1.55'\n",
      " ''\n",
      " ''\n",
      " ''\n",
      "GH72:2.4.1.* 'GH72:2.4.1.*'\n",
      "CE9:3.5.1.25 'CE9:3.5.1.25'\n",
      " ''\n",
      "GH62:3.2.1.55 'GH62:3.2.1.55'\n",
      "CE5:3.1.1.72 'CE5:3.1.1.72'\n",
      "GT1:2.4.1.173 'GT1:2.4.1.173'\n",
      " ''\n",
      "GH47:3.2.1.113 'GH47:3.2.1.113'\n",
      "CE1:3.1.1.72 'CE1:3.1.1.72'\n",
      "AA3:1.1.99.18 'AA3:1.1.99.18'\n",
      "GH11:3.2.1.8 'GH11:3.2.1.8'\n",
      "GH71:3.2.1.59 'GH71:3.2.1.59'\n",
      "GT57:2.4.1.* 'GT57:2.4.1.*'\n",
      "GH13:2.4.1.18 'GH13:2.4.1.18'\n",
      "GH13:2.4.1.18 'GH13:2.4.1.18'\n",
      "GT3:2.4.1.11 'GT3:2.4.1.11'\n",
      "GH3:3.2.1.21 'GH3:3.2.1.21'\n",
      "GH16:2.4.1.*&3.2.1.*-GH16:3.2.1.39 'GH16:2.4.1.*&3.2.1.*-GH16:3.2.1.39'\n",
      "CE5:3.1.1.74 'CE5:3.1.1.74'\n",
      "AA9:1.*.*.* 'AA9:1.*.*.*'\n",
      " ''\n",
      "GT39:2.4.1.109 'GT39:2.4.1.109'\n",
      "GT35:2.4.1.1 'GT35:2.4.1.1'\n",
      "GT59:2.4.1.256 'GT59:2.4.1.256'\n",
      " ''\n",
      "GH72:2.4.1.* 'GH72:2.4.1.*'\n",
      "GH76:3.2.1.101 'GH76:3.2.1.101'\n",
      "GT1:2.4.1.173 'GT1:2.4.1.173'\n",
      "GH71:3.2.1.59 'GH71:3.2.1.59'\n",
      "GT69:2.4.1.* 'GT69:2.4.1.*'\n",
      "GH18:3.2.1.14 'GH18:3.2.1.14'\n",
      "GH63:3.2.1.106 'GH63:3.2.1.106'\n",
      "GH5:3.2.1.4 'GH5:3.2.1.4'\n",
      " ''\n",
      "GH1:3.2.1.21&3.2.1.23 'GH1:3.2.1.21&3.2.1.23'\n",
      "GH16:3.2.1.6-GH16:3.2.1.73 'GH16:3.2.1.6-GH16:3.2.1.73'\n",
      " ''\n",
      "GH47:3.2.1.113 'GH47:3.2.1.113'\n",
      "GH47:3.2.1.113 'GH47:3.2.1.113'\n",
      "GH32:3.2.1.7 'GH32:3.2.1.7'\n"
     ]
    }
   ],
   "source": [
    "for line in log_file:\n",
    "    # separate the data fields\n",
    "    prediction_output = line.split(\"\\t\")\n",
    "    print(prediction_output[-2], repr(prediction_output[-2]))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The above shows that when no EC number is predicted an empty string is provided (in the same manner as discussed when retrieving the predicted CAZy subfamily). It is better to store missing data as a null value instead. Furthermore, sometimes instead of writing out an empty string the word 'Unknown' is written out. For standardisation of the output this to will be converted to a null value.\n",
    "\n",
    "Additionally, it is more common practise to annotate missing digits in an EC number with `-` rather than `*`. Changing `*` or `-` would also bring the EC number output from CUPP in line with the other CAZyme prediction tools.\n",
    "\n",
    "For human readability, when multiple EC numbers are predicted, the '&' will be changed to separate the EC numbers by a comma and match the formating of presenting multiple predicted CAZy subfamilies.\n",
    "\n",
    "Another issue with the retrieval of the predicted EC numbers is that the above method does not retrieval _all_ predicted EC numbers. If multiple EC numbers are predicted these predictions may be separated by '-'.\n",
    "\n",
    "For example, protein XP_001561324.1 BCIN_01g06010, has the following output in the log file:\n",
    "```\n",
    "['XP_001561324.1 BCIN_01g06010', 'GH16', 'GH16:21.1', '14.8', '4.11', '123', '89..206', '56', 'GH16:2.4.1.*&3.2.1.*-GH16:3.2.1.39', 'GH16:2.4.1.*&3.2.1.*-GH16:3.2.1.39', '']\n",
    "```\n",
    "Where, `GH16:2.4.1.*&3.2.1.*-GH16:3.2.1.39` and `GH16:2.4.1.*&3.2.1.*-GH16:3.2.1.39`, contain multiple listed EC numbers, separated by '-'.\n",
    "\n",
    "Therefore, additional checks for multiple predicated EC numbers and quality checking of EC number formating to ensure only the EC number is retrieved was added to the method of parsing the log file.\n",
    "\n",
    "The below examples look at the first few lines to highlight some important examples of data."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Checking EC# parsing\n",
      "ec_split= ['2.4.1.109'] \n",
      "all_ec_numbers= ['2.4.1.109'] \n",
      "ec_number= 2.4.1.109\n",
      "Checking EC# parsing\n",
      "ec_split= ['2.4.1.1'] \n",
      "all_ec_numbers= ['2.4.1.1'] \n",
      "ec_number= 2.4.1.1\n",
      "Checking EC# parsing\n",
      "ec_split= ['2.4.1.256'] \n",
      "all_ec_numbers= ['2.4.1.256'] \n",
      "ec_number= 2.4.1.256\n",
      "Checking EC# parsing\n",
      "ec_split= ['2.4.1.*'] \n",
      "all_ec_numbers= ['2.4.1.-'] \n",
      "ec_number= 2.4.1.-\n",
      "Checking EC# parsing\n",
      "ec_split= ['3.2.1.101'] \n",
      "all_ec_numbers= ['3.2.1.101'] \n",
      "ec_number= 3.2.1.101\n",
      "Checking EC# parsing\n",
      "ec_split= ['2.4.1.173'] \n",
      "all_ec_numbers= ['2.4.1.173'] \n",
      "ec_number= 2.4.1.173\n",
      "Checking EC# parsing\n",
      "ec_split= ['3.2.1.59'] \n",
      "all_ec_numbers= ['3.2.1.59'] \n",
      "ec_number= 3.2.1.59\n",
      "Checking EC# parsing\n",
      "ec_split= ['2.4.1.*'] \n",
      "all_ec_numbers= ['2.4.1.-'] \n",
      "ec_number= 2.4.1.-\n",
      "Checking EC# parsing\n",
      "ec_split= ['3.2.1.14'] \n",
      "all_ec_numbers= ['3.2.1.14'] \n",
      "ec_number= 3.2.1.14\n",
      "Checking EC# parsing\n",
      "ec_split= ['3.2.1.106'] \n",
      "all_ec_numbers= ['3.2.1.106'] \n",
      "ec_number= 3.2.1.106\n",
      "Checking EC# parsing\n",
      "ec_split= ['3.2.1.4'] \n",
      "all_ec_numbers= ['3.2.1.4'] \n",
      "ec_number= 3.2.1.4\n",
      "Checking EC# parsing\n",
      "ec_split= ['3.2.1.21', '3.2.1.23'] \n",
      "all_ec_numbers= ['3.2.1.21', '3.2.1.23'] \n",
      "ec_number= 3.2.1.21, 3.2.1.23\n",
      "Checking EC# parsing\n",
      "ec_split= ['3.2.1.113'] \n",
      "all_ec_numbers= ['3.2.1.113'] \n",
      "ec_number= 3.2.1.113\n",
      "Checking EC# parsing\n",
      "ec_split= ['3.2.1.113'] \n",
      "all_ec_numbers= ['3.2.1.113'] \n",
      "ec_number= 3.2.1.113\n",
      "Checking EC# parsing\n",
      "ec_split= ['3.2.1.7'] \n",
      "all_ec_numbers= ['3.2.1.7'] \n",
      "ec_number= 3.2.1.7\n"
     ]
    }
   ],
   "source": [
    "for line in log_file[462:]:\n",
    "    # separate the data fields\n",
    "    prediction_output = line.split(\"\\t\")\n",
    "    \n",
    "    # retrieve the predicted EC number\n",
    "    ec_data = prediction_output[-2].split(\":\")\n",
    "\n",
    "    if len(ec_data) == 1:  # no EC number predicted\n",
    "        ec_number = np.nan\n",
    "\n",
    "    elif len(ec_data) == 2:\n",
    "        # element 0 is the CAZy family\n",
    "        all_ec_numbers = []\n",
    "        ec_split = ec_data[1].split(\"&\")  # if multiple EC numbers are predicted, separate them\n",
    "\n",
    "        for item in ec_split:\n",
    "            # check formating, somtimes 'Unknown' is written in stead of an EC number\n",
    "            try:\n",
    "                re.match(r\"\\d+?\\.(\\d+?|\\*)\\.(\\d+?|\\*)\\.(\\d+?|\\*)\", item)\n",
    "                item = item.replace(\"*\",\"-\")  # standardise missing digit format\n",
    "            except AttributeError:\n",
    "                pass\n",
    "            all_ec_numbers.append(item)\n",
    "\n",
    "        if len(all_ec_numbers) == 0:\n",
    "            ec_number = np.nan\n",
    "        else:\n",
    "            ec_number = \", \".join(all_ec_numbers)\n",
    "        print(\"Checking EC# parsing\\nec_split=\", ec_split, \"\\nall_ec_numbers=\", all_ec_numbers, \"\\nec_number=\", ec_number)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "From this we can define a function to retrieved the predicted EC numbers for the CAZy family that CUPP has listed as the 'Best function' for the protein."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1.1.3.38\n",
      "2.4.1.117\n",
      "nan\n",
      "1.10.3.2\n",
      "nan\n",
      "nan\n",
      "2.4.1.109\n",
      "2.4.1.16\n",
      "1.-.-.-\n",
      "3.2.1.59\n",
      "3.2.1.59\n",
      "3.2.1.24\n",
      "3.2.1.24\n",
      "2.4.1.16\n",
      "2.4.1.16\n",
      "nan\n",
      "3.2.1.113\n",
      "nan\n",
      "3.2.1.109\n",
      "3.2.1.-\n",
      "2.4.1.-, 3.2.1.-, 3.2.1.39\n",
      "3.2.1.113\n",
      "3.1.1.-\n",
      "1.10.3.2\n",
      "nan\n",
      "3.2.1.67\n",
      "nan\n",
      "3.2.1.15\n",
      "3.2.1.-\n",
      "1.10.3.2\n",
      "3.2.1.49\n",
      "3.2.1.-\n",
      "2.4.1.-\n",
      "3.1.1.6\n",
      "nan\n",
      "2.4.1.15\n",
      "nan\n",
      "3.1.1.74\n",
      "nan\n",
      "3.5.1.25\n",
      "2.4.1.25, 3.2.1.33\n",
      "2.4.1.25, 3.2.1.33\n",
      "nan\n",
      "3.1.1.74\n",
      "nan\n",
      "3.1.1.11\n",
      "3.2.1.-\n",
      "nan\n",
      "3.2.1.101\n",
      "nan\n",
      "3.2.1.-\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "3.2.1.59\n",
      "3.2.1.59\n",
      "nan\n",
      "3.2.1.96\n",
      "2.4.1.80\n",
      "nan\n",
      "nan\n",
      "3.2.1.22, 3.2.1.49\n",
      "3.2.1.4\n",
      "2.4.1.109\n",
      "nan\n",
      "1.-.-.-\n",
      "2.4.99.18\n",
      "nan\n",
      "1.10.3.2\n",
      "1.10.3.2\n",
      "3.2.1.21, 3.2.1.23\n",
      "2.4.1.131\n",
      "nan\n",
      "3.2.1.6, 3.2.1.73\n",
      "nan\n",
      "nan\n",
      "2.4.1.259, 2.4.1.261\n",
      "2.4.1.-\n",
      "3.2.1.20, 5.4.99.11\n",
      "3.2.1.15\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "2.4.1.34\n",
      "2.4.1.-\n",
      "nan\n",
      "3.1.1.-\n",
      "3.2.1.6, 3.2.1.73\n",
      "3.2.1.6, 3.2.1.73\n",
      "nan\n",
      "1.10.3.2\n",
      "3.2.1.99\n",
      "3.2.1.101\n",
      "3.2.1.45\n",
      "2.4.1.15\n",
      "3.2.1.-\n",
      "3.2.1.23\n",
      "3.2.1.59\n",
      "4.2.2.10\n",
      "nan\n",
      "3.2.1.8\n",
      "nan\n",
      "1.1.3.13\n",
      "nan\n",
      "3.2.1.78\n",
      "nan\n",
      "nan\n",
      "3.2.1.15\n",
      "2.4.1.-\n",
      "nan\n",
      "3.2.1.22\n",
      "nan\n",
      "nan\n",
      "3.2.1.8\n",
      "nan\n",
      "3.2.1.151\n",
      "3.1.1.11\n",
      "nan\n",
      "3.2.1.4\n",
      "2.4.1.-\n",
      "nan\n",
      "3.1.1.11\n",
      "3.2.1.39\n",
      "nan\n",
      "nan\n",
      "4.2.2.2\n",
      "1.-.-.-\n",
      "2.4.1.-\n",
      "nan\n",
      "2.4.1.-, 2.4.2.-\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "3.2.1.3\n",
      "3.2.1.59\n",
      "4.2.2.2\n",
      "2.4.1.-\n",
      "nan\n",
      "nan\n",
      "2.4.1.16\n",
      "3.2.1.3\n",
      "3.2.1.3\n",
      "3.2.1.15\n",
      "3.2.1.23\n",
      "2.4.1.-\n",
      "nan\n",
      "nan\n",
      "2.4.1.131\n",
      "3.2.1.99\n",
      "3.2.1.113\n",
      "3.2.1.25\n",
      "3.2.1.39\n",
      "2.4.1.173\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "3.2.1.67\n",
      "nan\n",
      "2.4.1.-\n",
      "nan\n",
      "1.10.3.2\n",
      "nan\n",
      "nan\n",
      "2.4.1.-\n",
      "1.1.3.-\n",
      "nan\n",
      "3.2.1.8\n",
      "1.1.99.18\n",
      "nan\n",
      "nan\n",
      "3.2.1.39\n",
      "nan\n",
      "3.2.1.101\n",
      "3.2.1.-, 3.2.1.131\n",
      "nan\n",
      "3.2.1.67\n",
      "nan\n",
      "3.2.1.50\n",
      "2.4.1.-\n",
      "2.4.1.-\n",
      "2.4.1.173\n",
      "1.-.-.-\n",
      "nan\n",
      "nan\n",
      "3.2.1.58\n",
      "1.1.3.-\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "3.2.1.145\n",
      "3.2.1.-, 3.2.1.31\n",
      "nan\n",
      "3.2.1.40\n",
      "2.4.1.-\n",
      "3.2.1.-\n",
      "3.2.1.101\n",
      "nan\n",
      "3.1.1.6\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "3.2.1.113\n",
      "3.2.1.75\n",
      "3.2.1.21\n",
      "3.2.1.177\n",
      "1.-.-.-\n",
      "3.2.1.39\n",
      "2.4.1.-\n",
      "2.4.1.131\n",
      "3.2.1.113\n",
      "3.2.1.-\n",
      "3.2.1.-\n",
      "1.6.5.6\n",
      "3.1.1.6\n",
      "3.2.1.40\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "1.-.-.-\n",
      "2.4.1.142\n",
      "3.2.1.59\n",
      "3.2.1.14\n",
      "2.4.1.-, 2.4.2.-\n",
      "2.4.1.-\n",
      "3.2.1.14\n",
      "3.1.3.12\n",
      "3.1.3.12\n",
      "3.1.3.12\n",
      "3.1.1.74\n",
      "2.4.1.16\n",
      "nan\n",
      "nan\n",
      "3.2.1.3\n",
      "3.2.1.3\n",
      "3.2.1.164\n",
      "3.2.1.59\n",
      "3.2.1.58\n",
      "nan\n",
      "2.4.1.-\n",
      "2.4.1.-\n",
      "2.4.1.-\n",
      "1.1.3.13\n",
      "3.1.1.6\n",
      "3.2.1.35\n",
      "3.2.1.35\n",
      "1.-.-.-\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "3.2.1.52\n",
      "3.2.1.52\n",
      "2.4.1.-, 3.2.1.-, 3.2.1.39\n",
      "nan\n",
      "nan\n",
      "3.2.1.113\n",
      "3.2.1.113\n",
      "nan\n",
      "3.2.1.-\n",
      "3.2.1.21\n",
      "3.2.1.21\n",
      "3.2.1.63\n",
      "1.10.3.2\n",
      "nan\n",
      "3.2.1.37\n",
      "2.4.1.173\n",
      "2.4.1.-, 2.4.2.-\n",
      "3.2.1.55\n",
      "2.4.1.-\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "3.2.1.4\n",
      "nan\n",
      "nan\n",
      "2.4.1.-\n",
      "2.4.1.-\n",
      "3.1.1.11\n",
      "1.10.3.2\n",
      "4.2.2.10\n",
      "2.4.1.-\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "2.4.1.131\n",
      "3.2.1.6\n",
      "3.2.1.63\n",
      "2.4.1.-\n",
      "3.2.1.101\n",
      "3.2.1.89\n",
      "nan\n",
      "3.2.1.14\n",
      "3.2.1.14\n",
      "3.2.1.101\n",
      "nan\n",
      "nan\n",
      "3.2.1.-\n",
      "3.2.1.-\n",
      "2.4.1.16\n",
      "3.2.1.28\n",
      "nan\n",
      "nan\n",
      "3.2.1.39\n",
      "1.10.3.2\n",
      "3.2.1.101\n",
      "nan\n",
      "2.4.1.173\n",
      "nan\n",
      "3.2.1.6, 3.2.1.73\n",
      "1.1.3.9\n",
      "1.1.3.9\n",
      "nan\n",
      "3.2.1.55\n",
      "3.2.1.28\n",
      "nan\n",
      "3.2.1.6, 3.2.1.73\n",
      "1.10.3.2\n",
      "1.10.3.2\n",
      "1.10.3.2\n",
      "nan\n",
      "nan\n",
      "2.4.1.-\n",
      "1.-.-.-\n",
      "1.-.-.-\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "3.2.1.58\n",
      "3.2.1.58\n",
      "nan\n",
      "nan\n",
      "2.4.1.186\n",
      "3.2.1.21\n",
      "nan\n",
      "3.2.1.185\n",
      "nan\n",
      "nan\n",
      "3.2.1.171\n",
      "nan\n",
      "nan\n",
      "4.2.2.2\n",
      "nan\n",
      "3.2.1.26\n",
      "3.2.1.40\n",
      "3.2.1.21\n",
      "nan\n",
      "3.2.1.58\n",
      "nan\n",
      "2.4.1.83\n",
      "nan\n",
      "3.2.1.23\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "2.4.2.40\n",
      "1.-.-.-\n",
      "3.2.1.20\n",
      "nan\n",
      "3.2.1.23\n",
      "2.4.1.131\n",
      "3.2.1.78\n",
      "3.2.1.171\n",
      "nan\n",
      "2.4.1.16\n",
      "2.4.1.16\n",
      "nan\n",
      "3.2.1.39\n",
      "3.2.1.-\n",
      "nan\n",
      "2.4.1.-\n",
      "nan\n",
      "nan\n",
      "1.1.99.18\n",
      "nan\n",
      "3.2.1.3\n",
      "nan\n",
      "1.-.-.-\n",
      "3.2.1.173\n",
      "2.4.1.258\n",
      "nan\n",
      "3.2.1.4\n",
      "2.4.1.16\n",
      "2.4.1.16\n",
      "nan\n",
      "nan\n",
      "3.2.1.-\n",
      "3.2.1.6, 3.2.1.73\n",
      "3.2.1.149\n",
      "3.2.1.4\n",
      "1.10.3.2\n",
      "2.4.1.-\n",
      "nan\n",
      "3.2.1.151\n",
      "2.4.1.-\n",
      "3.2.1.171\n",
      "3.2.1.58\n",
      "3.2.1.101\n",
      "2.4.1.-, 3.2.1.-, 3.2.1.39\n",
      "2.4.1.132, 2.4.1.257\n",
      "nan\n",
      "nan\n",
      "3.2.1.174, 3.2.1.40\n",
      "nan\n",
      "2.4.1.173\n",
      "3.2.1.174\n",
      "nan\n",
      "2.4.1.-\n",
      "1.1.3.9\n",
      "3.1.1.74\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "3.2.1.15\n",
      "3.2.1.15\n",
      "3.2.1.15\n",
      "3.2.1.15\n",
      "3.2.1.15\n",
      "nan\n",
      "3.2.1.15\n",
      "3.1.1.11\n",
      "2.4.1.-\n",
      "4.2.2.2\n",
      "3.2.1.4\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "1.10.3.2\n",
      "3.2.1.55\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "2.4.1.-\n",
      "3.5.1.25\n",
      "nan\n",
      "3.2.1.55\n",
      "3.1.1.72\n",
      "2.4.1.173\n",
      "nan\n",
      "3.2.1.113\n",
      "3.1.1.72\n",
      "1.1.99.18\n",
      "3.2.1.8\n",
      "3.2.1.59\n",
      "2.4.1.-\n",
      "2.4.1.18\n",
      "2.4.1.18\n",
      "2.4.1.11\n",
      "3.2.1.21\n",
      "2.4.1.-, 3.2.1.-, 3.2.1.39\n",
      "3.1.1.74\n",
      "1.-.-.-\n",
      "nan\n",
      "2.4.1.109\n",
      "2.4.1.1\n",
      "2.4.1.256\n",
      "nan\n",
      "2.4.1.-\n",
      "3.2.1.101\n",
      "2.4.1.173\n",
      "3.2.1.59\n",
      "2.4.1.-\n",
      "3.2.1.14\n",
      "3.2.1.106\n",
      "3.2.1.4\n",
      "nan\n",
      "3.2.1.21, 3.2.1.23\n",
      "3.2.1.6, 3.2.1.73\n",
      "nan\n",
      "3.2.1.113\n",
      "3.2.1.113\n",
      "3.2.1.7\n"
     ]
    }
   ],
   "source": [
    "def get_cupp_ec_number(prediction_output):\n",
    "    \"\"\"Retrieve predicted EC numbers from CUPP log file.\n",
    "    \n",
    "    EC numbers are represented as \"CAZy_fam:EC_number\" in the log file.\n",
    "    If multiple EC numbers are predicated and the CAZy families of each are\n",
    "    given then these are separated by '-'. If multiple EC numbers are predicated\n",
    "    for the same CAZy family, these are separated by '&'.\n",
    "    \n",
    "    :param prediciton_output: list of items from log file line.\n",
    "    \n",
    "    Return string if EC numbers are given, or null value if not.\n",
    "    \"\"\"       \n",
    "    # retrieve the data from the CUPP 'Best function' prediction\n",
    "    ec_data = prediction_output[-2].split(\":\")\n",
    "    \n",
    "    if len(ec_data) == 1:  # Result of \":\" not being present, caused by no EC number predictions\n",
    "        ec_number = np.nan\n",
    "        return ec_number\n",
    "\n",
    "    # create empty list to store all predicted EC numbers\n",
    "    all_ec_numbers = []\n",
    "\n",
    "    for item in ec_data:\n",
    "        # check if the item may be an EC number, which start with a digit\n",
    "        try:\n",
    "            re.match(r\"\\d.+\", item).group()\n",
    "        except AttributeError:  # not an EC number\n",
    "            continue\n",
    "\n",
    "        # if multiple best function and/or EC number predictions were made they may\n",
    "        # be separated by a dash '-'\n",
    "        dash_separated_data = item.split(\"-\")\n",
    "\n",
    "        for data in dash_separated_data:\n",
    "            # if multiple EC numbers are predicated they are separate by '&\n",
    "            split_data = data.split(\"&\")\n",
    "            \n",
    "            for string in split_data:\n",
    "                # check if the string is an EC number\n",
    "                try:\n",
    "                    re.match(r\"\\d+?\\.(\\d+?|\\*)\\.(\\d+?|\\*)\\.(\\d+?|\\*)\", string). group()\n",
    "                    all_ec_numbers.append(string)\n",
    "                except AttributeError:  # not an EC number\n",
    "                    continue\n",
    "                    \n",
    "    # standardise missing digits in the EC numbers from '*' to '-'                \n",
    "    index = 0\n",
    "    for index in range(len(all_ec_numbers)):\n",
    "        all_ec_numbers[index] = all_ec_numbers[index].replace(\"*\",\"-\")\n",
    "        # Sometimes 'Unknown' is written out by CUPP, ensure this is removed\n",
    "        if all_ec_numbers[index].find(\"Unknown\") != -1:\n",
    "            all_ec_numbers.remove(all_ec_numbers[index])\n",
    "                    \n",
    "    # write out the EC numbers in easy human readable format or create null value if none were predicted\n",
    "    if len(all_ec_numbers) == 0:\n",
    "        ec_number = np.nan\n",
    "    else:\n",
    "        ec_number = \", \".join(all_ec_numbers) \n",
    "\n",
    "    return ec_number\n",
    "\n",
    "for line in log_file:\n",
    "    # separate the data fields\n",
    "    prediction_output = line.split(\"\\t\")\n",
    "    ecs = get_cupp_ec_number(prediction_output)\n",
    "    print(ecs)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<a id=\"cupp_subfam\"></a>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Getting the predicted CAZy subfamily\n",
    "\n",
    "Sometimes when no subfamily is predicted, an empty string is written out the output log file. However, it is better if the resulting Pandas dataframe contains a null value instead.\n",
    "\n",
    "At this stage we can also create an empty dataframe, to which the data from each parsed protein is added."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>protein_accession</th>\n",
       "      <th>cazy_family</th>\n",
       "      <th>cazy_subfamily</th>\n",
       "      <th>ec_number</th>\n",
       "      <th>domain_range</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>XP_001545605.1 BCIN_01g00260</td>\n",
       "      <td>AA4</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.1.3.38</td>\n",
       "      <td>122..568</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>XP_024545922.1 BCIN_01g00340</td>\n",
       "      <td>GT2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2.4.1.117</td>\n",
       "      <td>116..300</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>XP_001547235.1 BCIN_01g00640</td>\n",
       "      <td>AA3</td>\n",
       "      <td>AA3_1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>26..553</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>XP_001547254.1 BCIN_01g00800</td>\n",
       "      <td>AA1</td>\n",
       "      <td>AA1_3</td>\n",
       "      <td>1.10.3.2</td>\n",
       "      <td>202..455</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>XP_024546029.1 BCIN_01g01760</td>\n",
       "      <td>GH18</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>116..258</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>XP_024546029.1 BCIN_01g01760</td>\n",
       "      <td>PL7</td>\n",
       "      <td>PL7:2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>989..1040</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>XP_001548518.1 BCIN_01g02310</td>\n",
       "      <td>GT39</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2.4.1.109</td>\n",
       "      <td>62..264</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>XP_001548545.1 BCIN_01g02520</td>\n",
       "      <td>GT2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2.4.1.16</td>\n",
       "      <td>213..452</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>XP_001549934.1 BCIN_01g02970</td>\n",
       "      <td>AA11</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.*.*.*</td>\n",
       "      <td>55..160</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>XP_024546135.1 BCIN_01g03210</td>\n",
       "      <td>GH71</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3.2.1.59</td>\n",
       "      <td>71..381</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "              protein_accession cazy_family cazy_subfamily  ec_number  \\\n",
       "0  XP_001545605.1 BCIN_01g00260         AA4            NaN   1.1.3.38   \n",
       "0  XP_024545922.1 BCIN_01g00340         GT2            NaN  2.4.1.117   \n",
       "0  XP_001547235.1 BCIN_01g00640         AA3          AA3_1        NaN   \n",
       "0  XP_001547254.1 BCIN_01g00800         AA1          AA1_3   1.10.3.2   \n",
       "0  XP_024546029.1 BCIN_01g01760        GH18            NaN        NaN   \n",
       "0  XP_024546029.1 BCIN_01g01760         PL7          PL7:2        NaN   \n",
       "0  XP_001548518.1 BCIN_01g02310        GT39            NaN  2.4.1.109   \n",
       "0  XP_001548545.1 BCIN_01g02520         GT2            NaN   2.4.1.16   \n",
       "0  XP_001549934.1 BCIN_01g02970        AA11            NaN    1.*.*.*   \n",
       "0  XP_024546135.1 BCIN_01g03210        GH71            NaN   3.2.1.59   \n",
       "\n",
       "  domain_range  \n",
       "0     122..568  \n",
       "0     116..300  \n",
       "0      26..553  \n",
       "0     202..455  \n",
       "0     116..258  \n",
       "0    989..1040  \n",
       "0      62..264  \n",
       "0     213..452  \n",
       "0      55..160  \n",
       "0      71..381  "
      ]
     },
     "execution_count": 55,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# build an empty dataframe\n",
    "cupp_df = pd.DataFrame(columns=[\n",
    "    \"protein_accession\",\n",
    "    \"cazy_family\",\n",
    "    \"cazy_subfamily\",\n",
    "    \"ec_number\",\n",
    "    \"domain_range\",\n",
    "])\n",
    "\n",
    "for line in log_file[:10]:\n",
    "    # separate the data fields\n",
    "    prediction_output = line.split(\"\\t\")\n",
    "    \n",
    "    # retrieve the predicted domain range if given\n",
    "    for item in prediction_output:\n",
    "        if item.find(\"..\") != -1:\n",
    "            domain_range = item\n",
    "            break\n",
    "        else:\n",
    "            domain_range = np.nan\n",
    "    \n",
    "    # retrieve the predicted EC number\n",
    "    try:\n",
    "        ec_number = prediction_output[-2].split(\":\")[1]\n",
    "    except IndexError:\n",
    "        ec_number = np.nan\n",
    "\n",
    "    # store the prediction output in a dictionary\n",
    "    prediction = {\n",
    "        \"protein_accession\": [prediction_output[0]],\n",
    "        \"cazy_family\": [prediction_output[1]],\n",
    "        \"cazy_subfamily\": [prediction_output[-1]],\n",
    "        \"ec_number\": ec_number,\n",
    "        \"domain_range\": [domain_range],\n",
    "    }\n",
    "\n",
    "    # if no CAZy subfamily was predicted, change empty string to null value\n",
    "    if len(prediction[\"cazy_subfamily\"][0]) == 0:\n",
    "        prediction[\"cazy_subfamily\"][0] = np.nan\n",
    "\n",
    "    # create dataframe to represent the current working protein\n",
    "    prediction_df = pd.DataFrame(prediction)\n",
    "    # add protein to CUPP dataframe\n",
    "    cupp_df = cupp_df.append(prediction_df)\n",
    "\n",
    "cupp_df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<a id=\"cupp_std_subfam\"></a>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Standardising CAZy subfamilies (PL7:2)\n",
    "\n",
    "Sometimes the CAZy subfamily is not written in the standard CAZy format of family_subfam#. For example, for protein \"XP_024546029.1 BCIN_01g01760\" the CAZy subfamily is listed as 'PL7:2'.\n",
    "\n",
    "Looking further through the CUPP output log file, there are several instances of a this occuring, and often occuring for CAZy families for which CAZy has not created a subfamily with number outputted by CUPP. For example, protein XP_001549864.1 BCIN_03g07360 is predicted to have the subfamily 'PL1:4', however, there is not CAZy subfamily 'PL1:4'.\n",
    "\n",
    "It was hypothesised that listing a CAZy subfamily with a colon instead of the traditional CAZy underscore was done to indicate when the predict CAZy subfamily is one created by CUPP and is not in CAZy.\n",
    "\n",
    "Looking at lines 2890 and 2914 in the `CUPP.CUPPprediction.py` script, the `:` in the predicated subfamily is replaced by `_` when writing the output FASTA file. Therefore, the predicted CAZy subfamily is written as 'PL7_2' in the output FASTA file. However, in line 2909, when writing out to the log file, the replacement of `:` to `_` is not performed. Hence, the predicted CAZy subfamily is written as 'PL7:2' in the output log file.\n",
    "\n",
    "For standardistion, the colon in the predicted CAZy subfamily was changed to an underscore.\n",
    "\n",
    "Moreover, when looking through the log file, multiple subfamilies can be predicted for a given protein. These subfamilies are separated by '+'. To improve human readability this was changed to separate the subfamilies by commas (', '). For example, CUPP's analysis of protein sequences from the genome GCA143535.1 predicted two CAZy subfamilies within the GH13 CAZyme domain of the protein XP_001552718.2.  \n",
    "`XP_001552718.2 BCIN_02g05690\tGH13\tGH13:0.2\t15.7\t5.93\t151\t70..391\t76\tGH13:3.2.1.10&3.2.1.20\tGH13:3.2.1.20-GH13:5.4.99.11\tGH13:31+GH13:40`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "subfam= nan\n",
      "subfam= nan\n",
      "subfam= AA3_1\n",
      "subfam= AA1_3\n",
      "subfam= nan\n",
      "subfam= PL7_2\n",
      "subfam= nan\n",
      "subfam= nan\n",
      "subfam= nan\n",
      "subfam= nan\n"
     ]
    }
   ],
   "source": [
    "for line in log_file[:10]:\n",
    "    # separate the data fields\n",
    "    prediction_output = line.split(\"\\t\")\n",
    "\n",
    "    # retrieve predicated CAZy subfamily\n",
    "    subfam = prediction_output[-1] \n",
    "    # if no CAZy subfamily was predicted, change empty string to null value\n",
    "    if len(subfam) == 0:\n",
    "        subfam = np.nan\n",
    "    else:\n",
    "        # write subfamilies using CAZy standard format\n",
    "        subfam = subfam.replace(\":\",\"_\")\n",
    "        subfam = subfam.replace(\"+\", \", \")  # separate multiple subfams with ', ' not '+'\n",
    "\n",
    "    print(\"subfam=\", subfam)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<a id=\"cupp_functions\"></a>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Combing functions and populating the data model, and adding non-CAZymes\n",
    "\n",
    "The above can be combined into functions, that are then included in `pyrewton.cazymes.prediction.parse`. When it comes to evaluating CUPP's performance when differentiating between CAZymes and non-CAZymes, a binary classification of CAZymes (1) and non-CAZyme (0) is needed for each protein. If a protein is included in the CUPP output CUPP has predicted that the protein is a CAZyme, and thus an additional column with a value of 1 can be added to the standardised CUPP output, to represent the cazyme/non-cazyme classification of proteins by CUPP.\n",
    "\n",
    "For the evaluation CUPP's performance the information of none CAZymes is also needed. CUPP only includes the data for proteins in it's log file that have classified as CAZymes therefore, the remaining protein sequences are retrieved from the FASTA file."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "a54661b8b2a747948ed56eed7a586dda",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "HBox(children=(HTML(value='parsing cupp output'), FloatProgress(value=0.0, max=282.0), HTML(value='')))"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n"
     ]
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "05dd326593a047d9ad0a315e0ed4da52",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "HBox(children=(HTML(value='Adding non-CAZymes'), FloatProgress(value=0.0, max=93219.0), HTML(value='')))"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "type= <class 'list'>\n",
      "len= 9287\n",
      "index 0= -CUPPprediction, protein=ATY67253.1 A9K55_000001, non-CAZyme-\n"
     ]
    }
   ],
   "source": [
    "def parse_cupp_output(log_file_path, fasta_path):\n",
    "    \"\"\"Parse the output log file from CUPP and write out a dataframe.\n",
    "    \n",
    "    Retrieves the protein accession, predicted CAZy families, CAZy families EC\n",
    "    numbers and domain ranges.\n",
    "    \n",
    "    :param log_file_path: Path, path to the output log file\n",
    "    :fasta path: Path, path to the FASTA containing the query sequences for CUPP\n",
    "    \n",
    "    Return a list of CUPPprediction instances, where each instance represented one protein.\n",
    "    \"\"\"\n",
    "    # open the CUPP output log file\n",
    "    with open(log_file_path, \"r\") as lfh:\n",
    "        log_file = lfh.read().splitlines()\n",
    "    \n",
    "    cupp_predictions = {}  # stores proteins {protein_accession:CUPPprediction_instance}\n",
    "\n",
    "    for line in tqdm(log_file, desc=\"parsing cupp output\"):\n",
    "        # separate the data fields\n",
    "        prediction_output = line.split(\"\\t\")\n",
    "\n",
    "        # retrieve the predicted domain range if given\n",
    "        domain_range = get_cupp_domain_range(prediction_output)  # list of strs\n",
    "        \n",
    "        # retrieve predicted EC number if given\n",
    "        ec_numbers = get_cupp_ec_number(prediction_output)  # list of strs\n",
    "        \n",
    "        # retrieve predicted CAZy subfamily if given\n",
    "        cazy_subfamily = prediction_output[-1] \n",
    "        # if no CAZy subfamily was predicted, change empty string to null value\n",
    "        if len(cazy_subfamily) == 0:\n",
    "            cazy_subfamily = np.nan\n",
    "        else:\n",
    "            # write subfamilies using CAZy standard format\n",
    "            cazy_subfamily = cazy_subfamily.replace(\":\",\"_\")\n",
    "            cazy_subfamily = cazy_subfamily.replace(\"+\", \", \")  # separate multiple subfams with ', ' not '+'\n",
    "\n",
    "        # retrieve the protein accession and check if it already has a corresponding CUPPprediction instance\n",
    "        protein_accession = prediction_output[0]\n",
    "\n",
    "        # retrieve the CAZyme classification and predicted CAZy family\n",
    "        cazyme_classification = 1  # all proteins included in CUPP log file are identified as CAZymes\n",
    "        cazy_family = prediction_output[1]  # This is selected from the CUPP 'Best function'\n",
    "        \n",
    "        # check if a CUPPprediction instance already exists for the protein\n",
    "        try:\n",
    "            existing_prediction = cupp_predictions[protein_accession]\n",
    "            \n",
    "            # check if the CAZyme domain has been been parsed before\n",
    "            existing_cazyme_domains = existing_prediction.cazyme_domains\n",
    "            existance = False\n",
    "            while existance == False:\n",
    "                for domain in existing_cazyme_domains:\n",
    "                    if (domain.cazy_family == cazy_family) and (domain.cazy_subfamily == cazy_subfamily):\n",
    "                        for ec in ec_numbers:\n",
    "                                domain.ec_numbers.append(ec)\n",
    "                        for drange in domain_range:\n",
    "                                domain.domain_range.append(drange)\n",
    "                        existance = True\n",
    "            \n",
    "            if existance is False:\n",
    "                # create new CAZyme domain\n",
    "                new_cazyme_domain = CUPPCazymeDomain(\n",
    "                    protein_accession,\n",
    "                    cazy_family,\n",
    "                    cazy_subfamily,\n",
    "                    ec_numbers,\n",
    "                    domain_range,\n",
    "                )\n",
    "                existing_prediction.cazyme_domains.add(new_cazyme_domain)\n",
    "        \n",
    "        except KeyError:  # raised if there is not instance for the protein\n",
    "            new_cazyme_domain = CUPPCazymeDomain(\n",
    "                protein_accession,\n",
    "                cazy_family,\n",
    "                cazy_subfamily,\n",
    "                ec_numbers,\n",
    "                domain_range,\n",
    "            )\n",
    "            \n",
    "            new_protein = CUPPprediction(\n",
    "                protein_accession,\n",
    "                cazyme_classification,\n",
    "            )\n",
    "    \n",
    "            new_protein.cazyme_domains.add(new_cazyme_domain)\n",
    "\n",
    "    # add non-CAZymes\n",
    "    cupp_predictions = add_non_cazymes(fasta_path, cupp_predictions)\n",
    "    \n",
    "    cupp_predictions = list(cupp_predictions.values())\n",
    "\n",
    "    return cupp_predictions\n",
    "\n",
    "\n",
    "def get_cupp_domain_range(prediction_output):\n",
    "    \"\"\"Retrieve the predicted amino acid range of predicted CAZyme domain from CUPP log file.\n",
    "    \n",
    "    :param prediction_output: list of items from log file line.\n",
    "    \n",
    "    List of predicted domain ranges or null value in a list if not.\n",
    "    \"\"\"\n",
    "    domain_range = []  # store as a list in case multiple domain ranges are given\n",
    "\n",
    "    for item in prediction_output:\n",
    "        if item.find(\"..\") != -1:\n",
    "            domain_range.append(item)\n",
    "\n",
    "    # check retrieved items are definetly the domain ranges\n",
    "    for item in domain_range:\n",
    "        try:\n",
    "            re.match(r\"\\d+\\.\\.\\d+\", item).group()\n",
    "        except AttributeError:\n",
    "            # write as logger in pyrewton\n",
    "            print(f\"RAISED ATTRIBUTE ERROR: {item} misidentified as domain range\")\n",
    "            domain_range.remove(item)\n",
    "\n",
    "    if len(domain_range) == 0:\n",
    "        domain_range = [np.nan]\n",
    "    \n",
    "    return domain_range\n",
    "\n",
    "\n",
    "def get_cupp_ec_number(prediction_output):\n",
    "    \"\"\"Retrieve predicted EC numbers from CUPP log file.\n",
    "    \n",
    "    EC numbers are represented as \"CAZy_fam:EC_number\" in the log file.\n",
    "    If multiple EC numbers are predicated and the CAZy families of each are\n",
    "    given then these are separated by '-'. If multiple EC numbers are predicated\n",
    "    for the same CAZy family, these are separated by '&'.\n",
    "    \n",
    "    :param prediciton_output: list of items from log file line.\n",
    "    \n",
    "    List of predicted EC numbers or null value in a list if not.\n",
    "    \"\"\"       \n",
    "    # retrieve the data from the CUPP 'Best function' prediction\n",
    "    ec_data = prediction_output[-2].split(\":\")\n",
    "    \n",
    "    if len(ec_data) == 1:  # Result of \":\" not being present, caused by no EC number predictions\n",
    "        ec_number = np.nan\n",
    "        return ec_number\n",
    "\n",
    "    # create empty list to store all predicted EC numbers\n",
    "    all_ec_numbers = []\n",
    "\n",
    "    for item in ec_data:\n",
    "        # check if the item may be an EC number, which start with a digit\n",
    "        try:\n",
    "            re.match(r\"\\d.+\", item).group()\n",
    "        except AttributeError:  # not an EC number\n",
    "            continue\n",
    "\n",
    "        # if multiple best function and/or EC number predictions were made they may\n",
    "        # be separated by a dash '-'\n",
    "        dash_separated_data = item.split(\"-\")\n",
    "\n",
    "        for data in dash_separated_data:\n",
    "            # if multiple EC numbers are predicated they are separate by '&\n",
    "            split_data = data.split(\"&\")\n",
    "            \n",
    "            for string in split_data:\n",
    "                # check if the string is an EC number\n",
    "                try:\n",
    "                    re.match(r\"\\d+?\\.(\\d+?|\\*)\\.(\\d+?|\\*)\\.(\\d+?|\\*)\", string). group()\n",
    "                    all_ec_numbers.append(string)\n",
    "                except AttributeError:  # not an EC number\n",
    "                    continue\n",
    "                    \n",
    "    # standardise missing digits in the EC numbers from '*' to '-'                \n",
    "    index = 0\n",
    "    for index in range(len(all_ec_numbers)):\n",
    "        all_ec_numbers[index] = all_ec_numbers[index].replace(\"*\",\"-\")\n",
    "        # Sometimes 'Unknown' is written out by CUPP, ensure this is removed\n",
    "        if all_ec_numbers[index].find(\"Unknown\") != -1:\n",
    "            all_ec_numbers.remove(all_ec_numbers[index])\n",
    "                    \n",
    "    # write out the EC numbers in easy human readable format or create null value if none were predicted\n",
    "    if len(all_ec_numbers) == 0:\n",
    "        all_ec_numbers = [np.nan]\n",
    "\n",
    "    return all_ec_numbers\n",
    "\n",
    "\n",
    "def add_non_cazymes(fasta_path, cupp_predictions):\n",
    "    \"\"\"Add proteins that CUPP identified as non-CAZymes to the collection of CUPPprediction instances.\n",
    "    \n",
    "    :param fasta_path: Path, FASTA file used as input for CUPP.\n",
    "    :param cupp_predictions: dict, key=protein_accession, value=CUPPprediction instance\n",
    "    \n",
    "    Return a dictionary valued by protein accessions and keyed by their respective CUPPprediction instance.\n",
    "    \"\"\"\n",
    "    # open the FASTA path\n",
    "    with open(fasta_path) as fh:\n",
    "        fasta = fh.read().splitlines()\n",
    "\n",
    "    for line in tqdm(fasta, desc=\"Adding non-CAZymes\"):\n",
    "        if line.startswith(\">\"):\n",
    "            protein_accession = line[1:].strip()\n",
    "            \n",
    "            # check if the protein has been listed as a CAZyme by CUPP\n",
    "            try:\n",
    "                cupp_predictions[protein_accession]\n",
    "            except KeyError:\n",
    "                # raised if protein not in cupp_predictions, inferring proten was not labelled as CAZyme\n",
    "                cazyme_classification = 0\n",
    "                cupp_predictions[protein_accession] = CUPPprediction(\n",
    "                    protein_accession,\n",
    "                    cazyme_classification,\n",
    "                )\n",
    "        \n",
    "    return cupp_predictions\n",
    "    \n",
    "\n",
    "fasta_path = \"genbank_proteins_txid73501_GCA_008080495_1.fasta\"\n",
    "result = parse_cupp_output(\"gca_008080495_cupp.fasta.log\", fasta_path)\n",
    "\n",
    "print(\"type=\", type(result))\n",
    "print(\"len=\", len(result))\n",
    "print(\"index 0=\", result[0])"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
