{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Parsing eCAMI output\n",
    "\n",
    "The CAZyme prediction tools write out their output files in different formats, and do not present the data in a standardised manner across all tools. Additionally, the output from the prediction tools does not enable easy parsing of the data by the [Sklearn](https://scikit-learn.org/stable/index.html) library, which performs the statistical analysis of the tools preformances.\n",
    "\n",
    "This notebook explains the theorlogy and method used by the `pyrewton.cazymes.prediction.parse` submodule to parse the output from the CAZyme prediction tool eCAMI.\n",
    "\n",
    "[eCAMI](https://academic.oup.com/bioinformatics/article/36/7/2068/5651014?login=true) is a k-mer based CAZyme prediction tool. The information defining the data included in each field of the output files came from the [README](https://github.com/zhanglabNKU/eCAMI) and the Python script `prediction.py`.\n",
    "\n",
    "The following data is collected from the eCAMI output:\n",
    "- protein_accession\n",
    "- cazy_family (all predicated CAZy families)\n",
    "- cazy_subfamily (all predicated CAZy subfamilies)\n",
    "- ec_number (all predicted EC numbers)\n",
    "\n",
    "### Notebook structure\n",
    "\n",
    "The section titled 'exploring the output' shows the exploration of the tools output files. This is to demonstrate how the data is constructed and presented, and forms a foundation for explaining the developed method of parsing the data.\n",
    "\n",
    "The section 'parsing the output' shows the method of parsing the data from the CAZyme prediction tool.\n",
    "\n",
    "The final section shows any additional parsing that was performed to make the statistical analysis of the tool's performance easier. This typically includes removing duplicates, and combining rows that represent data for the same predicted CAZyme.\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Contents\n",
    "\n",
    "- [Imports](#notebook_imports)\n",
    "- [Exploring the eCAMI output text file](#text_file) -- This explains what data is taken from the eCAMI output, and why\n",
    "- [Definding the data model](#data_model)\n",
    "- [Getting the predicted CAZy family for the CAZyme domain](#cazy_fam)\n",
    "- [Getting the predicted CAZy subfamily for the CAZyme domain](#cazy_subfam)\n",
    "- [Getting the predicted EC numbers](#ecnum)\n",
    "- [Combining data for a single protein](#combining_proteins)\n",
    "- [Adding non-CAZymes](#no_caz)\n",
    "- [Final functions](#final)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<a id=\"notebook_imports\"></a>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Imports"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Requirement already satisfied: pandas in /home/em/.local/lib/python3.8/site-packages (1.1.4)\n",
      "Requirement already satisfied: pytz>=2017.2 in /home/em/anaconda3/lib/python3.8/site-packages (from pandas) (2020.1)\n",
      "Requirement already satisfied: numpy>=1.15.4 in /home/em/.local/lib/python3.8/site-packages (from pandas) (1.19.4)\n",
      "Requirement already satisfied: python-dateutil>=2.7.3 in /home/em/anaconda3/lib/python3.8/site-packages (from pandas) (2.8.1)\n",
      "Requirement already satisfied: six>=1.5 in /home/em/anaconda3/lib/python3.8/site-packages (from python-dateutil>=2.7.3->pandas) (1.15.0)\n",
      "Requirement already satisfied: numpy in /home/em/.local/lib/python3.8/site-packages (1.19.4)\n"
     ]
    }
   ],
   "source": [
    "# Install necessary libraries\n",
    "!pip3 install pandas\n",
    "!pip3 install numpy\n",
    "\n",
    "import re\n",
    "\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "\n",
    "from tqdm.notebook import tqdm"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<a id=\"text_file\"></a>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Exploring the eCAMI output text file\n",
    "\n",
    "eCAMI produces a single output file, which is a plain text file in the style of a FASTA-like file. The definitions and explanations of the data in the output file is taken from the multiple REAMEs in the eCAMI [GitHub repository](https://github.com/zhanglabNKU/eCAMI)  -- there is a `readme.md` file each in the `examples/prediction` and `examples/clustering` directories.\n",
    "\n",
    "Each predicted CAZyme can be predicted to be annotated with multiple CAZy families. Each unique CAZy familiy/CAZy subfamily combination is interpreted as a unique CAZyme domain prediction. Each CAZyme domain prediction is listed as a separate item in the output text file from eCAMI.\n",
    "\n",
    "Each CAZyme domain prediction is stored across two lines, with the first line starting with '>'.\n",
    "\n",
    "For example:\n",
    "\n",
    "```\n",
    ">ATY67280.1 A9K55_000020 \tGH17:43\tGH17:24|3.2.1.-:1\n",
    "LHKVFPGM(369),HKVFPGMD(370),KVFPGMDY(371),VFPGMDYT(372),FPGMDYTP(373),PPSQNNVT(392),PSQNNVTR(393),SQNNVTRD(394),QNNVTRDV(395),NNVTRDVA(396),NVTRDVAV(397),VTRDVAVL(398),TRDVAVLS(399),RDVAVLSQ(400),DVAVLSQL(401),AVLSQLTN(403),IRLYGTDC(412),RLYGTDCN(413),LYGTDCNQ(414),YGTDCNQT(415),GTDCNQTQ(416),TDCNQTQM(417),IVANEILF(476),VANEILFR(477),LPVATSDL(510),PVATSDLG(511),VATSDLGD(512),ATSDLGDD(513),TSDLGDDW(514),IMANIHPF(532),MANIHPFF(533),ISETGWPS(573),NGTNYFWF(618),GTNYFWFE(619),TNYFWFEA(620),NYFWFEAF(621),YFWFEAFD(622),FWFEAFDE(623),WFEAFDEP(624),FEAFDEPW(625),EAFDEPWK(626),WEDKWGLL(643),GVKIPDCG(659),VKIPDCGG(660),\n",
    "```\n",
    "\n",
    "The first line (denoted with the '>' prefix, contains the following:\n",
    "- 'protein_name' the protein accession, taken straight from the description/identifier line from the input FASTA file of all query protein sequences parsed by eCAMI\n",
    "- 'fam_name:group_number' -- this is the `GH17:43` in the above example\n",
    "    - the 'fam_name' is the CAZy family assigned to the CAZyme by ecAMI\n",
    "    - the 'group_number' is the cluster number (which acts as an ID number for the kmer cluster) for the respective CAZy family\n",
    "- 'subfam_name_of_the_group:subfam_name_count' -- in the above example this is `GH17:24|3.2.1.-:1`\n",
    "    - this data is extracted from the kmer library\n",
    "    - each 'label' (where each label is separated with a '|') is taken from the 'known labels' of CAZymes used to create the kmer clusters. Specifically, to create the kmer clusters, eCAMI parses a FASTA file of all protein sequences to be used for the clustering. In the protein data line (the line prefixed with '>') for each protein contains the protein name/accession and 'known labels' of that proteins. These known labels include CAZy family, CAZy subfamily and EC number annotations/labels. When the kmer cluster is created, the known lables that a sufficiently high ocrreuncy frequency in the cluster are stored within the cluster as a `kmer_message` (this name was taken from the `prediction.py` script from eCAMI. This `kmer_message` is then written out as the 'subfam_name_of_the_group:subfam_name_count' in the output file.\n",
    "    - the 'subfam_name_count' is the number of times the respective known label appears in the kmer cluster\n",
    "\n",
    "The second line contains the kmers from the predicted CAZy family's kmer cluster that were found in the query protein sequence, and includes the amino acid starting position of the kmer in the quer protein sequence, in brackets.\n",
    "\n",
    "To determine the CAZy family annotation predictions for each query protein sequence, eCAMI performs its search of all kmers in every CAZyme cluster they have created. eCAMI scores the number of kmers for each CAZy family that appear in the query protein sequence, and scores the sum frequency of the kmers from the cluster, this is assigned to a variable called `score`. To be more explicit, during the generation of the kmer clusters, a frequency score is calculated for each identified kmer, and this is stored along side the kmer amino acid sequence in the kmer library. Then for each CAZy family, and for each of its charactistic kmers found in the query protein sequence, the sum of these kmer frequencies from the cluster library is calculated, and this assigned to a variabled called `number_score`. As long for the associated CAZy family the `score` is greater than the `important_k_mer_number` (the minimum number of times kmers for a family need to appear in the query protein sequence), and the `number_score` is greater than `beta` (the minimum sum of the kmer frequencies, default is 2) then the CAZy family and its associated score is stored within the `sort_fam`, which is a list of lists, with each internal list representing one unqiue CAZy family. These CAZy families are then ranked by their `score`. The highest scoring CAZy family is listed as the `first_fam` and is written to the ouput file in the formate displayed above.\n",
    "\n",
    "To find other predicted CAZyme domains (CAZy family annotations), eCAMI then retrieves the `kmer_message` for the `first_fam`. It then iterates through the 'known labels' in the `kmer_message`, and for each CAZy family found in the 'known labels', checks if the CAZy family is in `sort_fam`. If it is then the CAZy family is written out to the output file for the query protein sequence in the same format written out above, often with an extremely similar `kmer_message` to the first CAZy family.\n",
    "\n",
    "For CAZy families in the `kmer_message` but not in the `sort_fam`, these CAZy families are **not** written out to the output file. Likewise, CAZy families in the `sort_fam` not in the `kmer_message`, these CAZy families are **not** written out to the output file.\n",
    "\n",
    "This is why it is possible to find an output like follows in the output file of eCAMI (where the lines of kmers have been removed:\n",
    "```\n",
    ">ATY67289.1 A9K55_000120     CBM48:20    CBM48:339|GH13_8:324|2.4.1.18:32|GH13:4|GT4:2\n",
    "\n",
    ">ATY67289.1 A9K55_000120     GH13:203    GH13_8:577|CBM48:328|2.4.1.18:32|GH13:13|GT4:1\n",
    "```\n",
    "In this example, 'GT4' has been listed in both `kmer_messages`, and the `kmer_messages` share similar content, but the family 'GT4' is not listed in 'fam_name:group_number' for the protein in the output file. Therefore, the actualy CAZy family prediction from eCAMI is written out in the 'fam_name:group_number' field. This field always contains a CAZy family, and not a subfamily. However, to retrieve a potential subfamily prediction, we can check the `kmer_message` on the respective row and check if it is the child of the CAZy family listed in the 'fam_name:group_number' field. If it is, then we can retrive it as the CAZy subfamily prediction.\n",
    "\n",
    "The EC numbers listed for each predicted CAZyme domain (unique CAZy family) are not predicted EC numbers. To predict the EC numbers can be done using eCAMI but that is not of interest to this project. However, the EC numbers from the `kmer_message` will be retrieved and assigned to their respective CAZyme domain. But, it will be noted that these associations are potentially weak and only indicative and there is not the same confidence level in them as would be if using eCAMI to specifically predict the EC numbers. These EC numbers are retrieved chiefly for the reason that other researchers may want them, thus the functionality is included now to save potential time in the future.\n",
    "\n",
    "**NOTE:** In the eCAMI README the `important_k_mer_number` is said to be default to 3, this is incorrect. In the script `eCAMI.prediction.py` it is set at a default value of 5.\n",
    "\n",
    "\n",
    "**Therefore, all the desired data we want from eCAMI is in the lines prefixed with '>'.**"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<a id=\"data_model\"></a>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Defining the data model\n",
    "\n",
    "The explanation of the design of the data model is explained in the note book prediction_data_model.ipynb.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [],
   "source": [
    "class CazymeDomain:\n",
    "    \"\"\"Single CAZyme domain in a protein, predicted by a CAZyme prediction tool.\n",
    "    \n",
    "    Each unique predicted CAZy family-subfamily combination by a CAZyme prediction\n",
    "    tool for a query protein sequence is viewed as a CAZyme domain.]\n",
    "    \n",
    "    Every CAZyme domain has a source CAZyme prediction tool that predicted the CAZyme\n",
    "    domain, a parent CAZyme protein (represented by the protein accession), and CAZy\n",
    "    family and subfamily combination. If no CAZy subfamily is predicted, the \n",
    "    subfamily will be listed as a null value.\n",
    "\n",
    "    Hotpep, CUPP and eCAMI predict EC numbers for each CAZyme domain. HMMER and CUPP\n",
    "    predict amino acid domain ranges. Multiple EC numbers and domain ranges can be\n",
    "    predicted for a single CAZyme domain, therefore, these attritbutes are stored as\n",
    "    lists.\n",
    "    \n",
    "    If an item of data is not or was not predicted by a CAZyme prediction tool, the\n",
    "    respective data will be stored as a null value.\n",
    "    \"\"\"\n",
    "    \n",
    "    def __init__(\n",
    "        self,\n",
    "        prediction_tool,\n",
    "        protein_accession,\n",
    "        cazy_family,\n",
    "        cazy_subfamily=None,\n",
    "        ec_numbers=None,\n",
    "        domain_range=None,\n",
    "    ):\n",
    "        \"\"\"Initiate instance\n",
    "        \n",
    "        :attr prediction_tool: str, CAZyme prediciton tool which predicted the domain\n",
    "        :attr protein_accession: str\n",
    "        :attr cazy_family: str\n",
    "        :attr cazy_subfamily: str\n",
    "        :attr ec_numbers: list (list of str, each str contains a unique EC number)\n",
    "        :attr domain_range: list (list of str, each str contains a unique domain range)\n",
    "        \"\"\"\n",
    "        self.prediction_tool = prediction_tool\n",
    "        self.protein_accession = protein_accession\n",
    "        self.cazy_family = cazy_family\n",
    "        \n",
    "        # not all CAZyme domans are catalogued under a CAZy subfamily\n",
    "        if cazy_subfamily is None:\n",
    "            self.cazy_subfamiy = np.nan\n",
    "        else:\n",
    "            self.cazy_subfamily = cazy_subfamily\n",
    "        \n",
    "        # EC numbers are not predicted for all CAZyme domains\n",
    "        if ec_numbers is None:\n",
    "            self.ec_numbers = []  # enables adding in EC numbers included in another line of the output file\n",
    "        else:\n",
    "            self.ec_numbers = ec_numbers\n",
    "        \n",
    "        # Not all prediction tools predict CAZyme domains\n",
    "        if domain_range is None:\n",
    "            self.domain_range = []  # enables adding domain range listed in another line of the ouput file\n",
    "        else:\n",
    "            self.domain_range = domain_range\n",
    "    \n",
    "    def __str__(self):\n",
    "        return f\"-CazymeDomain in {self.protein_accession}, fam={self.cazy_family}, subfam={self.cazy_subfamily}-\"\n",
    "    \n",
    "    def __repr__(self):\n",
    "        return f\"<CazymeDomain parent={self.protein_accession} fam={self.cazy_family}, subfam={self.cazy_subfamily}>\"\n",
    "\n",
    "    \n",
    "class CazymeProteinPrediction:\n",
    "    \"\"\"Single protein and CAZyme/non-CAZyme prediction by a CAZyme prediction tool\"\"\"\n",
    "    \n",
    "    def __init__(self, prediction_tool, protein_accession, cazyme_classification, cazyme_domains=None):\n",
    "        \"\"\"Initate class instance.\n",
    "        \n",
    "        :attr prediction_tool: str, name of CAZyme prediction tool\n",
    "        :attr protein_accession: str\n",
    "        :attr cazyme_classification: int, 1=CAZyme, 0=non-CAZyme\n",
    "        :attr cazyme_domains: list of CazymeDomain instances, domain predicted to be in the CAZyme\n",
    "        \"\"\"\n",
    "        self.prediction_tool = prediction_tool\n",
    "        self.protein_accession = protein_accession\n",
    "        self.cazyme_classification = cazyme_classification  # CAZyme=1, non-CAZyme=0\n",
    "        \n",
    "        # non-CAZymes will have no cazyme_domains\n",
    "        if cazyme_domains is None:\n",
    "            self.cazyme_domains = None\n",
    "        else:\n",
    "            self.cazyme_domains = cazyme_domains\n",
    "    \n",
    "    def __str__(self):\n",
    "        if self.cazyme_classification == 0:\n",
    "            return f\"-CazymeProteinPrediction, protein={self.protein_accession}, non-CAZyme-\"\n",
    "        else:\n",
    "            return f\"-CazymeProteinPrediction, protein={self.protein_accession}, CAZyme domains={len(self.cazyme_domains)}-\"\n",
    "    \n",
    "    def __repr__(self):\n",
    "        return f\"<CazymeProteinPrediction, protein={self.protein_accession}>\"\n",
    "\n",
    "    def get_cazyme_classification_dicts(self):\n",
    "        \"\"\"Returns a dictionary that contains the protein accession and its CAZyme/non-CAZyme classification.\n",
    "        \n",
    "        This dictionary is used to create the cazyme_classification dataframe for evaluating the performance\n",
    "        of the CAZyme prediction tools in terms of differnetiating between CAZymes (identified with a score of 1)\n",
    "        and non-CAZymes (identified with a score of 0). The CAZy classes under which the protein is classifed are\n",
    "        retrieved as well to enable evaluating the CAZyme/non-CAZyme differneitation per CAZy class, and obsever\n",
    "        differences between this performance between classes. A score of 0 incidates a protein was not classified\n",
    "        under the class, a score of 1 indicates the protein as classified under the class.\n",
    "        \"\"\"\n",
    "        cazyme_classification_dict = {\n",
    "            \"protein_accession\": self.protein_accession,\n",
    "            \"cazyme_classification\": self.cazyme_classification,\n",
    "            \"GH\": 0,\n",
    "            \"GT\": 0,\n",
    "            \"CE\": 0,\n",
    "            \"PL\": 0,\n",
    "            \"AA\": 0,\n",
    "            \"CBM\": 0,\n",
    "        }\n",
    "        \n",
    "        # check which CAZy classes the predicted CAZyme domains were classified under\n",
    "        for domain in cazyme_domains:\n",
    "            if domain.cazy_family[:3] == \"CBM\":\n",
    "                cazyme_classification_dict[\"CBM\"] = 1\n",
    "            else:\n",
    "                cazyme_classification_dict[family[:2]] = 1\n",
    "\n",
    "        return cazyme_classification_dict\n",
    "\n",
    "    def get_cazy_family_dict(self):\n",
    "        \"\"\"Return a dictionary of the predicted CAZy families for a protein.\n",
    "        \n",
    "        This dictionary is used to create a dataframe of CAZy family predictions for evaluating the performance\n",
    "        of the prediction tools to predict the CAZy family of a CAZyme. The dictionary is keyed by the CAZy\n",
    "        classes, and valued by a list of the corresponding families predicted for the CAZyme. The is to enable\n",
    "        evaluating the CAZy family prediction performance per class.\"\"\"\n",
    "        cazy_family_dict = {\n",
    "            \"GH\": [],\n",
    "            \"GT\": [],\n",
    "            \"CE\": [],\n",
    "            \"PL\": [],\n",
    "            \"AA\": [],\n",
    "            \"CBM\": [],\n",
    "        }\n",
    "        \n",
    "        for domain in self.cazyme_domains:\n",
    "            if domain.cazy_family[:3] == \"CBM\":\n",
    "                cazy_family_dict[\"CBM\"].append(domain.cazy_family)\n",
    "            else:\n",
    "                cazy_family_dict[domain.cazy_family[:2].append(domain.cazy_family)]\n",
    "        \n",
    "        return cazy_family_dict\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<a id=\"explore_one\"></a>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Exploring the first line of protein data\n",
    "\n",
    "First lets check out the data is separated and presented when using Python to read the text file. Spliting up the data weill prepare it for adding it to the data model. Additionally, we create a dictionary, keyed by protein accession and valued by its corresponding ECAMIprediciton instance. This is to enable quickly checking if a ECAMIprediction instance has already been created for a protein."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      ">ATY67280.1 A9K55_000020 \tGH17:43\tGH17:24|3.2.1.-:1\n",
      "LHKVFPGM(369),HKVFPGMD(370),KVFPGMDY(371),VFPGMDYT(372),FPGMDYTP(373),PPSQNNVT(392),PSQNNVTR(393),SQNNVTRD(394),QNNVTRDV(395),NNVTRDVA(396),NVTRDVAV(397),VTRDVAVL(398),TRDVAVLS(399),RDVAVLSQ(400),DVAVLSQL(401),AVLSQLTN(403),IRLYGTDC(412),RLYGTDCN(413),LYGTDCNQ(414),YGTDCNQT(415),GTDCNQTQ(416),TDCNQTQM(417),IVANEILF(476),VANEILFR(477),LPVATSDL(510),PVATSDLG(511),VATSDLGD(512),ATSDLGDD(513),TSDLGDDW(514),IMANIHPF(532),MANIHPFF(533),ISETGWPS(573),NGTNYFWF(618),GTNYFWFE(619),TNYFWFEA(620),NYFWFEAF(621),YFWFEAFD(622),FWFEAFDE(623),WFEAFDEP(624),FEAFDEPW(625),EAFDEPWK(626),WEDKWGLL(643),GVKIPDCG(659),VKIPDCGG(660),\n",
      ">ATY67467.1 A9K55_000078 \tGT90:15\tGT90:8\n",
      "RRGRHPPP(101),RGRHPPPG(102),GRHPPPGF(103),RHPPPGFD(104),AQDPCLQP(311),QDPCLQPH(312),DPCLQPHL(313),PCLQPHLR(314),PLFGGSKL(341),LPDVDGNS(503),DGNSYSAR(507),GNSYSARW(508),NSYSARWR(509),SMPLKATI(523),MPLKATIY(524),TPWVHFVP(540),PWVHFVPF(541),WVHFVPFD(542),LYVWRLLL(592),YVWRLLLE(593),\n"
     ]
    }
   ],
   "source": [
    "# Read the output file from eCAMI\n",
    "with open(\"8080495_output.txt\", \"r\") as fh:\n",
    "    ecami_file = fh.read().splitlines()\n",
    "\n",
    "for line in ecami_file[:4]:\n",
    "    print(line)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We are only interested in first line, which is identifable by starting with a '>'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      ">ATY67280.1 A9K55_000020 \tGH17:43\tGH17:24|3.2.1.-:1\n",
      ">ATY67467.1 A9K55_000078 \tGT90:15\tGT90:8\n"
     ]
    }
   ],
   "source": [
    "# Read the output file from eCAMI\n",
    "with open(\"8080495_output.txt\", \"r\") as fh:\n",
    "    ecami_file = fh.read().splitlines()\n",
    "\n",
    "for line in ecami_file[:4]:\n",
    "    if line.startswith('>'):\n",
    "        print(line)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['>ATY67280.1 A9K55_000020 ', 'GH17:43', 'GH17:24|3.2.1.-:1']\n",
      "['>ATY67467.1 A9K55_000078 ', 'GT90:15', 'GT90:8']\n"
     ]
    }
   ],
   "source": [
    "# Read the output file from eCAMI\n",
    "with open(\"8080495_output.txt\", \"r\") as fh:\n",
    "    ecami_file = fh.read().splitlines()\n",
    "\n",
    "ecami_predictions = {}  # stores proteins {protein_accession:ECAMIprediction_instance}\n",
    "\n",
    "for line in ecami_file[:4]:\n",
    "    if not line.startswith('>'):  # only want the first line for a protien not the listed k-mers\n",
    "        continue\n",
    "    \n",
    "    prediction_output = line.split(\"\\t\")\n",
    "    print(prediction_output)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<a id=\"cazy_fam\"></a>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Getting the predicted CAZy family for the CAZyme domain\n",
    "\n",
    "The CAZy family of the CAZyme domain of interest in the line starting with '>' is stored in the second element in `prediction_output` list."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "cazy_fam= GH17:43\n",
      "cazy_fam= GT90:15\n",
      "cazy_fam= CBM48:20\n",
      "cazy_fam= GH13:203\n"
     ]
    }
   ],
   "source": [
    "# Read the output file from eCAMI\n",
    "with open(\"8080495_output.txt\", \"r\") as fh:\n",
    "    ecami_file = fh.read().splitlines()\n",
    "\n",
    "ecami_predictions = {}  # stores proteins {protein_accession:ECAMIprediction_instance}\n",
    "\n",
    "for line in ecami_file[:8]:\n",
    "    if not line.startswith('>'):  # only want the first line for a protien not the listed k-mers\n",
    "        continue\n",
    "    \n",
    "    prediction_output = line.split(\"\\t\")\n",
    "    \n",
    "    # retrieve the CAZy family\n",
    "    cazy_fam = prediction_output[1]\n",
    "    print(\"cazy_fam=\", cazy_fam)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The CAZy family is stored in the format \"fam_name:group_number\", where the group_number is the number of the k-mer group, which is data that is not needed for evaluating the performance of eCAMI. We can remove the k-mer group number by spliting the \"fam_name:group_number\" string at the colon, and retrieving the first item."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "cazy_fam= GH17\n",
      "cazy_fam= GT90\n",
      "cazy_fam= CBM48\n",
      "cazy_fam= GH13\n"
     ]
    }
   ],
   "source": [
    "# Read the output file from eCAMI\n",
    "with open(\"8080495_output.txt\", \"r\") as fh:\n",
    "    ecami_file = fh.read().splitlines()\n",
    "\n",
    "ecami_predictions = {}  # stores proteins {protein_accession:ECAMIprediction_instance}\n",
    "\n",
    "for line in ecami_file[:8]:\n",
    "    if not line.startswith('>'):  # only want the first line for a protien not the listed k-mers\n",
    "        continue\n",
    "    \n",
    "    prediction_output = line.split(\"\\t\")\n",
    "    \n",
    "    # retrieve the CAZy family\n",
    "    cazy_fam = prediction_output[1].split(\":\")[0]\n",
    "    print(\"cazy_fam=\", cazy_fam)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<a id=\"cazy_subfam\"></a>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Getting the predicted CAZy subfamily\n",
    "\n",
    "**Note:** For details about why the CAZy subfamily is retrieved from please refere back to the section [\n",
    "Exploring the eCAMI output text file](#text_file).\n",
    "\n",
    "The CAZy subfamily is stored in the third element of the first protein line (when the line is split up by tabs), this referred to as the 'subfam_name_of_the_group:subfam_name_count' in the eCAMI documentation. This string is composed of known labels associated with the proteins used to create the kmer cluster library, which was then used to predicted the CAZy families of our query protein sequences.\n",
    "\n",
    "Therefore, in the line  \n",
    "`>ATY67289.1 A9K55_000120     GH13:203    GH13_8:577|CBM48:328|2.4.1.18:32|GH13:13|GT4:1`, the string `GH13_8:577|CBM48:328|2.4.1.18:32|GH13:13|GT4:1`   \n",
    "contains the CAZy subfamily of the CAZyme domain represented by the line.\n",
    "\n",
    "We need to check if the third element of the `prediction_output` list contains any subfamiles (identified by the general format `\\D{2,3}\\d+_\\d+`) and that the parent family matches the family listed in the second element of the `prediction_output` list.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "GH17:24|3.2.1.-:1\n",
      "GT90:8\n",
      "CBM48:339|GH13_8:324|2.4.1.18:32|GH13:4|GT4:2\n",
      "GH13_8:577|CBM48:328|2.4.1.18:32|GH13:13|GT4:1\n"
     ]
    }
   ],
   "source": [
    "# Read the output file from eCAMI\n",
    "with open(\"8080495_output.txt\", \"r\") as fh:\n",
    "    ecami_file = fh.read().splitlines()\n",
    "\n",
    "ecami_predictions = {}  # stores proteins {protein_accession:ECAMIprediction_instance}\n",
    "\n",
    "for line in ecami_file[:8]:\n",
    "    if not line.startswith('>'):  # only want the first line for a protien not the listed k-mers\n",
    "        continue\n",
    "    \n",
    "    prediction_output = line.split(\"\\t\")\n",
    "    \n",
    "    # retrieve the CAZy family\n",
    "    cazy_fam = prediction_output[1].split(\":\")[0]\n",
    "    \n",
    "    # retrieve the CAZy subfamily\n",
    "    subfam_group = prediction_output[2]\n",
    "    print(subfam_group)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The items in the \"subfam_name_of_the_group:subfam_name_count\" section are separated by '|'. Therefore, we need to separte the predicted items and then check which are CAZy subfamilies."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['GH17:24', '3.2.1.-:1']\n",
      "['GT90:8']\n",
      "['CBM48:339', 'GH13_8:324', '2.4.1.18:32', 'GH13:4', 'GT4:2']\n",
      "['GH13_8:577', 'CBM48:328', '2.4.1.18:32', 'GH13:13', 'GT4:1']\n"
     ]
    }
   ],
   "source": [
    "# Read the output file from eCAMI\n",
    "with open(\"8080495_output.txt\", \"r\") as fh:\n",
    "    ecami_file = fh.read().splitlines()\n",
    "\n",
    "ecami_predictions = {}  # stores proteins {protein_accession:CazymeProteinPrediction_instance}\n",
    "\n",
    "for line in ecami_file[:8]:\n",
    "    if not line.startswith('>'):  # only want the first line for a protien not the listed k-mers\n",
    "        continue\n",
    "    \n",
    "    prediction_output = line.split(\"\\t\")\n",
    "    \n",
    "    # retrieve the CAZy family\n",
    "    cazy_fam = prediction_output[1].split(\":\")[0]\n",
    "    \n",
    "    # retrieve the CAZy subfamily\n",
    "    subfam_group = prediction_output[2]\n",
    "    subfam_group = subfam_group.split(\"|\")\n",
    "    print(subfam_group)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The EC numbers and CAZy (sub)families are also stored with the associated group number, e.g. the group number of the predicted EC number. These group numbers are separated by a colon from the EC numbers and CAZy (sub)families."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "cazy_family= GH17\n",
      "cazy_subfamilies= []\n",
      "cazy_family= GT90\n",
      "cazy_subfamilies= []\n",
      "cazy_family= CBM48\n",
      "cazy_subfamilies= []\n",
      "cazy_family= GH13\n",
      "cazy_subfamilies= ['GH13_8']\n"
     ]
    }
   ],
   "source": [
    "# Read the output file from eCAMI\n",
    "with open(\"8080495_output.txt\", \"r\") as fh:\n",
    "    ecami_file = fh.read().splitlines()\n",
    "\n",
    "ecami_predictions = {}  # stores proteins {protein_accession:CazymeProteinPrediction_instance}\n",
    "\n",
    "for line in ecami_file[:8]:\n",
    "    if not line.startswith('>'):  # only want the first line for a protien not the listed k-mers\n",
    "        continue\n",
    "    \n",
    "    prediction_output = line.split(\"\\t\")\n",
    "    \n",
    "    # retrieve the CAZy family\n",
    "    cazy_family = prediction_output[1].split(\":\")[0]\n",
    "    \n",
    "    # retrieve the CAZy subfamily\n",
    "    subfam_group = prediction_output[2]\n",
    "    subfam_group = subfam_group.split(\"|\")\n",
    "    cazy_subfamilies = []  # store all listed predicted CAZy subfamilies\n",
    "    # check if the subfam_group contains a predicted CAZy subfamily\n",
    "    for item in subfam_group:\n",
    "        # remove the group number of the predicted item\n",
    "        item = item.split(\":\")[0]\n",
    "        try:\n",
    "            re.match(r\"\\D{2,3}\\d+_\\d+\", item)\n",
    "            # check if the subfamily belongs to the CAZy family listed for the current CAZyme domain\n",
    "            if item[:item.find(\"_\")] == cazy_family:\n",
    "                cazy_subfamilies.append(item)\n",
    "        except AttributeError:  # raised if the item is not a CAZy subfamily\n",
    "            pass\n",
    "    \n",
    "    print(\"cazy_family=\", cazy_family)\n",
    "    print(\"cazy_subfamilies=\", cazy_subfamilies)\n",
    "            "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<a id=\"ecnum\"></a>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Getting the EC numbers\n",
    "\n",
    "As noted and explained in the section [Exploring the eCAMI output text file](#text_file), these are not the EC numbers specifically predicted by eCAMI. To predict the EC numbers of the query proteins, the analysis would need to be run again using the 'EC' kmer library. However, the EC numbers listed in the `kmer_message` (for this definition see the section 'Exploring the eCAMI output text file')....\n",
    "\n",
    "The predicted EC numbers for the current working CAZyme domain are listed in the subfam_group (the third element of the `prediction_output` list), along with the CAZy subfamilies. Therefore, the identification of the EC numbers can be achieved at the same time as getting the CAZy subfamilies. If digits are missing in the EC number, they are replaced with a dash, from this we are able to check by the format of the string if a string includes an EC number."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "cazy_family= GH17\n",
      "cazy_subfamilies= []\n",
      "ec_numbers= ['3.2.1.-']\n",
      "cazy_family= GT90\n",
      "cazy_subfamilies= []\n",
      "ec_numbers= []\n",
      "cazy_family= CBM48\n",
      "cazy_subfamilies= []\n",
      "ec_numbers= ['2.4.1.18']\n",
      "cazy_family= GH13\n",
      "cazy_subfamilies= ['GH13_8']\n",
      "ec_numbers= ['2.4.1.18']\n"
     ]
    }
   ],
   "source": [
    "# Read the output file from eCAMI\n",
    "with open(\"8080495_output.txt\", \"r\") as fh:\n",
    "    ecami_file = fh.read().splitlines()\n",
    "\n",
    "ecami_predictions = {}  # stores proteins {protein_accession:ECAMIprediction_instance}\n",
    "\n",
    "for line in ecami_file[:8]:\n",
    "    if not line.startswith('>'):  # only want the first line for a protien not the listed k-mers\n",
    "        continue\n",
    "    \n",
    "    prediction_output = line.split(\"\\t\")\n",
    "    \n",
    "    # retrieve the CAZy family\n",
    "    cazy_family = prediction_output[1].split(\":\")[0]\n",
    "    \n",
    "    # retrieve the CAZy subfamily and EC numbers\n",
    "    subfam_group = prediction_output[2]\n",
    "    subfam_group = subfam_group.split(\"|\")\n",
    "\n",
    "    cazy_subfamilies = []  # store all listed predicted CAZy subfamilies\n",
    "    ec_numbers = []  # store all predicted EC numbers\n",
    "\n",
    "    # check if the subfam_group contains a predicted CAZy subfamily\n",
    "    for item in subfam_group:\n",
    "        # remove the group number of the predicted item\n",
    "        item = item.split(\":\")[0]\n",
    "\n",
    "        # check if the item contains a CAZy subfamily\n",
    "        try:\n",
    "            re.match(r\"\\D{2,3}\\d+_\\d+\", item).group()\n",
    "            # check if the subfamily belongs to the CAZy family listed for the current CAZyme domain\n",
    "            if item[:item.find(\"_\")] == cazy_family:\n",
    "                cazy_subfamilies.append(item)\n",
    "        except AttributeError:  # raised if the item is not a CAZy subfamily\n",
    "            pass\n",
    "        \n",
    "        # check if item contains an EC number\n",
    "        try:\n",
    "            re.match(r\"\\d+\\.(\\d+|-)\\.(\\d+|-)\\.(\\d+|-)\", item).group()\n",
    "            ec_numbers.append(item)\n",
    "        except AttributeError:  # raised if the item is not an EC number\n",
    "            pass\n",
    "    \n",
    "    print(\"cazy_family=\", cazy_family)\n",
    "    print(\"cazy_subfamilies=\", cazy_subfamilies)\n",
    "    print(\"ec_numbers=\", ec_numbers)\n",
    "            "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<a id=\"combining_proteins\"></a>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Combining data for a single protein\n",
    "\n",
    "As discussed before, a unique protein can be listed multiple times in the eCAMI output file. To combine data together so a single ECAMIprediction instance for a protein, we check if a corresponding key (defined by the protein's accession) is in the ecami_prediction dictionary."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "97740f2da7b340a685c7689aea00f134",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "HBox(children=(HTML(value='Parsing eCAMI output'), FloatProgress(value=0.0, max=644.0), HTML(value='')))"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "existing_prediction= -CazymeProteinPrediction, protein=ATY67289.1 A9K55_000120, CAZyme domains=1-\n",
      "existing_prediction=[2] -CazymeProteinPrediction, protein=ATY67289.1 A9K55_000120, CAZyme domains=2-\n",
      "existing_prediction= -CazymeProteinPrediction, protein=ATY66529.1 A9K55_000499, CAZyme domains=1-\n",
      "existing_prediction=[2] -CazymeProteinPrediction, protein=ATY66529.1 A9K55_000499, CAZyme domains=2-\n",
      "existing_prediction= -CazymeProteinPrediction, protein=ATY66664.1 A9K55_000997, CAZyme domains=1-\n",
      "existing_prediction=[2] -CazymeProteinPrediction, protein=ATY66664.1 A9K55_000997, CAZyme domains=2-\n",
      "existing_prediction= -CazymeProteinPrediction, protein=ATY66444.1 A9K55_001136, CAZyme domains=1-\n",
      "existing_prediction=[2] -CazymeProteinPrediction, protein=ATY66444.1 A9K55_001136, CAZyme domains=2-\n",
      "existing_prediction= -CazymeProteinPrediction, protein=ATY65637.1 A9K55_001791, CAZyme domains=1-\n",
      "existing_prediction=[2] -CazymeProteinPrediction, protein=ATY65637.1 A9K55_001791, CAZyme domains=2-\n",
      "existing_prediction= -CazymeProteinPrediction, protein=ATY59023.1 A9K55_002239, CAZyme domains=1-\n",
      "existing_prediction=[2] -CazymeProteinPrediction, protein=ATY59023.1 A9K55_002239, CAZyme domains=2-\n",
      "existing_prediction= -CazymeProteinPrediction, protein=ATY59188.1 A9K55_002493, CAZyme domains=1-\n",
      "existing_prediction=[2] -CazymeProteinPrediction, protein=ATY59188.1 A9K55_002493, CAZyme domains=2-\n",
      "existing_prediction= -CazymeProteinPrediction, protein=ATY59188.1 A9K55_002493, CAZyme domains=2-\n",
      "existing_prediction=[2] -CazymeProteinPrediction, protein=ATY59188.1 A9K55_002493, CAZyme domains=3-\n",
      "existing_prediction= -CazymeProteinPrediction, protein=ATY58486.1 A9K55_003070, CAZyme domains=1-\n",
      "existing_prediction=[2] -CazymeProteinPrediction, protein=ATY58486.1 A9K55_003070, CAZyme domains=2-\n",
      "existing_prediction= -CazymeProteinPrediction, protein=ATY59698.1 A9K55_003119, CAZyme domains=1-\n",
      "existing_prediction=[2] -CazymeProteinPrediction, protein=ATY59698.1 A9K55_003119, CAZyme domains=2-\n",
      "existing_prediction= -CazymeProteinPrediction, protein=ATY58848.1 A9K55_003123, CAZyme domains=1-\n",
      "existing_prediction=[2] -CazymeProteinPrediction, protein=ATY58848.1 A9K55_003123, CAZyme domains=2-\n",
      "existing_prediction= -CazymeProteinPrediction, protein=ATY58848.1 A9K55_003123, CAZyme domains=2-\n",
      "existing_prediction=[2] -CazymeProteinPrediction, protein=ATY58848.1 A9K55_003123, CAZyme domains=3-\n",
      "existing_prediction= -CazymeProteinPrediction, protein=ATY65223.1 A9K55_003996, CAZyme domains=1-\n",
      "existing_prediction=[2] -CazymeProteinPrediction, protein=ATY65223.1 A9K55_003996, CAZyme domains=2-\n",
      "existing_prediction= -CazymeProteinPrediction, protein=ATY65223.1 A9K55_003996, CAZyme domains=2-\n",
      "existing_prediction=[2] -CazymeProteinPrediction, protein=ATY65223.1 A9K55_003996, CAZyme domains=3-\n",
      "existing_prediction= -CazymeProteinPrediction, protein=ATY64248.1 A9K55_004102, CAZyme domains=1-\n",
      "existing_prediction=[2] -CazymeProteinPrediction, protein=ATY64248.1 A9K55_004102, CAZyme domains=2-\n",
      "existing_prediction= -CazymeProteinPrediction, protein=ATY64693.1 A9K55_005188, CAZyme domains=1-\n",
      "existing_prediction=[2] -CazymeProteinPrediction, protein=ATY64693.1 A9K55_005188, CAZyme domains=2-\n",
      "existing_prediction= -CazymeProteinPrediction, protein=ATY64918.1 A9K55_005214, CAZyme domains=1-\n",
      "existing_prediction=[2] -CazymeProteinPrediction, protein=ATY64918.1 A9K55_005214, CAZyme domains=2-\n",
      "existing_prediction= -CazymeProteinPrediction, protein=ATY64953.1 A9K55_005326, CAZyme domains=1-\n",
      "existing_prediction=[2] -CazymeProteinPrediction, protein=ATY64953.1 A9K55_005326, CAZyme domains=2-\n",
      "existing_prediction= -CazymeProteinPrediction, protein=ATY64976.1 A9K55_005332, CAZyme domains=1-\n",
      "existing_prediction=[2] -CazymeProteinPrediction, protein=ATY64976.1 A9K55_005332, CAZyme domains=2-\n",
      "existing_prediction= -CazymeProteinPrediction, protein=ATY61215.1 A9K55_005410, CAZyme domains=1-\n",
      "existing_prediction=[2] -CazymeProteinPrediction, protein=ATY61215.1 A9K55_005410, CAZyme domains=2-\n",
      "existing_prediction= -CazymeProteinPrediction, protein=ATY61276.1 A9K55_005467, CAZyme domains=1-\n",
      "existing_prediction=[2] -CazymeProteinPrediction, protein=ATY61276.1 A9K55_005467, CAZyme domains=2-\n",
      "existing_prediction= -CazymeProteinPrediction, protein=ATY60620.1 A9K55_005793, CAZyme domains=1-\n",
      "existing_prediction=[2] -CazymeProteinPrediction, protein=ATY60620.1 A9K55_005793, CAZyme domains=2-\n",
      "existing_prediction= -CazymeProteinPrediction, protein=ATY59890.1 A9K55_006062, CAZyme domains=1-\n",
      "existing_prediction=[2] -CazymeProteinPrediction, protein=ATY59890.1 A9K55_006062, CAZyme domains=2-\n",
      "existing_prediction= -CazymeProteinPrediction, protein=ATY61069.1 A9K55_006802, CAZyme domains=1-\n",
      "existing_prediction=[2] -CazymeProteinPrediction, protein=ATY61069.1 A9K55_006802, CAZyme domains=2-\n",
      "existing_prediction= -CazymeProteinPrediction, protein=ATY61824.1 A9K55_006988, CAZyme domains=1-\n",
      "existing_prediction=[2] -CazymeProteinPrediction, protein=ATY61824.1 A9K55_006988, CAZyme domains=2-\n",
      "existing_prediction= -CazymeProteinPrediction, protein=ATY62856.1 A9K55_007728, CAZyme domains=1-\n",
      "existing_prediction=[2] -CazymeProteinPrediction, protein=ATY62856.1 A9K55_007728, CAZyme domains=2-\n",
      "existing_prediction= -CazymeProteinPrediction, protein=ATY61893.1 A9K55_007772, CAZyme domains=1-\n",
      "existing_prediction=[2] -CazymeProteinPrediction, protein=ATY61893.1 A9K55_007772, CAZyme domains=2-\n",
      "existing_prediction= -CazymeProteinPrediction, protein=ATY63454.1 A9K55_009028, CAZyme domains=1-\n",
      "existing_prediction=[2] -CazymeProteinPrediction, protein=ATY63454.1 A9K55_009028, CAZyme domains=2-\n",
      "existing_prediction= -CazymeProteinPrediction, protein=ATY62564.1 A9K55_009060, CAZyme domains=1-\n",
      "existing_prediction=[2] -CazymeProteinPrediction, protein=ATY62564.1 A9K55_009060, CAZyme domains=2-\n",
      "existing_prediction= -CazymeProteinPrediction, protein=ATY61851.1 A9K55_009292, CAZyme domains=1-\n",
      "existing_prediction=[2] -CazymeProteinPrediction, protein=ATY61851.1 A9K55_009292, CAZyme domains=2-\n",
      "existing_prediction= -CazymeProteinPrediction, protein=ATY61851.1 A9K55_009292, CAZyme domains=2-\n",
      "existing_prediction=[2] -CazymeProteinPrediction, protein=ATY61851.1 A9K55_009292, CAZyme domains=3-\n",
      "existing_prediction= -CazymeProteinPrediction, protein=ATY61718.1 A9K55_009347, CAZyme domains=1-\n",
      "existing_prediction=[2] -CazymeProteinPrediction, protein=ATY61718.1 A9K55_009347, CAZyme domains=2-\n",
      "existing_prediction= -CazymeProteinPrediction, protein=ATY61718.1 A9K55_009347, CAZyme domains=2-\n",
      "existing_prediction=[2] -CazymeProteinPrediction, protein=ATY61718.1 A9K55_009347, CAZyme domains=3-\n",
      "\n",
      "length= 289\n"
     ]
    }
   ],
   "source": [
    "# Read the output file from eCAMI\n",
    "with open(\"8080495_output.txt\", \"r\") as fh:\n",
    "    ecami_file = fh.read().splitlines()\n",
    "\n",
    "ecami_predictions = {}  # stores proteins {protein_accession:CazymeProteinPrediction_instance}\n",
    "\n",
    "for line in tqdm(ecami_file, desc=\"Parsing eCAMI output\"):\n",
    "    if not line.startswith('>'):  # only want the first line for a protien not the listed k-mers\n",
    "        continue\n",
    "    \n",
    "    prediction_output = line.split(\"\\t\")\n",
    "    \n",
    "    # retrieve protein accession, removing the '>' prefix and stripping white space\n",
    "    protein_accession = prediction_output[0][1:].strip()\n",
    "    cazyme_classification = 1\n",
    "    \n",
    "    # retrieve the CAZy family\n",
    "    cazy_family = prediction_output[1].split(\":\")[0]\n",
    "    \n",
    "    # retrieve the CAZy subfamily and EC numbers\n",
    "    subfam_group = prediction_output[2]\n",
    "    subfam_group = subfam_group.split(\"|\")\n",
    "\n",
    "    cazy_subfamilies = []  # store all listed predicted CAZy subfamilies\n",
    "    ec_numbers = []  # store all predicted EC numbers\n",
    "\n",
    "    # check if the subfam_group contains a predicted CAZy subfamily\n",
    "    for item in subfam_group:\n",
    "        # remove the group number of the predicted item\n",
    "        item = item.split(\":\")[0]\n",
    "\n",
    "        # check if the item contains a CAZy subfamily\n",
    "        try:\n",
    "            re.match(r\"\\D{2,3}\\d+_\\d+\", item).group()\n",
    "            # check if the subfamily belongs to the CAZy family listed for the current CAZyme domain\n",
    "            if item[:item.find(\"_\")] == cazy_family:\n",
    "                cazy_subfamilies.append(item)\n",
    "            continue  # it was a CAZy subfamily, don't need to check if it was an EC#\n",
    "    \n",
    "        except AttributeError:  # raised if the item is not a CAZy subfamily\n",
    "            pass\n",
    "        \n",
    "        # check if item contains an EC number\n",
    "        try:\n",
    "            re.match(r\"\\d+\\.(\\d+|-)\\.(\\d+|-)\\.(\\d+|-)\", item).group()\n",
    "            ec_numbers.append(item)\n",
    "        except AttributeError:  # raised if the item is not an EC number\n",
    "            # check if it was a CAZy family, this is for detecting unexpected data types\n",
    "            try:\n",
    "                re.match(r\"\\D{2,3}\\d+\", item).group()\n",
    "            except AttributeError:\n",
    "                print(\n",
    "                    f\"WARNING -- unexpected data type of item '{item}' for protein {protein_accession} \"\n",
    "                    \"by eCAMI. Not adding data to parsed data from eCAMI.\"\n",
    "                )\n",
    "    \n",
    "    # add ECAMIprediction instance to ecami_prediction\n",
    "    cazyme_classification = 1  # all proteins included in eCAMI output file are identified as CAZymes\n",
    "    \n",
    "    if len(cazy_subfamilies) == 0:\n",
    "        cazy_subfamilies = [np.nan]\n",
    "    else:\n",
    "        cazy_subfamilies = \", \".join(cazy_subfamilies)  # convert to string\n",
    "        \n",
    "    if len(ec_numbers) == 0:\n",
    "        ec_numbers = [np.nan]\n",
    "    \n",
    "    try:\n",
    "        # add new predicted CAZyme domain to an existing ECAMIprediction instance\n",
    "        existing_prediction = ecami_predictions[protein_accession]\n",
    "        print(\"existing_prediction=\", existing_prediction)\n",
    "        \n",
    "        # check if the CAZyme domain has already been passed\n",
    "        existing_cazyme_domains = existing_prediction.cazyme_domains\n",
    "        existance = False  # becomes true if existing instance found\n",
    "        for domain in existing_cazyme_domains:\n",
    "            if (domain.cazy_family == cazy_family) and (domain.cazy_subfamily == cazy_subfamily):\n",
    "                domain.ec_numbers.append(np.nan)\n",
    "                existance = True\n",
    "                print(f\"WARNING -- existing cazyme domain found, {protein_accession}, {domain.cazy_family}, {domain.cazy_subfamily}\")\n",
    "                break\n",
    "        \n",
    "        if existance is False:\n",
    "            # create new CAZyme domain\n",
    "            new_cazyme_domain = CazymeDomain(\n",
    "                \"eCAMI\",\n",
    "                protein_accession,\n",
    "                cazy_family,\n",
    "                cazy_subfamilies,\n",
    "                ec_numbers,\n",
    "            )\n",
    "            existing_prediction.cazyme_domains.append(new_cazyme_domain)\n",
    "        print(\"existing_prediction=[2]\", existing_prediction)\n",
    "    \n",
    "    except KeyError:  # raised if no corresponding ECAMIprediction instance was found\n",
    "\n",
    "            new_cazyme_domain = CazymeDomain(\n",
    "                \"eCAMI\",\n",
    "                protein_accession,\n",
    "                cazy_family,\n",
    "                cazy_subfamilies,\n",
    "                ec_numbers,\n",
    "            )\n",
    "            \n",
    "            new_protein = CazymeProteinPrediction(\n",
    "                \"eCAMI\",\n",
    "                protein_accession,\n",
    "                cazyme_classification,\n",
    "            )\n",
    "    \n",
    "            new_protein.cazyme_domains = [new_cazyme_domain]\n",
    "            \n",
    "            ecami_predictions[protein_accession] = new_protein\n",
    "\n",
    "\n",
    "\n",
    "print(\"length=\", len(list(ecami_predictions.values())))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "From the print out we can see we are picking up when multiple CAZyme domains are predicted for the same protein. Remember that each CAZyme domain prediction covers 2 lines, therefore, the 644 lines from the output file contains 322 predicted CAZyme domains, some of which are from the same CAZyme."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<a id=\"no_caz\"></a>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Addining non-CAZymes\n",
    "\n",
    "The eCAMI output includes all the proteins that eCAMI has predicted at CAZymes. For evaluating the non-CAZyme/CAZyme differentiation performance we need the CAZyme and non-CAZyme classification. Therefore, we can parse the FASTA file that containing the protein sequences used as input by eCAMI and add those that have not been including in the eCAMI output file as non-CAZymes."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "65fe73cf61514801b8fff5ffc5446457",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "HBox(children=(HTML(value='Parsing eCAMI output'), FloatProgress(value=0.0, max=644.0), HTML(value='')))"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "length pre-non-CAZymes = 289\n"
     ]
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "804d853475304150ae24ab5471cc4680",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "HBox(children=(HTML(value='Adding non-CAZymes'), FloatProgress(value=0.0, max=93219.0), HTML(value='')))"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "proteins parsed = 9287\n",
      "proteins in the input FASTA file= 9287\n"
     ]
    }
   ],
   "source": [
    "# Read the output file from eCAMI\n",
    "with open(\"8080495_output.txt\", \"r\") as fh:\n",
    "    ecami_file = fh.read().splitlines()\n",
    "\n",
    "ecami_predictions = {}  # stores proteins {protein_accession:CazymeProteinPrediction_instance}\n",
    "\n",
    "for line in tqdm(ecami_file, desc=\"Parsing eCAMI output\"):\n",
    "    if not line.startswith('>'):  # only want the first line for a protien not the listed k-mers\n",
    "        continue\n",
    "    \n",
    "    prediction_output = line.split(\"\\t\")\n",
    "    \n",
    "    # retrieve protein accession, removing the '>' prefix and stripping white space\n",
    "    protein_accession = prediction_output[0][1:].strip()\n",
    "    cazyme_classification = 1\n",
    "    \n",
    "    # retrieve the CAZy family\n",
    "    cazy_family = prediction_output[1].split(\":\")[0]\n",
    "    \n",
    "    # retrieve the CAZy subfamily and EC numbers\n",
    "    subfam_group = prediction_output[2]\n",
    "    subfam_group = subfam_group.split(\"|\")\n",
    "\n",
    "    cazy_subfamilies = []  # store all listed predicted CAZy subfamilies\n",
    "    ec_numbers = []  # store all predicted EC numbers\n",
    "\n",
    "    # check if the subfam_group contains a predicted CAZy subfamily\n",
    "    for item in subfam_group:\n",
    "        # remove the group number of the predicted item\n",
    "        item = item.split(\":\")[0]\n",
    "\n",
    "        # check if the item contains a CAZy subfamily\n",
    "        try:\n",
    "            re.match(r\"\\D{2,3}\\d+_\\d+\", item).group()\n",
    "            # check if the subfamily belongs to the CAZy family listed for the current CAZyme domain\n",
    "            if item[:item.find(\"_\")] == cazy_family:\n",
    "                cazy_subfamilies.append(item)\n",
    "            continue  # it was a CAZy subfamily, don't need to check if it was an EC#\n",
    "    \n",
    "        except AttributeError:  # raised if the item is not a CAZy subfamily\n",
    "            pass\n",
    "        \n",
    "        # check if item contains an EC number\n",
    "        try:\n",
    "            re.match(r\"\\d+\\.(\\d+|-)\\.(\\d+|-)\\.(\\d+|-)\", item).group()\n",
    "            ec_numbers.append(item)\n",
    "        except AttributeError:  # raised if the item is not an EC number\n",
    "            # check if it was a CAZy family, this is for detecting unexpected data types\n",
    "            try:\n",
    "                re.match(r\"\\D{2,3}\\d+\", item).group()\n",
    "            except AttributeError:\n",
    "                print(\n",
    "                    f\"WARNING -- unexpected data type of item '{item}' for protein {protein_accession} \"\n",
    "                    \"by eCAMI. Not adding data to parsed data from eCAMI.\"\n",
    "                )\n",
    "    \n",
    "    # add ECAMIprediction instance to ecami_prediction\n",
    "    cazyme_classification = 1  # all proteins included in eCAMI output file are identified as CAZymes\n",
    "    \n",
    "    if len(cazy_subfamilies) == 0:\n",
    "        cazy_subfamilies = [np.nan]\n",
    "    else:\n",
    "        cazy_subfamilies = \", \".join(cazy_subfamilies)  # convert to string\n",
    "        \n",
    "    if len(ec_numbers) == 0:\n",
    "        ec_numbers = [np.nan]\n",
    "    \n",
    "    try:\n",
    "        # add new predicted CAZyme domain to an existing ECAMIprediction instance\n",
    "        existing_prediction = ecami_predictions[protein_accession]\n",
    "        \n",
    "        # check if the CAZyme domain has already been passed\n",
    "        existing_cazyme_domains = existing_prediction.cazyme_domains\n",
    "        existance = False  # becomes true if existing instance found\n",
    "        for domain in existing_cazyme_domains:\n",
    "            if (domain.cazy_family == cazy_family) and (domain.cazy_subfamily == cazy_subfamily):\n",
    "                domain.ec_numbers.append(np.nan)\n",
    "                existance = True\n",
    "                print(f\"WARNING -- existing cazyme domain found, {protein_accession}, {domain.cazy_family}, {domain.cazy_subfamily}\")\n",
    "                break\n",
    "        \n",
    "        if existance is False:\n",
    "            # create new CAZyme domain\n",
    "            new_cazyme_domain = CazymeDomain(\n",
    "                \"eCAMI\",\n",
    "                protein_accession,\n",
    "                cazy_family,\n",
    "                cazy_subfamilies,\n",
    "                ec_numbers,\n",
    "            )\n",
    "            existing_prediction.cazyme_domains.append(new_cazyme_domain)\n",
    "    \n",
    "    except KeyError:  # raised if no corresponding ECAMIprediction instance was found\n",
    "\n",
    "            new_cazyme_domain = CazymeDomain(\n",
    "                \"eCAMI\",\n",
    "                protein_accession,\n",
    "                cazy_family,\n",
    "                cazy_subfamilies,\n",
    "                ec_numbers,\n",
    "            )\n",
    "            \n",
    "            new_protein = CazymeProteinPrediction(\n",
    "                \"eCAMI\",\n",
    "                protein_accession,\n",
    "                cazyme_classification,\n",
    "            )\n",
    "    \n",
    "            new_protein.cazyme_domains = [new_cazyme_domain]\n",
    "            \n",
    "            ecami_predictions[protein_accession] = new_protein\n",
    "\n",
    "print(\"length pre-non-CAZymes =\", len(list(ecami_predictions.values())))\n",
    "            \n",
    "# Add non-CAZymes\n",
    "# open the FASTA file containing the input protein sequences\n",
    "with open(\"genbank_proteins_txid73501_GCA_008080495_1.fasta\", \"r\") as fh:\n",
    "    fasta = fh.read().splitlines()\n",
    "count = 0\n",
    "for line in tqdm(fasta, desc=\"Adding non-CAZymes\"):\n",
    "    if line.startswith(\">\"):\n",
    "        count += 1\n",
    "        # remove '>' prefix and white space\n",
    "        protein_accession = line[1:].strip()\n",
    "        \n",
    "        # check if the protein is already listed in the eCAMI predictions\n",
    "        try:\n",
    "            ecami_predictions[protein_accession]\n",
    "        except KeyError:  # raised of protein not in ecami_predictions\n",
    "            cazyme_classification = 0\n",
    "            ecami_predictions[protein_accession] = CazymeProteinPrediction(\n",
    "                \"eCAMI\",\n",
    "                protein_accession,\n",
    "                cazyme_classification,\n",
    "            )\n",
    "\n",
    "print(\"proteins parsed =\", len(list(ecami_predictions.values())))\n",
    "print(\"proteins in the input FASTA file=\", count)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "From counting the proteins in the FASTA file we can see that all proteins are retrieved, and every protein is represented by a `CazymeProteinPrediction` instance."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<a id=\"final\"></a>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Final functions\n",
    "\n",
    "We can see above that all the proteins that were contained in the FASTA file, which was used as input by eCAMI, are parsed by the method above. Now we can break this up into functions that are suitable for dropping straigt into `pyrewton` and are easier to follow."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def parse_ecami_output(ecami_output_path, fasta_path):\n",
    "    \"\"\"Parse the output from eCAMI, retrieving predicted CAZyme domains, CAZy (sub)families and EC numbers.\n",
    "    \n",
    "    :param ecami_output_path: Path, output text file from eCAMI\n",
    "    :param fasta_path: Path, FASTA file used as input by eCAMI\n",
    "    \n",
    "    Returns a list of ECAMIprediction instances. One instance per protein in the FASTA file used as eCAMI input\"\"\"\n",
    "    # Read the output file from eCAMI\n",
    "    with open(ecami_output_path, \"r\") as fh:\n",
    "        ecami_file = fh.read().splitlines()\n",
    "\n",
    "    ecami_predictions = {}  # stores proteins {protein_accession:ECAMIprediction_instance}\n",
    "\n",
    "    for line in tqdm(ecami_file, desc=\"Parsing eCAMI output\"):\n",
    "        if not line.startswith('>'):  # only want the first line for a protien not the listed k-mers\n",
    "            continue\n",
    "\n",
    "        prediction_output = line.split(\"\\t\")\n",
    "\n",
    "        # retrieve protein accession, removing the '>' prefix and stripping white space\n",
    "        protein_accession = prediction_output[0][1:].strip()\n",
    "\n",
    "        # retrieve the CAZy family\n",
    "        cazy_family = prediction_output[1].split(\":\")[0]\n",
    "\n",
    "        # retrieve the CAZy subfamily and EC numbers\n",
    "        cazy_subfamilies, ec_numbers = get_subfamily_ec_numbers(prediction_output[2], cazy_family)\n",
    "\n",
    "        # add ECAMIprediction instance to ecami_prediction\n",
    "        cazyme_classification = 1  # all proteins included in eCAMI output file are identified as CAZymes\n",
    "\n",
    "    try:\n",
    "        # add new predicted CAZyme domain to an existing ECAMIprediction instance\n",
    "        ecami_predictions[protein_accession]\n",
    "        \n",
    "        # check if the CAZyme domain has already been passed\n",
    "        existing_cazyme_domains = existing_prediction.cazyme_domains\n",
    "        existance = False\n",
    "        while existance == False:\n",
    "            for domain in existing_cazyme_domains:\n",
    "                if (domain.cazy_family == cazy_family) and (domain.cazy_subfamily == cazy_subfamily):\n",
    "                    for ec in ec_numbers:\n",
    "                        if ec not in domain.ec_numbers:\n",
    "                            domain.ec_numbers.append(ec)\n",
    "                    existance = True\n",
    "        \n",
    "        if existance is False:\n",
    "            # create new CAZyme domain\n",
    "            new_cazyme_domain = CazymeDomain(\n",
    "                protein_accession,\n",
    "                cazy_family,\n",
    "                cazy_subfamily,\n",
    "                ec_numbers,\n",
    "            )\n",
    "            existing_prediction.cazyme_domains.append(new_cazyme_domain)\n",
    "    \n",
    "    except KeyError:  # raised if no corresponding ECAMIprediction instance was found\n",
    "\n",
    "            new_cazyme_domain = CazymeDomain(\n",
    "                protein_accession,\n",
    "                cazy_family,\n",
    "                cazy_subfamily,\n",
    "                ec_numbers,\n",
    "                domain_range,\n",
    "            )\n",
    "            \n",
    "            new_protein = CazymeProteinPrediction(\n",
    "                \"eCAMI\",\n",
    "                protein_accession,\n",
    "                cazyme_classification,\n",
    "            )\n",
    "    \n",
    "            new_protein.cazyme_domains.append(new_cazyme_domain)\n",
    "\n",
    "    \n",
    "    ecami_predictions = add_non_cazymes(ecami_predictions, fasta_path)\n",
    "    \n",
    "    return list(ecami_predictions.values())\n",
    "\n",
    "                \n",
    "def get_subfamily_ec_numbers(subfam_group, cazy_family):\n",
    "    \"\"\"Retrieve the predicted CAZy subfamily and associated EC numbers.\n",
    "    \n",
    "    Retrieves only the child CAZy subfamilies for the CAZy familiy of the current working\n",
    "    CAZyme domain in the protein. Returns the CAZy subfamilies as a string, and EC numbers\n",
    "    as a list of string, with each string containing a unique EC number. The 'subfam_group'\n",
    "    refers to the 'subfam_name_of_the_group:subfam_name_count' in the eCAMI output file.\n",
    "    \n",
    "    :param subfam_group: string, the 'subfam_name_of_the_group:subfam_name_count'\n",
    "    :param cazy_family: string, CAZy family for the current working CAZyme domain\n",
    "    \n",
    "    Return the CAZy subfamily for the CAZyme domain (str) and list of associated EC numbers.\n",
    "    \"\"\"\n",
    "    cazy_subfamilies = []  # store all listed predicted CAZy subfamilies\n",
    "    ec_numbers = []  # store all predicted EC numbers\n",
    "\n",
    "    # individual items are separated by \"|\"\n",
    "    subfam_group = subfam_group.split(\"|\")\n",
    "\n",
    "    # check if the subfam_group contains a predicted CAZy subfamily\n",
    "    for item in subfam_group:\n",
    "        # remove the group number of the predicted item\n",
    "        item = item.split(\":\")[0]\n",
    "\n",
    "        # check if the item contains a CAZy subfamily\n",
    "        try:\n",
    "            re.match(r\"\\D{2,3}\\d+_\\d+\", item).group()\n",
    "            # check if the subfamily belongs to the CAZy family listed for the current CAZyme domain\n",
    "            if item[:item.find(\"_\")] == cazy_family:\n",
    "                cazy_subfamilies.append(item)\n",
    "\n",
    "        except AttributeError:  # raised if the item is not a CAZy subfamily\n",
    "            pass\n",
    "\n",
    "        # check if item contains an EC number\n",
    "        try:\n",
    "            re.match(r\"\\d+\\.(\\d+|-)\\.(\\d+|-)\\.(\\d+|-)\", item).group()\n",
    "            ec_numbers.append(item)\n",
    "\n",
    "        except AttributeError:  # raised if the item is not an EC number\n",
    "            pass\n",
    "\n",
    "    cazy_subfamilies = \", \".join(cazy_subfamilies)  # convert to string\n",
    "\n",
    "    return cazy_subfamilies, ec_numbers\n",
    "\n",
    "\n",
    "def add_non_cazymes(ecami_predictions, fasta_path):\n",
    "    \"\"\"Add non-CAZymes to the parsed eCAMI output.\n",
    "    \n",
    "    The eCAMI output only includes the protein it predicts are CAZymes. Therefore, this function\n",
    "    goes through the FASTA file that was used as input by eCAMI and adds the proteins not\n",
    "    classified as CAZymes to the parsed proteins.\n",
    "    \n",
    "    :param ecami_predictions: dict, keyed by protein accession and valued by ECAMIprediction instance\n",
    "    :param fasta_path: Path, FASTA file used as input by eCAMI\n",
    "    \n",
    "    Return a dictionary keyed by protein accession and valued by corresponding ECAMIprediction instance.\n",
    "    \"\"\"\n",
    "    # Add non-CAZymes\n",
    "    # open the FASTA file containing the input protein sequences\n",
    "    with open(fasta_path, \"r\") as fh:\n",
    "        fasta = fh.read().splitlines()\n",
    "\n",
    "    for line in tqdm(fasta, desc=\"Adding non-CAZymes\"):\n",
    "        if line.startswith(\">\"):\n",
    "\n",
    "            # remove '>' prefix and white space\n",
    "            protein_accession = line[1:].strip()\n",
    "\n",
    "            # check if the protein is already listed in the eCAMI predictions\n",
    "            try:\n",
    "                ecami_predictions[protein_accession]\n",
    "\n",
    "            except KeyError:  # raised of protein not in ecami_predictions\n",
    "                cazyme_classification = 0\n",
    "                ecami_predictions[protein_accession] = CazymeProteinPrediction(\n",
    "                    \"eCAMI\",\n",
    "                    protein_accession,\n",
    "                    cazyme_classification,\n",
    "                )\n",
    "    \n",
    "    return ecami_predictions\n",
    "\n",
    "results = parse_ecami_output(\"8080495_output.txt\", \"genbank_proteins_txid73501_GCA_008080495_1.fasta\")\n",
    "print(\"proteins parsed=\", len(results))\n",
    "print(\"proteins in the input FASTA file=\", 9287)\n",
    "print(\"output=\", results[0])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "From before we know there are 9,287 proteins in the FASTA file that was parsed by eCAMI, so we can see that all proteins are now stored in the list `results`."
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
