
<!DOCTYPE html>

<html>
  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0">
    <title>The genbank module &#8212; pyrewton 0.1.1 documentation</title>
    <link rel="stylesheet" href="_static/alabaster.css" type="text/css" />
    <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
    <script id="documentation_options" data-url_root="./" src="_static/documentation_options.js"></script>
    <script src="_static/jquery.js"></script>
    <script src="_static/underscore.js"></script>
    <script src="_static/doctools.js"></script>
    <script src="_static/language_data.js"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
   
  <link rel="stylesheet" href="_static/custom.css" type="text/css" />
  
  
  <meta name="viewport" content="width=device-width, initial-scale=0.9, maximum-scale=0.9" />

  </head><body>
  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          

          <div class="body" role="main">
            
  <div class="section" id="the-genbank-module">
<h1>The genbank module<a class="headerlink" href="#the-genbank-module" title="Permalink to this headline">¶</a></h1>
<p>The module <code class="docutils literal notranslate"><span class="pre">genbank</span></code> contains submodules that handle GenBank files. This includes the retrieval of GenBank files from
the NCBI Assembly database, and retrieval of cazyme annotations from GenBank files.</p>
<div class="line-block">
<div class="line">- <a class="reference internal" href="#genbank-mod-get-ncbi-genomes-def"><span class="std std-ref">get_ncbi_genomes</span></a></div>
<div class="line">- <a class="reference internal" href="#genbank-mod-get-cazyme-annotations-def"><span class="std std-ref">get_cazyme_annotations</span></a></div>
</div>
<div class="section" id="get-ncbi-genomes">
<span id="genbank-mod-get-ncbi-genomes-def"></span><h2>get_ncbi_genomes<a class="headerlink" href="#get-ncbi-genomes" title="Permalink to this headline">¶</a></h2>
<p><code class="docutils literal notranslate"><span class="pre">get_ncbi_genomes</span></code> is a submodule of the <code class="docutils literal notranslate"><span class="pre">genbank</span></code>, it takes a plain text file containing species scientific names
or NCBI taxonomy as input. The script finds the corresponding taxonomy ID or scientific name, as appropriate, as well
as retrieve all directly linked accession numbers. The submodule generates a dataframe containing ‘Genus’, ‘Species’,
‘NCBI Taxonomy ID’, and ‘NCBI Accession Number’.</p>
<div class="section" id="invoking-the-script-from-the-command-line">
<h3>Invoking the script from the command line<a class="headerlink" href="#invoking-the-script-from-the-command-line" title="Permalink to this headline">¶</a></h3>
<p><strong>Compulsary arguments</strong></p>
<p><code class="docutils literal notranslate"><span class="pre">-u,</span> <span class="pre">--user</span></code> - Although indicated as optional, Entrez requires an email address must be provided.
If not provided the programme will log this as an error and terminate.</p>
<p><strong>Optional arguments</strong></p>
<p><code class="docutils literal notranslate"><span class="pre">-d,</span> <span class="pre">--`dataframe</span></code> - Specify output path for dataframe, which will be saved as a .csv file
(inclusion of file extensions is optional). If not provided dataframe will be written out to STDOUT.</p>
<p><code class="docutils literal notranslate"><span class="pre">-f,</span> <span class="pre">--force</span></code> - Enable writting in specificed output directory if output directory already exists.</p>
<p><code class="docutils literal notranslate"><span class="pre">-g,</span> <span class="pre">--genbank</span></code> - Enable or disable downloading of GenBank files.</p>
<p><code class="docutils literal notranslate"><span class="pre">-h,</span> <span class="pre">--help</span></code> - Display help messages and exit</p>
<p><code class="docutils literal notranslate"><span class="pre">-i,</span> <span class="pre">--input</span></code> - Specify input filename (with extension) input file.If only the filename is supplied
<code class="docutils literal notranslate"><span class="pre">get_ncbi_genomes.py</span></code> will only look in the curent working directory, otherwise provide path to
input file. If no option is given the default input is taken from STDIN.</p>
<p><code class="docutils literal notranslate"><span class="pre">-l,</span> <span class="pre">--log</span></code> - Specify name of log file (With extension). If only filename is given, log file will
be written out to the current working directory, otherwise provide path including filename. If not
option is given no log file will be written out, however, logs will still be printed to the terminal.</p>
<p><code class="docutils literal notranslate"><span class="pre">-n,</span> <span class="pre">--nodelete</span></code> - Enable not deleting files in existing output directory. If not enabled, output
directory exists and writing in output directory is ‘forced’ then files in output directory will not
be deleted, and new files will be written to the output directory.</p>
<p><code class="docutils literal notranslate"><span class="pre">-o,</span> <span class="pre">--output</span></code> - Specify filename (with extension) of output file. If not option is given output
will be written to STDOUT.</p>
<p><code class="docutils literal notranslate"><span class="pre">-r,</span> <span class="pre">--retries</span></code> - Specifiy maximum number of retries before cancelling call to NCBI
if a network error is encountered. The default is a maximum of 10 retries. When
maximum is reached, a value of ‘NA’ is returned.</p>
<p><code class="docutils literal notranslate"><span class="pre">-t,</span> <span class="pre">--timeout</span></code> - Specify timeout limit of URL connection when downloading GenBank files.
Default is 10 seconds.</p>
<p><code class="docutils literal notranslate"><span class="pre">-v,</span> <span class="pre">--verbose</span></code> - Enable verbose logging - changes logger level from WARNING to INFO.</p>
</div>
</div>
<div class="section" id="get-cazyme-annotations">
<span id="genbank-mod-get-cazyme-annotations-def"></span><h2>get_cazyme_annotations<a class="headerlink" href="#get-cazyme-annotations" title="Permalink to this headline">¶</a></h2>
<p><code class="docutils literal notranslate"><span class="pre">get_cazyme_annotations</span></code> is a submodule of the <code class="docutils literal notranslate"><span class="pre">genbank</span></code>, which reads NCBI GenBank files and
retrieves annotations of encoded cazymes. Where applicable, protein data from UniProt for each
cazyme is also retrieved. All cazyme data is written out to a dataframe, with a unique protein
per row, containing:</p>
<ul class="simple">
<li><p>‘Genus’</p></li>
<li><p>‘Species’</p></li>
<li><p>‘NCBI Taxonomy ID’</p></li>
<li><p>‘NCBI Accession Number’</p></li>
<li><p>‘NCBI Protein ID’</p></li>
<li><p>‘Locus Tag’</p></li>
<li><p>‘Gene Locus’</p></li>
<li><p>‘NCBI Recorded Function’</p></li>
<li><p>‘Protein Sequence’</p></li>
<li><p>‘UniProtKB Entry ID’</p></li>
<li><p>‘UniProt Entry Name’</p></li>
<li><p>‘UniProtKB Protein Names’</p></li>
<li><p>‘EC number’</p></li>
<li><p>‘Length (Aa)’</p></li>
<li><p>‘Mass (Da)’</p></li>
<li><p>‘Domains’</p></li>
<li><p>‘Domain count’</p></li>
<li><p>‘UniProtKB Linked Protein Families’</p></li>
<li><p>‘Gene ontology IDs’</p></li>
<li><p>‘Gene ontology (molecular function)’</p></li>
<li><p>‘Gene ontology (biological process)’</p></li>
</ul>
<div class="section" id="id1">
<h3>Invoking the script from the command line<a class="headerlink" href="#id1" title="Permalink to this headline">¶</a></h3>
<p><code class="docutils literal notranslate"><span class="pre">-d,</span> <span class="pre">--df_input</span></code> - Path to input dataframe.</p>
<p><code class="docutils literal notranslate"><span class="pre">-f,</span> <span class="pre">--force</span></code> - Enable writting in specificed output directory if output directory already exists.</p>
<p><code class="docutils literal notranslate"><span class="pre">-g,</span> <span class="pre">--genbank</span></code> - Path to director containing GenBank files.</p>
<p><code class="docutils literal notranslate"><span class="pre">-l,</span> <span class="pre">--log</span></code> - Specify name of log file (With extension). If only filename is given, log file will
be written out to the current working directory, otherwise provide path including filename. If not
option is given no log file will be written out, however, logs will still be printed to the terminal.</p>
<p><code class="docutils literal notranslate"><span class="pre">-n,</span> <span class="pre">--nodelete</span></code> - Enable not deleting files in existing output directory. If not enabled, output
directory exists and writing in output directory is ‘forced’ then files in output directory will not
be deleted, and new files will be written to the output directory.</p>
<p><code class="docutils literal notranslate"><span class="pre">-o,</span> <span class="pre">--output</span></code> - Specify filename (with extension) of output file. If not option is given output
will be written to STDOUT.</p>
<p><code class="docutils literal notranslate"><span class="pre">-v,</span> <span class="pre">--verbose</span></code> - Enable verbose logging - changes logger level from WARNING to INFO.</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>get_cazyme_annotations is still under development.
Please see the <a class="reference external" href="https://github.com/HobnobMancer/PhD_Project_Scripts/tree/master">GitHub repository</a> for the latest developments.</p>
</div>
</div>
</div>
</div>


          </div>
          
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
<h1 class="logo"><a href="index.html">pyrewton</a></h1>








<h3>Navigation</h3>

<div class="relations">
<h3>Related Topics</h3>
<ul>
  <li><a href="index.html">Documentation overview</a><ul>
  </ul></li>
</ul>
</div>
<div id="searchbox" style="display: none" role="search">
  <h3 id="searchlabel">Quick search</h3>
    <div class="searchformwrapper">
    <form class="search" action="search.html" method="get">
      <input type="text" name="q" aria-labelledby="searchlabel" />
      <input type="submit" value="Go" />
    </form>
    </div>
</div>
<script>$('#searchbox').show(0);</script>








        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="footer">
      &copy;2020, Emma E. M. Hobbs.
      
      |
      Powered by <a href="http://sphinx-doc.org/">Sphinx 3.1.1</a>
      &amp; <a href="https://github.com/bitprophet/alabaster">Alabaster 0.7.12</a>
      
      |
      <a href="_sources/genbank.rst.txt"
          rel="nofollow">Page source</a>
    </div>

    

    
  </body>
</html>